diff --git a/web/tests/compose.test.js b/web/tests/compose.test.js
index be61661f60..5bc7d97d91 100644
--- a/web/tests/compose.test.js
+++ b/web/tests/compose.test.js
@@ -235,8 +235,8 @@ test_ui("send_message", ({override, override_rewire, mock_template}) => {
         override(compose_pm_pill, "get_emails", () => "alice@example.com");
 
         const server_message_id = 127;
-        override(markdown, "apply_markdown", noop);
-        override(markdown, "add_topic_links", noop);
+        override(markdown, "render", noop);
+        override(markdown, "get_topic_links", noop);
 
         override_rewire(echo, "try_deliver_locally", (message_request) => {
             const local_id_float = 123.04;
@@ -730,7 +730,7 @@ test_ui("on_events", ({override, override_rewire}) => {
         assert.ok(make_indicator_called);
         assert_visibilities();
 
-        let apply_markdown_called = false;
+        let render_called = false;
         $("textarea#compose-textarea").val("foobarfoobar");
         setup_visibilities();
         setup_mock_markdown_contains_backend_only_syntax("foobarfoobar", false);
@@ -738,15 +738,14 @@ test_ui("on_events", ({override, override_rewire}) => {
 
         current_message = "foobarfoobar";
 
-        override(markdown, "apply_markdown", (msg) => {
-            assert.equal(msg.raw_content, "foobarfoobar");
-            apply_markdown_called = true;
-            return msg;
+        override(markdown, "render", (raw_content) => {
+            assert.equal(raw_content, "foobarfoobar");
+            render_called = true;
         });
 
         handler(event);
 
-        assert.ok(apply_markdown_called);
+        assert.ok(render_called);
         assert_visibilities();
         assert.equal($("#compose .preview_content").html(), "Server: foobarfoobar");
     })();
diff --git a/web/tests/drafts.test.js b/web/tests/drafts.test.js
index 3d1d7c5148..16c0a8137e 100644
--- a/web/tests/drafts.test.js
+++ b/web/tests/drafts.test.js
@@ -31,7 +31,7 @@ set_global("setTimeout", (f, delay) => {
     f();
 });
 mock_esm("../src/markdown", {
-    apply_markdown: noop,
+    render: noop,
 });
 mock_esm("../src/overlays", {
     open_overlay: noop,
diff --git a/web/tests/echo.test.js b/web/tests/echo.test.js
index 61666909d3..80b72d7a2a 100644
--- a/web/tests/echo.test.js
+++ b/web/tests/echo.test.js
@@ -221,16 +221,16 @@ run_test("insert_local_message streams", ({override}) => {
 
     const local_id_float = 101.01;
 
-    let apply_markdown_called = false;
-    let add_topic_links_called = false;
+    let render_called = false;
+    let get_topic_links_called = false;
     let insert_message_called = false;
 
-    override(markdown, "apply_markdown", () => {
-        apply_markdown_called = true;
+    override(markdown, "render", () => {
+        render_called = true;
     });
 
-    override(markdown, "add_topic_links", () => {
-        add_topic_links_called = true;
+    override(markdown, "get_topic_links", () => {
+        get_topic_links_called = true;
     });
 
     const insert_new_messages = ([message]) => {
@@ -245,14 +245,15 @@ run_test("insert_local_message streams", ({override}) => {
     const message_request = {
         type: "stream",
         stream_id: general_sub.stream_id,
+        topic: "important note",
         sender_email: "iago@zulip.com",
         sender_full_name: "Iago",
         sender_id: 123,
     };
     echo.insert_local_message(message_request, local_id_float, insert_new_messages);
 
-    assert.ok(apply_markdown_called);
-    assert.ok(add_topic_links_called);
+    assert.ok(render_called);
+    assert.ok(get_topic_links_called);
     assert.ok(insert_message_called);
 });
 
@@ -273,8 +274,7 @@ run_test("insert_local_message direct message", ({override}) => {
     params.cross_realm_bots = [];
     people.initialize(page_params.user_id, params);
 
-    let add_topic_links_called = false;
-    let apply_markdown_called = false;
+    let render_called = false;
     let insert_message_called = false;
 
     const insert_new_messages = ([message]) => {
@@ -282,12 +282,8 @@ run_test("insert_local_message direct message", ({override}) => {
         insert_message_called = true;
     };
 
-    override(markdown, "apply_markdown", () => {
-        apply_markdown_called = true;
-    });
-
-    override(markdown, "add_topic_links", () => {
-        add_topic_links_called = true;
+    override(markdown, "render", () => {
+        render_called = true;
     });
 
     const message_request = {
@@ -298,16 +294,14 @@ run_test("insert_local_message direct message", ({override}) => {
         sender_id: 123,
     };
     echo.insert_local_message(message_request, local_id_float, insert_new_messages);
-    assert.ok(add_topic_links_called);
-    assert.ok(apply_markdown_called);
+    assert.ok(render_called);
     assert.ok(insert_message_called);
 });
 
 run_test("test reify_message_id", ({override}) => {
     const local_id_float = 103.01;
 
-    override(markdown, "apply_markdown", noop);
-    override(markdown, "add_topic_links", noop);
+    override(markdown, "render", noop);
 
     const message_request = {
         type: "stream",
diff --git a/web/tests/markdown.test.js b/web/tests/markdown.test.js
index 6e3c66cd9b..c1a5c57de3 100644
--- a/web/tests/markdown.test.js
+++ b/web/tests/markdown.test.js
@@ -263,9 +263,12 @@ test("marked_shared", () => {
             continue;
         }
 
-        const message = {raw_content: test.input};
+        let message = {raw_content: test.input};
         user_settings.translate_emoticons = test.translate_emoticons || false;
-        markdown.apply_markdown(message);
+        message = {
+            ...message,
+            ...markdown.render(message.raw_content),
+        };
         const output = message.content;
         const error_message = `Failure in test: ${test.name}`;
         if (test.marked_expected_output) {
@@ -281,20 +284,32 @@ test("marked_shared", () => {
 
 test("message_flags", () => {
     let message = {raw_content: "@**Leo**"};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.ok(!message.flags.includes("mentioned"));
 
     message = {raw_content: "@**Cordelia, Lear's daughter**"};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.ok(message.flags.includes("mentioned"));
 
     message = {raw_content: "@**all**"};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.ok(message.flags.includes("stream_wildcard_mentioned"));
     assert.ok(!message.flags.includes("topic_wildcard_mentioned"));
 
     message = {raw_content: "@**topic**"};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.ok(!message.flags.includes("stream_wildcard_mentioned"));
     assert.ok(message.flags.includes("topic_wildcard_mentioned"));
 });
@@ -617,8 +632,7 @@ test("marked", () => {
         const input = test_case.input;
         const expected = test_case.expected;
 
-        const message = {raw_content: input};
-        markdown.apply_markdown(message);
+        const message = markdown.render(input);
         const output = message.content;
         assert.equal(output, expected);
     }
@@ -626,11 +640,11 @@ test("marked", () => {
 
 test("topic_links", () => {
     let message = {type: "stream", topic: "No links here"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 0);
 
     message = {type: "stream", topic: "One #123 link here"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 1);
     assert.deepEqual(message.topic_links[0], {
         url: "https://trac.example.com/ticket/123",
@@ -638,7 +652,7 @@ test("topic_links", () => {
     });
 
     message = {type: "stream", topic: "Two #123 #456 link here"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 2);
     assert.deepEqual(message.topic_links[0], {
         url: "https://trac.example.com/ticket/123",
@@ -650,7 +664,7 @@ test("topic_links", () => {
     });
 
     message = {type: "stream", topic: "New ZBUG_123 link here"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 1);
     assert.deepEqual(message.topic_links[0], {
         url: "https://trac2.zulip.net/ticket/123",
@@ -658,7 +672,7 @@ test("topic_links", () => {
     });
 
     message = {type: "stream", topic: "New ZBUG_123 with #456 link here"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 2);
     assert.deepEqual(message.topic_links[0], {
         url: "https://trac2.zulip.net/ticket/123",
@@ -670,7 +684,7 @@ test("topic_links", () => {
     });
 
     message = {type: "stream", topic: "One ZGROUP_123:45 link here"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 1);
     assert.deepEqual(message.topic_links[0], {
         url: "https://zone_45.zulip.net/ticket/123",
@@ -678,7 +692,7 @@ test("topic_links", () => {
     });
 
     message = {type: "stream", topic: "Hello https://google.com"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 1);
     assert.deepEqual(message.topic_links[0], {
         url: "https://google.com",
@@ -686,7 +700,7 @@ test("topic_links", () => {
     });
 
     message = {type: "stream", topic: "#456 https://google.com https://github.com"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 3);
     assert.deepEqual(message.topic_links[0], {
         url: "https://trac.example.com/ticket/456",
@@ -702,11 +716,11 @@ test("topic_links", () => {
     });
 
     message = {type: "not-stream"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 0);
 
     message = {type: "stream", topic: "FOO_abcde;e;zulip;luxembourg;foo;23;testing"};
-    markdown.add_topic_links(message);
+    message.topic_links = markdown.get_topic_links(message.topic);
     assert.equal(message.topic_links.length, 1);
     assert.deepEqual(message.topic_links[0], {
         url: "https://zone_e.zulip.net/ticket/luxembourg/abcde?name=foo&chapter=23#testi",
@@ -717,20 +731,29 @@ test("topic_links", () => {
 test("message_flags", () => {
     let input = "/me is testing this";
     let message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
 
     assert.equal(message.is_me_message, true);
     assert.ok(!message.unread);
 
     input = "/me is testing\nthis";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
 
     assert.equal(message.is_me_message, true);
 
     input = "testing this @**all** @**Cordelia, Lear's daughter**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.is_me_message, false);
     assert.equal(message.flags.includes("mentioned"), true);
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), true);
@@ -738,7 +761,10 @@ test("message_flags", () => {
 
     input = "test @**everyone**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.is_me_message, false);
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), true);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
@@ -746,7 +772,10 @@ test("message_flags", () => {
 
     input = "test @**stream**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.is_me_message, false);
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), true);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
@@ -754,7 +783,10 @@ test("message_flags", () => {
 
     input = "test @**topic**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.is_me_message, false);
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), true);
@@ -762,98 +794,140 @@ test("message_flags", () => {
 
     input = "test @all";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @everyone";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @topic";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @any";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @alleycat.com";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @*hamletcharacters*";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), true);
 
     input = "test @*backend*";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @**invalid_user**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @_**all**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "> test @**all**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @_**topic**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "> test @**topic**";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "test @_*hamletcharacters*";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
 
     input = "> test @*hamletcharacters*";
     message = {topic: "No links here", raw_content: input};
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.flags.includes("stream_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("topic_wildcard_mentioned"), false);
     assert.equal(message.flags.includes("mentioned"), false);
@@ -929,9 +1003,12 @@ test("parse_non_message", () => {
 });
 
 test("missing unicode emojis", ({override}) => {
-    const message = {raw_content: "\u{1F6B2}"};
+    let message = {raw_content: "\u{1F6B2}"};
 
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(
         message.content,
         '<p><span aria-label="bike" class="emoji emoji-1f6b2" role="img" title="bike">:bike:</span></p>',
@@ -941,7 +1018,10 @@ test("missing unicode emojis", ({override}) => {
     override(emoji_codes.codepoint_to_name, "1f6b2", undefined);
 
     markdown.initialize(markdown_config.get_helpers());
-    markdown.apply_markdown(message);
+    message = {
+        ...message,
+        ...markdown.render(message.raw_content),
+    };
     assert.equal(message.content, "<p>\u{1F6B2}</p>");
 });
 
@@ -952,7 +1032,7 @@ test("katex_throws_unexpected_exceptions", ({override_rewire}) => {
             throw new Error("some-exception");
         },
     });
-    assert.throws(() => markdown.apply_markdown(message), {
+    assert.throws(() => markdown.render(message.raw_content), {
         name: "Error",
         message: "some-exception\nPlease report this to https://zulip.com/development-community/",
     });
diff --git a/web/tests/markdown_parse.test.js b/web/tests/markdown_parse.test.js
index 1e53500078..274902fc69 100644
--- a/web/tests/markdown_parse.test.js
+++ b/web/tests/markdown_parse.test.js
@@ -223,15 +223,15 @@ run_test("linkifiers", () => {
 });
 
 function assert_topic_links(topic, expected_links) {
-    const topic_links = markdown.get_topic_links({
-        topic,
-        get_linkifier_map: linkifiers.get_linkifier_map,
-    });
+    const topic_links = markdown.get_topic_links(topic);
     assert.deepEqual(topic_links, expected_links);
 }
 
 run_test("topic links", () => {
     linkifiers.initialize([{pattern: "#foo(?P<id>\\d+)", url_template: "http://foo.com/{id}"}]);
+    markdown.initialize({
+        get_linkifier_map: linkifiers.get_linkifier_map,
+    });
     assert_topic_links("progress on #foo101 and #foo102", [
         {
             text: "#foo101",
