diff --git a/lib/segment/src/index/field_index/numeric_index/tests.rs b/lib/segment/src/index/field_index/numeric_index/tests.rs
index d2a7b3634..9d5765171 100644
--- a/lib/segment/src/index/field_index/numeric_index/tests.rs
+++ b/lib/segment/src/index/field_index/numeric_index/tests.rs
@@ -166,6 +166,7 @@ fn cardinality_request(
             &hw_counter,
         )
         .unwrap()
+        .unwrap()
         .unique()
         .collect_vec();
 
@@ -607,7 +608,11 @@ fn test_cond<
     let condition = FieldCondition::new_range(JsonPath::new("unused"), ordered_range);
     let hw_acc = HwMeasurementAcc::new();
     let hw_counter = hw_acc.get_counter_cell();
-    let offsets = index.filter(&condition, &hw_counter).unwrap().collect_vec();
+    let offsets = index
+        .filter(&condition, &hw_counter)
+        .unwrap()
+        .unwrap()
+        .collect_vec();
     assert_eq!(offsets, result);
 }
