diff --git a/pkg/configs/tidb/config_test.go b/pkg/configs/tidb/config_test.go
index ae36616d0..811746be5 100644
--- a/pkg/configs/tidb/config_test.go
+++ b/pkg/configs/tidb/config_test.go
@@ -73,6 +73,9 @@ func TestOverlay(t *testing.T) {
 	cluster := &v1alpha1.Cluster{
 		Spec: v1alpha1.ClusterSpec{
 			TLSCluster: &v1alpha1.TLSCluster{Enabled: true},
+			BootstrapSQL: &corev1.LocalObjectReference{
+				Name: "bootstrap-sql",
+			},
 		},
 		Status: v1alpha1.ClusterStatus{
 			PD: "https://basic-pd.ns1:2379",
@@ -95,9 +98,6 @@ func TestOverlay(t *testing.T) {
 							Name: "auth-token-jwks",
 						},
 					},
-					BootstrapSQL: &corev1.LocalObjectReference{
-						Name: "bootstrap-sql",
-					},
 					TLS: &v1alpha1.TiDBTLS{
 						MySQL: &v1alpha1.TLS{
 							Enabled: true,
diff --git a/tests/e2e/cluster/cluster.go b/tests/e2e/cluster/cluster.go
index 8172808cd..0289d2da3 100644
--- a/tests/e2e/cluster/cluster.go
+++ b/tests/e2e/cluster/cluster.go
@@ -55,7 +55,6 @@ import (
 	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
 	"github.com/pingcap/tidb-operator/tests/e2e/config"
 	"github.com/pingcap/tidb-operator/tests/e2e/utils/data"
-	"github.com/pingcap/tidb-operator/tests/e2e/utils/jwt"
 	"github.com/pingcap/tidb-operator/tests/e2e/utils/k8s"
 	utiltidb "github.com/pingcap/tidb-operator/tests/e2e/utils/tidb"
 )
@@ -1357,144 +1356,6 @@ var _ = Describe("TiDB Cluster", func() {
 	})
 
 	Context("TiDB Feature", func() {
-		It("should init a cluster with bootstrap SQL specified", func() {
-			By("Creating a ConfigMap with bootstrap SQL")
-			bsqlCm := corev1.ConfigMap{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      "bootstrap-sql",
-					Namespace: ns.Name,
-				},
-				Data: map[string]string{
-					v1alpha1.BootstrapSQLConfigMapKey: "SET PASSWORD FOR 'root'@'%' = 'pingcap';",
-				},
-			}
-			Expect(k8sClient.Create(ctx, &bsqlCm)).To(Succeed())
-
-			By("Creating the components with bootstrap SQL")
-			pdg := data.NewPDGroup(ns.Name, "pdg", tc.Name, ptr.To(int32(1)), nil)
-			kvg := data.NewTiKVGroup(ns.Name, "kvg", tc.Name, ptr.To(int32(1)), nil)
-			dbg := data.NewTiDBGroup(ns.Name, "dbg", tc.Name, ptr.To(int32(1)), func(group *v1alpha1.TiDBGroup) {
-				group.Spec.Template.Spec.Security = &v1alpha1.TiDBSecurity{
-					BootstrapSQL: &corev1.LocalObjectReference{
-						Name: bsqlCm.Name,
-					},
-				}
-			})
-			Expect(k8sClient.Create(ctx, pdg)).To(Succeed())
-			Expect(k8sClient.Create(ctx, kvg)).To(Succeed())
-			Expect(k8sClient.Create(ctx, dbg)).To(Succeed())
-
-			By("Checking the status of the cluster and the connection to the TiDB service")
-			Eventually(func(g Gomega) {
-				tcGet, ready := utiltidb.IsClusterReady(k8sClient, tc.Name, tc.Namespace)
-				g.Expect(ready).To(BeTrue())
-				g.Expect(len(tcGet.Status.PD)).NotTo(BeZero())
-				for _, compStatus := range tcGet.Status.Components {
-					switch compStatus.Kind {
-					case v1alpha1.ComponentKindPD, v1alpha1.ComponentKindTiKV, v1alpha1.ComponentKindTiDB:
-						g.Expect(compStatus.Replicas).To(Equal(int32(1)))
-					default:
-						g.Expect(compStatus.Replicas).To(BeZero())
-					}
-				}
-
-				g.Expect(utiltidb.AreAllInstancesReady(k8sClient, pdg,
-					[]*v1alpha1.TiKVGroup{kvg}, []*v1alpha1.TiDBGroup{dbg}, nil)).To(Succeed())
-
-				// connect with the password set in bootstrap SQL
-				g.Expect(utiltidb.IsTiDBConnectable(ctx, k8sClient, fw,
-					tc.Namespace, tc.Name, dbg.Name, "root", "pingcap", "")).To(Succeed())
-			}).WithTimeout(createClusterTimeout).WithPolling(createClusterPolling).Should(Succeed())
-		})
-
-		It("should connect to the TiDB cluster with JWT authentication", func() {
-			const (
-				kid   = "the-key-id-0"
-				sub   = "user@pingcap.com"
-				email = "user@pingcap.com"
-				iss   = "issuer-abc"
-			)
-			token, err := jwt.GenerateJWT(kid, sub, email, iss)
-			if err != nil {
-				Skip(fmt.Sprintf("failed to generate JWT token: %v", err))
-			}
-
-			// use Bootstrap SQL to create the a user with JWT authentication
-			bsqlCm := corev1.ConfigMap{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      "bootstrap-sql",
-					Namespace: ns.Name,
-				},
-				Data: map[string]string{
-					v1alpha1.BootstrapSQLConfigMapKey: fmt.Sprintf(
-						`CREATE USER '%s' IDENTIFIED WITH 'tidb_auth_token' REQUIRE TOKEN_ISSUER '%s' ATTRIBUTE '{"email": "%s"}';
-GRANT ALL PRIVILEGES ON *.* TO '%s'@'%s';`, sub, iss, email, sub, "%"),
-				},
-			}
-
-			// The `tidb_auth_token` authentication method requires clients to support
-			// the `mysql_clear_password` plugin to send the token to TiDB in plain text.
-			// Therefore, it's better to enale TLS between clients and servers before using `tidb_auth_token`.
-			By("Installing the certificates for conmutication between MySQL client and TiDB server")
-			Expect(installTiDBIssuer(ctx, yamlApplier, ns.Name, tc.Name)).To(Succeed())
-			Expect(installTiDBCertificates(ctx, yamlApplier, ns.Name, tc.Name, "dbg")).To(Succeed())
-
-			pdg := data.NewPDGroup(ns.Name, "pdg", tc.Name, ptr.To(int32(1)), nil)
-			kvg := data.NewTiKVGroup(ns.Name, "kvg", tc.Name, ptr.To(int32(1)), nil)
-			dbg := data.NewTiDBGroup(ns.Name, "dbg", tc.Name, ptr.To(int32(1)), func(group *v1alpha1.TiDBGroup) {
-				group.Spec.Template.Spec.Security = &v1alpha1.TiDBSecurity{
-					TLS: &v1alpha1.TiDBTLS{
-						MySQL: &v1alpha1.TLS{
-							Enabled: true,
-						},
-					},
-					BootstrapSQL: &corev1.LocalObjectReference{
-						Name: bsqlCm.Name,
-					},
-					AuthToken: &v1alpha1.TiDBAuthToken{
-						JWKs: corev1.LocalObjectReference{
-							Name: "jwks-secret",
-						},
-					},
-				}
-			})
-
-			By("Creating the JWKS secret")
-			jwksSecret := jwt.GenerateJWKSSecret(dbg.Namespace, "jwks-secret")
-			Expect(k8sClient.Create(ctx, &jwksSecret)).To(Succeed())
-
-			By("Creating the ConfigMap with a user created by Bootstrap SQL")
-			Expect(k8sClient.Create(ctx, &bsqlCm)).To(Succeed())
-
-			By("Creating the components with JWT authentication")
-			Expect(k8sClient.Create(ctx, pdg)).To(Succeed())
-			Expect(k8sClient.Create(ctx, kvg)).To(Succeed())
-			Expect(k8sClient.Create(ctx, dbg)).To(Succeed())
-
-			By("Checking the status of the cluster and the connection to the TiDB service")
-			Eventually(func(g Gomega) {
-				tcGet, ready := utiltidb.IsClusterReady(k8sClient, tc.Name, tc.Namespace)
-				g.Expect(ready).To(BeTrue())
-				g.Expect(len(tcGet.Status.PD)).NotTo(BeZero())
-				for _, compStatus := range tcGet.Status.Components {
-					switch compStatus.Kind {
-					case v1alpha1.ComponentKindPD, v1alpha1.ComponentKindTiKV, v1alpha1.ComponentKindTiDB:
-						g.Expect(compStatus.Replicas).To(Equal(int32(1)))
-					default:
-						g.Expect(compStatus.Replicas).To(BeZero())
-					}
-				}
-
-				g.Expect(utiltidb.AreAllInstancesReady(k8sClient, pdg,
-					[]*v1alpha1.TiKVGroup{kvg}, []*v1alpha1.TiDBGroup{dbg}, nil)).To(Succeed())
-
-				// connect with the JWT token
-				// TODO(liubo02): extract to common namer pkg
-				g.Expect(utiltidb.IsTiDBConnectable(ctx, k8sClient, fw,
-					tc.Namespace, tc.Name, dbg.Name, sub, token, dbg.Name+"-tidb-client-secret")).To(Succeed())
-			}).WithTimeout(createClusterTimeout).WithPolling(createClusterPolling).Should(Succeed())
-		})
-
 		It("should set store labels for TiKV and TiFlash, set server lables for TiDB", func() {
 			By("Creating the components with location labels")
 			pdg := data.NewPDGroup(ns.Name, "pdg", tc.Name, ptr.To(int32(1)), func(group *v1alpha1.PDGroup) {
@@ -1677,6 +1538,7 @@ location-labels = ["region", "zone", "host"]`
 									Name:  "testing-workload",
 									Image: "pingcap/testing-workload:latest",
 									Args: []string{
+										"--action", "workload",
 										"--host", clusterIP,
 										"--duration", "8",
 										"--max-connections", "30",
diff --git a/tests/e2e/data/cluster.go b/tests/e2e/data/cluster.go
index 1f6e2555f..60c37c7cc 100644
--- a/tests/e2e/data/cluster.go
+++ b/tests/e2e/data/cluster.go
@@ -15,13 +15,18 @@
 package data
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
 )
 
-func NewCluster(namespace string) *v1alpha1.Cluster {
-	return &v1alpha1.Cluster{
+const (
+	BootstrapSQLName = "bootstrap-sql"
+)
+
+func NewCluster(namespace string, patches ...ClusterPatch) *v1alpha1.Cluster {
+	c := &v1alpha1.Cluster{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: namespace,
 			Name:      defaultClusterName,
@@ -30,4 +35,18 @@ func NewCluster(namespace string) *v1alpha1.Cluster {
 			UpgradePolicy: v1alpha1.UpgradePolicyDefault,
 		},
 	}
+
+	for _, p := range patches {
+		p(c)
+	}
+
+	return c
+}
+
+func WithBootstrapSQL() ClusterPatch {
+	return func(obj *v1alpha1.Cluster) {
+		obj.Spec.BootstrapSQL = &corev1.LocalObjectReference{
+			Name: BootstrapSQLName,
+		}
+	}
 }
diff --git a/tests/e2e/data/data.go b/tests/e2e/data/data.go
index 54f049415..9dbc9c1db 100644
--- a/tests/e2e/data/data.go
+++ b/tests/e2e/data/data.go
@@ -32,6 +32,14 @@ const (
 	defaultTiFlashGroupName = "fg"
 
 	defaultVersion = "v8.1.0"
+
+	defaultImageRegistry = "gcr.io/pingcap-public/dbaas/"
+	defaultHelperImage   = "gcr.io/pingcap-public/dbaas/busybox:1.36.0"
+)
+
+const (
+	// TODO(liubo02): extract to namer
+	DefaultTiDBServiceName = defaultTiDBGroupName + "-tidb"
 )
 
 func WithReplicas[G runtime.Group](replicas int32) GroupPatch[G] {
diff --git a/tests/e2e/data/pd.go b/tests/e2e/data/pd.go
index 712eefb97..49890b038 100644
--- a/tests/e2e/data/pd.go
+++ b/tests/e2e/data/pd.go
@@ -17,6 +17,7 @@ package data
 import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
 
 	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
 	"github.com/pingcap/tidb-operator/pkg/runtime"
@@ -29,10 +30,12 @@ func NewPDGroup(ns string, patches ...GroupPatch[*runtime.PDGroup]) *v1alpha1.PD
 			Name:      defaultPDGroupName,
 		},
 		Spec: v1alpha1.PDGroupSpec{
-			Cluster: v1alpha1.ClusterReference{Name: defaultClusterName},
-			Version: defaultVersion,
+			Cluster:  v1alpha1.ClusterReference{Name: defaultClusterName},
+			Version:  defaultVersion,
+			Replicas: ptr.To[int32](1),
 			Template: v1alpha1.PDTemplate{
 				Spec: v1alpha1.PDTemplateSpec{
+					Image: ptr.To(defaultImageRegistry + "pd"),
 					Volumes: []v1alpha1.Volume{
 						{
 							Name:    "data",
diff --git a/tests/e2e/data/tidb.go b/tests/e2e/data/tidb.go
new file mode 100644
index 000000000..fe68b3edb
--- /dev/null
+++ b/tests/e2e/data/tidb.go
@@ -0,0 +1,83 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package data
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/runtime"
+)
+
+const (
+	JWKsSecretName = "jwks-secret"
+)
+
+func NewTiDBGroup(ns string, patches ...GroupPatch[*runtime.TiDBGroup]) *v1alpha1.TiDBGroup {
+	kvg := &runtime.TiDBGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      defaultTiDBGroupName,
+		},
+		Spec: v1alpha1.TiDBGroupSpec{
+			Cluster:  v1alpha1.ClusterReference{Name: defaultClusterName},
+			Version:  defaultVersion,
+			Replicas: ptr.To[int32](1),
+			Template: v1alpha1.TiDBTemplate{
+				Spec: v1alpha1.TiDBTemplateSpec{
+					Image: ptr.To(defaultImageRegistry + "tidb"),
+					SlowLog: &v1alpha1.TiDBSlowLog{
+						Image: ptr.To(defaultHelperImage),
+					},
+				},
+			},
+		},
+	}
+	for _, p := range patches {
+		p(kvg)
+	}
+
+	return runtime.ToTiDBGroup(kvg)
+}
+
+func WithAuthToken() GroupPatch[*runtime.TiDBGroup] {
+	return func(obj *runtime.TiDBGroup) {
+		if obj.Spec.Template.Spec.Security == nil {
+			obj.Spec.Template.Spec.Security = &v1alpha1.TiDBSecurity{}
+		}
+
+		obj.Spec.Template.Spec.Security.AuthToken = &v1alpha1.TiDBAuthToken{
+			JWKs: corev1.LocalObjectReference{
+				Name: JWKsSecretName,
+			},
+		}
+	}
+}
+
+func WithTLS() GroupPatch[*runtime.TiDBGroup] {
+	return func(obj *runtime.TiDBGroup) {
+		if obj.Spec.Template.Spec.Security == nil {
+			obj.Spec.Template.Spec.Security = &v1alpha1.TiDBSecurity{}
+		}
+
+		obj.Spec.Template.Spec.Security.TLS = &v1alpha1.TiDBTLS{
+			MySQL: &v1alpha1.TLS{
+				Enabled: true,
+			},
+		}
+	}
+}
diff --git a/tests/e2e/data/tikv.go b/tests/e2e/data/tikv.go
new file mode 100644
index 000000000..5b7b8838c
--- /dev/null
+++ b/tests/e2e/data/tikv.go
@@ -0,0 +1,55 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package data
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/runtime"
+)
+
+func NewTiKVGroup(ns string, patches ...GroupPatch[*runtime.TiKVGroup]) *v1alpha1.TiKVGroup {
+	kvg := &runtime.TiKVGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      defaultTiKVGroupName,
+		},
+		Spec: v1alpha1.TiKVGroupSpec{
+			Cluster:  v1alpha1.ClusterReference{Name: defaultClusterName},
+			Replicas: ptr.To[int32](1),
+			Version:  defaultVersion,
+			Template: v1alpha1.TiKVTemplate{
+				Spec: v1alpha1.TiKVTemplateSpec{
+					Image: ptr.To(defaultImageRegistry + "tikv"),
+					Volumes: []v1alpha1.Volume{
+						{
+							Name:    "data",
+							Mounts:  []v1alpha1.VolumeMount{{Type: "data"}},
+							Storage: resource.MustParse("1Gi"),
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, p := range patches {
+		p(kvg)
+	}
+
+	return runtime.ToTiKVGroup(kvg)
+}
diff --git a/tests/e2e/e2e_test.go b/tests/e2e/e2e_test.go
index 706949a7b..907bfde09 100644
--- a/tests/e2e/e2e_test.go
+++ b/tests/e2e/e2e_test.go
@@ -25,6 +25,7 @@ import (
 
 	_ "github.com/pingcap/tidb-operator/tests/e2e/cluster"
 	_ "github.com/pingcap/tidb-operator/tests/e2e/pd"
+	_ "github.com/pingcap/tidb-operator/tests/e2e/tidb"
 )
 
 func TestE2E(t *testing.T) {
diff --git a/tests/e2e/framework/framework.go b/tests/e2e/framework/framework.go
index ddc012b3f..0fe9659a1 100644
--- a/tests/e2e/framework/framework.go
+++ b/tests/e2e/framework/framework.go
@@ -20,9 +20,12 @@ import (
 	"github.com/onsi/ginkgo/v2"
 	"github.com/onsi/gomega"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
 
 	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
 	"github.com/pingcap/tidb-operator/pkg/client"
+	"github.com/pingcap/tidb-operator/pkg/runtime"
 	"github.com/pingcap/tidb-operator/tests/e2e/data"
 	"github.com/pingcap/tidb-operator/tests/e2e/utils/waiter"
 )
@@ -32,6 +35,8 @@ type Framework struct {
 	Cluster   *v1alpha1.Cluster
 
 	Client client.Client
+
+	podLogClient rest.Interface
 }
 
 func New() *Framework {
@@ -40,36 +45,90 @@ func New() *Framework {
 
 func (f *Framework) Setup() {
 	// TODO: get context and config path from options
-	c, err := newClient("", "")
+	cfg, err := newConfig("", "")
+	gomega.Expect(err).To(gomega.Succeed())
+
+	c, err := newClient(cfg)
 	gomega.Expect(err).To(gomega.Succeed())
 	f.Client = c
 
+	podLogClient, err := newRESTClientForPod(cfg)
+	gomega.Expect(err).To(gomega.Succeed())
+	f.podLogClient = podLogClient
+
 	ginkgo.BeforeEach(func(ctx context.Context) {
 		ns := data.NewNamespace()
 
 		f.Namespace = ns
-		f.Cluster = data.NewCluster(ns.Name)
+		ginkgo.By("Creating a namespace")
+		f.Must(f.Client.Create(ctx, f.Namespace))
+
+		ginkgo.DeferCleanup(func(ctx context.Context) {
+			ginkgo.By("Delete the namespace")
+			f.Must(f.Client.Delete(ctx, f.Namespace))
 
-		gomega.Expect(
-			f.Client.Create(ctx, f.Namespace),
-		).To(gomega.Succeed())
+			ginkgo.By("Ensure the namespace can be deleted")
+			f.Must(waiter.WaitForObjectDeleted(ctx, f.Client, f.Namespace, waiter.ShortTaskTimeout))
+		})
+	})
+}
 
-		gomega.Expect(
-			f.Client.Create(ctx, f.Cluster),
-		).To(gomega.Succeed())
+func (f *Framework) SetupCluster(ps ...data.ClusterPatch) {
+	ginkgo.BeforeEach(func(ctx context.Context) {
+		f.Cluster = data.NewCluster(f.Namespace.Name, ps...)
+		ginkgo.By("Creating a cluster")
+		f.Must(f.Client.Create(ctx, f.Cluster))
 
 		ginkgo.DeferCleanup(func(ctx context.Context) {
-			ginkgo.By("Delete the namespace")
+			ginkgo.By("Delete the cluster")
+			f.Must(f.Client.Delete(ctx, f.Cluster))
 
-			gomega.Expect(
-				f.Client.Delete(ctx, f.Namespace),
-			).To(gomega.Succeed())
+			ginkgo.By("Ensure the cluster can be deleted")
+			f.Must(waiter.WaitForObjectDeleted(ctx, f.Client, f.Cluster, waiter.ShortTaskTimeout))
+		})
+	})
+}
 
-			ginkgo.By("Ensure the namespace can be deleted")
+func (f *Framework) MustCreatePD(ctx context.Context, ps ...data.GroupPatch[*runtime.PDGroup]) *v1alpha1.PDGroup {
+	pdg := data.NewPDGroup(f.Namespace.Name, ps...)
+	ginkgo.By("Creating a pd group")
+	f.Must(f.Client.Create(ctx, pdg))
+
+	return pdg
+}
+
+func (f *Framework) MustCreateTiKV(ctx context.Context, ps ...data.GroupPatch[*runtime.TiKVGroup]) *v1alpha1.TiKVGroup {
+	kvg := data.NewTiKVGroup(f.Namespace.Name, ps...)
+	ginkgo.By("Creating a tikv group")
+	f.Must(f.Client.Create(ctx, kvg))
 
-			gomega.Expect(
-				waiter.WaitForObjectDeleted(ctx, f.Client, f.Namespace, waiter.ShortTaskTimeout),
-			).To(gomega.Succeed())
+	return kvg
+}
+
+func (f *Framework) MustCreateTiDB(ctx context.Context, ps ...data.GroupPatch[*runtime.TiDBGroup]) *v1alpha1.TiDBGroup {
+	dbg := data.NewTiDBGroup(f.Namespace.Name, ps...)
+	ginkgo.By("Creating a tidb group")
+	f.Must(f.Client.Create(ctx, dbg))
+
+	return dbg
+}
+
+func (f *Framework) SetupBootstrapSQL(sql string) {
+	ginkgo.BeforeEach(func(ctx context.Context) {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      data.BootstrapSQLName,
+				Namespace: f.Namespace.Name,
+			},
+			Data: map[string]string{
+				v1alpha1.BootstrapSQLConfigMapKey: sql,
+			},
+		}
+		ginkgo.By("Creating a bootstrap sql configmap")
+		f.Must(f.Client.Create(ctx, cm))
+		ginkgo.DeferCleanup(func(ctx context.Context) {
+			ginkgo.By("Delete the bootstrap sql configmap")
+			f.Must(f.Client.Delete(ctx, cm))
 		})
 	})
 }
diff --git a/tests/e2e/framework/pd.go b/tests/e2e/framework/pd.go
index d7b651f67..cdc436dee 100644
--- a/tests/e2e/framework/pd.go
+++ b/tests/e2e/framework/pd.go
@@ -20,6 +20,8 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	"github.com/onsi/ginkgo/v2"
+
 	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
 	"github.com/pingcap/tidb-operator/pkg/client"
 	"github.com/pingcap/tidb-operator/pkg/runtime"
@@ -28,6 +30,7 @@ import (
 
 func (f *Framework) WaitForPDGroupReady(ctx context.Context, pdg *v1alpha1.PDGroup) {
 	// TODO: maybe wait for cluster ready
+	ginkgo.By("wait for pd group ready")
 	f.Must(waiter.WaitForPDsHealthy(ctx, f.Client, pdg, waiter.LongTaskTimeout))
 	f.Must(waiter.WaitForPodsReady(ctx, f.Client, runtime.FromPDGroup(pdg), waiter.LongTaskTimeout))
 }
diff --git a/tests/e2e/framework/tidb.go b/tests/e2e/framework/tidb.go
new file mode 100644
index 000000000..d51afb69b
--- /dev/null
+++ b/tests/e2e/framework/tidb.go
@@ -0,0 +1,32 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"context"
+
+	"github.com/onsi/ginkgo/v2"
+
+	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/runtime"
+	"github.com/pingcap/tidb-operator/tests/e2e/utils/waiter"
+)
+
+func (f *Framework) WaitForTiDBGroupReady(ctx context.Context, dbg *v1alpha1.TiDBGroup) {
+	// TODO: maybe wait for cluster ready
+	ginkgo.By("wait for tidb group ready")
+	f.Must(waiter.WaitForTiDBsHealthy(ctx, f.Client, dbg, waiter.LongTaskTimeout))
+	f.Must(waiter.WaitForPodsReady(ctx, f.Client, runtime.FromTiDBGroup(dbg), waiter.LongTaskTimeout))
+}
diff --git a/tests/e2e/framework/tikv.go b/tests/e2e/framework/tikv.go
new file mode 100644
index 000000000..132a76095
--- /dev/null
+++ b/tests/e2e/framework/tikv.go
@@ -0,0 +1,32 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"context"
+
+	"github.com/onsi/ginkgo/v2"
+
+	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/runtime"
+	"github.com/pingcap/tidb-operator/tests/e2e/utils/waiter"
+)
+
+func (f *Framework) WaitForTiKVGroupReady(ctx context.Context, kvg *v1alpha1.TiKVGroup) {
+	// TODO: maybe wait for cluster ready
+	ginkgo.By("wait for tikv group ready")
+	f.Must(waiter.WaitForTiKVsHealthy(ctx, f.Client, kvg, waiter.LongTaskTimeout))
+	f.Must(waiter.WaitForPodsReady(ctx, f.Client, runtime.FromTiKVGroup(kvg), waiter.LongTaskTimeout))
+}
diff --git a/tests/e2e/framework/util.go b/tests/e2e/framework/util.go
index 8d63bc565..b6045c8b0 100644
--- a/tests/e2e/framework/util.go
+++ b/tests/e2e/framework/util.go
@@ -15,15 +15,20 @@
 package framework
 
 import (
+	"context"
 	"fmt"
+	"io"
 
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 
 	"github.com/pingcap/tidb-operator/pkg/client"
 	"github.com/pingcap/tidb-operator/pkg/scheme"
 )
 
-func newClient(configPath, ctxName string) (client.Client, error) {
+func newConfig(configPath, ctxName string) (*rest.Config, error) {
 	rule := clientcmd.NewDefaultClientConfigLoadingRules()
 	rule.ExplicitPath = configPath
 
@@ -33,7 +38,12 @@ func newClient(configPath, ctxName string) (client.Client, error) {
 	if err != nil {
 		return nil, fmt.Errorf("can't parse kubeconfig: %w", err)
 	}
-	c, err := client.New(kubeconfig, client.Options{
+
+	return kubeconfig, nil
+}
+
+func newClient(cfg *rest.Config) (client.Client, error) {
+	c, err := client.New(cfg, client.Options{
 		Scheme: scheme.Scheme,
 	})
 	if err != nil {
@@ -41,3 +51,21 @@ func newClient(configPath, ctxName string) (client.Client, error) {
 	}
 	return c, nil
 }
+
+func newRESTClientForPod(cfg *rest.Config) (rest.Interface, error) {
+	httpClient, err := rest.HTTPClientFor(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("cannot new http client: %w", err)
+	}
+	return apiutil.RESTClientForGVK(corev1.SchemeGroupVersion.WithKind("Pod"), false, cfg, scheme.Codecs, httpClient)
+}
+
+func logPod(ctx context.Context, c rest.Interface, pod *corev1.Pod) (io.ReadCloser, error) {
+	req := c.Get().
+		Namespace(pod.Namespace).
+		Name(pod.Name).
+		Resource("pods").
+		SubResource("log").
+		VersionedParams(&corev1.PodLogOptions{}, scheme.ParameterCodec)
+	return req.Stream(ctx)
+}
diff --git a/tests/e2e/framework/workload.go b/tests/e2e/framework/workload.go
new file mode 100644
index 000000000..16facb866
--- /dev/null
+++ b/tests/e2e/framework/workload.go
@@ -0,0 +1,119 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"bytes"
+	"context"
+	"io"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+
+	"github.com/pingcap/tidb-operator/pkg/client"
+	"github.com/pingcap/tidb-operator/tests/e2e/utils/waiter"
+)
+
+const (
+	workloadJobName = "testing-workload-job"
+)
+
+type Workload struct {
+	f *Framework
+
+	job *batchv1.Job
+}
+
+func (f *Framework) SetupWorkload() *Workload {
+	w := &Workload{
+		f: f,
+	}
+	w.DeferPrintLogs()
+
+	return w
+}
+
+func (w *Workload) MustPing(ctx context.Context, host, user, password string) {
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workloadJobName,
+			Namespace: w.f.Namespace.Name,
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": workloadJobName,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "testing-workload",
+							Image: "pingcap/testing-workload:latest",
+							Args: []string{
+								"--action", "ping",
+								"--host", host,
+								"--user", user,
+								"--password", password,
+								"--duration", "8",
+								"--max-connections", "30",
+							},
+							ImagePullPolicy: corev1.PullIfNotPresent,
+						},
+					},
+					RestartPolicy: corev1.RestartPolicyNever,
+				},
+			},
+			BackoffLimit: ptr.To[int32](0),
+		},
+	}
+
+	ginkgo.By("Creating ping job")
+	w.f.Must(w.f.Client.Create(ctx, job))
+	w.job = job
+
+	w.f.Must(waiter.WaitForJobComplete(ctx, w.f.Client, job, waiter.ShortTaskTimeout))
+}
+
+func (w *Workload) DeferPrintLogs() {
+	ginkgo.JustAfterEach(func(ctx context.Context) {
+		if ginkgo.CurrentSpecReport().Failed() && w.job != nil {
+			podList := corev1.PodList{}
+			ginkgo.By("Try to get the workload pod")
+			w.f.Must(w.f.Client.List(ctx, &podList, client.InNamespace(w.f.Namespace.Name), client.MatchingLabels{
+				"app": w.job.Name,
+			}))
+			gomega.Expect(len(podList.Items)).To(gomega.Equal(1))
+
+			pod := &podList.Items[0]
+			logs, err := logPod(ctx, w.f.podLogClient, pod)
+			gomega.Expect(err).To(gomega.Succeed())
+			defer logs.Close()
+
+			buf := bytes.Buffer{}
+			_, err = io.Copy(&buf, logs)
+			gomega.Expect(err).To(gomega.Succeed())
+
+			// TODO(liubo02): add color for logs
+			ginkgo.AddReportEntry("WorkloadLogs", buf.String())
+		}
+	})
+}
diff --git a/tests/e2e/pd/pd.go b/tests/e2e/pd/pd.go
index 244e7026d..2b6bb0b89 100644
--- a/tests/e2e/pd/pd.go
+++ b/tests/e2e/pd/pd.go
@@ -33,6 +33,7 @@ import (
 var _ = ginkgo.Describe("PD", label.PD, func() {
 	f := framework.New()
 	f.Setup()
+	f.SetupCluster()
 
 	ginkgo.Context("Basic", label.P0, func() {
 		ginkgo.It("support create PD with 1 replica", func(ctx context.Context) {
diff --git a/tests/e2e/tidb/tidb.go b/tests/e2e/tidb/tidb.go
new file mode 100644
index 000000000..e730f9ef4
--- /dev/null
+++ b/tests/e2e/tidb/tidb.go
@@ -0,0 +1,90 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tidb
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/ginkgo/v2"
+
+	"github.com/pingcap/tidb-operator/tests/e2e/data"
+	"github.com/pingcap/tidb-operator/tests/e2e/framework"
+	"github.com/pingcap/tidb-operator/tests/e2e/label"
+	"github.com/pingcap/tidb-operator/tests/e2e/utils/jwt"
+)
+
+var _ = ginkgo.Describe("TiDB", label.TiDB, func() {
+	f := framework.New()
+	f.Setup()
+
+	ginkgo.Context("Bootstrap SQL", label.P1, func() {
+		sql := "SET PASSWORD FOR 'root'@'%' = 'pingcap';"
+
+		f.SetupBootstrapSQL(sql)
+		f.SetupCluster(data.WithBootstrapSQL())
+		workload := f.SetupWorkload()
+
+		ginkgo.It("support init a cluster with bootstrap SQL specified", func(ctx context.Context) {
+			ginkgo.By("Creating components")
+			pdg := f.MustCreatePD(ctx)
+			kvg := f.MustCreateTiKV(ctx)
+			dbg := f.MustCreateTiDB(ctx)
+
+			f.WaitForPDGroupReady(ctx, pdg)
+			f.WaitForTiKVGroupReady(ctx, kvg)
+			f.WaitForTiDBGroupReady(ctx, dbg)
+
+			workload.MustPing(ctx, data.DefaultTiDBServiceName, "root", "pingcap")
+		})
+	})
+
+	ginkgo.Context("Auth token", label.P1, func() {
+		const (
+			kid   = "the-key-id-0"
+			sub   = "user@pingcap.com"
+			email = "user@pingcap.com"
+			iss   = "issuer-abc"
+		)
+		sql := fmt.Sprintf(
+			`CREATE USER '%s' IDENTIFIED WITH 'tidb_auth_token' REQUIRE TOKEN_ISSUER '%s' ATTRIBUTE '{"email": "%s"}';
+GRANT ALL PRIVILEGES ON *.* TO '%s'@'%s';`, sub, iss, email, sub, "%")
+
+		f.SetupBootstrapSQL(sql)
+		f.SetupCluster(data.WithBootstrapSQL())
+		workload := f.SetupWorkload()
+
+		ginkgo.It("should connect to the TiDB cluster with JWT authentication", func(ctx context.Context) {
+			token, err := jwt.GenerateJWT(kid, sub, email, iss)
+			if err != nil {
+				// ??
+				ginkgo.Skip(fmt.Sprintf("failed to generate JWT token: %v", err))
+			}
+			jwksSecret := jwt.GenerateJWKSSecret(f.Namespace.Name, data.JWKsSecretName)
+			f.Must(f.Client.Create(ctx, &jwksSecret))
+
+			ginkgo.By("Creating components")
+			pdg := f.MustCreatePD(ctx)
+			kvg := f.MustCreateTiKV(ctx)
+			dbg := f.MustCreateTiDB(ctx, data.WithAuthToken())
+
+			f.WaitForPDGroupReady(ctx, pdg)
+			f.WaitForTiKVGroupReady(ctx, kvg)
+			f.WaitForTiDBGroupReady(ctx, dbg)
+
+			workload.MustPing(ctx, data.DefaultTiDBServiceName, sub, token)
+		})
+	})
+})
diff --git a/tests/e2e/utils/waiter/job.go b/tests/e2e/utils/waiter/job.go
new file mode 100644
index 000000000..ccbd44ccf
--- /dev/null
+++ b/tests/e2e/utils/waiter/job.go
@@ -0,0 +1,40 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package waiter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	batchv1 "k8s.io/api/batch/v1"
+
+	"github.com/pingcap/tidb-operator/pkg/client"
+)
+
+func WaitForJobComplete(ctx context.Context, c client.Client, job *batchv1.Job, timeout time.Duration) error {
+	return WaitForObject(ctx, c, job, func() error {
+		for _, cond := range job.Status.Conditions {
+			switch cond.Type {
+			case batchv1.JobComplete:
+				return nil
+			case batchv1.JobFailed:
+				return fmt.Errorf("job is failed: %v", cond.Message)
+			}
+		}
+
+		return fmt.Errorf("job status is unknown")
+	}, timeout)
+}
diff --git a/tests/e2e/utils/waiter/tidb.go b/tests/e2e/utils/waiter/tidb.go
new file mode 100644
index 000000000..3ba83f867
--- /dev/null
+++ b/tests/e2e/utils/waiter/tidb.go
@@ -0,0 +1,51 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package waiter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/client"
+)
+
+func WaitForTiDBsHealthy(ctx context.Context, c client.Client, dbg *v1alpha1.TiDBGroup, timeout time.Duration) error {
+	list := v1alpha1.TiDBList{}
+	return WaitForList(ctx, c, &list, func() error {
+		if len(list.Items) != int(*dbg.Spec.Replicas) {
+			return fmt.Errorf("db %s/%s replicas %d not equal to %d", dbg.Namespace, dbg.Name, len(list.Items), *dbg.Spec.Replicas)
+		}
+		for i := range list.Items {
+			db := &list.Items[i]
+			if db.Generation != db.Status.ObservedGeneration {
+				return fmt.Errorf("db %s/%s is not synced", db.Namespace, db.Name)
+			}
+			if !meta.IsStatusConditionPresentAndEqual(db.Status.Conditions, v1alpha1.TiDBCondHealth, metav1.ConditionTrue) {
+				return fmt.Errorf("db %s/%s is not healthy", db.Namespace, db.Name)
+			}
+		}
+
+		return nil
+	}, timeout, client.InNamespace(dbg.Namespace), client.MatchingLabels{
+		v1alpha1.LabelKeyCluster:   dbg.Spec.Cluster.Name,
+		v1alpha1.LabelKeyGroup:     dbg.Name,
+		v1alpha1.LabelKeyComponent: v1alpha1.LabelValComponentTiDB,
+	})
+}
diff --git a/tests/e2e/utils/waiter/tikv.go b/tests/e2e/utils/waiter/tikv.go
new file mode 100644
index 000000000..319be6968
--- /dev/null
+++ b/tests/e2e/utils/waiter/tikv.go
@@ -0,0 +1,51 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package waiter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/pingcap/tidb-operator/apis/core/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/client"
+)
+
+func WaitForTiKVsHealthy(ctx context.Context, c client.Client, kvg *v1alpha1.TiKVGroup, timeout time.Duration) error {
+	list := v1alpha1.TiKVList{}
+	return WaitForList(ctx, c, &list, func() error {
+		if len(list.Items) != int(*kvg.Spec.Replicas) {
+			return fmt.Errorf("kv %s/%s replicas %d not equal to %d", kvg.Namespace, kvg.Name, len(list.Items), *kvg.Spec.Replicas)
+		}
+		for i := range list.Items {
+			kv := &list.Items[i]
+			if kv.Generation != kv.Status.ObservedGeneration {
+				return fmt.Errorf("kv %s/%s is not synced", kv.Namespace, kv.Name)
+			}
+			if !meta.IsStatusConditionPresentAndEqual(kv.Status.Conditions, v1alpha1.TiKVCondHealth, metav1.ConditionTrue) {
+				return fmt.Errorf("kv %s/%s is not healthy", kv.Namespace, kv.Name)
+			}
+		}
+
+		return nil
+	}, timeout, client.InNamespace(kvg.Namespace), client.MatchingLabels{
+		v1alpha1.LabelKeyCluster:   kvg.Spec.Cluster.Name,
+		v1alpha1.LabelKeyGroup:     kvg.Name,
+		v1alpha1.LabelKeyComponent: v1alpha1.LabelValComponentTiKV,
+	})
+}
