diff --git a/client/client_test.go b/client/client_test.go
index 83cc70dac..beafa0b22 100644
--- a/client/client_test.go
+++ b/client/client_test.go
@@ -23,6 +23,7 @@ import (
 	"runtime"
 	"strconv"
 	"strings"
+	"sync/atomic"
 	"syscall"
 	"testing"
 	"time"
@@ -151,6 +152,7 @@ var allTests = []func(t *testing.T, sb integration.Sandbox){
 	testFrontendMetadataReturn,
 	testFrontendUseSolveResults,
 	testSSHMount,
+	testRawSocketMount,
 	testStdinClosed,
 	testHostnameLookup,
 	testHostnameSpecifying,
@@ -843,6 +845,61 @@ func testSSHMount(t *testing.T, sb integration.Sandbox) {
 	require.Contains(t, string(dt), "(RSA)")
 }
 
+func testRawSocketMount(t *testing.T, sb integration.Sandbox) {
+	integration.SkipOnPlatform(t, "windows")
+
+	c, err := New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	dir := t.TempDir()
+	sockPath := filepath.Join(dir, "test.sock")
+	l, err := net.Listen("unix", sockPath)
+	require.NoError(t, err)
+	defer l.Close()
+
+	var called atomic.Bool
+	srv := &http.Server{
+		ReadHeaderTimeout: 10 * time.Second,
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			called.Store(true)
+			w.WriteHeader(http.StatusOK)
+		}),
+	}
+	go srv.Serve(l)
+	defer srv.Close()
+
+	st := llb.Image("alpine:latest").
+		Run(llb.Shlex(`apk add --no-cache curl`)).
+		Run(llb.Shlex(`curl --unix-socket /tmp/test.sock http://./foo`),
+			llb.AddSSHSocket(llb.SSHSocketTarget("/tmp/test.sock")),
+		).Root()
+
+	def, err := st.Marshal(sb.Context())
+	require.NoError(t, err)
+
+	raw, err := sshprovider.NewSSHAgentProvider([]sshprovider.AgentConfig{{
+		Paths: []string{sockPath},
+		Raw:   true,
+	}})
+	require.NoError(t, err)
+
+	ch := make(chan *SolveStatus)
+	go func() {
+		for status := range ch {
+			for _, l := range status.Logs {
+				t.Log(string(l.Data))
+			}
+		}
+	}()
+
+	_, err = c.Solve(sb.Context(), def, SolveOpt{
+		Session: []session.Attachable{raw},
+	}, ch)
+	require.NoError(t, err)
+	require.True(t, called.Load(), "server should have been called")
+}
+
 func testExtraHosts(t *testing.T, sb integration.Sandbox) {
 	integration.SkipOnPlatform(t, "windows")
 	c, err := New(sb.Context(), sb.Address())
diff --git a/session/sshforward/sshprovider/agentprovider_test.go b/session/sshforward/sshprovider/agentprovider_test.go
index 733fe9ffe..5c893e424 100644
--- a/session/sshforward/sshprovider/agentprovider_test.go
+++ b/session/sshforward/sshprovider/agentprovider_test.go
@@ -1,11 +1,15 @@
 package sshprovider_test
 
 import (
+	"net"
+	"os"
+	"path/filepath"
 	"strings"
 	"testing"
 
 	"github.com/moby/buildkit/cmd/buildctl/build"
 	"github.com/moby/buildkit/session/sshforward/sshprovider"
+	"github.com/stretchr/testify/require"
 )
 
 func TestToAgentSource(t *testing.T) {
@@ -18,4 +22,35 @@ func TestToAgentSource(t *testing.T) {
 	if !ok {
 		t.Fatal(err)
 	}
+
+	_, err = build.ParseSSH([]string{"default=raw=true"})
+	require.ErrorContains(t, err, "raw mode must supply exactly one socket path")
+
+	dir := t.TempDir()
+	normalFilePath := filepath.Join(dir, "not-a-socket")
+	f, err := os.Create(normalFilePath)
+	require.NoError(t, err)
+	f.Close()
+
+	configs, err = build.ParseSSH([]string{"default=raw=true," + normalFilePath})
+	require.NoError(t, err)
+
+	_, err = sshprovider.NewSSHAgentProvider(configs)
+	require.ErrorContains(t, err, "raw mode only supported with socket paths")
+
+	sockPath := filepath.Join(dir, "test.sock")
+	l, err := net.Listen("unix", sockPath)
+	require.NoError(t, err)
+	defer l.Close()
+
+	configs, err = build.ParseSSH([]string{"default=raw=true," + sockPath})
+	require.NoError(t, err)
+
+	_, err = sshprovider.NewSSHAgentProvider(configs)
+	require.NoError(t, err)
+
+	configs, err = build.ParseSSH([]string{"default=" + sockPath + ",raw=true"})
+	require.NoError(t, err)
+	_, err = sshprovider.NewSSHAgentProvider(configs)
+	require.NoError(t, err)
 }
diff --git a/session/sshforward/sshprovider/pipe_test.go b/session/sshforward/sshprovider/pipe_test.go
new file mode 100644
index 000000000..f812f685b
--- /dev/null
+++ b/session/sshforward/sshprovider/pipe_test.go
@@ -0,0 +1,116 @@
+package sshprovider
+
+import (
+	"context"
+	"net"
+	"sync"
+)
+
+// pipeListener is a net.Listener that uses net.Pipe to create connections
+// and is useful for in-memory uses of [newRawProvider].
+//
+// PipelListener can be created using the zero value.
+type pipeListener struct {
+	mu     sync.Mutex
+	closed bool
+	conns  []net.Conn
+
+	chConn chan net.Conn
+
+	closedCh chan struct{}
+}
+
+var _ net.Listener = (*pipeListener)(nil)
+
+func (pl *pipeListener) Accept() (net.Conn, error) {
+	pl.mu.Lock()
+	if pl.chConn == nil {
+		pl.chConn = make(chan net.Conn)
+	}
+	if pl.closedCh == nil {
+		pl.closedCh = make(chan struct{})
+	}
+	pl.mu.Unlock()
+
+	var conn net.Conn
+	select {
+	case <-pl.closedCh:
+		return nil, net.ErrClosed
+	case conn = <-pl.chConn:
+	}
+
+	if conn == nil {
+		// channel is closed, no more connections can be accepted
+		return nil, net.ErrClosed
+	}
+
+	pl.mu.Lock()
+	defer pl.mu.Unlock()
+
+	if pl.closed {
+		conn.Close()
+		return nil, net.ErrClosed
+	}
+
+	pl.conns = append(pl.conns, conn)
+
+	return conn, nil
+}
+
+// Dialer implements the [DialerFn] type for [ProxyHandler].
+func (pl *pipeListener) Dialer(ctx context.Context) (net.Conn, error) {
+	pl.mu.Lock()
+	if pl.closed {
+		pl.mu.Unlock()
+		return nil, net.ErrClosed
+	}
+
+	if pl.chConn == nil {
+		pl.chConn = make(chan net.Conn)
+	}
+
+	pl.mu.Unlock()
+
+	c1, c2 := net.Pipe()
+
+	select {
+	case <-ctx.Done():
+		return nil, context.Cause(ctx)
+	case pl.chConn <- c2:
+	}
+
+	return c1, nil
+}
+
+func (pl *pipeListener) Addr() net.Addr {
+	return &pipeAddr{}
+}
+
+type pipeAddr struct{}
+
+func (pa *pipeAddr) Network() string {
+	return "pipe"
+}
+
+func (pa *pipeAddr) String() string {
+	return "pipe"
+}
+
+func (pl *pipeListener) Close() error {
+	pl.mu.Lock()
+	defer pl.mu.Unlock()
+
+	if pl.closed {
+		return nil
+	}
+
+	pl.closed = true
+	if pl.chConn != nil {
+		close(pl.chConn)
+	}
+	for _, conn := range pl.conns {
+		conn.Close()
+	}
+	pl.conns = pl.conns[:0]
+	return nil
+}
diff --git a/session/sshforward/sshprovider/raw_provider_test.go b/session/sshforward/sshprovider/raw_provider_test.go
new file mode 100644
index 000000000..f5798af12
--- /dev/null
+++ b/session/sshforward/sshprovider/raw_provider_test.go
@@ -0,0 +1,186 @@
+package sshprovider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"net"
+	"strings"
+	"testing"
+
+	"github.com/moby/buildkit/session/sshforward"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+// echoServer implements a message based echo server
+// It expects to send and receive discreete json messages over a net.Conn.
+// Using discreete messages is helpful for testing purposes over a literal echo.
+type echoServer struct{}
+
+type echoRequest struct {
+	Data string `json:"data"`
+}
+
+type echoResponse struct {
+	Recvd string `json:"recvd"`
+	Count int    `json:"count"`
+}
+
+func (es *echoServer) Serve(l net.Listener) error {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				return nil
+			}
+			return err
+		}
+		go func() {
+			dec := json.NewDecoder(conn)
+			enc := json.NewEncoder(conn)
+
+			var req echoRequest
+			var resp echoResponse
+
+			for i := 1; ; i++ {
+				if err := dec.Decode(&req); err != nil {
+					conn.Close()
+					return
+				}
+
+				resp.Recvd = req.Data
+				resp.Count = i
+				if err := enc.Encode(&resp); err != nil {
+					conn.Close()
+					return
+				}
+			}
+		}()
+	}
+}
+
+func dialerFnToGRPCDialer(dialer func(ctx context.Context) (net.Conn, error)) grpc.DialOption {
+	return grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+		return dialer(ctx)
+	})
+}
+
+func TestRawProvider(t *testing.T) {
+	handlerListener := &pipeListener{}
+	defer handlerListener.Close()
+
+	echoListener := &pipeListener{}
+	defer echoListener.Close()
+	echo := &echoServer{}
+	go echo.Serve(echoListener)
+
+	handler := &socketProvider{
+		m: map[string]dialerFn{
+			"test": echoListener.Dialer,
+		},
+	}
+
+	srv := grpc.NewServer()
+	handler.Register(srv)
+
+	// Start proxy handler service
+	go srv.Serve(handlerListener)
+
+	// passthrough:// is a special scheme that allows us to use the handlerListener's Dialer directly
+	// otherwise grpc will try to resolve whatever we put in there.
+	c, err := grpc.NewClient("passthrough://", dialerFnToGRPCDialer(handlerListener.Dialer), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	require.NoError(t, err)
+	defer c.Close()
+
+	client := sshforward.NewSSHClient(c)
+
+	ctx := context.Background()
+
+	_, err = client.CheckAgent(ctx, &sshforward.CheckAgentRequest{ID: "does-not-exist"})
+	require.ErrorContains(t, err, "does-not-exis")
+
+	_, err = client.CheckAgent(ctx, &sshforward.CheckAgentRequest{ID: "test"})
+	require.NoError(t, err)
+
+	ctx = metadata.AppendToOutgoingContext(ctx, sshforward.KeySSHID, "test")
+	stream, err := client.ForwardAgent(ctx)
+	require.NoError(t, err)
+
+	defer stream.CloseSend()
+
+	req := echoRequest{Data: "hello, world!"}
+	sw := &streamWriter{stream: stream}
+
+	enc := json.NewEncoder(sw)
+	err = enc.Encode(&req)
+	require.NoError(t, err)
+
+	sr := &streamReader{stream: stream}
+	var resp echoResponse
+	dec := json.NewDecoder(sr)
+
+	err = dec.Decode(&resp)
+	require.NoError(t, err)
+	assert.Equal(t, req.Data, resp.Recvd)
+	assert.Equal(t, 1, resp.Count)
+
+	req.Data = "another message"
+	err = enc.Encode(&req)
+	require.NoError(t, err)
+
+	err = dec.Decode(&resp)
+	require.NoError(t, err)
+	assert.Equal(t, resp.Recvd, req.Data)
+	assert.Equal(t, 2, resp.Count)
+
+	// Test larger message
+	req.Data = strings.Repeat("x", 10000)
+	err = enc.Encode(&req)
+	require.NoError(t, err)
+
+	err = dec.Decode(&resp)
+	require.NoError(t, err)
+	assert.Equal(t, resp.Recvd, req.Data)
+	assert.Equal(t, 3, resp.Count)
+}
+
+type streamWriter struct {
+	stream grpc.BidiStreamingClient[sshforward.BytesMessage, sshforward.BytesMessage]
+}
+
+func (sw *streamWriter) Write(p []byte) (n int, err error) {
+	msg := &sshforward.BytesMessage{Data: p}
+	if err := sw.stream.Send(msg); err != nil {
+		return 0, err
+	}
+
+	return len(p), nil
+}
+
+type streamReader struct {
+	stream grpc.BidiStreamingClient[sshforward.BytesMessage, sshforward.BytesMessage]
+	buf    []byte
+}
+
+func (sr *streamReader) Read(p []byte) (int, error) {
+	if len(sr.buf) > 0 {
+		n := copy(p, sr.buf)
+		sr.buf = sr.buf[n:]
+		return n, nil
+	}
+
+	msg, err := sr.stream.Recv()
+	if err != nil {
+		return 0, err
+	}
+
+	n := copy(p, msg.Data)
+	if n < len(msg.Data) {
+		sr.buf = append(sr.buf, msg.Data[n:]...)
+	}
+	return n, nil
+}
