diff --git a/crates/cubecl-linalg/src/matmul/tests/test_macros/cmma/matmul_algorithm.rs b/crates/cubecl-linalg/src/matmul/tests/test_macros/cmma/matmul_algorithm.rs
index a5f0e063..2b16b08b 100644
--- a/crates/cubecl-linalg/src/matmul/tests/test_macros/cmma/matmul_algorithm.rs
+++ b/crates/cubecl-linalg/src/matmul/tests/test_macros/cmma/matmul_algorithm.rs
@@ -1,6 +1,6 @@
 // Tests nomenclature:
 // batch: b[o=one_to_one, m=one_to_many][batch dims, optional]
-// global: g[m]x[n]x[k], with m,n,k the whole matrix dimensions
+// global: g[h=homogeneous, pc=producer_consumer][m]x[n]x[k], with m,n,k the whole matrix dimensions
 // stage: s[m]x[n]x[k], with m,n,k the number of tiles along those dims
 // tile: t[m]x[n]x[k], with m,n,k the tile dimensions. tile algorithm is given by macro arguments
 // layouts: [r/c][r/c], r=row, c=col, respectively for lhs and rhs
@@ -19,8 +19,63 @@ macro_rules! matmul_test_define {
         $ea:ty,
         $plane_dim:expr
     ) => {
+        pub fn bo1_gpc16x16x480_s1x1x3_t16x16x16_rr_ln4() {
+            // Triple buffering
+            let problem = MatmulProblem {
+                m: 16,
+                n: 16,
+                k: 480,
+                batches: vec![],
+                lhs_layout: MatrixLayout::RowMajor,
+                rhs_layout: MatrixLayout::RowMajor,
+                lhs_line_size: 4,
+                rhs_line_size: 4,
+                out_line_size: 4,
+            };
+
+            struct Test {}
+            impl matmul::Algorithm<$eg> for Test {
+                const PLANE_DIM: u32 = $plane_dim;
+                type EG = $eg;
+                type ES = $es;
+                type EA = $ea;
+                type StageSize = S1x1x3;
+
+                type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
+                type StageMatmul = stage::single_buffer::Matmul<
+                    Self::ES,
+                    Self::EG,
+                    Self::EA,
+                    Self::TileMatmul,
+                    Self::StageSize,
+                >;
+                type GlobalMatmul =
+                    global::producer_consumer::Matmul<Self::EG, Self::ES, Self::StageMatmul>;
+                type BatchMatmul = batch::one_to_one::Matmul<
+                    Self::EG,
+                    Self::ES,
+                    Self::GlobalMatmul,
+                    batch::NaturalDispatch,
+                >;
+
+                fn cube_dim() -> CubeDim {
+                    CubeDim::new($plane_dim, 2, 1)
+                }
+
+                fn cube_count(_problem: &MatmulProblem) -> CubeCount {
+                    CubeCount::Static(1, 1, 1)
+                }
+            }
+
+            let advanced_config = AdvancedConfig {
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::RowMajor,
+                ..Default::default()
+            };
+        }
+
         #[test]
-        pub fn bo1_g1000x16x16_s1x1x1_t16x16x16_rr_ln4_transposed_dispatch() {
+        pub fn bo1_gh1000x16x16_s1x1x1_t16x16x16_rr_ln4_transposed_dispatch() {
             let problem = MatmulProblem {
                 m: 1024,
                 n: 16,
@@ -42,7 +97,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -77,7 +132,311 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm1_g16x16x16_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo3x4_gpc300x300x300_s4x4x2_t16x16x16_cc_ln4() {
+            let problem = MatmulProblem {
+                m: 300,
+                n: 300,
+                k: 300,
+                batches: vec![3, 4],
+                lhs_layout: MatrixLayout::ColMajor,
+                rhs_layout: MatrixLayout::ColMajor,
+                lhs_line_size: 4,
+                rhs_line_size: 4,
+                out_line_size: 4,
+            };
+
+            struct Test {}
+            impl matmul::Algorithm<$eg> for Test {
+                const PLANE_DIM: u32 = $plane_dim;
+                type EG = $eg;
+                type ES = $es;
+                type EA = $ea;
+                type StageSize = S4x4x2;
+
+                type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
+                type StageMatmul = stage::single_buffer::Matmul<
+                    Self::ES,
+                    Self::EG,
+                    Self::EA,
+                    Self::TileMatmul,
+                    Self::StageSize,
+                >;
+                type GlobalMatmul =
+                    global::producer_consumer::Matmul<Self::EG, Self::ES, Self::StageMatmul>;
+                type BatchMatmul = batch::one_to_one::Matmul<
+                    Self::EG,
+                    Self::ES,
+                    Self::GlobalMatmul,
+                    batch::NaturalDispatch,
+                >;
+
+                fn cube_dim() -> CubeDim {
+                    CubeDim::new($plane_dim, 8, 1)
+                }
+                fn cube_count(_problem: &MatmulProblem) -> CubeCount {
+                    CubeCount::Static(5, 5, 12)
+                }
+            }
+
+            let advanced_config = AdvancedConfig {
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::RowMajor,
+                ..Default::default()
+            };
+
+            test_matmul_algorithm::<Test, $eg, $es, TestRuntime>(
+                problem,
+                advanced_config,
+                &<<TestRuntime as Runtime>::Device>::default(),
+            );
+        }
+
+        #[test]
+        pub fn bo1_gpc16x32x32_s1x2x2_t16x16x16_rr_ln4() {
+            let problem = MatmulProblem {
+                m: 16,
+                n: 32,
+                k: 32,
+                batches: vec![],
+                lhs_layout: MatrixLayout::RowMajor,
+                rhs_layout: MatrixLayout::RowMajor,
+                lhs_line_size: 4,
+                rhs_line_size: 4,
+                out_line_size: 4,
+            };
+
+            struct Test {}
+            impl matmul::Algorithm<$eg> for Test {
+                const PLANE_DIM: u32 = $plane_dim;
+                type EG = $eg;
+                type ES = $es;
+                type EA = $ea;
+                type StageSize = S1x2x2;
+
+                type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
+                type StageMatmul = stage::single_buffer::Matmul<
+                    Self::ES,
+                    Self::EG,
+                    Self::EA,
+                    Self::TileMatmul,
+                    Self::StageSize,
+                >;
+                type GlobalMatmul =
+                    global::producer_consumer::Matmul<Self::EG, Self::ES, Self::StageMatmul>;
+                type BatchMatmul = batch::one_to_one::Matmul<
+                    Self::EG,
+                    Self::ES,
+                    Self::GlobalMatmul,
+                    batch::NaturalDispatch,
+                >;
+
+                fn cube_dim() -> CubeDim {
+                    CubeDim::new($plane_dim, 2, 1)
+                }
+
+                fn cube_count(_problem: &MatmulProblem) -> CubeCount {
+                    CubeCount::Static(1, 1, 1)
+                }
+            }
+
+            let advanced_config = AdvancedConfig {
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::RowMajor,
+                ..Default::default()
+            };
+
+            test_matmul_algorithm::<Test, $eg, $es, TestRuntime>(
+                problem,
+                advanced_config,
+                &<<TestRuntime as Runtime>::Device>::default(),
+            );
+        }
+
+        #[test]
+        pub fn bo1_gpc32x16x32_s2x1x2_t16x16x16_rr_ln4() {
+            let problem = MatmulProblem {
+                m: 32,
+                n: 16,
+                k: 32,
+                batches: vec![],
+                lhs_layout: MatrixLayout::RowMajor,
+                rhs_layout: MatrixLayout::RowMajor,
+                lhs_line_size: 4,
+                rhs_line_size: 4,
+                out_line_size: 4,
+            };
+
+            struct Test {}
+            impl matmul::Algorithm<$eg> for Test {
+                const PLANE_DIM: u32 = $plane_dim;
+                type EG = $eg;
+                type ES = $es;
+                type EA = $ea;
+                type StageSize = S2x1x2;
+
+                type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
+                type StageMatmul = stage::single_buffer::Matmul<
+                    Self::ES,
+                    Self::EG,
+                    Self::EA,
+                    Self::TileMatmul,
+                    Self::StageSize,
+                >;
+                type GlobalMatmul =
+                    global::producer_consumer::Matmul<Self::EG, Self::ES, Self::StageMatmul>;
+                type BatchMatmul = batch::one_to_one::Matmul<
+                    Self::EG,
+                    Self::ES,
+                    Self::GlobalMatmul,
+                    batch::NaturalDispatch,
+                >;
+
+                fn cube_dim() -> CubeDim {
+                    CubeDim::new($plane_dim, 3, 1)
+                }
+
+                fn cube_count(_problem: &MatmulProblem) -> CubeCount {
+                    CubeCount::Static(1, 1, 1)
+                }
+            }
+
+            let advanced_config = AdvancedConfig {
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::RowMajor,
+                ..Default::default()
+            };
+
+            test_matmul_algorithm::<Test, $eg, $es, TestRuntime>(
+                problem,
+                advanced_config,
+                &<<TestRuntime as Runtime>::Device>::default(),
+            );
+        }
+
+        #[test]
+        pub fn bo1_gpc16x16x128_s1x1x2_t16x16x16_rr_ln4() {
+            let problem = MatmulProblem {
+                m: 16,
+                n: 16,
+                k: 128,
+                batches: vec![],
+                lhs_layout: MatrixLayout::RowMajor,
+                rhs_layout: MatrixLayout::RowMajor,
+                lhs_line_size: 4,
+                rhs_line_size: 4,
+                out_line_size: 4,
+            };
+
+            struct Test {}
+            impl matmul::Algorithm<$eg> for Test {
+                const PLANE_DIM: u32 = $plane_dim;
+                type EG = $eg;
+                type ES = $es;
+                type EA = $ea;
+                type StageSize = S1x1x2;
+
+                type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
+                type StageMatmul = stage::single_buffer::Matmul<
+                    Self::ES,
+                    Self::EG,
+                    Self::EA,
+                    Self::TileMatmul,
+                    Self::StageSize,
+                >;
+                type GlobalMatmul =
+                    global::producer_consumer::Matmul<Self::EG, Self::ES, Self::StageMatmul>;
+                type BatchMatmul = batch::one_to_one::Matmul<
+                    Self::EG,
+                    Self::ES,
+                    Self::GlobalMatmul,
+                    batch::NaturalDispatch,
+                >;
+
+                fn cube_dim() -> CubeDim {
+                    CubeDim::new($plane_dim, 2, 1)
+                }
+
+                fn cube_count(_problem: &MatmulProblem) -> CubeCount {
+                    CubeCount::Static(1, 1, 1)
+                }
+            }
+
+            let advanced_config = AdvancedConfig {
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::RowMajor,
+                ..Default::default()
+            };
+
+            test_matmul_algorithm::<Test, $eg, $es, TestRuntime>(
+                problem,
+                advanced_config,
+                &<<TestRuntime as Runtime>::Device>::default(),
+            );
+        }
+
+        #[test]
+        pub fn bo1_gpc16x16x32_s1x1x2_t16x16x16_rr_ln4() {
+            let problem = MatmulProblem {
+                m: 16,
+                n: 16,
+                k: 32,
+                batches: vec![],
+                lhs_layout: MatrixLayout::RowMajor,
+                rhs_layout: MatrixLayout::RowMajor,
+                lhs_line_size: 4,
+                rhs_line_size: 4,
+                out_line_size: 4,
+            };
+
+            struct Test {}
+            impl matmul::Algorithm<$eg> for Test {
+                const PLANE_DIM: u32 = $plane_dim;
+                type EG = $eg;
+                type ES = $es;
+                type EA = $ea;
+                type StageSize = S1x1x2;
+
+                type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
+                type StageMatmul = stage::single_buffer::Matmul<
+                    Self::ES,
+                    Self::EG,
+                    Self::EA,
+                    Self::TileMatmul,
+                    Self::StageSize,
+                >;
+                type GlobalMatmul =
+                    global::producer_consumer::Matmul<Self::EG, Self::ES, Self::StageMatmul>;
+                type BatchMatmul = batch::one_to_one::Matmul<
+                    Self::EG,
+                    Self::ES,
+                    Self::GlobalMatmul,
+                    batch::NaturalDispatch,
+                >;
+
+                fn cube_dim() -> CubeDim {
+                    CubeDim::new($plane_dim, 2, 1)
+                }
+
+                fn cube_count(_problem: &MatmulProblem) -> CubeCount {
+                    CubeCount::Static(1, 1, 1)
+                }
+            }
+
+            let advanced_config = AdvancedConfig {
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::RowMajor,
+                ..Default::default()
+            };
+
+            test_matmul_algorithm::<Test, $eg, $es, TestRuntime>(
+                problem,
+                advanced_config,
+                &<<TestRuntime as Runtime>::Device>::default(),
+            );
+        }
+
+        #[test]
+        pub fn bm1_gh16x16x16_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -99,7 +458,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -135,7 +494,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm1_g32x16x16_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bm1_gh32x16x16_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 16,
@@ -157,7 +516,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -193,7 +552,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm1_g16x32x16_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bm1_gh16x32x16_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 32,
@@ -215,7 +574,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -251,7 +610,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm1_g16x16x32_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bm1_gh16x16x32_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -273,7 +632,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -309,7 +668,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm6_g16x16x16_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bm6_gh16x16x16_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -331,7 +690,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -367,7 +726,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm2_g32x32x32_s1x1x1_t16x16x16_rr_ln4_colspan() {
+        pub fn bm2_gh32x32x32_s1x1x1_t16x16x16_rr_ln4_colspan() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -389,7 +748,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -425,7 +784,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm2_g32x32x32_s1x1x1_t16x16x16_rr_ln4_swizzlespan() {
+        pub fn bm2_gh32x32x32_s1x1x1_t16x16x16_rr_ln4_swizzlespan() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -447,7 +806,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -483,7 +842,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm2_g32x32x32_s1x1x1_t16x16x16_rr_ln4_transposed_dispatch() {
+        pub fn bm2_gh32x32x32_s1x1x1_t16x16x16_rr_ln4_transposed_dispatch() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -505,7 +864,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -541,7 +900,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm2_g160x256x16_s1x1x1_t16x16x16_rr_ln4_swizzle_x_dispatch() {
+        pub fn bm2_gh160x256x16_s1x1x1_t16x16x16_rr_ln4_swizzle_x_dispatch() {
             let problem = MatmulProblem {
                 m: 160,
                 n: 256,
@@ -563,7 +922,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -599,7 +958,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm2_g160x256x16_s1x1x1_t16x16x16_rr_ln4_swizzle_y_dispatch() {
+        pub fn bm2_gh160x256x16_s1x1x1_t16x16x16_rr_ln4_swizzle_y_dispatch() {
             let problem = MatmulProblem {
                 m: 160,
                 n: 256,
@@ -621,7 +980,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -657,7 +1016,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bm5_g16x16x16_s1x1x1_t16x16x16_rr_ln4_cubez2() {
+        pub fn bm5_gh16x16x16_s1x1x1_t16x16x16_rr_ln4_cubez2() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -679,7 +1038,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S1x1x1;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -715,7 +1074,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo3x4_g300x300x300_s4x4x2_t16x16x16_cc_ln4() {
+        pub fn bo3x4_gh300x300x300_s4x4x2_t16x16x16_cc_ln4() {
             let problem = MatmulProblem {
                 m: 300,
                 n: 300,
@@ -737,7 +1096,7 @@ macro_rules! matmul_test_define {
                 type StageSize = S4x4x2;
 
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -771,7 +1130,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo3x4_g108x108x243_s4x4x2_t16x16x16_cr_ln4() {
+        pub fn bo3x4_gh108x108x243_s4x4x2_t16x16x16_cr_ln4() {
             let problem = MatmulProblem {
                 m: 108,
                 n: 108,
@@ -792,7 +1151,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -825,7 +1184,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo3x4_g256x256x256_s4x4x2_t16x16x16_cr_ln2x2x4() {
+        pub fn bo3x4_gh256x256x256_s4x4x2_t16x16x16_cr_ln2x2x4() {
             let problem = MatmulProblem {
                 m: 256,
                 n: 256,
@@ -846,7 +1205,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -879,7 +1238,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo3_g256x256x256_s4x4x2_t16x16x16_rc_ln4() {
+        pub fn bo3_gh256x256x256_s4x4x2_t16x16x16_rc_ln4() {
             let problem = MatmulProblem {
                 m: 256,
                 n: 256,
@@ -900,7 +1259,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -933,7 +1292,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo3_g16x16x16_s1x1x1_t16x16x16_cc_ln4() {
+        pub fn bo3_gh16x16x16_s1x1x1_t16x16x16_cc_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -954,7 +1313,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -987,7 +1346,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo3_g16x16x16_s1x1x1_t16x16x16_R() {
+        pub fn bo3_gh16x16x16_s1x1x1_t16x16x16_R() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1008,7 +1367,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1041,7 +1400,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g256x256x256_s4x4x2_t16x16x16_rc_ln4_ymajor() {
+        pub fn bo_gh256x256x256_s4x4x2_t16x16x16_rc_ln4_ColMajor() {
             let problem = MatmulProblem {
                 m: 256,
                 n: 256,
@@ -1062,7 +1421,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1087,7 +1446,8 @@ macro_rules! matmul_test_define {
             }
 
             let advanced_config = AdvancedConfig {
-                tiling_order: TilingOrderConfig::YMajor,
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::ColMajor,
                 ..Default::default()
             };
 
@@ -1099,7 +1459,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x32_s1x1x1_t16x16x16_cc_ln4_ymajor() {
+        pub fn bo_gh32x32x32_s1x1x1_t16x16x16_cc_ln4_ColMajor() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -1120,7 +1480,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1145,7 +1505,8 @@ macro_rules! matmul_test_define {
             }
 
             let advanced_config = AdvancedConfig {
-                tiling_order: TilingOrderConfig::YMajor,
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::ColMajor,
                 ..Default::default()
             };
 
@@ -1157,7 +1518,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x32_s1x1x1_t16x16x16_R() {
+        pub fn bo_gh32x32x32_s1x1x1_t16x16x16_R() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -1178,7 +1539,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1211,7 +1572,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x14x16_s1x1x1_t16x16x16_rc_ln4x4x2() {
+        pub fn bo_gh16x14x16_s1x1x1_t16x16x16_rc_ln4x4x2() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 14,
@@ -1232,7 +1593,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1265,7 +1626,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x12x16_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh16x12x16_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 12,
@@ -1286,7 +1647,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1319,7 +1680,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x12_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh16x16x12_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1340,7 +1701,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1373,7 +1734,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g60x60x120_s4x4x2_t16x16x16_rr_ln4() {
+        pub fn bo_gh60x60x120_s4x4x2_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 60,
                 n: 60,
@@ -1394,7 +1755,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1427,7 +1788,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x36_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh16x16x36_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1448,7 +1809,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1481,7 +1842,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g12x12x16_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh12x12x16_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 12,
                 n: 12,
@@ -1502,7 +1863,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1535,7 +1896,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_rr_ln1() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_rr_ln1() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1556,7 +1917,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1589,7 +1950,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1610,7 +1971,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1643,7 +2004,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_rc_ln4() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_rc_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1664,7 +2025,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1697,7 +2058,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_rr_ln4_lhs_col_enforced() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_rr_ln4_lhs_col_enforced() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1718,7 +2079,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1755,7 +2116,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_rr_ln4_rhs_col_enforced() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_rr_ln4_rhs_col_enforced() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1776,7 +2137,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1812,7 +2173,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_rc_ln4_rhs_row_enforced() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_rc_ln4_rhs_row_enforced() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -1833,7 +2194,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1869,7 +2230,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_rr_ln4_lhs_col_enforced() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_rr_ln4_lhs_col_enforced() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -1890,7 +2251,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1926,7 +2287,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_cr_ln4_lhs_row_enforced() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_cr_ln4_lhs_row_enforced() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -1947,7 +2308,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -1983,7 +2344,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_rr_ln4_rhs_col_enforced() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_rr_ln4_rhs_col_enforced() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -2004,7 +2365,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2040,7 +2401,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_rc_ln4_rhs_row_enforced() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_rc_ln4_rhs_row_enforced() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -2061,7 +2422,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2097,7 +2458,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_rr_ln4_lhs_col_enforced() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_rr_ln4_lhs_col_enforced() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2118,7 +2479,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2154,7 +2515,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_cr_ln4_lhs_row_enforced() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_cr_ln4_lhs_row_enforced() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2175,7 +2536,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2211,7 +2572,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_rr_ln4_rhs_col_enforced() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_rr_ln4_rhs_col_enforced() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2232,7 +2593,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2268,7 +2629,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_rc_ln4_rhs_row_enforced() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_rc_ln4_rhs_row_enforced() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2289,7 +2650,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2325,7 +2686,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g256x256x256_s4x4x2_t16x16x16_cr_ln2x2x4_lhs_row_rhs_col_enforced() {
+        pub fn bo_gh256x256x256_s4x4x2_t16x16x16_cr_ln2x2x4_lhs_row_rhs_col_enforced() {
             let problem = MatmulProblem {
                 m: 256,
                 n: 256,
@@ -2346,7 +2707,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2382,7 +2743,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_cr_ln4() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_cr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -2403,7 +2764,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2436,7 +2797,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_rr_ln4() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -2457,7 +2818,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2490,7 +2851,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_rc_ln1() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_rc_ln1() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -2511,7 +2872,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2544,7 +2905,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_cr_ln4() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_cr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -2565,7 +2926,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2598,7 +2959,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_cc_ln4() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_cc_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -2619,7 +2980,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2652,7 +3013,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_rr_ln4() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2673,7 +3034,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2706,7 +3067,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_rc_ln4() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_rc_ln4() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2727,7 +3088,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2760,7 +3121,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_cr_ln4() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_cr_ln4() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2781,7 +3142,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2814,7 +3175,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g8x32x16_s1x1x1_t8x32x16_cc_ln4() {
+        pub fn bo_gh8x32x16_s1x1x1_t8x32x16_cc_ln4() {
             let problem = MatmulProblem {
                 m: 8,
                 n: 32,
@@ -2835,7 +3196,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_8x32x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2868,7 +3229,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_rr_ln2() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_rr_ln2() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -2889,7 +3250,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2922,7 +3283,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x32_s2x2x2_t16x16x16_rr_ln4_ymajor() {
+        pub fn bo_gh32x32x32_s2x2x2_t16x16x16_rr_ln4_ColMajor() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -2943,7 +3304,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x2x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -2968,7 +3329,8 @@ macro_rules! matmul_test_define {
             }
 
             let advanced_config = AdvancedConfig {
-                tiling_order: TilingOrderConfig::YMajor,
+                lhs_tiling_order: TilingOrderConfig::ColMajor,
+                rhs_tiling_order: TilingOrderConfig::ColMajor,
                 ..Default::default()
             };
             test_matmul_algorithm::<Test, $eg, $es, TestRuntime>(
@@ -2979,7 +3341,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x32_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh16x16x32_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -3000,7 +3362,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3033,7 +3395,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x16_s1x1x1_t16x16x16_cc_ln4() {
+        pub fn bo_gh16x16x16_s1x1x1_t16x16x16_cc_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -3054,7 +3416,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3087,7 +3449,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x16x128_s1x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh16x16x128_s1x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 16,
@@ -3108,7 +3470,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3141,7 +3503,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x16x128_s2x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh32x16x128_s2x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 16,
@@ -3162,7 +3524,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3195,7 +3557,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x224_s2x2x2_t16x16x16_rr_ln4() {
+        pub fn bo_gh32x32x224_s2x2x2_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -3216,7 +3578,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x2x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3249,7 +3611,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g16x32x16_s1x2x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh16x32x16_s1x2x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 16,
                 n: 32,
@@ -3270,7 +3632,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x2x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3303,7 +3665,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x32_s2x2x2_t16x16x16_rr_ln4() {
+        pub fn bo_gh32x32x32_s2x2x2_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -3324,7 +3686,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x2x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3357,7 +3719,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x16_s2x2x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh32x32x16_s2x2x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -3378,7 +3740,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x2x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3411,7 +3773,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x32_s2x2x2_t16x16x16_rc_ln4() {
+        pub fn bo_gh32x32x32_s2x2x2_t16x16x16_rc_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -3432,7 +3794,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x2x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3465,7 +3827,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x32_s2x2x2_t16x16x16_cr_ln4() {
+        pub fn bo_gh32x32x32_s2x2x2_t16x16x16_cr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -3486,7 +3848,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x2x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3519,7 +3881,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x32x32_s2x2x2_t16x16x16_cc_ln4() {
+        pub fn bo_gh32x32x32_s2x2x2_t16x16x16_cc_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 32,
@@ -3540,7 +3902,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x2x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3573,7 +3935,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x16x16_s2x1x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh32x16x16_s2x1x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 16,
@@ -3594,7 +3956,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S2x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3627,7 +3989,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g32x8x16_s1x1x1_t32x8x16_cc_ln1() {
+        pub fn bo_gh32x8x16_s1x1x1_t32x8x16_cc_ln1() {
             let problem = MatmulProblem {
                 m: 32,
                 n: 8,
@@ -3648,7 +4010,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S1x1x1;
                 type TileMatmul = $t_32x8x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3681,7 +4043,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g128x16x16_s8x1x1_t16x16x16_rr_ln1() {
+        pub fn bo_gh128x16x16_s8x1x1_t16x16x16_rr_ln1() {
             let problem = MatmulProblem {
                 m: 128,
                 n: 16,
@@ -3702,7 +4064,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S8x1x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3735,7 +4097,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g64x64x16_s4x4x1_t16x16x16_rr_ln4() {
+        pub fn bo_gh64x64x16_s4x4x1_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 64,
                 n: 64,
@@ -3756,7 +4118,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x1;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
@@ -3789,7 +4151,7 @@ macro_rules! matmul_test_define {
         }
 
         #[test]
-        pub fn bo_g64x64x32_s4x4x2_t16x16x16_rr_ln4() {
+        pub fn bo_gh64x64x32_s4x4x2_t16x16x16_rr_ln4() {
             let problem = MatmulProblem {
                 m: 64,
                 n: 64,
@@ -3810,7 +4172,7 @@ macro_rules! matmul_test_define {
                 type EA = $ea;
                 type StageSize = S4x4x2;
                 type TileMatmul = $t_16x16x16<Self::ES, Self::EA>;
-                type StageMatmul = stage::row_accumulate::Matmul<
+                type StageMatmul = stage::multi_buffer::Matmul<
                     Self::ES,
                     Self::EG,
                     Self::EA,
