diff --git a/chains/chains_test.go b/chains/chains_test.go
index 1c808082..103fa946 100644
--- a/chains/chains_test.go
+++ b/chains/chains_test.go
@@ -34,10 +34,6 @@ func (spv stringPromptValue) Messages() []schema.ChatMessage {
 	return nil
 }
 
-func (l *testLanguageModel) GetNumTokens(_ string) int {
-	return -1
-}
-
 func (l *testLanguageModel) Call(_ context.Context, prompt string, _ ...llms.CallOption) (string, error) {
 	l.recordedPrompt = []schema.PromptValue{
 		stringPromptValue{s: prompt},
