diff --git a/packages/wonder-blocks-dropdown/src/components/__tests__/action-menu.test.tsx b/packages/wonder-blocks-dropdown/src/components/__tests__/action-menu.test.tsx
index 61388ddaa..e2f5e2fbb 100644
--- a/packages/wonder-blocks-dropdown/src/components/__tests__/action-menu.test.tsx
+++ b/packages/wonder-blocks-dropdown/src/components/__tests__/action-menu.test.tsx
@@ -523,7 +523,7 @@ describe("ActionMenu", () => {
             expect(opener).not.toHaveAttribute("data-testid");
         });
 
-        it("passes the menu text to the custom opener", async () => {
+        it("passes the menu content to the custom opener", async () => {
             // Arrange
             render(
                 <ActionMenu
diff --git a/packages/wonder-blocks-dropdown/src/components/__tests__/multi-select.test.tsx b/packages/wonder-blocks-dropdown/src/components/__tests__/multi-select.test.tsx
index 31b06b4bf..dad6274f6 100644
--- a/packages/wonder-blocks-dropdown/src/components/__tests__/multi-select.test.tsx
+++ b/packages/wonder-blocks-dropdown/src/components/__tests__/multi-select.test.tsx
@@ -19,7 +19,7 @@ import OptionItem from "../option-item";
 import MultiSelect from "../multi-select";
 import {defaultLabels as builtinLabels} from "../../util/constants";
 
-import type {Labels} from "../multi-select";
+import type {LabelsValues} from "../multi-select";
 
 const doRender = (element: React.ReactElement) => {
     return {
@@ -31,7 +31,7 @@ const doRender = (element: React.ReactElement) => {
     };
 };
 
-const defaultLabels: Labels = {
+const defaultLabels: LabelsValues = {
     ...builtinLabels,
     selectAllLabel: (numOptions: any) => `Select all (${numOptions})`,
     noneSelected: "Choose",
@@ -77,7 +77,7 @@ describe("MultiSelect", () => {
             const {userEvent} = doRender(uncontrolledMultiSelect);
 
             // Act
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             // Assert
             expect(
@@ -109,7 +109,7 @@ describe("MultiSelect", () => {
             doRender(uncontrolledMultiSelect);
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             // No items are selected, display placeholder because there is one
@@ -133,7 +133,7 @@ describe("MultiSelect", () => {
             );
 
             // Assert
-            expect(await screen.findByRole("button")).toHaveTextContent(
+            expect(await screen.findByRole("combobox")).toHaveTextContent(
                 "item 1",
             );
         });
@@ -155,7 +155,7 @@ describe("MultiSelect", () => {
             );
 
             // Assert
-            expect(await screen.findByRole("button")).toHaveTextContent(
+            expect(await screen.findByRole("combobox")).toHaveTextContent(
                 "1 student",
             );
         });
@@ -177,7 +177,7 @@ describe("MultiSelect", () => {
             );
 
             // Assert
-            expect(await screen.findByRole("button")).toHaveTextContent(
+            expect(await screen.findByRole("combobox")).toHaveTextContent(
                 "Choose",
             );
         });
@@ -200,7 +200,7 @@ describe("MultiSelect", () => {
 
             // Assert
             // More than one item is selected, display n itemTypes
-            expect(await screen.findByRole("button")).toHaveTextContent(
+            expect(await screen.findByRole("combobox")).toHaveTextContent(
                 "2 students",
             );
         });
@@ -223,7 +223,7 @@ describe("MultiSelect", () => {
 
             // Assert
             // All items are selected
-            expect(await screen.findByRole("button")).toHaveTextContent(
+            expect(await screen.findByRole("combobox")).toHaveTextContent(
                 "All students",
             );
         });
@@ -246,7 +246,7 @@ describe("MultiSelect", () => {
             );
 
             // Assert
-            expect(await screen.findByRole("button")).toHaveTextContent(
+            expect(await screen.findByRole("combobox")).toHaveTextContent(
                 "All students",
             );
         });
@@ -265,7 +265,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("data-testid", "some-test-id");
@@ -286,12 +286,49 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             const menuLabel = within(opener).getByText("custom item 1");
 
             // Assert
             expect(menuLabel).toBeVisible();
         });
+
+        it("applies an aria-label to the opener", async () => {
+            doRender(
+                <MultiSelect onChange={jest.fn()} aria-label="Select a school">
+                    <OptionItem label="school 1" value="1" />
+                    <OptionItem label="school 2" value="2" />
+                    <OptionItem label="school 3" value="3" />
+                </MultiSelect>,
+            );
+
+            // Act
+            const opener = await screen.findByRole("combobox");
+
+            // Assert
+            expect(opener).toHaveAccessibleName("Select a school");
+        });
+
+        it("keeps the original aria-label when items are selected", async () => {
+            doRender(
+                <MultiSelect
+                    onChange={jest.fn()}
+                    selectedValues={["1", "2"]}
+                    labels={defaultLabels}
+                    aria-label="Select a school"
+                >
+                    <OptionItem label="school 1" value="1" />
+                    <OptionItem label="school 2" value="2" />
+                    <OptionItem label="school 3" value="3" />
+                </MultiSelect>,
+            );
+
+            // Act
+            const opener = await screen.findByRole("combobox");
+
+            // Assert
+            expect(opener).toHaveAccessibleName("Select a school");
+        });
     });
 
     describe("Controlled component", () => {
@@ -301,7 +338,7 @@ describe("MultiSelect", () => {
             shortcuts?: boolean;
         };
 
-        const labels: Labels = {
+        const labels: LabelsValues = {
             ...builtinLabels,
             selectAllLabel: (numOptions: any) => `Select all (${numOptions})`,
             allSelected: "All fruits",
@@ -552,7 +589,7 @@ describe("MultiSelect", () => {
 
             const {userEvent} = doRender(<ControlledMultiSelect />);
 
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             // Act
@@ -625,7 +662,7 @@ describe("MultiSelect", () => {
             // Arrange
             const {userEvent} = doRender(filterableMultiSelect);
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             // Act
             const searchInput = await screen.findByPlaceholderText("Filter");
@@ -638,7 +675,7 @@ describe("MultiSelect", () => {
             // Arrange
             const {userEvent} = doRender(filterableMultiSelect);
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             const searchInput = await screen.findByPlaceholderText("Filter");
 
@@ -655,7 +692,7 @@ describe("MultiSelect", () => {
             // Arrange
             const {userEvent} = doRender(filterableMultiSelect);
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             const searchInput = await screen.findByPlaceholderText("Filter");
 
@@ -686,7 +723,7 @@ describe("MultiSelect", () => {
 
             // Act
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             // Assert
             expect(
@@ -709,7 +746,7 @@ describe("MultiSelect", () => {
                 </MultiSelect>,
             );
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             const searchInput = await screen.findByPlaceholderText("Filter");
 
@@ -724,7 +761,7 @@ describe("MultiSelect", () => {
             // Arrange
             const {userEvent} = doRender(filterableMultiSelect);
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             // Act
             const searchInput = await screen.findByPlaceholderText("Filter");
@@ -751,7 +788,7 @@ describe("MultiSelect", () => {
 
             // Act
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             // Assert
             expect(
@@ -775,7 +812,7 @@ describe("MultiSelect", () => {
                 </MultiSelect>,
             );
 
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             const searchInput = await screen.findByPlaceholderText("Filter");
             await userEvent.type(searchInput, "Item 2");
@@ -807,7 +844,7 @@ describe("MultiSelect", () => {
                 </MultiSelect>,
             );
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             const searchInput = await screen.findByPlaceholderText("Filter");
             await userEvent.type(searchInput, "some text");
@@ -824,7 +861,7 @@ describe("MultiSelect", () => {
             // Arrange
             const {userEvent} = doRender(filterableMultiSelect);
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             const searchInput = await screen.findByPlaceholderText("Filter");
             await userEvent.type(searchInput, "Should be cleared");
@@ -842,7 +879,7 @@ describe("MultiSelect", () => {
             // Arrange
             const {userEvent} = doRender(filterableMultiSelect);
             // open the dropdown menu
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             const searchInput = await screen.findByPlaceholderText("Filter");
@@ -863,7 +900,7 @@ describe("MultiSelect", () => {
 
         it("should find an option after using the search filter", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 someSelected: (numOptions: number): string =>
                     numOptions <= 1
@@ -897,7 +934,7 @@ describe("MultiSelect", () => {
         // The Venus option is still in the document.
         it.skip("should filter out an option if it's not part of the results", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 someSelected: (numOptions: number): string =>
                     numOptions <= 1
@@ -931,6 +968,53 @@ describe("MultiSelect", () => {
     });
 
     describe("Custom Opener", () => {
+        it("labels the custom opener with `aria-label` on MultiSelect", async () => {
+            // Arrange
+            doRender(
+                <MultiSelect
+                    aria-label="Search"
+                    onChange={jest.fn()}
+                    opener={(eventState: any) => <button onClick={jest.fn()} />}
+                >
+                    <OptionItem label="item 1" value="1" />
+                    <OptionItem label="item 2" value="2" />
+                    <OptionItem label="item 3" value="3" />
+                </MultiSelect>,
+            );
+
+            // Act
+            const opener = await screen.findByRole("combobox");
+
+            // Assert
+            expect(opener).toHaveAccessibleName("Search");
+        });
+
+        it("prioritizes `aria-label` on the custom opener", async () => {
+            // Arrange
+            doRender(
+                <MultiSelect
+                    aria-label="Not winning the label race"
+                    onChange={jest.fn()}
+                    opener={(eventState: any) => (
+                        <button
+                            aria-label="Search button"
+                            onClick={jest.fn()}
+                        />
+                    )}
+                >
+                    <OptionItem label="item 1" value="1" />
+                    <OptionItem label="item 2" value="2" />
+                    <OptionItem label="item 3" value="3" />
+                </MultiSelect>,
+            );
+
+            // Act
+            const opener = await screen.findByRole("combobox");
+
+            // Assert
+            expect(opener).toHaveAccessibleName("Search button");
+        });
+
         it("opens the menu when clicking on the custom opener", async () => {
             // Arrange
             const {userEvent} = doRender(
@@ -1029,9 +1113,9 @@ describe("MultiSelect", () => {
             expect(opener).not.toHaveAttribute("data-testid", "custom-opener");
         });
 
-        it("passes the current label to the custom opener (no items selected)", async () => {
+        it("passes the current value to the custom opener (no items selected)", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 noneSelected: "No items selected",
             };
@@ -1060,7 +1144,7 @@ describe("MultiSelect", () => {
             expect(opener).toHaveTextContent("No items selected");
         });
 
-        it("passes the current label to the custom opener (1 item selected)", async () => {
+        it("passes the current value to the custom opener (1 item selected)", async () => {
             // Arrange
             const ControlledMultiSelect = () => {
                 const [selected, setSelected] = React.useState([]);
@@ -1101,7 +1185,7 @@ describe("MultiSelect", () => {
             expect(opener).toHaveTextContent("item 1");
         });
 
-        it("passes the current label to the custom opener (2 items selected)", async () => {
+        it("passes the current value to the custom opener (2 items selected)", async () => {
             // Arrange
             const ControlledMultiSelect = () => {
                 const [selected, setSelected] = React.useState([]);
@@ -1142,7 +1226,7 @@ describe("MultiSelect", () => {
             expect(opener).toHaveTextContent("2 items");
         });
 
-        it("passes the current label to the custom opener (all items selected)", async () => {
+        it("passes the current value to the custom opener (all items selected)", async () => {
             // Arrange
             const ControlledMultiSelect = () => {
                 const [selected, setSelected] = React.useState([]);
@@ -1185,10 +1269,10 @@ describe("MultiSelect", () => {
         });
     });
 
-    describe("Custom labels", () => {
-        it("passes the custom label to the opener", async () => {
+    describe("Custom labels and values", () => {
+        it("passes the custom value to the opener", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 noneSelected: "0 escuelas",
             };
@@ -1213,7 +1297,7 @@ describe("MultiSelect", () => {
 
         it("passes the custom label to the opener (2 items selected)", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 someSelected: (numSelectedValues: any) =>
                     `${numSelectedValues} escuelas`,
@@ -1238,9 +1322,9 @@ describe("MultiSelect", () => {
             expect(opener).toHaveTextContent("2 escuelas");
         });
 
-        it("passes the custom label to the opener (all items selected)", async () => {
+        it("passes the custom value to the opener (all items selected)", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 allSelected: "Todas las escuelas",
             };
@@ -1266,7 +1350,7 @@ describe("MultiSelect", () => {
 
         it("passes the custom label to the dismiss icon", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 clearSearch: "Limpiar busqueda",
             };
@@ -1303,7 +1387,7 @@ describe("MultiSelect", () => {
 
         it("passes the custom label to the search input field", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 filter: "Filtrar",
             };
@@ -1331,9 +1415,9 @@ describe("MultiSelect", () => {
             expect(searchInput).toBeInTheDocument();
         });
 
-        it("passes the custom label to the no results label", async () => {
+        it("passes the custom value for no results", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 noResults: "No hay resultados",
             };
@@ -1368,7 +1452,7 @@ describe("MultiSelect", () => {
 
         it("passes the custom label to the select all shortcut", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 selectAllLabel: (numOptions: any) =>
                     `Seleccionar todas las escuelas (${numOptions})`,
@@ -1400,7 +1484,7 @@ describe("MultiSelect", () => {
 
         it("passes the custom label to the select none shortcut", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 selectNoneLabel: "Deseleccionar todas las escuelas",
             };
@@ -1431,12 +1515,12 @@ describe("MultiSelect", () => {
 
         it("verifies a custom label is updated when props change", async () => {
             // Arrange
-            const initialLabels: Labels = {
+            const initialLabels: LabelsValues = {
                 ...defaultLabels,
                 selectNoneLabel: "Deseleccionar todas las escuelas",
             };
 
-            const TranslatedComponent = ({labels}: {labels: Labels}) => (
+            const TranslatedComponent = ({labels}: {labels: LabelsValues}) => (
                 <MultiSelect
                     onChange={jest.fn()}
                     isFilterable={true}
@@ -1454,7 +1538,7 @@ describe("MultiSelect", () => {
             );
 
             // update label value
-            const updatedLabels: Labels = {
+            const updatedLabels: LabelsValues = {
                 ...defaultLabels,
                 selectNoneLabel: "Ninguna seleccionada",
             };
@@ -1462,7 +1546,7 @@ describe("MultiSelect", () => {
             rerender(<TranslatedComponent labels={updatedLabels} />);
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             // open dropdown
             await userEvent.click(opener);
 
@@ -1511,7 +1595,7 @@ describe("MultiSelect", () => {
     describe("a11y > Live region", () => {
         it("should announce the number of options when the listbox is open", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 someSelected: (numOptions: number): string =>
                     numOptions <= 1
@@ -1541,7 +1625,7 @@ describe("MultiSelect", () => {
         // output is now: "2 planets0 items".
         it.skip("should change the number of options after using the search filter", async () => {
             // Arrange
-            const labels: Labels = {
+            const labels: LabelsValues = {
                 ...builtinLabels,
                 someSelected: (numOptions: number): string =>
                     numOptions <= 1
@@ -1682,7 +1766,7 @@ describe("MultiSelect", () => {
 
             // Act
             // Press the button
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -1710,7 +1794,7 @@ describe("MultiSelect", () => {
 
             // Act
             // Press the button
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -1825,7 +1909,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("id", expect.any(String));
@@ -1842,7 +1926,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("id", id);
@@ -1859,7 +1943,7 @@ describe("MultiSelect", () => {
 
             // Act
             // Open the dropdown
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             // Assert
@@ -1880,7 +1964,7 @@ describe("MultiSelect", () => {
 
             // Act
             // Open the dropdown
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             // Assert
@@ -1902,7 +1986,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
             const dropdown = await screen.findByRole("listbox", {hidden: true});
 
@@ -1921,7 +2005,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
             const dropdown = await screen.findByRole("listbox", {hidden: true});
 
@@ -1992,7 +2076,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("aria-haspopup", "listbox");
@@ -2031,7 +2115,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("aria-expanded", "false");
@@ -2047,7 +2131,7 @@ describe("MultiSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             // Assert
@@ -2110,7 +2194,7 @@ describe("MultiSelect", () => {
                 doRender(<MultiSelect onChange={jest.fn()} error={error} />);
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveAttribute("aria-invalid", ariaInvalid);
@@ -2136,7 +2220,7 @@ describe("MultiSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveAttribute("aria-invalid", ariaInvalid);
@@ -2152,7 +2236,11 @@ describe("MultiSelect", () => {
         it("should not have any violations", async () => {
             // Arrange
             const {container} = doRender(
-                <MultiSelect onChange={jest.fn()} opened={true}>
+                <MultiSelect
+                    onChange={jest.fn()}
+                    opened={true}
+                    aria-label="Item selector"
+                >
                     <OptionItem label="item 1" value="1" />
                     <OptionItem label="item 2" value="2" />
                 </MultiSelect>,
@@ -2173,7 +2261,11 @@ describe("MultiSelect", () => {
         it("should not have any violations when it is open", async () => {
             // Arrange
             const {container} = doRender(
-                <MultiSelect onChange={jest.fn()} opened={true}>
+                <MultiSelect
+                    onChange={jest.fn()}
+                    opened={true}
+                    aria-label="Item selector"
+                >
                     <OptionItem label="item 1" value="1" />
                     <OptionItem label="item 2" value="2" />
                 </MultiSelect>,
@@ -2235,7 +2327,7 @@ describe("MultiSelect", () => {
                     const {userEvent} = doRender(
                         <ControlledMultiSelect validate={validate} />,
                     );
-                    const opener = await screen.findByRole("button");
+                    const opener = await screen.findByRole("combobox");
                     await userEvent.click(opener);
                     await userEvent.click(screen.getByText("item 1"));
                     validate.mockClear(); // Clear any calls
@@ -2257,7 +2349,7 @@ describe("MultiSelect", () => {
                             onValidate={onValidate}
                         />,
                     );
-                    const opener = await screen.findByRole("button");
+                    const opener = await screen.findByRole("combobox");
                     await userEvent.click(opener);
                     await userEvent.click(screen.getByText("item 1"));
                     onValidate.mockClear(); // Clear any calls
@@ -2276,7 +2368,7 @@ describe("MultiSelect", () => {
                     const {userEvent} = doRender(
                         <ControlledMultiSelect validate={() => "Error"} />,
                     );
-                    const opener = await screen.findByRole("button");
+                    const opener = await screen.findByRole("combobox");
                     await userEvent.click(opener);
                     await userEvent.click(screen.getByText("item 1"));
 
@@ -2320,7 +2412,7 @@ describe("MultiSelect", () => {
                 const {userEvent} = doRender(
                     <ControlledMultiSelect validate={validate} />,
                 );
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
                 await userEvent.click(opener);
 
                 // Act
@@ -2339,7 +2431,7 @@ describe("MultiSelect", () => {
                         onValidate={onValidate}
                     />,
                 );
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
                 await userEvent.click(opener);
 
                 // Act
@@ -2354,7 +2446,7 @@ describe("MultiSelect", () => {
                 const {userEvent} = doRender(
                     <ControlledMultiSelect validate={() => "Error"} />,
                 );
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
                 await userEvent.click(opener);
 
                 // Act
@@ -2395,7 +2487,7 @@ describe("MultiSelect", () => {
                     />,
                 );
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "true",
                 );
@@ -2415,7 +2507,7 @@ describe("MultiSelect", () => {
                     />,
                 );
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -2446,7 +2538,7 @@ describe("MultiSelect", () => {
                     />,
                 );
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -2467,10 +2559,10 @@ describe("MultiSelect", () => {
                 );
 
                 // Act
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "true",
                 );
@@ -2487,13 +2579,13 @@ describe("MultiSelect", () => {
                         selectedValues={["1"]}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                 // Act
                 await userEvent.click(await screen.findByText("item 2")); // Pick a value
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -2510,14 +2602,14 @@ describe("MultiSelect", () => {
                         selectedValues={["1"]}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
                 await userEvent.click(await screen.findByText("item 2")); // Pick a value
 
                 // Act
-                await userEvent.click(await screen.findByRole("button")); // Close the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Close the dropdown
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "true",
                 );
@@ -2536,7 +2628,7 @@ describe("MultiSelect", () => {
                         onValidate={onValidate}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
                 onValidate.mockClear(); // Clear any calls
 
                 // Act
@@ -2559,7 +2651,7 @@ describe("MultiSelect", () => {
                     />,
                 );
                 // Open the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
                 // Need to find item text within the listbox since the item text is also used on the opener
                 const listbox = await screen.findByRole("listbox", {
                     hidden: true,
@@ -2575,10 +2667,10 @@ describe("MultiSelect", () => {
 
                 // Act
                 // Close the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "true",
                 );
@@ -2598,7 +2690,7 @@ describe("MultiSelect", () => {
                         shortcuts={true}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
                 onValidate.mockClear(); // Clear any calls
 
                 // Act
@@ -2620,13 +2712,13 @@ describe("MultiSelect", () => {
                         shortcuts={true}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                 // Act
                 await userEvent.click(await screen.findByText("Select none")); // Select none
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -2646,7 +2738,7 @@ describe("MultiSelect", () => {
                         shortcuts={true}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
                 onValidate.mockClear(); // Clear any calls
 
                 // Act
@@ -2670,7 +2762,7 @@ describe("MultiSelect", () => {
                         shortcuts={true}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                 // Act
                 await userEvent.click(
@@ -2678,7 +2770,7 @@ describe("MultiSelect", () => {
                 ); // Select all
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -2720,7 +2812,7 @@ describe("MultiSelect", () => {
                     await userEvent.tab(); // leave the select
 
                     // Assert
-                    expect(screen.getByRole("button")).toHaveAttribute(
+                    expect(screen.getByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -2814,7 +2906,7 @@ describe("MultiSelect", () => {
                     await userEvent.tab(); // leave the select
 
                     // Assert
-                    expect(screen.getByRole("button")).toHaveAttribute(
+                    expect(screen.getByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "false",
                     );
@@ -2832,10 +2924,10 @@ describe("MultiSelect", () => {
                             required={requiredMessage}
                         />,
                     );
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Close the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Close the dropdown
 
                     // Assert
                     expect(onValidate).toHaveBeenCalledExactlyOnceWith(
@@ -2849,13 +2941,13 @@ describe("MultiSelect", () => {
                     const {userEvent} = doRender(
                         <ControlledMultiSelect required={requiredMessage} />,
                     );
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Close the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Close the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -2873,7 +2965,7 @@ describe("MultiSelect", () => {
                     );
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Assert
                     expect(onValidate).not.toHaveBeenCalled();
@@ -2887,10 +2979,10 @@ describe("MultiSelect", () => {
                     );
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "false",
                     );
@@ -2933,7 +3025,7 @@ describe("MultiSelect", () => {
                     await userEvent.keyboard("{escape}"); // Close the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -2986,10 +3078,10 @@ describe("MultiSelect", () => {
                     await userEvent.tab(); // Tab through the select to trigger error
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -3003,13 +3095,13 @@ describe("MultiSelect", () => {
                     );
                     await userEvent.tab();
                     await userEvent.tab(); // Tab through the select to trigger error
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Act
                     await userEvent.click(await screen.findByText("item 1")); // Pick a value
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "false",
                     );
@@ -3049,7 +3141,7 @@ describe("MultiSelect", () => {
                     />,
                 );
                 // Open the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
 
                 // Unselect selected option
                 const listbox = await screen.findByRole("listbox", {
@@ -3061,10 +3153,10 @@ describe("MultiSelect", () => {
 
                 // Act
                 // Close the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "true",
                 );
@@ -3083,7 +3175,7 @@ describe("MultiSelect", () => {
                     />,
                 );
                 // Open the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
 
                 // Unselect selected option
                 const listbox = await screen.findByRole("listbox", {
@@ -3096,7 +3188,7 @@ describe("MultiSelect", () => {
 
                 // Act
                 // Close the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
 
                 // Assert
                 expect(onValidate.mock.calls).toStrictEqual([
@@ -3118,7 +3210,7 @@ describe("MultiSelect", () => {
                     />,
                 );
                 // Open the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
                 // Unselect selected option
                 const listbox = await screen.findByRole("listbox", {
                     hidden: true,
@@ -3130,7 +3222,7 @@ describe("MultiSelect", () => {
 
                 // Act
                 // Close the dropdown
-                await userEvent.click(await screen.findByRole("button"));
+                await userEvent.click(await screen.findByRole("combobox"));
 
                 // Assert
                 expect(onValidate).toHaveBeenCalledExactlyOnceWith(
diff --git a/packages/wonder-blocks-dropdown/src/components/__tests__/select-opener.test.tsx b/packages/wonder-blocks-dropdown/src/components/__tests__/select-opener.test.tsx
index 5bfeb047a..1dcb91a75 100644
--- a/packages/wonder-blocks-dropdown/src/components/__tests__/select-opener.test.tsx
+++ b/packages/wonder-blocks-dropdown/src/components/__tests__/select-opener.test.tsx
@@ -26,7 +26,7 @@ describe("SelectOpener", () => {
 
             // Act
             // Press the button.
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             // Assert
             expect(onOpenMock).toHaveBeenCalledTimes(1);
@@ -51,7 +51,7 @@ describe("SelectOpener", () => {
 
             // Act
             // Press the button.
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -87,7 +87,7 @@ describe("SelectOpener", () => {
 
             // Act
             // Press the button.
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -131,7 +131,7 @@ describe("SelectOpener", () => {
 
             // Act
             // Press the button.
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             // Assert
             expect(onOpenMock).toHaveBeenCalledTimes(0);
@@ -156,7 +156,7 @@ describe("SelectOpener", () => {
 
             // Act
             // Press the button.
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -192,7 +192,7 @@ describe("SelectOpener", () => {
 
             // Act
             // Press the button.
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -238,7 +238,7 @@ describe("SelectOpener", () => {
                     </SelectOpener>,
                 );
                 // Assert
-                expect(screen.getByRole("button")).toHaveAttribute(
+                expect(screen.getByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     ariaInvalid,
                 );
diff --git a/packages/wonder-blocks-dropdown/src/components/__tests__/single-select.test.tsx b/packages/wonder-blocks-dropdown/src/components/__tests__/single-select.test.tsx
index a622a3202..8da80b701 100644
--- a/packages/wonder-blocks-dropdown/src/components/__tests__/single-select.test.tsx
+++ b/packages/wonder-blocks-dropdown/src/components/__tests__/single-select.test.tsx
@@ -9,7 +9,7 @@ import {
 import {PropsFor} from "@khanacademy/wonder-blocks-core";
 import OptionItem from "../option-item";
 import SingleSelect from "../single-select";
-import type {SingleSelectLabels} from "../single-select";
+import type {SingleSelectLabelsValues} from "../single-select";
 
 const doRender = (element: React.ReactElement) => {
     return {
@@ -64,12 +64,55 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveTextContent("Default placeholder");
             });
 
+            it("should render an aria-label if passed in as a prop", async () => {
+                // Arrange
+                doRender(
+                    <SingleSelect
+                        placeholder="Default placeholder"
+                        onChange={jest.fn()}
+                        aria-label="Select a thing"
+                    >
+                        <OptionItem label="" value="" />
+                        <OptionItem label="Toggle A" value="toggle_a" />
+                        <OptionItem label="Toggle B" value="toggle_b" />
+                    </SingleSelect>,
+                );
+
+                // Act
+                const opener = await screen.findByRole("combobox");
+
+                // Assert
+                expect(opener).toHaveAccessibleName("Select a thing");
+            });
+
+            it("should preserve the original aria-label when an option is selected", async () => {
+                // Arrange
+                doRender(
+                    <SingleSelect
+                        placeholder="Default placeholder"
+                        onChange={jest.fn()}
+                        aria-label="Select a thing"
+                        selectedValue="toggle_a"
+                    >
+                        <OptionItem label="" value="" />
+                        <OptionItem label="Toggle A" value="toggle_a" />
+                        <OptionItem label="Toggle B" value="toggle_b" />
+                    </SingleSelect>,
+                );
+
+                // Act
+                const opener = await screen.findByRole("combobox");
+
+                // Assert
+                expect(opener).toHaveAccessibleName("Select a thing");
+            });
+
             it("should render empty if the selected option has an empty value", async () => {
                 // Arrange
                 doRender(
@@ -85,7 +128,7 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveTextContent("");
@@ -105,13 +148,13 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveTextContent("Toggle A");
             });
 
-            it("should render labelAsText of the selected option", async () => {
+            it("should render labelAsText of the selected option for opener value", async () => {
                 // Arrange
                 doRender(
                     <SingleSelect
@@ -133,13 +176,13 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveTextContent("Plain Toggle A");
             });
 
-            it("can render a Node as a label", async () => {
+            it("can render a Node as an opener value", async () => {
                 // Arrange
                 doRender(
                     <SingleSelect
@@ -162,7 +205,7 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
                 const menuLabel = within(opener).getByText("custom item A");
 
                 // Assert
@@ -253,13 +296,13 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                await userEvent.click(await screen.findByRole("textbox"));
+                await userEvent.click(await screen.findByRole("combobox"));
 
                 // wait for the dropdown to open
                 await screen.findByRole("listbox", {hidden: true});
 
                 // Assert
-                expect(await screen.findByRole("textbox")).toHaveFocus();
+                expect(await screen.findByRole("combobox")).toHaveFocus();
             });
         });
 
@@ -432,6 +475,7 @@ describe("SingleSelect", () => {
                         onToggle={handleToggleMenu}
                         onChange={onChange}
                         placeholder="Choose"
+                        aria-label="Choose"
                     >
                         <OptionItem label="item 1" value="1" />
                         <OptionItem label="item 2" value="2" />
@@ -454,7 +498,7 @@ describe("SingleSelect", () => {
 
             // Act
             await userEvent.click(
-                await screen.findByRole("button", {name: "Choose"}),
+                await screen.findByRole("combobox", {name: /choose/i}),
             );
 
             // Assert
@@ -469,7 +513,7 @@ describe("SingleSelect", () => {
             );
             // open the menu from the outside
             await userEvent.click(
-                await screen.findByRole("button", {name: "Choose"}),
+                await screen.findByRole("combobox", {name: /choose/i}),
             );
 
             // Act
@@ -513,7 +557,9 @@ describe("SingleSelect", () => {
             const {userEvent} = doRender(<ControlledComponent />);
 
             // Act
-            const opener = await screen.findByRole("button", {name: "Choose"});
+            const opener = await screen.findByRole("combobox", {
+                name: /choose/i,
+            });
             // open the menu from the outside
             await userEvent.click(opener);
             // click on the dropdown anchor to hide the menu
@@ -527,6 +573,55 @@ describe("SingleSelect", () => {
     });
 
     describe("Custom Opener", () => {
+        it("labels the custom opener with `aria-label` on SingleSelect", async () => {
+            // Arrange
+            doRender(
+                <SingleSelect
+                    aria-label="Custom opener"
+                    placeholder="e.g. item 1"
+                    onChange={jest.fn()}
+                    opener={(eventState: any) => <button onClick={jest.fn()} />}
+                >
+                    <OptionItem label="item 1" value="1" />
+                    <OptionItem label="item 2" value="2" />
+                    <OptionItem label="item 3" value="3" />
+                </SingleSelect>,
+            );
+
+            // Act
+            const opener = await screen.findByRole("combobox");
+
+            // Assert
+            expect(opener).toHaveAccessibleName("Custom opener");
+        });
+
+        it("prioritizes `aria-label` on the custom opener", async () => {
+            // Arrange
+            doRender(
+                <SingleSelect
+                    aria-label="Not winning the label race"
+                    placeholder="Choose an item"
+                    onChange={jest.fn()}
+                    opener={(eventState: any) => (
+                        <button
+                            aria-label="Custom opener button"
+                            onClick={jest.fn()}
+                        />
+                    )}
+                >
+                    <OptionItem label="item 1" value="1" />
+                    <OptionItem label="item 2" value="2" />
+                    <OptionItem label="item 3" value="3" />
+                </SingleSelect>,
+            );
+
+            // Act
+            const opener = await screen.findByRole("combobox");
+
+            // Assert
+            expect(opener).toHaveAccessibleName("Custom opener button");
+        });
+
         it("opens the menu when clicking on the custom opener", async () => {
             // Arrange
             const {userEvent} = doRender(
@@ -606,7 +701,7 @@ describe("SingleSelect", () => {
             expect(opener).not.toHaveAttribute("data-testid");
         });
 
-        it("passes the placeholder text to the custom opener", async () => {
+        it("passes the placeholder content to the custom opener", async () => {
             // Arrange
             const {userEvent} = doRender(
                 <SingleSelect
@@ -623,7 +718,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             // open dropdown
             await userEvent.click(opener);
 
@@ -631,7 +726,7 @@ describe("SingleSelect", () => {
             expect(opener).toHaveTextContent("Custom placeholder");
         });
 
-        it("passes the selected label to the custom opener", async () => {
+        it("passes the selected value to the custom opener", async () => {
             // Arrange
             type Props = Record<any, any>;
 
@@ -668,13 +763,13 @@ describe("SingleSelect", () => {
             const {userEvent} = doRender(<ControlledComponent />);
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             // open dropdown
             await userEvent.click(opener);
             await userEvent.click(await screen.findByText("Toggle B"));
 
             // Assert
-            // NOTE: the opener text is only updated in response to changes to the
+            // NOTE: the opener value text is only updated in response to changes to the
             // `selectedValue` prop.
             expect(opener).toHaveTextContent("Toggle B");
         });
@@ -823,7 +918,7 @@ describe("SingleSelect", () => {
                 </SingleSelect>,
             );
             // open the dropdown menu
-            await userEvent.click(await screen.findByRole("button"));
+            await userEvent.click(await screen.findByRole("combobox"));
 
             const searchInput = await screen.findByPlaceholderText("Filter");
             await userEvent.click(searchInput);
@@ -1017,7 +1112,7 @@ describe("SingleSelect", () => {
             <OptionItem label="Mango" value="mango" />,
         ];
 
-        const enLabels: SingleSelectLabels = {
+        const enLabels: SingleSelectLabelsValues = {
             clearSearch: "Clear Search",
             filter: "Filter",
             noResults: "No Results",
@@ -1026,7 +1121,7 @@ describe("SingleSelect", () => {
 
         it("passes the custom label to the search input field", async () => {
             // Arrange
-            const labels: SingleSelectLabels = {
+            const labels: SingleSelectLabelsValues = {
                 ...enLabels,
                 filter: "Filtrar",
             };
@@ -1052,7 +1147,7 @@ describe("SingleSelect", () => {
 
         it("passes the custom label to the dismiss filter icon", async () => {
             // Arrange
-            const labels: SingleSelectLabels = {
+            const labels: SingleSelectLabelsValues = {
                 ...enLabels,
                 clearSearch: "Limpiar busqueda",
                 filter: "Filtrar",
@@ -1085,7 +1180,7 @@ describe("SingleSelect", () => {
 
         it("passes the custom label to the no results label", async () => {
             // Arrange
-            const labels: SingleSelectLabels = {
+            const labels: SingleSelectLabelsValues = {
                 ...enLabels,
                 filter: "Filtrar",
                 noResults: "No hay resultados",
@@ -1241,7 +1336,7 @@ describe("SingleSelect", () => {
 
             // Act
             // Press the button
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -1272,7 +1367,7 @@ describe("SingleSelect", () => {
 
             // Act
             // Press the button
-            const button = await screen.findByRole("button");
+            const button = await screen.findByRole("combobox");
             // NOTE: we need to use fireEvent here because await userEvent doesn't
             // support keyUp/Down events and we use these handlers to override
             // the default behavior of the button.
@@ -1394,7 +1489,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("id", expect.any(String));
@@ -1411,7 +1506,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("id", id);
@@ -1428,7 +1523,7 @@ describe("SingleSelect", () => {
 
             // Act
             // Open the dropdown
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             // Assert
@@ -1453,7 +1548,7 @@ describe("SingleSelect", () => {
 
             // Act
             // Open the dropdown
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             // Assert
@@ -1479,7 +1574,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
             const dropdown = await screen.findByRole("listbox", {hidden: true});
 
@@ -1498,7 +1593,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
             const dropdown = await screen.findByRole("listbox", {hidden: true});
 
@@ -1571,7 +1666,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("aria-haspopup", "listbox");
@@ -1611,7 +1706,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
 
             // Assert
             expect(opener).toHaveAttribute("aria-expanded", "false");
@@ -1627,7 +1722,7 @@ describe("SingleSelect", () => {
             );
 
             // Act
-            const opener = await screen.findByRole("button");
+            const opener = await screen.findByRole("combobox");
             await userEvent.click(opener);
 
             // Assert
@@ -1698,7 +1793,7 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveAttribute("aria-invalid", ariaInvalid);
@@ -1725,7 +1820,7 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
 
                 // Assert
                 expect(opener).toHaveAttribute("aria-invalid", ariaInvalid);
@@ -1745,6 +1840,7 @@ describe("SingleSelect", () => {
                     onChange={jest.fn()}
                     opened={true}
                     placeholder="Choose"
+                    aria-label="Choose"
                 >
                     <OptionItem label="item 1" value="1" />
                     <OptionItem label="item 2" value="2" />
@@ -1770,6 +1866,7 @@ describe("SingleSelect", () => {
                     onChange={jest.fn()}
                     opened={true}
                     placeholder="Choose"
+                    aria-label="Choose"
                 >
                     <OptionItem label="item 1" value="1" />
                     <OptionItem label="item 2" value="2" />
@@ -1799,6 +1896,7 @@ describe("SingleSelect", () => {
             return (
                 <SingleSelect
                     {...props}
+                    aria-label="Choose"
                     placeholder="Choose"
                     selectedValue={value}
                     onChange={setValue}
@@ -1815,7 +1913,7 @@ describe("SingleSelect", () => {
                 const {userEvent} = doRender(
                     <ControlledSingleSelect validate={validate} />,
                 );
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
                 await userEvent.click(opener);
 
                 // Act
@@ -1834,7 +1932,7 @@ describe("SingleSelect", () => {
                         onValidate={onValidate}
                     />,
                 );
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
                 await userEvent.click(opener);
 
                 // Act
@@ -1851,7 +1949,7 @@ describe("SingleSelect", () => {
                 const {userEvent} = doRender(
                     <ControlledSingleSelect validate={() => "Error"} />,
                 );
-                const opener = await screen.findByRole("button");
+                const opener = await screen.findByRole("combobox");
                 await userEvent.click(opener);
 
                 // Act
@@ -1911,7 +2009,7 @@ describe("SingleSelect", () => {
                     />,
                 );
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "true",
                 );
@@ -1931,7 +2029,7 @@ describe("SingleSelect", () => {
                     />,
                 );
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -1962,7 +2060,7 @@ describe("SingleSelect", () => {
                     />,
                 );
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -1983,10 +2081,10 @@ describe("SingleSelect", () => {
                 );
 
                 // Act
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "true",
                 );
@@ -2003,13 +2101,13 @@ describe("SingleSelect", () => {
                         selectedValue={"1"}
                     />,
                 );
-                await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                 // Act
                 await userEvent.click(await screen.findByText("item 2")); // Pick a value
 
                 // Assert
-                expect(await screen.findByRole("button")).toHaveAttribute(
+                expect(await screen.findByRole("combobox")).toHaveAttribute(
                     "aria-invalid",
                     "false",
                 );
@@ -2051,7 +2149,7 @@ describe("SingleSelect", () => {
                     await userEvent.tab(); // leave the select
 
                     // Assert
-                    expect(screen.getByRole("button")).toHaveAttribute(
+                    expect(screen.getByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -2145,7 +2243,7 @@ describe("SingleSelect", () => {
                     await userEvent.tab(); // leave the select
 
                     // Assert
-                    expect(screen.getByRole("button")).toHaveAttribute(
+                    expect(screen.getByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "false",
                     );
@@ -2164,8 +2262,8 @@ describe("SingleSelect", () => {
                     );
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
-                    await userEvent.click(await screen.findByRole("button")); // Close the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Close the dropdown
 
                     // Assert
                     expect(onValidate).toHaveBeenCalledExactlyOnceWith(
@@ -2181,11 +2279,11 @@ describe("SingleSelect", () => {
                     );
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
-                    await userEvent.click(await screen.findByRole("button")); // Close the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Close the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -2203,7 +2301,7 @@ describe("SingleSelect", () => {
                     );
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Assert
                     expect(onValidate).not.toHaveBeenCalled();
@@ -2217,10 +2315,10 @@ describe("SingleSelect", () => {
                     );
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "false",
                     );
@@ -2263,7 +2361,7 @@ describe("SingleSelect", () => {
                     await userEvent.keyboard("{escape}"); // Close the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -2318,10 +2416,10 @@ describe("SingleSelect", () => {
                     await userEvent.tab(); // Tab through the select to trigger error
 
                     // Act
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "true",
                     );
@@ -2335,13 +2433,13 @@ describe("SingleSelect", () => {
                     );
                     await userEvent.tab();
                     await userEvent.tab(); // Tab through the select to trigger error
-                    await userEvent.click(await screen.findByRole("button")); // Open the dropdown
+                    await userEvent.click(await screen.findByRole("combobox")); // Open the dropdown
 
                     // Act
                     await userEvent.click(await screen.findByText("item 1")); // Pick a value
 
                     // Assert
-                    expect(await screen.findByRole("button")).toHaveAttribute(
+                    expect(await screen.findByRole("combobox")).toHaveAttribute(
                         "aria-invalid",
                         "false",
                     );
