diff --git a/packages/mui-joy/src/Input/Input.test.js b/packages/mui-joy/src/Input/Input.test.js
index 77d37fa9f4..991362f952 100644
--- a/packages/mui-joy/src/Input/Input.test.js
+++ b/packages/mui-joy/src/Input/Input.test.js
@@ -7,6 +7,7 @@ import {
   createRenderer,
   screen,
   act,
+  fireEvent,
 } from 'test/utils';
 import Input, { inputClasses as classes } from '@mui/joy/Input';
 import { ThemeProvider } from '@mui/joy/styles';
@@ -34,6 +35,11 @@ describe('Joy <Input />', () => {
 
   describeJoyColorInversion(<Input />, { muiName: 'JoyInput', classes });
 
+  it('should have placeholder', () => {
+    const { getByPlaceholderText } = render(<Input placeholder="Placeholder" />);
+    expect(getByPlaceholderText('Placeholder')).toBeVisible();
+  });
+
   it('should have error classes', () => {
     const { container } = render(<Input error />);
     expect(container.firstChild).to.have.class(classes.error);
@@ -71,10 +77,10 @@ describe('Joy <Input />', () => {
     it('should reset the focused state if getting disabled', () => {
       const handleBlur = spy();
       const handleFocus = spy();
-      const { container, setProps } = render(<Input onBlur={handleBlur} onFocus={handleFocus} />);
+      const { getByRole, setProps } = render(<Input onBlur={handleBlur} onFocus={handleFocus} />);
 
       act(() => {
-        container.querySelector('input').focus();
+        getByRole('textbox').focus();
       });
       expect(handleFocus.callCount).to.equal(1);
 
@@ -84,4 +90,54 @@ describe('Joy <Input />', () => {
       expect(handleFocus.callCount).to.equal(1);
     });
   });
+
+  describe('slotProps: input', () => {
+    it('`onKeyDown` and `onKeyUp` should work', () => {
+      const handleKeyDown = spy();
+      const handleKeyUp = spy();
+      const { getByRole } = render(
+        <Input slotProps={{ input: { onKeyDown: handleKeyDown, onKeyUp: handleKeyUp } }} />,
+      );
+
+      act(() => {
+        getByRole('textbox').focus();
+      });
+      fireEvent.keyDown(getByRole('textbox'));
+      fireEvent.keyUp(getByRole('textbox'));
+
+      expect(handleKeyDown.callCount).to.equal(1);
+      expect(handleKeyUp.callCount).to.equal(1);
+    });
+
+    it('should call focus and blur', () => {
+      const handleBlur = spy();
+      const handleFocus = spy();
+      const { getByRole } = render(
+        <Input slotProps={{ input: { onFocus: handleFocus, onBlur: handleBlur } }} />,
+      );
+
+      act(() => {
+        getByRole('textbox').focus();
+      });
+      expect(handleFocus.callCount).to.equal(1);
+      act(() => {
+        getByRole('textbox').blur();
+      });
+      expect(handleFocus.callCount).to.equal(1);
+    });
+
+    it('should override outer handlers', () => {
+      const handleFocus = spy();
+      const handleSlotFocus = spy();
+      const { getByRole } = render(
+        <Input onFocus={handleFocus} slotProps={{ input: { onFocus: handleSlotFocus } }} />,
+      );
+
+      act(() => {
+        getByRole('textbox').focus();
+      });
+      expect(handleFocus.callCount).to.equal(0);
+      expect(handleSlotFocus.callCount).to.equal(1);
+    });
+  });
 });
diff --git a/packages/mui-joy/src/Textarea/Textarea.test.tsx b/packages/mui-joy/src/Textarea/Textarea.test.tsx
index 2710f4f0f3..dfe176c484 100644
--- a/packages/mui-joy/src/Textarea/Textarea.test.tsx
+++ b/packages/mui-joy/src/Textarea/Textarea.test.tsx
@@ -7,6 +7,7 @@ import {
   createRenderer,
   screen,
   act,
+  fireEvent,
 } from 'test/utils';
 import Textarea, { textareaClasses as classes } from '@mui/joy/Textarea';
 import { ThemeProvider } from '@mui/joy/styles';
@@ -34,6 +35,11 @@ describe('Joy <Textarea />', () => {
 
   describeJoyColorInversion(<Textarea />, { muiName: 'JoyTextarea', classes });
 
+  it('should have placeholder', () => {
+    const { getByPlaceholderText } = render(<Textarea placeholder="Placeholder" />);
+    expect(getByPlaceholderText('Placeholder')).toBeVisible();
+  });
+
   it('should have error classes', () => {
     const { container } = render(<Textarea error />);
     expect(container.firstChild).to.have.class(classes.error);
@@ -88,4 +94,54 @@ describe('Joy <Textarea />', () => {
       expect(handleFocus.callCount).to.equal(1);
     });
   });
+
+  describe('slotProps: input', () => {
+    it('`onKeyDown` and `onKeyUp` should work', () => {
+      const handleKeyDown = spy();
+      const handleKeyUp = spy();
+      const { container } = render(
+        <Textarea slotProps={{ textarea: { onKeyDown: handleKeyDown, onKeyUp: handleKeyUp } }} />,
+      );
+
+      act(() => {
+        container.querySelector('textarea')!.focus();
+      });
+      fireEvent.keyDown(container.querySelector('textarea')!);
+      fireEvent.keyUp(container.querySelector('textarea')!);
+
+      expect(handleKeyDown.callCount).to.equal(1);
+      expect(handleKeyUp.callCount).to.equal(1);
+    });
+
+    it('should call focus and blur', () => {
+      const handleBlur = spy();
+      const handleFocus = spy();
+      const { container } = render(
+        <Textarea slotProps={{ textarea: { onFocus: handleFocus, onBlur: handleBlur } }} />,
+      );
+
+      act(() => {
+        container.querySelector('textarea')!.focus();
+      });
+      expect(handleFocus.callCount).to.equal(1);
+      act(() => {
+        container.querySelector('textarea')!.blur();
+      });
+      expect(handleFocus.callCount).to.equal(1);
+    });
+
+    it('should override outer handlers', () => {
+      const handleFocus = spy();
+      const handleSlotFocus = spy();
+      const { container } = render(
+        <Textarea onFocus={handleFocus} slotProps={{ textarea: { onFocus: handleSlotFocus } }} />,
+      );
+
+      act(() => {
+        container.querySelector('textarea')!.focus();
+      });
+      expect(handleFocus.callCount).to.equal(0);
+      expect(handleSlotFocus.callCount).to.equal(1);
+    });
+  });
 });
diff --git a/packages/mui-joy/test/integration/TabsIntegration.test.tsx b/packages/mui-joy/test/integration/TabsIntegration.test.tsx
new file mode 100644
index 0000000000..b99a03e927
--- /dev/null
+++ b/packages/mui-joy/test/integration/TabsIntegration.test.tsx
@@ -0,0 +1,27 @@
+import * as React from 'react';
+import { expect } from 'chai';
+import { createRenderer } from 'test/utils';
+import Tabs from '@mui/joy/Tabs';
+import TabList from '@mui/joy/TabList';
+import Tab from '@mui/joy/Tab';
+import TabPanel from '@mui/joy/TabPanel';
+
+describe('Joy <Tabs />', () => {
+  const { render } = createRenderer();
+
+  it('display first tab', () => {
+    const { getByText, queryByText } = render(
+      <Tabs>
+        <TabList>
+          <Tab>1</Tab>
+          <Tab>2</Tab>
+        </TabList>
+        <TabPanel>Panel 1</TabPanel>
+        <TabPanel value={1}>Panel 2</TabPanel>
+      </Tabs>,
+    );
+
+    expect(getByText('Panel 1')).toBeVisible();
+    expect(queryByText('Panel 2')).to.equal(null);
+  });
+});
