diff --git a/pkg/cluster/fetch_test.go b/pkg/cluster/fetch_test.go
index 1817e8641..769ef5fd2 100644
--- a/pkg/cluster/fetch_test.go
+++ b/pkg/cluster/fetch_test.go
@@ -294,6 +294,14 @@ func wantKubeDistroForEksdRelease() (*eksdv1.Release, *cluster.KubeDistro) {
 				{
 					Name:   "etcd",
 					GitTag: "v3.4.14",
+					Assets: []eksdv1.Asset{
+						{
+							Arch: []string{"amd64"},
+							Archive: &eksdv1.AssetArchive{
+								URI: "https://distro.eks.amazonaws.com/kubernetes-1-19/releases/4/artifacts/etcd/v3.4.14/etcd-linux-amd64-v3.4.14.tar.gz",
+							},
+						},
+					},
 				},
 				{
 					Name: "comp-1",
@@ -405,6 +413,7 @@ func wantKubeDistroForEksdRelease() (*eksdv1.Release, *cluster.KubeDistro) {
 			URI: "public.ecr.aws/eks-distro/kubernetes/kube-proxy:v1.19.8",
 		},
 		EtcdVersion: "3.4.14",
+		EtcdURL:     "https://distro.eks.amazonaws.com/kubernetes-1-19/releases/4/artifacts/etcd/v3.4.14/etcd-linux-amd64-v3.4.14.tar.gz",
 	}
 
 	return eksdRelease, kubeDistro
diff --git a/pkg/clusterapi/apibuilder_test.go b/pkg/clusterapi/apibuilder_test.go
index 82f240a88..72c9fb005 100644
--- a/pkg/clusterapi/apibuilder_test.go
+++ b/pkg/clusterapi/apibuilder_test.go
@@ -120,6 +120,8 @@ func newApiBuilerTest(t *testing.T) apiBuilerTest {
 					EtcdImage: v1alpha1.Image{
 						URI: "public.ecr.aws/eks-distro/etcd-io/etcd:0.0.1",
 					},
+					EtcdURL:     "https://distro.eks.amazonaws.com/kubernetes-1-19/releases/4/artifacts/etcd/v3.4.14/etcd-linux-amd64-v3.4.14.tar.gz",
+					EtcdVersion: "3.4.14",
 					Pause: v1alpha1.Image{
 						URI: "public.ecr.aws/eks-distro/kubernetes/pause:0.0.1",
 					},
diff --git a/pkg/clusterapi/etcd_test.go b/pkg/clusterapi/etcd_test.go
index fc0ed7613..2b7c5dd48 100644
--- a/pkg/clusterapi/etcd_test.go
+++ b/pkg/clusterapi/etcd_test.go
@@ -14,15 +14,36 @@ import (
 
 func TestSetUbuntuConfigInEtcdCluster(t *testing.T) {
 	g := newApiBuilerTest(t)
+	eksaVersion := anywherev1.EksaVersion("v0.19.2")
+	g.clusterSpec.Cluster.Spec.EksaVersion = &eksaVersion
 	got := wantEtcdCluster()
-	v := "0.0.1"
+	versionBundle := g.clusterSpec.VersionsBundles["1.21"]
+
+	want := got.DeepCopy()
+	want.Spec.EtcdadmConfigSpec.Format = etcdbootstrapv1.Format("cloud-config")
+	want.Spec.EtcdadmConfigSpec.CloudInitConfig = &etcdbootstrapv1.CloudInitConfig{
+		Version:        versionBundle.KubeDistro.EtcdVersion,
+		InstallDir:     "/usr/bin",
+		EtcdReleaseURL: versionBundle.KubeDistro.EtcdURL,
+	}
+	clusterapi.SetUbuntuConfigInEtcdCluster(got, versionBundle, string(eksaVersion))
+	g.Expect(got).To(Equal(want))
+}
+
+func TestSetUbuntuConfigInEtcdClusterNoEtcdUrl(t *testing.T) {
+	g := newApiBuilerTest(t)
+	eksaVersion := anywherev1.EksaVersion("v0.18.2")
+	g.clusterSpec.Cluster.Spec.EksaVersion = &eksaVersion
+	got := wantEtcdCluster()
+	versionBundle := g.clusterSpec.VersionsBundles["1.21"]
+
 	want := got.DeepCopy()
 	want.Spec.EtcdadmConfigSpec.Format = etcdbootstrapv1.Format("cloud-config")
 	want.Spec.EtcdadmConfigSpec.CloudInitConfig = &etcdbootstrapv1.CloudInitConfig{
-		Version:    v,
+		Version:    versionBundle.KubeDistro.EtcdVersion,
 		InstallDir: "/usr/bin",
 	}
-	clusterapi.SetUbuntuConfigInEtcdCluster(got, v)
+	clusterapi.SetUbuntuConfigInEtcdCluster(got, versionBundle, string(eksaVersion))
 	g.Expect(got).To(Equal(want))
 }
 
diff --git a/pkg/providers/common/common_test.go b/pkg/providers/common/common_test.go
index 2004bd246..1b337de80 100644
--- a/pkg/providers/common/common_test.go
+++ b/pkg/providers/common/common_test.go
@@ -7,6 +7,7 @@ import (
 	. "github.com/onsi/gomega"
 	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
 
+	"github.com/aws/eks-anywhere/internal/test"
 	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
 	"github.com/aws/eks-anywhere/pkg/providers/common"
 )
@@ -136,3 +137,46 @@ func TestGetCAPIBottlerocketSettingsConfig(t *testing.T) {
 		})
 	}
 }
+
+func TestGetExternalEtcdReleaseURL(t *testing.T) {
+	g := NewWithT(t)
+	testcases := []struct {
+		name           string
+		clusterVersion string
+		etcdURL        string
+		err            error
+	}{
+		{
+			name:           "Pre-etcd url enabled version < 0.19.0",
+			clusterVersion: "v0.15.2",
+			etcdURL:        "",
+			err:            nil,
+		},
+		{
+			name:           "Etcd url enabled version = 0.19.0",
+			clusterVersion: "v0.19.0",
+			etcdURL:        test.VersionBundle().KubeDistro.EtcdURL,
+		},
+		{
+			name:           "Post etcd url enabled version > 0.19.0",
+			clusterVersion: "v0.21.0",
+			etcdURL:        test.VersionBundle().KubeDistro.EtcdURL,
+		},
+		{
+			name:           "Invalid semver for eks-a cluster version",
+			clusterVersion: "a.12.4.3.78",
+			err:            fmt.Errorf("invalid semver for clusterVersion: invalid major version in semver a.12.4.3.78: strconv.ParseInt: parsing \"\": invalid syntax"),
+		},
+	}
+	for _, tt := range testcases {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := common.GetExternalEtcdReleaseURL(tt.clusterVersion, test.VersionBundle())
+			if tt.err == nil {
+				g.Expect(err).ToNot(HaveOccurred())
+			} else {
+				g.Expect(err.Error()).To(Equal(tt.err.Error()))
+			}
+			g.Expect(got).To(Equal(tt.etcdURL))
+		})
+	}
+}
diff --git a/pkg/providers/docker/controlplane_test.go b/pkg/providers/docker/controlplane_test.go
index e379cd09f..7df338f3c 100644
--- a/pkg/providers/docker/controlplane_test.go
+++ b/pkg/providers/docker/controlplane_test.go
@@ -301,6 +301,7 @@ func TestControlPlaneUpgradeRolloutStrategy(t *testing.T) {
 func testClusterSpec(opts ...test.ClusterSpecOpt) *cluster.Spec {
 	name := "test"
 	namespace := "test-namespace"
+	devVersion := test.DevEksaVersion()
 
 	clusterOpts := make([]test.ClusterSpecOpt, 0)
 	clusterOpts = append(clusterOpts, func(s *cluster.Spec) {
@@ -322,6 +323,7 @@ func testClusterSpec(opts ...test.ClusterSpecOpt) *cluster.Spec {
 					Count: 3,
 				},
 				KubernetesVersion: "1.23",
+				EksaVersion:       &devVersion,
 				WorkerNodeGroupConfigurations: []anywherev1.WorkerNodeGroupConfiguration{
 					{
 						Count:           ptr.Int(3),
diff --git a/pkg/providers/docker/docker_test.go b/pkg/providers/docker/docker_test.go
index 78d519931..153c31c7a 100644
--- a/pkg/providers/docker/docker_test.go
+++ b/pkg/providers/docker/docker_test.go
@@ -603,6 +603,7 @@ var versionsBundle = &cluster.VersionsBundle{
 			Tag:        "v3.4.14-eks-1-19-2",
 		},
 		EtcdVersion: "3.4.14",
+		EtcdURL:     "https://distro.eks.amazonaws.com/kubernetes-1-21/releases/4/artifacts/etcd/v3.4.16/etcd-linux-amd64-v3.4.16.tar.gz",
 	},
 	VersionsBundle: &releasev1alpha1.VersionsBundle{
 		EksD: releasev1alpha1.EksDRelease{
diff --git a/pkg/providers/snow/apibuilder_test.go b/pkg/providers/snow/apibuilder_test.go
index 1cb0c731b..6d1bc13bb 100644
--- a/pkg/providers/snow/apibuilder_test.go
+++ b/pkg/providers/snow/apibuilder_test.go
@@ -1074,8 +1074,9 @@ func wantEtcdClusterUbuntu() *etcdv1.EtcdadmCluster {
 	etcd := wantEtcdCluster()
 	etcd.Spec.EtcdadmConfigSpec.Format = etcdbootstrapv1.Format("cloud-config")
 	etcd.Spec.EtcdadmConfigSpec.CloudInitConfig = &etcdbootstrapv1.CloudInitConfig{
-		Version:    "3.4.16",
-		InstallDir: "/usr/bin",
+		Version:        "3.4.16",
+		InstallDir:     "/usr/bin",
+		EtcdReleaseURL: "https://distro.eks.amazonaws.com/kubernetes-1-21/releases/4/artifacts/etcd/v3.4.16/etcd-linux-amd64-v3.4.16.tar.gz",
 	}
 	etcd.Spec.EtcdadmConfigSpec.PreEtcdadmCommands = []string{
 		"/etc/eks/bootstrap.sh",
@@ -1119,6 +1120,8 @@ func wantEtcdClusterBottlerocket() *etcdv1.EtcdadmCluster {
 
 func TestEtcdadmClusterUbuntu(t *testing.T) {
 	tt := newApiBuilerTest(t)
+	eksaVersion := v1alpha1.EksaVersion("v0.19.0")
+	tt.clusterSpec.Cluster.Spec.EksaVersion = &eksaVersion
 	tt.clusterSpec.Cluster.Spec.ExternalEtcdConfiguration = &v1alpha1.ExternalEtcdConfiguration{
 		Count: 3,
 		MachineGroupRef: &v1alpha1.Ref{
diff --git a/pkg/providers/snow/snow_test.go b/pkg/providers/snow/snow_test.go
index 1085d39e4..c481df3fd 100644
--- a/pkg/providers/snow/snow_test.go
+++ b/pkg/providers/snow/snow_test.go
@@ -113,6 +113,7 @@ func givenVersionsBundle() *cluster.VersionsBundle {
 				URI: "public.ecr.aws/eks-distro/kubernetes/pause:0.0.1",
 			},
 			EtcdVersion: "3.4.16",
+			EtcdURL:     "https://distro.eks.amazonaws.com/kubernetes-1-21/releases/4/artifacts/etcd/v3.4.16/etcd-linux-amd64-v3.4.16.tar.gz",
 		},
 		VersionsBundle: &releasev1alpha1.VersionsBundle{
 			KubeVersion: "1.21",
@@ -166,6 +167,7 @@ func givenManagementCluster() *types.Cluster {
 }
 
 func givenClusterConfig() *v1alpha1.Cluster {
+	devVersion := test.DevEksaVersion()
 	return &v1alpha1.Cluster{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "snow-test",
@@ -202,6 +204,7 @@ func givenClusterConfig() *v1alpha1.Cluster {
 				},
 			},
 			KubernetesVersion: "1.21",
+			EksaVersion:       &devVersion,
 			WorkerNodeGroupConfigurations: []v1alpha1.WorkerNodeGroupConfiguration{
 				{
 					Name:  "md-0",
diff --git a/pkg/providers/vsphere/controlplane_test.go b/pkg/providers/vsphere/controlplane_test.go
index 031c361a1..a331690e3 100644
--- a/pkg/providers/vsphere/controlplane_test.go
+++ b/pkg/providers/vsphere/controlplane_test.go
@@ -783,6 +783,7 @@ spec:
     cloudInitConfig:
       version: 3.4.14
       installDir: "/usr/bin"
+      etcdReleaseURL: https://distro.eks.amazonaws.com/kubernetes-1-19/releases/4/artifacts/etcd/v3.4.14/etcd-linux-amd64-v3.4.14.tar.gz
     preEtcdadmCommands:
       - hostname "{{ ds.meta_data.hostname }}"
       - echo "::1         ipv6-localhost ipv6-loopback" >/etc/hosts
