diff --git a/packages/components/Button/src/FAB/FAB.test.tsx b/packages/components/Button/src/FAB/FAB.test.tsx
index f8c7102c5..e8391b241 100644
--- a/packages/components/Button/src/FAB/FAB.test.tsx
+++ b/packages/components/Button/src/FAB/FAB.test.tsx
@@ -15,6 +15,12 @@ it('Default FAB (iOS)', () => {
   expect(tree).toMatchSnapshot();
 });
 
+it('Custom FAB with no shadow(iOS)', () => {
+  const CustomFABNoShadow = FAB.customize({ shadowToken: undefined });
+  const tree = renderer.create(<CustomFABNoShadow>Custom FAB with no shadow(iOS)</CustomFABNoShadow>).toJSON();
+  expect(tree).toMatchSnapshot();
+});
+
 it('Button simple rendering does not invalidate styling', () => {
   checkRenderConsistency(() => <FAB>Default FAB</FAB>, 2);
 });
