diff --git a/e2e/release/src/circular-dependencies.test.ts b/e2e/release/src/circular-dependencies.test.ts
index 225c813550..f6b091334b 100644
--- a/e2e/release/src/circular-dependencies.test.ts
+++ b/e2e/release/src/circular-dependencies.test.ts
@@ -701,11 +701,6 @@ describe('nx release circular dependencies', () => {
 
         Would stage files in git with the following command, but --dry-run was set:
         git add {project-name}/package.json
-
-        NX   Your filter "{project-name}" matched the following projects:
-
-        - {project-name}
-
         Determined workspace --from ref from the first commit in the workspace: {SHASUM}
         Determined --from ref for {project-name} from the first commit in which it exists: {COMMIT_SHA}
 
@@ -722,11 +717,6 @@ describe('nx release circular dependencies', () => {
         Would stage files in git with the following command, but --dry-run was set:
         git add {project-name}/CHANGELOG.md
 
-        NX   Your filter "{project-name}" matched the following projects:
-
-        - {project-name}
-
-
         NX   Committing changes with git
 
         Would commit all previously staged files in git with the following command, but --dry-run was set:
@@ -737,11 +727,6 @@ describe('nx release circular dependencies', () => {
         Would tag the current commit in git with the following command, but --dry-run was set:
         git tag --annotate {project-name}@2.0.0 --message {project-name}@2.0.0
 
-        NX   Your filter "{project-name}" matched the following projects:
-
-        - {project-name}
-
-
         NX   Running target nx-release-publish for project {project-name}:
 
         - {project-name}
@@ -1037,11 +1022,6 @@ describe('nx release circular dependencies', () => {
 
         Would stage files in git with the following command, but --dry-run was set:
         git add {project-name}/package.json {project-name}/package.json
-
-        NX   Your filter "{project-name}" matched the following projects:
-
-        - {project-name}
-
         Determined workspace --from ref from the first commit in the workspace: {SHASUM}
         Determined --from ref for {project-name} from the first commit in which it exists: {COMMIT_SHA}
 
@@ -1071,11 +1051,6 @@ describe('nx release circular dependencies', () => {
         Would stage files in git with the following command, but --dry-run was set:
         git add {project-name}/CHANGELOG.md {project-name}/CHANGELOG.md
 
-        NX   Your filter "{project-name}" matched the following projects:
-
-        - {project-name}
-
-
         NX   Committing changes with git
 
         Would commit all previously staged files in git with the following command, but --dry-run was set:
@@ -1086,11 +1061,6 @@ describe('nx release circular dependencies', () => {
         Would tag the current commit in git with the following command, but --dry-run was set:
         git tag --annotate {project-name}@2.0.0 --message {project-name}@2.0.0
 
-        NX   Your filter "{project-name}" matched the following projects:
-
-        - {project-name}
-
-
         NX   Running target nx-release-publish for project {project-name}:
 
         - {project-name}
diff --git a/e2e/release/src/create-github-release.test.ts b/e2e/release/src/create-github-release.test.ts
index c00c90c652..9ee45a66b8 100644
--- a/e2e/release/src/create-github-release.test.ts
+++ b/e2e/release/src/create-github-release.test.ts
@@ -114,7 +114,8 @@ describe('nx release create github release', () => {
     const result = runCLI('release patch -d --first-release --verbose');
 
     expect(
-      result.match(new RegExp(`NX   Pushing to git remote`, 'g')).length
+      result.match(new RegExp(`NX   Pushing to git remote "origin"`, 'g'))
+        .length
     ).toEqual(1);
     expect(
       result.match(new RegExp(`NX   Creating GitHub Release`, 'g')).length
@@ -148,7 +149,8 @@ describe('nx release create github release', () => {
     const result = runCLI('release -d --first-release --verbose');
 
     expect(
-      result.match(new RegExp(`NX   Pushing to git remote`, 'g')).length
+      result.match(new RegExp(`NX   Pushing to git remote "origin"`, 'g'))
+        .length
     ).toEqual(1);
     expect(
       result.match(new RegExp(`NX   Creating GitHub Release`, 'g')).length
diff --git a/e2e/release/src/release.test.ts b/e2e/release/src/release.test.ts
index 8cb139a05d..32f15e23da 100644
--- a/e2e/release/src/release.test.ts
+++ b/e2e/release/src/release.test.ts
@@ -697,7 +697,7 @@ describe('nx release', () => {
       NX   Tagging commit with git
 
 
-      NX   Pushing to git remote
+      NX   Pushing to git remote "origin"
 
 
       NX   Creating GitHub Release
diff --git a/packages/nx/src/command-line/release/config/config.spec.ts b/packages/nx/src/command-line/release/config/config.spec.ts
index 7c75ef51d7..4216e0349c 100644
--- a/packages/nx/src/command-line/release/config/config.spec.ts
+++ b/packages/nx/src/command-line/release/config/config.spec.ts
@@ -137,6 +137,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -255,6 +256,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -289,6 +291,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -313,6 +316,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -431,6 +435,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -465,6 +470,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -492,6 +498,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -610,6 +617,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -644,6 +652,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -702,6 +711,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -820,6 +830,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -854,6 +865,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -896,6 +908,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -1014,6 +1027,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -1048,6 +1062,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -1099,6 +1114,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -1217,6 +1233,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -1250,6 +1267,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -1300,6 +1318,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -1418,6 +1437,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -1453,6 +1473,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -1486,6 +1507,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -1604,6 +1626,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -1636,6 +1659,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -1667,6 +1691,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -1785,6 +1810,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -1818,6 +1844,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -1861,6 +1888,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -1968,6 +1996,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -2017,6 +2046,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -2049,6 +2079,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -2167,6 +2198,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -2210,6 +2242,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -2247,6 +2280,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -2354,6 +2388,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -2423,6 +2458,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -2457,6 +2493,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -2564,6 +2601,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -2608,6 +2646,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -2642,6 +2681,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -2760,6 +2800,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -2798,6 +2839,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -2828,6 +2870,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "--no-verify",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -2946,6 +2989,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": true,
               "commitArgs": "--no-verify",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -2980,6 +3024,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "--no-verify",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -3013,6 +3058,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -3131,6 +3177,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -3165,6 +3212,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "--no-verify",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": true,
                 "tagArgs": "",
@@ -3194,6 +3242,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -3312,6 +3361,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -3346,6 +3396,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -3380,6 +3431,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -3498,6 +3550,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -3530,6 +3583,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -3579,6 +3633,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -3697,6 +3752,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -3729,6 +3785,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -3758,6 +3815,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -3876,6 +3934,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -3910,6 +3969,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -3949,6 +4009,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -4067,6 +4128,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -4113,6 +4175,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -4145,6 +4208,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -4252,6 +4316,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -4286,6 +4351,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -4326,6 +4392,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -4455,6 +4522,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -4500,6 +4568,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -4530,6 +4599,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -4659,6 +4729,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -4704,6 +4775,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -4734,6 +4806,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": false,
                 "tagArgs": "",
@@ -4852,6 +4925,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -4886,6 +4960,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -4920,6 +4995,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -5042,6 +5118,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -5076,6 +5153,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -5111,6 +5189,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -5233,6 +5312,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -5267,6 +5347,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -5364,6 +5445,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -5486,6 +5568,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -5535,6 +5618,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -5571,6 +5655,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -5693,6 +5778,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -5742,6 +5828,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -5816,6 +5903,134 @@ describe('createNxReleaseConfig()', () => {
         }
       `);
     });
+
+    it('should return an error if createRelease is enabled but git push is explicitly disabled', async () => {
+      const res = await createNxReleaseConfig(projectGraph, projectFileMap, {
+        changelog: {
+          workspaceChangelog: {
+            createRelease: 'github',
+          },
+        },
+        git: {
+          push: false,
+        },
+      });
+      expect(res.error).toMatchInlineSnapshot(`
+        {
+          "code": "GIT_PUSH_FALSE_WITH_CREATE_RELEASE",
+          "data": {},
+        }
+      `);
+
+      const res2 = await createNxReleaseConfig(projectGraph, projectFileMap, {
+        changelog: {
+          git: {
+            push: false,
+          },
+          workspaceChangelog: {
+            createRelease: 'github',
+          },
+        },
+      });
+      expect(res2.error).toMatchInlineSnapshot(`
+        {
+          "code": "GIT_PUSH_FALSE_WITH_CREATE_RELEASE",
+          "data": {},
+        }
+      `);
+
+      const res3 = await createNxReleaseConfig(projectGraph, projectFileMap, {
+        changelog: {
+          projectChangelogs: {
+            createRelease: 'github',
+          },
+        },
+        git: {
+          push: false,
+        },
+      });
+      expect(res3.error).toMatchInlineSnapshot(`
+        {
+          "code": "GIT_PUSH_FALSE_WITH_CREATE_RELEASE",
+          "data": {},
+        }
+      `);
+
+      const res4 = await createNxReleaseConfig(projectGraph, projectFileMap, {
+        changelog: {
+          git: {
+            push: false,
+          },
+          projectChangelogs: {
+            createRelease: 'github',
+          },
+        },
+      });
+      expect(res4.error).toMatchInlineSnapshot(`
+        {
+          "code": "GIT_PUSH_FALSE_WITH_CREATE_RELEASE",
+          "data": {},
+        }
+      `);
+
+      const res5 = await createNxReleaseConfig(projectGraph, projectFileMap, {
+        changelog: {
+          workspaceChangelog: {
+            createRelease: 'github',
+          },
+        },
+        version: {
+          git: {
+            push: false,
+          },
+        },
+      });
+      expect(res5.error).toMatchInlineSnapshot(`
+        {
+          "code": "GIT_PUSH_FALSE_WITH_CREATE_RELEASE",
+          "data": {},
+        }
+      `);
+
+      const res6 = await createNxReleaseConfig(projectGraph, projectFileMap, {
+        changelog: {
+          projectChangelogs: {
+            createRelease: 'github',
+          },
+        },
+        version: {
+          git: {
+            push: false,
+          },
+        },
+      });
+      expect(res6.error).toMatchInlineSnapshot(`
+        {
+          "code": "GIT_PUSH_FALSE_WITH_CREATE_RELEASE",
+          "data": {},
+        }
+      `);
+
+      const res7 = await createNxReleaseConfig(projectGraph, projectFileMap, {
+        groups: {
+          'group-1': {
+            projects: ['lib-a'],
+            changelog: {
+              createRelease: 'github',
+            },
+          },
+        },
+        git: {
+          push: false,
+        },
+      });
+      expect(res7.error).toMatchInlineSnapshot(`
+        {
+          "code": "GIT_PUSH_FALSE_WITH_CREATE_RELEASE",
+          "data": {},
+        }
+      `);
+    });
   });
 
   describe('user config -> top level conventional commits configuration', () => {
@@ -5834,6 +6049,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -5952,6 +6168,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -5986,6 +6203,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -6014,6 +6232,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -6132,6 +6351,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -6166,6 +6386,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -6216,6 +6437,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -6348,6 +6570,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -6382,6 +6605,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -6415,6 +6639,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -6540,6 +6765,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -6574,6 +6800,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -6610,6 +6837,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -6735,6 +6963,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -6769,6 +6998,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -6806,6 +7036,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -6931,6 +7162,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -6965,6 +7197,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -7005,6 +7238,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -7130,6 +7364,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -7164,6 +7399,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -7223,6 +7459,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -7341,6 +7578,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -7429,6 +7667,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -7471,6 +7710,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -7608,6 +7848,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -7655,6 +7896,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -7713,6 +7955,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -7831,6 +8074,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -7878,6 +8122,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -7917,6 +8162,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": true,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -8035,6 +8281,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -8082,6 +8329,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -8354,6 +8602,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -8472,6 +8721,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -8506,6 +8756,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -8728,6 +8979,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -8835,6 +9087,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -8883,6 +9136,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -8912,6 +9166,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -9019,6 +9274,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -9052,6 +9308,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -9086,6 +9343,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -9204,6 +9462,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -9244,6 +9503,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -9274,6 +9534,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -9392,6 +9653,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -9432,6 +9694,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -9461,6 +9724,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -9579,6 +9843,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -9619,6 +9884,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -9656,6 +9922,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -9774,6 +10041,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -9809,6 +10077,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -9841,6 +10110,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -9959,6 +10229,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -10001,6 +10272,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -10073,6 +10345,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -10191,6 +10464,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -10229,6 +10503,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -10261,6 +10536,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -10379,6 +10655,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -10421,6 +10698,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -10466,6 +10744,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -10573,6 +10852,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -10622,6 +10902,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -10663,6 +10944,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -10770,6 +11052,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -10823,6 +11106,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -10864,6 +11148,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -10971,6 +11256,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -11022,6 +11308,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
@@ -11065,6 +11352,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": true,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": false,
                 "tag": true,
                 "tagArgs": "",
@@ -11172,6 +11460,7 @@ describe('createNxReleaseConfig()', () => {
               "commit": false,
               "commitArgs": "",
               "commitMessage": "chore(release): publish {version}",
+              "push": false,
               "stageChanges": false,
               "tag": false,
               "tagArgs": "",
@@ -11227,6 +11516,7 @@ describe('createNxReleaseConfig()', () => {
                 "commit": false,
                 "commitArgs": "",
                 "commitMessage": "chore(release): publish {version}",
+                "push": false,
                 "stageChanges": true,
                 "tag": false,
                 "tagArgs": "",
diff --git a/packages/nx/src/command-line/release/config/filter-release-groups.spec.ts b/packages/nx/src/command-line/release/config/filter-release-groups.spec.ts
index 12d0ec341e..c95f8ac695 100644
--- a/packages/nx/src/command-line/release/config/filter-release-groups.spec.ts
+++ b/packages/nx/src/command-line/release/config/filter-release-groups.spec.ts
@@ -20,6 +20,7 @@ describe('filterReleaseGroups()', () => {
           tagMessage: '',
           tagArgs: '',
           stageChanges: false,
+          push: false,
         },
         workspaceChangelog: false,
         projectChangelogs: false,
@@ -37,6 +38,7 @@ describe('filterReleaseGroups()', () => {
           tag: false,
           tagMessage: '',
           tagArgs: '',
+          push: false,
         },
         preVersionCommand: '',
       },
@@ -49,6 +51,7 @@ describe('filterReleaseGroups()', () => {
         tagMessage: '',
         tagArgs: '',
         stageChanges: false,
+        push: false,
       },
       conventionalCommits: DEFAULT_CONVENTIONAL_COMMITS_CONFIG,
       versionPlans: false,
