diff --git a/api/tests/unit/api/test_unit_api.py b/api/tests/unit/api/test_unit_api.py
index e4959b34d..fdcf58a42 100644
--- a/api/tests/unit/api/test_unit_api.py
+++ b/api/tests/unit/api/test_unit_api.py
@@ -6,9 +6,17 @@ from rest_framework.test import APIClient
 
 
 @pytest.mark.parametrize(
-    "client", (lazy_fixture("api_client"), lazy_fixture("admin_client"))
+    "url",
+    (
+        reverse("api-v1:schema-json", args=[".json"]),
+        reverse("api-v1:schema-json", args=[".yaml"]),
+        reverse("api-v1:schema-swagger-ui"),
+    ),
 )
-def test_swagger_docs_generation(client: APIClient) -> None:
-    url = reverse("api-v1:schema-swagger-ui")
+@pytest.mark.parametrize(
+    "client",
+    (lazy_fixture("api_client"), lazy_fixture("admin_client")),
+)
+def test_swagger_docs_generation(url: str, client: APIClient) -> None:
     response = client.get(url)
     assert response.status_code == status.HTTP_200_OK
diff --git a/api/tests/unit/audit/test_unit_audit_models.py b/api/tests/unit/audit/test_unit_audit_models.py
index a6b49229b..07d6ca67c 100644
--- a/api/tests/unit/audit/test_unit_audit_models.py
+++ b/api/tests/unit/audit/test_unit_audit_models.py
@@ -1,3 +1,4 @@
+import pytest
 from pytest_mock import MockerFixture
 
 from audit.models import AuditLog
@@ -228,3 +229,15 @@ def test_creating_audit_logs_for_change_request_does_not_trigger_process_environ
     # Then
     process_environment_update.delay.assert_not_called()
     assert audit_log.created_date != environment.updated_at
+
+
+@pytest.mark.django_db
+def test_audit_log__organisation__empty_instance__return_expected() -> None:
+    # Given
+    audit_log = AuditLog.objects.create()
+
+    # When
+    organisation = audit_log.organisation
+
+    # Then
+    assert organisation is None
diff --git a/api/tests/unit/audit/test_unit_audit_signals.py b/api/tests/unit/audit/test_unit_audit_signals.py
index 6265276f2..2b7bd6c55 100644
--- a/api/tests/unit/audit/test_unit_audit_signals.py
+++ b/api/tests/unit/audit/test_unit_audit_signals.py
@@ -3,8 +3,17 @@ from pytest_mock import MockerFixture
 
 from audit.models import AuditLog
 from audit.related_object_type import RelatedObjectType
-from audit.signals import call_webhooks, send_audit_log_event_to_grafana
-from integrations.grafana.models import GrafanaConfiguration
+from audit.signals import (
+    call_webhooks,
+    send_audit_log_event_to_dynatrace,
+    send_audit_log_event_to_grafana,
+)
+from environments.models import Environment
+from integrations.dynatrace.models import DynatraceConfiguration
+from integrations.grafana.models import (
+    GrafanaOrganisationConfiguration,
+    GrafanaProjectConfiguration,
+)
 from organisations.models import Organisation, OrganisationWebhook
 from projects.models import Project
 from webhooks.webhooks import WebhookEventType
@@ -90,7 +99,7 @@ def test_send_audit_log_event_to_grafana__project_grafana_config__calls_expected
     project: Project,
 ) -> None:
     # Given
-    grafana_config = GrafanaConfiguration(base_url="test.com", api_key="test")
+    grafana_config = GrafanaProjectConfiguration(base_url="test.com", api_key="test")
     project.grafana_config = grafana_config
     audit_log_record = AuditLog.objects.create(
         project=project,
@@ -113,3 +122,71 @@ def test_send_audit_log_event_to_grafana__project_grafana_config__calls_expected
     grafana_wrapper_instance_mock.track_event_async.assert_called_once_with(
         event=grafana_wrapper_instance_mock.generate_event_data.return_value
     )
+
+
+def test_send_audit_log_event_to_grafana__organisation_grafana_config__calls_expected(
+    mocker: MockerFixture,
+    organisation: Organisation,
+    project: Project,
+) -> None:
+    # Given
+    grafana_config = GrafanaOrganisationConfiguration(
+        base_url="test.com", api_key="test"
+    )
+    organisation.grafana_config = grafana_config
+    audit_log_record = AuditLog.objects.create(
+        project=project,
+        related_object_type=RelatedObjectType.FEATURE.name,
+    )
+    grafana_wrapper_mock = mocker.patch("audit.signals.GrafanaWrapper", autospec=True)
+    grafana_wrapper_instance_mock = grafana_wrapper_mock.return_value
+
+    # When
+    send_audit_log_event_to_grafana(AuditLog, audit_log_record)
+
+    # Then
+    grafana_wrapper_mock.assert_called_once_with(
+        base_url=grafana_config.base_url,
+        api_key=grafana_config.api_key,
+    )
+    grafana_wrapper_instance_mock.generate_event_data.assert_called_once_with(
+        audit_log_record
+    )
+    grafana_wrapper_instance_mock.track_event_async.assert_called_once_with(
+        event=grafana_wrapper_instance_mock.generate_event_data.return_value
+    )
+
+
+def test_send_audit_log_event_to_dynatrace__environment_dynatrace_config__calls_expected(
+    mocker: MockerFixture,
+    environment: Environment,
+) -> None:
+    # Given
+    dynatrace_config = DynatraceConfiguration.objects.create(
+        base_url="http://test.com", api_key="api_123", environment=environment
+    )
+    environment.refresh_from_db()
+    audit_log_record = AuditLog.objects.create(
+        environment=environment,
+        related_object_type=RelatedObjectType.FEATURE.name,
+    )
+    dynatrace_wrapper_mock = mocker.patch(
+        "audit.signals.DynatraceWrapper", autospec=True
+    )
+    dynatrace_wrapper_instance_mock = dynatrace_wrapper_mock.return_value
+
+    # When
+    send_audit_log_event_to_dynatrace(AuditLog, audit_log_record)
+
+    # Then
+    dynatrace_wrapper_mock.assert_called_once_with(
+        base_url=dynatrace_config.base_url,
+        api_key=dynatrace_config.api_key,
+        entity_selector=dynatrace_config.entity_selector,
+    )
+    dynatrace_wrapper_instance_mock.generate_event_data.assert_called_once_with(
+        audit_log_record
+    )
+    dynatrace_wrapper_instance_mock.track_event_async.assert_called_once_with(
+        event=dynatrace_wrapper_instance_mock.generate_event_data.return_value
+    )
diff --git a/api/tests/unit/integrations/grafana/test_views.py b/api/tests/unit/integrations/grafana/test_views.py
new file mode 100644
index 000000000..aa68e0f33
--- /dev/null
+++ b/api/tests/unit/integrations/grafana/test_views.py
@@ -0,0 +1,194 @@
+import json
+
+import pytest
+from django.urls import reverse
+from rest_framework import status
+from rest_framework.test import APIClient
+
+from integrations.grafana.models import GrafanaOrganisationConfiguration
+from organisations.models import Organisation
+
+
+@pytest.fixture
+def grafana_organisation_configuration(
+    organisation: Organisation,
+) -> GrafanaOrganisationConfiguration:
+    return GrafanaOrganisationConfiguration.objects.create(
+        organisation=organisation,
+        base_url="http://test.com",
+        api_key="abc-123",
+    )
+
+
+def test_grafana_organisation_view__create_configuration__persist_expected(
+    admin_client_new: APIClient,
+    organisation: Organisation,
+) -> None:
+    # Given
+    data = {
+        "base_url": "http://test.com",
+        "api_key": "abc-123",
+    }
+    url = reverse(
+        "api-v1:organisations:integrations-grafana-list", args=[organisation.id]
+    )
+
+    # When
+    response = admin_client_new.post(
+        url,
+        data=json.dumps(data),
+        content_type="application/json",
+    )
+
+    # Then
+    assert response.status_code == status.HTTP_201_CREATED
+    assert (
+        GrafanaOrganisationConfiguration.objects.filter(
+            organisation=organisation
+        ).count()
+        == 1
+    )
+
+    created_config = GrafanaOrganisationConfiguration.objects.filter(
+        organisation=organisation
+    ).first()
+    assert created_config.base_url == data["base_url"]
+    assert created_config.api_key == data["api_key"]
+
+
+def test_grafana_organisation_view__create_configuration__existing__return_expected(
+    admin_client_new: APIClient,
+    organisation: Organisation,
+    grafana_organisation_configuration: GrafanaOrganisationConfiguration,
+) -> None:
+    # Given
+    data = {
+        "base_url": "http://test.com",
+        "api_key": "abc-123",
+    }
+    url = reverse(
+        "api-v1:organisations:integrations-grafana-list", args=[organisation.id]
+    )
+
+    # When
+    response = admin_client_new.post(
+        url,
+        data=json.dumps(data),
+        content_type="application/json",
+    )
+
+    # Then
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+
+def test_grafana_organisation_view__update_configuration__persist_expected(
+    admin_client_new: APIClient,
+    organisation: Organisation,
+    grafana_organisation_configuration: GrafanaOrganisationConfiguration,
+) -> None:
+    # Given
+    data = {
+        "base_url": "http://updated.test.com",
+        "api_key": "updated-abc-123",
+    }
+    url = reverse(
+        "api-v1:organisations:integrations-grafana-detail",
+        args=[organisation.id, grafana_organisation_configuration.id],
+    )
+
+    # When
+    response = admin_client_new.put(
+        url,
+        data=json.dumps(data),
+        content_type="application/json",
+    )
+
+    # Then
+    assert response.status_code == status.HTTP_200_OK
+    updated_config = GrafanaOrganisationConfiguration.objects.filter(
+        organisation=organisation
+    ).first()
+    assert updated_config.base_url == data["base_url"]
+    assert updated_config.api_key == data["api_key"]
+
+
+def test_grafana_organisation_view__delete_configuration__return_expected(
+    admin_client_new: APIClient,
+    organisation: Organisation,
+    grafana_organisation_configuration: GrafanaOrganisationConfiguration,
+) -> None:
+    # Given
+    url = reverse(
+        "api-v1:organisations:integrations-grafana-detail",
+        args=[organisation.id, grafana_organisation_configuration.id],
+    )
+
+    # When
+    response = admin_client_new.delete(url)
+
+    # Then
+    assert response.status_code == status.HTTP_204_NO_CONTENT
+    assert not GrafanaOrganisationConfiguration.objects.filter(
+        organisation=organisation
+    ).exists()
+
+
+def test_grafana_organisation_view__create_configuration__non_admin__return_expected(
+    test_user_client: APIClient,
+    organisation: Organisation,
+    grafana_organisation_configuration: GrafanaOrganisationConfiguration,
+) -> None:
+    # Given
+    data = {
+        "base_url": "http://test.com",
+        "api_key": "abc-123",
+    }
+    url = reverse(
+        "api-v1:organisations:integrations-grafana-list", args=[organisation.id]
+    )
+
+    # When
+    response = test_user_client.post(
+        url,
+        data=json.dumps(data),
+        content_type="application/json",
+    )
+
+    # Then
+    assert response.status_code == status.HTTP_403_FORBIDDEN
+
+
+def test_grafana_organisation_view__get_configuration__non_admin__return_expected(
+    test_user_client: APIClient,
+    organisation: Organisation,
+    grafana_organisation_configuration: GrafanaOrganisationConfiguration,
+) -> None:
+    # Given
+    url = reverse(
+        "api-v1:organisations:integrations-grafana-detail",
+        args=[organisation.id, grafana_organisation_configuration.id],
+    )
+
+    # When
+    response = test_user_client.get(url)
+
+    # Then
+    assert response.status_code == status.HTTP_403_FORBIDDEN
+
+
+def test_grafana_organisation_view__delete_configuration__non_admin__return_expected(
+    test_user_client: APIClient,
+    organisation: Organisation,
+    grafana_organisation_configuration: GrafanaOrganisationConfiguration,
+) -> None:
+    # Given
+    url = reverse(
+        "api-v1:organisations:integrations-grafana-detail",
+        args=[organisation.id, grafana_organisation_configuration.id],
+    )
+
+    # When
+    response = test_user_client.delete(url)
+
+    # Then
+    assert response.status_code == status.HTTP_403_FORBIDDEN
