diff --git a/packages/gasket-plugin-command/test/configure.test.js b/packages/gasket-plugin-command/test/configure.test.js
index 019a4fea..fa217185 100644
--- a/packages/gasket-plugin-command/test/configure.test.js
+++ b/packages/gasket-plugin-command/test/configure.test.js
@@ -29,27 +29,30 @@ describe('configure', () => {
         env: 'development'
       }
     };
-    mockConfig = {};
+    mockConfig = {
+      commands: {
+        test: {
+          extra: 'test-only'
+        }
+      }
+    };
   });
 
   it('should be a function', () => {
     expect(configure).toEqual(expect.any(Function));
   });
 
-  it('should not exec commands if not a gasket command', () => {
-    configure(mockGasket, mockConfig);
-    expect(mockGasket.execSync).not.toHaveBeenCalled();
-  });
-
-  it('should execute on gasket command', () => {
-    process.argv = ['node', '/path/to/gasket.js'];
-    configure(mockGasket, mockConfig);
-    expect(mockGasket.execSync).toHaveBeenCalled();
+  it('adds command id to config if gasket command', () => {
+    process.argv = ['node', '/path/to/gasket.js', 'test'];
+    const result = configure(mockGasket, mockConfig);
+    expect(result).toEqual(expect.objectContaining({ command: 'test' }));
   });
 
-  it('should add commands to gasketBin', () => {
-    process.argv = ['node', '/path/to/gasket.js'];
-    configure(mockGasket, mockConfig);
-    expect(mockAddCommand).toHaveBeenCalled();
+  it('applies command overrides', () => {
+    process.argv = ['node', '/path/to/gasket.js', 'test'];
+    expect(mockConfig).toHaveProperty('commands');
+    const result = configure(mockGasket, mockConfig);
+    expect(result).toEqual(expect.objectContaining({ extra: 'test-only' }));
+    expect(result).not.toHaveProperty('commands');
   });
 });
diff --git a/packages/gasket-plugin-command/test/index.test.js b/packages/gasket-plugin-command/test/index.test.js
index 1de7443b..b1d6bed9 100644
--- a/packages/gasket-plugin-command/test/index.test.js
+++ b/packages/gasket-plugin-command/test/index.test.js
@@ -1,7 +1,4 @@
 import plugin from '../lib/index.js';
-import create from '../lib/create.js';
-import ready from '../lib/ready.js';
-import commands from '../lib/commands.js';
 import { createRequire } from 'module';
 
 const require = createRequire(import.meta.url);
@@ -19,23 +16,18 @@ describe('@gasket/plugin-command', () => {
     expect(plugin).toHaveProperty('hooks');
   });
 
-  it('should have a create hook', () => {
-    expect(plugin.hooks.create).toBe(create);
-    expect(plugin.hooks.create).toEqual(expect.any(Function));
-  });
-
-  it('should have a ready hook', () => {
-    expect(plugin.hooks.ready).toBe(ready);
-    expect(plugin.hooks.ready).toEqual(expect.any(Function));
-  });
-
-  it('should have a commands hook', () => {
-    expect(plugin.hooks.commands).toBe(commands);
-    expect(plugin.hooks.commands).toEqual(expect.any(Function));
-  });
-
-  it('should include a metadata hook', () => {
-    expect(plugin.hooks.metadata).toEqual(expect.any(Function));
+  it('should have expected hooks', () => {
+    const expectedHooks = [
+      'create',
+      'configure',
+      'prepare',
+      'commands',
+      'ready',
+      'metadata'
+    ];
+    expect(Object.keys(plugin.hooks)).toEqual(
+      expect.arrayContaining(expectedHooks)
+    );
   });
 
   it('should return metadata from the metadata hook', () => {
@@ -60,7 +52,7 @@ describe('@gasket/plugin-command', () => {
             method: 'exec',
             description: 'Add custom commands to the CLI',
             link: 'README.md#commands',
-            parent: 'ready'
+            parent: 'prepare'
           },
           {
             name: 'build',
@@ -73,11 +65,4 @@ describe('@gasket/plugin-command', () => {
       })
     );
   });
-
-  it('should have expected hooks', () => {
-    const expectedHooks = ['create', 'ready', 'commands', 'metadata'];
-    expect(Object.keys(plugin.hooks)).toEqual(
-      expect.arrayContaining(expectedHooks)
-    );
-  });
 });
diff --git a/packages/gasket-plugin-command/test/prepare.test.js b/packages/gasket-plugin-command/test/prepare.test.js
new file mode 100644
index 00000000..10d46f21
--- /dev/null
+++ b/packages/gasket-plugin-command/test/prepare.test.js
@@ -0,0 +1,62 @@
+/* eslint-disable no-sync */
+import { jest } from '@jest/globals';
+
+const mockAddCommand = jest.fn();
+const mockParse = jest.fn();
+const mockProcessCommand = jest.fn();
+
+jest.unstable_mockModule('../lib/cli.js', () => {
+  return {
+    gasketBin: {
+      addCommand: mockAddCommand,
+      parse: mockParse
+    }
+  };
+});
+
+jest.unstable_mockModule('../lib/utils/process-command.js', () => {
+  return {
+    processCommand: mockProcessCommand.mockReturnValue({ command: 'test', hidden: false, isDefault: false })
+  };
+});
+
+const prepare = ((await import('../lib/prepare.js')).default);
+
+describe('prepare', () => {
+  let mockGasket, mockConfig;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    mockGasket = {
+      execSync: jest.fn().mockReturnValue([{ id: 'test', description: 'test', action: jest.fn() }]),
+      config: {
+        env: 'development'
+      }
+    };
+    mockConfig = {
+      command: 'test'
+    };
+  });
+
+  it('should be function', () => {
+    expect(prepare).toEqual(expect.any(Function));
+  });
+
+  it('should not exec commands if no gasket command detected', async () => {
+    delete mockConfig.command;
+    await prepare(mockGasket, mockConfig);
+    expect(mockGasket.execSync).not.toHaveBeenCalled();
+  });
+
+  it('should execute command lifecycle', async () => {
+    process.argv = ['node', '/path/to/gasket.js'];
+    await prepare(mockGasket, mockConfig);
+    expect(mockGasket.execSync).toHaveBeenCalledWith('commands');
+  });
+
+  it('should add commands to gasketBin', async () => {
+    process.argv = ['node', '/path/to/gasket.js', 'bogus'];
+    await prepare(mockGasket, mockConfig);
+    expect(mockAddCommand).toHaveBeenCalledWith('test', expect.any(Object));
+  });
+});
diff --git a/packages/gasket-plugin-dynamic-plugins/test/__mocks__/plugin-custom.js b/packages/gasket-plugin-dynamic-plugins/test/__mocks__/plugin-custom.js
new file mode 100644
index 00000000..e69de29b
diff --git a/packages/gasket-plugin-dynamic-plugins/test/prepare-hook.test.js b/packages/gasket-plugin-dynamic-plugins/test/prepare-hook.test.js
deleted file mode 100644
index 7bc6132d..00000000
--- a/packages/gasket-plugin-dynamic-plugins/test/prepare-hook.test.js
+++ /dev/null
@@ -1,92 +0,0 @@
-import { jest } from '@jest/globals';
-import { makeGasket } from '@gasket/core';
-
-const mockPluginOne = { default: {
-  name: '@gasket/plugin-one',
-  hooks: {
-    hook: () => {}
-  }
-} };
-const mockPluginTwo = { default: {
-  name: '@gasket/plugin-two',
-  hooks: {
-    hook: () => {}
-  }
-} };
-
-jest.unstable_mockModule('@gasket/plugin-one', () => mockPluginOne);
-jest.unstable_mockModule('@gasket/plugin-two', () => mockPluginTwo);
-
-const pluginDynamicPlugins = (await import('../lib/index.js')).default;
-
-describe('Prepare Hook', () => {
-  let registerPluginsSpy, execApplySyncSpy, generateGasket, mockConfig;
-
-  beforeEach(() => {
-    generateGasket = function (gasketEnv) {
-      process.env.GASKET_ENV = gasketEnv;
-      const gasket = makeGasket({
-        plugins: [pluginDynamicPlugins],
-        environments: {
-          'local.nothing': {},
-          'local.both': {
-            dynamicPlugins: [
-              '@gasket/plugin-one',
-              '@gasket/plugin-two'
-            ]
-          },
-          'local.two': {
-            dynamicPlugins: [
-              '@gasket/plugin-two'
-            ]
-          }
-        }
-      });
-      mockConfig = gasket.config;
-      registerPluginsSpy = jest.spyOn(gasket.engine, 'registerPlugins');
-      execApplySyncSpy = jest.spyOn(gasket, 'execApplySync');
-      return gasket;
-    };
-
-  });
-
-  afterEach(() => {
-    jest.clearAllMocks();
-  });
-
-  it('does not register dynamic plugins if GASKET_ENV is not set', async () => {
-    const gasket = generateGasket();
-    await pluginDynamicPlugins.hooks.prepare(gasket, mockConfig);
-    expect(registerPluginsSpy).not.toHaveBeenCalled();
-    expect(execApplySyncSpy).not.toHaveBeenCalled();
-  });
-
-  it('does not add plugins if dynamicPlugins is not set on a GASKET_ENV', async () => {
-    jest.clearAllMocks();
-    const gasket = generateGasket('local.nothing');
-    await pluginDynamicPlugins.hooks.prepare(gasket, mockConfig);
-    expect(registerPluginsSpy).not.toHaveBeenCalled();
-    expect(execApplySyncSpy).not.toHaveBeenCalled();
-  });
-
-  it('registers dynamic plugins and calls hooks for a specific GASKET_ENV', async () => {
-    jest.clearAllMocks();
-    const gasket = generateGasket('local.both');
-    await pluginDynamicPlugins.hooks.prepare(gasket, mockConfig);
-    expect(gasket.engine.registerPlugins).toHaveBeenCalledWith(expect.arrayContaining([mockPluginOne.default, mockPluginTwo.default]));
-    expect(gasket.config.plugins).toContain(mockPluginOne.default);
-    expect(gasket.config.plugins).toContain(mockPluginTwo.default);
-    expect(execApplySyncSpy).toHaveBeenNthCalledWith(1, 'init', expect.any(Function));
-    expect(execApplySyncSpy).toHaveBeenNthCalledWith(2, 'configure', expect.any(Function));
-  });
-
-  it('registers dynamic plugins and calls hooks for another GASKET_ENV', async () => {
-    jest.clearAllMocks();
-    const gasket = generateGasket('local.two');
-    await pluginDynamicPlugins.hooks.prepare(gasket, mockConfig);
-    expect(gasket.engine.registerPlugins).toHaveBeenCalledWith(expect.arrayContaining([mockPluginTwo.default]));
-    expect(gasket.config.plugins).toContain(mockPluginTwo.default);
-    expect(execApplySyncSpy).toHaveBeenNthCalledWith(1, 'init', expect.any(Function));
-    expect(execApplySyncSpy).toHaveBeenNthCalledWith(2, 'configure', expect.any(Function));
-  });
-});
diff --git a/packages/gasket-plugin-dynamic-plugins/test/prepare.test.js b/packages/gasket-plugin-dynamic-plugins/test/prepare.test.js
new file mode 100644
index 00000000..cdb61015
--- /dev/null
+++ b/packages/gasket-plugin-dynamic-plugins/test/prepare.test.js
@@ -0,0 +1,156 @@
+import { jest } from '@jest/globals';
+import { makeGasket } from '@gasket/core';
+import * as path from 'path';
+import { setTimeout } from 'timers/promises';
+
+const __dirname = new URL('.', import.meta.url).pathname;
+const mockDir = path.join(__dirname, '__mocks__');
+
+const mockPluginStatic = {
+  default: {
+    name: '@gasket/plugin-static',
+    hooks: {
+      init: jest.fn(),
+      configure: jest.fn((g, c) => c),
+      prepare: jest.fn((g, c) => c)
+    }
+  }
+};
+const mockPluginOne = {
+  default: {
+    name: '@gasket/plugin-one',
+    hooks: {
+      hook: () => {
+      }
+    }
+  }
+};
+const mockPluginTwo = {
+  default: {
+    name: '@gasket/plugin-two',
+    hooks: {
+      hook: () => {
+      }
+    }
+  }
+};
+const mockPluginCustom = {
+  default: {
+    name: 'plugin-custom',
+    hooks: {
+      configure: jest.fn((g, c) => c)
+    }
+  }
+};
+
+jest.unstable_mockModule('@gasket/plugin-one', () => mockPluginOne);
+jest.unstable_mockModule('@gasket/plugin-two', () => mockPluginTwo);
+jest.unstable_mockModule(path.join(mockDir, 'plugin-custom.js'), () => mockPluginCustom);
+
+const pluginDynamicPlugins = (await import('../lib/index.js')).default;
+
+describe('Prepare Hook', () => {
+  let registerPluginsSpy, execApplySyncSpy, execApplySpy, traceSpy, generateGasket, mockConfig;
+  let execApplySyncHandler;
+
+  beforeEach(() => {
+    generateGasket = function (gasketEnv) {
+      // eslint-disable-next-line no-process-env
+      process.env.GASKET_ENV = gasketEnv;
+      const gasket = makeGasket({
+        root: mockDir,
+        plugins: [mockPluginStatic],
+        environments: {
+          'local.nothing': {},
+          'local.both': {
+            dynamicPlugins: [
+              '@gasket/plugin-one',
+              '@gasket/plugin-two'
+            ]
+          },
+          'local.two': {
+            dynamicPlugins: [
+              '@gasket/plugin-two',
+              null,
+              undefined // eslint-disable-line no-undefined
+            ]
+          },
+          'local.custom': {
+            dynamicPlugins: [
+              './plugin-custom.js'
+            ]
+          }
+        }
+      });
+      mockConfig = gasket.config;
+      registerPluginsSpy = jest.spyOn(gasket.engine, 'registerPlugins');
+      execApplySyncHandler = jest.fn();
+      execApplySyncSpy = jest.spyOn(gasket, 'execApplySync').mockImplementation(execApplySyncHandler);
+      execApplySpy = jest.spyOn(gasket, 'execApply');
+      traceSpy = gasket.trace = jest.fn().mockImplementation(() => {
+      });
+      return gasket;
+    };
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('does not add plugins if dynamicPlugins is not configured', async () => {
+    const gasket = generateGasket('local.nothing');
+    await pluginDynamicPlugins.hooks.prepare.handler(gasket, mockConfig);
+
+    expect(registerPluginsSpy).not.toHaveBeenCalled();
+    expect(execApplySyncSpy).not.toHaveBeenCalled();
+  });
+
+  it('registers multiple dynamic plugins and calls hooks if configured', async () => {
+    const gasket = generateGasket('local.both');
+    await pluginDynamicPlugins.hooks.prepare.handler(gasket, mockConfig);
+
+    expect(gasket.engine.registerPlugins).toHaveBeenCalledWith(
+      expect.arrayContaining([mockPluginOne.default, mockPluginTwo.default])
+    );
+    expect(gasket.config.plugins).toContain(mockPluginOne.default);
+    expect(gasket.config.plugins).toContain(mockPluginTwo.default);
+    expect(execApplySyncSpy).toHaveBeenNthCalledWith(1, 'init', expect.any(Function));
+    expect(execApplySyncSpy).toHaveBeenNthCalledWith(2, 'configure', expect.any(Function));
+    expect(execApplySpy).toHaveBeenNthCalledWith(1, 'prepare', expect.any(Function));
+  });
+
+  it('registers dynamic plugins ignoring falsy', async () => {
+    const gasket = generateGasket('local.two');
+    await pluginDynamicPlugins.hooks.prepare.handler(gasket, mockConfig);
+    expect(gasket.engine.registerPlugins).toHaveBeenCalledWith(expect.arrayContaining([mockPluginTwo.default]));
+    expect(gasket.config.plugins).toContain(mockPluginTwo.default);
+    expect(execApplySyncSpy).toHaveBeenNthCalledWith(1, 'init', expect.any(Function));
+    expect(execApplySyncSpy).toHaveBeenNthCalledWith(2, 'configure', expect.any(Function));
+    expect(execApplySpy).toHaveBeenNthCalledWith(1, 'prepare', expect.any(Function));
+  });
+
+  it('ignores setup lifecycles for non-dynamic plugins', async () => {
+    expect(mockPluginStatic.default.hooks.init).not.toHaveBeenCalled();
+    expect(mockPluginStatic.default.hooks.configure).not.toHaveBeenCalled();
+    expect(mockPluginStatic.default.hooks.prepare).not.toHaveBeenCalled();
+
+    const gasket = generateGasket('local.both');
+    await setTimeout(100);
+
+    expect(mockPluginStatic.default.hooks.init).toHaveBeenCalledTimes(1);
+    expect(mockPluginStatic.default.hooks.configure).toHaveBeenCalledTimes(1);
+    expect(mockPluginStatic.default.hooks.prepare).toHaveBeenCalledTimes(1);
+
+    await pluginDynamicPlugins.hooks.prepare.handler(gasket, mockConfig);
+
+    expect(traceSpy).toHaveBeenCalledWith(expect.stringContaining('deduped'));
+    expect(traceSpy).toHaveBeenCalledTimes(1);
+  });
+
+  it('loads local app plugins from relative path', async () => {
+    const gasket = generateGasket('local.custom');
+    await pluginDynamicPlugins.hooks.prepare.handler(gasket, mockConfig);
+
+    expect(gasket.engine.registerPlugins).toHaveBeenCalledWith(expect.arrayContaining([mockPluginCustom.default]));
+  });
+});
