diff --git a/mirrord/agent/iptables/src/error.rs b/mirrord/agent/iptables/src/error.rs
--- a/mirrord/agent/iptables/src/error.rs
+++ b/mirrord/agent/iptables/src/error.rs
@@ -40,3 +40,11 @@ impl Error for IPTablesError {
 }

 pub type IPTablesResult<T, E = IPTablesError> = Result<T, E>;
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn test_exclude_from_mesh_defined() {
+        let _ = crate::IPTABLE_EXCLUDE_FROM_MESH;
+    }
+}
