diff --git a/analysis/analysis_test.go b/analysis/analysis_test.go
index cec3446b8..9e7d3fe5f 100644
--- a/analysis/analysis_test.go
+++ b/analysis/analysis_test.go
@@ -3,6 +3,7 @@ package analysis
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"strings"
 	"testing"
@@ -14,12 +15,18 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8sschema "k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	k8stesting "k8s.io/client-go/testing"
 	"k8s.io/utils/pointer"
 
 	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
 	"github.com/argoproj/argo-rollouts/utils/defaults"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
 )
 
 func timePtr(t metav1.Time) *metav1.Time {
@@ -1985,3 +1992,412 @@ func TestInvalidMeasurementsRetentionConfigThrowsError(t *testing.T) {
 	assert.Equal(t, v1alpha1.AnalysisPhaseError, newRun.Status.Phase)
 	assert.Equal(t, "Analysis spec invalid: measurementRetention[0]: Rule didn't match any metric name(s)", newRun.Status.Message)
 }
+
+func TestExceededTtlChecked(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+
+	testTTLStrategy := func(
+		t *testing.T,
+		ttlStrategy *v1alpha1.TTLStrategy,
+		expiredStatus *v1alpha1.AnalysisRunStatus,
+		notExpiredStatus *v1alpha1.AnalysisRunStatus) {
+		testId := string(uuid.NewUUID())
+		ttlExpiredRun := &v1alpha1.AnalysisRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "expired-run" + testId,
+				Namespace: metav1.NamespaceDefault,
+			},
+			Spec: v1alpha1.AnalysisRunSpec{
+				TTLStrategy: ttlStrategy,
+			},
+			Status: *expiredStatus,
+		}
+		_ = c.reconcileAnalysisRun(ttlExpiredRun)
+		if notExpiredStatus != nil {
+			ttlNotExpiredRun := &v1alpha1.AnalysisRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "not-expired-run" + testId,
+					Namespace: metav1.NamespaceDefault,
+				},
+				Spec: v1alpha1.AnalysisRunSpec{
+					TTLStrategy: ttlStrategy,
+				},
+				Status: *notExpiredStatus,
+			}
+			_ = c.reconcileAnalysisRun(ttlNotExpiredRun)
+		}
+
+		pi := f.expectDeleteAnalysisRunAction(ttlExpiredRun)
+		assert.Equal(t, fmt.Sprintf("%s/%s", metav1.NamespaceDefault, "expired-run"+testId), f.getDeletedAnalysisRunNamespaceAndName(pi))
+		// Nothing else is deleted
+		assert.Equal(t, 1, len(filterInformerActions(f.client.Actions())))
+		// Clear actions to avoid affecting other test instances.
+		f.client.ClearActions()
+		f.actions = nil
+	}
+
+	ttlNotExpiredCompletedTime := f.now.Add(-86400 * time.Second)
+	ttlExpiredCompletedTime := ttlNotExpiredCompletedTime.Add(-1 * time.Second)
+	secondsOfOneDay := int32(86400)
+
+	// Test completed TTL.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterCompletion: &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	})
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterCompletion: &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	})
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterCompletion: &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseError,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseError,
+	})
+	// Test successful TTL.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterSuccess: &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	})
+	// Test failed TTL.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterFailure: &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	})
+
+	// Test success TTL does not affect failed run.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterSuccess: &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	})
+	// Test failed TTL does not affect successful run.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterFailure: &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	})
+	// Test success TTL overrides completed TTL.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterCompletion: pointer.Int32Ptr(100000),
+		SecondsAfterSuccess:    &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	})
+	// Test failed TTL overrides completed TTL.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterCompletion: pointer.Int32Ptr(100000),
+		SecondsAfterFailure:    &secondsOfOneDay,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	})
+	// Test completed TTL still evaluated when non-matching overrides exist.
+	testTTLStrategy(t, &v1alpha1.TTLStrategy{
+		SecondsAfterCompletion: &secondsOfOneDay,
+		SecondsAfterFailure:    pointer.Int32Ptr(86401),
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseSuccessful,
+	}, &v1alpha1.AnalysisRunStatus{
+		CompletedAt: timePtr(metav1.NewTime(ttlNotExpiredCompletedTime)),
+		Phase:       v1alpha1.AnalysisPhaseFailed,
+	})
+}
+
+func TestTtlNotGCInProgressAnalysisRun(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+
+	expectedCount := intstr.FromInt(3)
+	origRun := &v1alpha1.AnalysisRun{
+		Spec: v1alpha1.AnalysisRunSpec{
+			Metrics: []v1alpha1.Metric{
+				{
+					Name:     "metric1",
+					Interval: "60s",
+					Count:    &expectedCount,
+					Provider: v1alpha1.MetricProvider{
+						Job: &v1alpha1.JobMetric{},
+					},
+				},
+			},
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase:     v1alpha1.AnalysisPhaseRunning,
+			StartedAt: timePtr(metav1.NewTime(time.Now())),
+			MetricResults: []v1alpha1.MetricResult{
+				{
+					Name:  "metric1",
+					Phase: v1alpha1.AnalysisPhaseRunning,
+					Count: 1,
+					Measurements: []v1alpha1.Measurement{{
+						Value:      "1",
+						Phase:      v1alpha1.AnalysisPhaseSuccessful,
+						StartedAt:  timePtr(metav1.NewTime(time.Now().Add(-60 * time.Second))),
+						FinishedAt: timePtr(metav1.NewTime(time.Now().Add(-60 * time.Second))),
+					}},
+				},
+			},
+		},
+	}
+	newRun := c.reconcileAnalysisRun(origRun)
+	assert.Equal(t, v1alpha1.AnalysisPhaseRunning, newRun.Status.Phase)
+	assert.Nil(t, newRun.Status.CompletedAt)
+	// Nothing else is deleted
+	assert.Equal(t, 0, len(filterInformerActions(f.client.Actions())))
+}
+
+func TestCompletedTimeFilled(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+
+	expectedCount := intstr.FromInt(1)
+	origRun := &v1alpha1.AnalysisRun{
+		Spec: v1alpha1.AnalysisRunSpec{
+			Metrics: []v1alpha1.Metric{
+				{
+					Name:     "metric1",
+					Interval: "60s",
+					Count:    &expectedCount,
+					Provider: v1alpha1.MetricProvider{
+						Job: &v1alpha1.JobMetric{},
+					},
+				},
+			},
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase:     v1alpha1.AnalysisPhaseRunning,
+			StartedAt: timePtr(metav1.NewTime(time.Now())),
+			MetricResults: []v1alpha1.MetricResult{
+				{
+					Name:  "metric1",
+					Phase: v1alpha1.AnalysisPhaseSuccessful,
+					Count: 1,
+					Measurements: []v1alpha1.Measurement{{
+						Value:      "1",
+						Phase:      v1alpha1.AnalysisPhaseSuccessful,
+						StartedAt:  timePtr(metav1.NewTime(time.Now().Add(-60 * time.Second))),
+						FinishedAt: timePtr(metav1.NewTime(time.Now().Add(-60 * time.Second))),
+					}},
+				},
+			},
+		},
+	}
+	newRun := c.reconcileAnalysisRun(origRun)
+	assert.Equal(t, v1alpha1.AnalysisPhaseSuccessful, newRun.Status.Phase)
+	assert.NotNil(t, newRun.Status.CompletedAt)
+	assert.Equal(t, f.now, newRun.Status.CompletedAt.Time)
+}
+
+func TestReconcileAnalysisRunOnRunNotFound(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+	buf := bytes.NewBufferString("")
+	log.SetOutput(buf)
+
+	// Prepend since there is a default reaction that captures it.
+	f.client.Fake.PrependReactor("delete", "analysisruns", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+		return true, nil, k8serrors.NewNotFound(k8sschema.GroupResource{Resource: "analysisruns"}, "test")
+	})
+
+	origRun := &v1alpha1.AnalysisRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "run" + string(uuid.NewUUID()),
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: v1alpha1.AnalysisRunSpec{
+			TTLStrategy: &v1alpha1.TTLStrategy{
+				SecondsAfterCompletion: pointer.Int32Ptr(1),
+			},
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase:       v1alpha1.AnalysisPhaseSuccessful,
+			CompletedAt: timePtr(metav1.NewTime(f.now.Add(-2 * time.Second))),
+		},
+	}
+	_ = c.reconcileAnalysisRun(origRun)
+	logMessage := buf.String()
+	assert.Contains(t, logMessage, "Trying to cleanup TTL exceeded analysis run")
+	assert.NotContains(t, logMessage, "Failed to garbage collect analysis run")
+	// One deletion issued.
+	assert.Len(t, f.client.Fake.Actions(), 1)
+	assert.Equal(t, "delete", f.client.Fake.Actions()[0].GetVerb())
+	assert.Equal(t, "analysisruns", f.client.Fake.Actions()[0].GetResource().Resource)
+}
+
+func TestReconcileAnalysisRunOnOtherRunErrors(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+	buf := bytes.NewBufferString("")
+	log.SetOutput(buf)
+
+	// Prepend since there is a default reaction that captures it.
+	f.client.Fake.PrependReactor("delete", "analysisruns", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+		return true, nil, errors.New("some error")
+	})
+
+	origRun := &v1alpha1.AnalysisRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "run" + string(uuid.NewUUID()),
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: v1alpha1.AnalysisRunSpec{
+			TTLStrategy: &v1alpha1.TTLStrategy{
+				SecondsAfterCompletion: pointer.Int32Ptr(1),
+			},
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase:       v1alpha1.AnalysisPhaseSuccessful,
+			CompletedAt: timePtr(metav1.NewTime(f.now.Add(-2 * time.Second))),
+		},
+	}
+	_ = c.reconcileAnalysisRun(origRun)
+	logMessage := buf.String()
+	assert.Contains(t, logMessage, "Failed to garbage collect analysis run")
+	// One deletion issued.
+	assert.Len(t, f.client.Fake.Actions(), 1)
+	assert.Equal(t, "delete", f.client.Fake.Actions()[0].GetVerb())
+	assert.Equal(t, "analysisruns", f.client.Fake.Actions()[0].GetResource().Resource)
+}
+
+func TestMaybeGarbageCollectAnalysisRunNoGCIfNotCompleted(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+
+	origRun := &v1alpha1.AnalysisRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "run" + string(uuid.NewUUID()),
+			Namespace: metav1.NamespaceDefault,
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase: v1alpha1.AnalysisPhaseRunning,
+		},
+	}
+	logger := logutil.WithAnalysisRun(origRun)
+	err := c.maybeGarbageCollectAnalysisRun(origRun, logger)
+	// No error, no deletion issued.
+	assert.NoError(t, err)
+	assert.Empty(t, f.client.Fake.Actions())
+}
+
+func TestMaybeGarbageCollectAnalysisRunNoGCIfNoTTLStrategy(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+
+	origRun := &v1alpha1.AnalysisRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "run" + string(uuid.NewUUID()),
+			Namespace: metav1.NamespaceDefault,
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase: v1alpha1.AnalysisPhaseSuccessful,
+		},
+	}
+	logger := logutil.WithAnalysisRun(origRun)
+	err := c.maybeGarbageCollectAnalysisRun(origRun, logger)
+	// No error, no deletion issued.
+	assert.NoError(t, err)
+	assert.Empty(t, f.client.Fake.Actions())
+}
+
+func TestMaybeGarbageCollectAnalysisRunNoGCIfWithDeletionTimestamp(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+
+	origRun := &v1alpha1.AnalysisRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              "run" + string(uuid.NewUUID()),
+			Namespace:         metav1.NamespaceDefault,
+			DeletionTimestamp: timePtr(metav1.NewTime(f.now)),
+		},
+		Spec: v1alpha1.AnalysisRunSpec{
+			TTLStrategy: &v1alpha1.TTLStrategy{
+				SecondsAfterCompletion: pointer.Int32Ptr(1),
+			},
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase:       v1alpha1.AnalysisPhaseSuccessful,
+			CompletedAt: timePtr(metav1.NewTime(f.now.Add(-2 * time.Second))),
+		},
+	}
+	logger := logutil.WithAnalysisRun(origRun)
+	err := c.maybeGarbageCollectAnalysisRun(origRun, logger)
+	// No error, no deletion issued.
+	assert.NoError(t, err)
+	assert.Empty(t, f.client.Fake.Actions())
+}
+
+func TestMaybeGarbageCollectAnalysisRunNoGCIfNoCompletedAt(t *testing.T) {
+	f := newFixture(t)
+	defer f.Close()
+	c, _, _ := f.newController(noResyncPeriodFunc)
+
+	origRun := &v1alpha1.AnalysisRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "run" + string(uuid.NewUUID()),
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: v1alpha1.AnalysisRunSpec{
+			TTLStrategy: &v1alpha1.TTLStrategy{
+				SecondsAfterCompletion: pointer.Int32Ptr(1),
+			},
+		},
+		Status: v1alpha1.AnalysisRunStatus{
+			Phase: v1alpha1.AnalysisPhaseSuccessful,
+		},
+	}
+	logger := logutil.WithAnalysisRun(origRun)
+	err := c.maybeGarbageCollectAnalysisRun(origRun, logger)
+	// No error, no deletion issued.
+	assert.NoError(t, err)
+	assert.Empty(t, f.client.Fake.Actions())
+}
diff --git a/analysis/controller_test.go b/analysis/controller_test.go
index 032f5cf74..8282f09c2 100644
--- a/analysis/controller_test.go
+++ b/analysis/controller_test.go
@@ -55,6 +55,9 @@ type fixture struct {
 	unfreezeTime    func() error
 	// fake provider
 	provider *mocks.Provider
+
+	// Reference to frozen now
+	now time.Time
 }
 
 func newFixture(t *testing.T) *fixture {
@@ -62,9 +65,9 @@ func newFixture(t *testing.T) *fixture {
 	f.t = t
 	f.objects = []runtime.Object{}
 	f.enqueuedObjects = make(map[string]int)
-	now := time.Now()
+	f.now = time.Now()
 	timeutil.Now = func() time.Time {
-		return now
+		return f.now
 	}
 	f.unfreezeTime = func() error {
 		timeutil.Now = time.Now
@@ -286,6 +289,22 @@ func (f *fixture) getPatchedAnalysisRun(index int) v1alpha1.AnalysisRun { //noli
 	return ar
 }
 
+func (f *fixture) expectDeleteAnalysisRunAction(analysisRun *v1alpha1.AnalysisRun) int { //nolint:unused
+	action := core.NewDeleteAction(schema.GroupVersionResource{Resource: "analysisrun"}, analysisRun.Namespace, analysisRun.Name)
+	len := len(f.actions)
+	f.actions = append(f.actions, action)
+	return len
+}
+
+func (f *fixture) getDeletedAnalysisRunNamespaceAndName(index int) string { //nolint:unused
+	action := filterInformerActions(f.client.Actions())[index]
+	deleteAction, ok := action.(core.DeleteAction)
+	if !ok {
+		f.t.Fatalf("Expected Patch action, not %s", action.GetVerb())
+	}
+	return fmt.Sprintf("%s/%s", deleteAction.GetNamespace(), deleteAction.GetName())
+}
+
 func TestNoReconcileForNotFoundAnalysisRun(t *testing.T) {
 	f := newFixture(t)
 	defer f.Close()
