diff --git a/projects/batfish/src/test/java/org/batfish/grammar/cisco_nxos/CiscoNxosGrammarTest.java b/projects/batfish/src/test/java/org/batfish/grammar/cisco_nxos/CiscoNxosGrammarTest.java
index 5853e5f23e..81fc78ad46 100644
--- a/projects/batfish/src/test/java/org/batfish/grammar/cisco_nxos/CiscoNxosGrammarTest.java
+++ b/projects/batfish/src/test/java/org/batfish/grammar/cisco_nxos/CiscoNxosGrammarTest.java
@@ -80,6 +80,7 @@ import static org.batfish.datamodel.matchers.VniSettingsMatchers.hasSourceAddres
 import static org.batfish.datamodel.matchers.VniSettingsMatchers.hasUdpPort;
 import static org.batfish.datamodel.matchers.VniSettingsMatchers.hasVni;
 import static org.batfish.datamodel.matchers.VrfMatchers.hasL2VniSettings;
+import static org.batfish.datamodel.routing_policy.Common.SUMMARY_ONLY_SUPPRESSION_POLICY_NAME;
 import static org.batfish.datamodel.vendor_family.cisco_nxos.NexusPlatform.NEXUS_3000;
 import static org.batfish.datamodel.vendor_family.cisco_nxos.NexusPlatform.NEXUS_5000;
 import static org.batfish.datamodel.vendor_family.cisco_nxos.NexusPlatform.NEXUS_6000;
@@ -115,6 +116,7 @@ import static org.batfish.representation.cisco_nxos.OspfProcess.DEFAULT_TIMERS_L
 import static org.batfish.representation.cisco_nxos.OspfProcess.DEFAULT_TIMERS_THROTTLE_LSA_HOLD_INTERVAL_MS;
 import static org.batfish.representation.cisco_nxos.OspfProcess.DEFAULT_TIMERS_THROTTLE_LSA_MAX_INTERVAL_MS;
 import static org.batfish.representation.cisco_nxos.OspfProcess.DEFAULT_TIMERS_THROTTLE_LSA_START_INTERVAL_MS;
+import static org.hamcrest.Matchers.aMapWithSize;
 import static org.hamcrest.Matchers.allOf;
 import static org.hamcrest.Matchers.anEmptyMap;
 import static org.hamcrest.Matchers.any;
@@ -226,6 +228,7 @@ import org.batfish.datamodel.UniverseIpSpace;
 import org.batfish.datamodel.acl.AclLineMatchExpr;
 import org.batfish.datamodel.acl.AclLineMatchExprs;
 import org.batfish.datamodel.answers.ConvertConfigurationAnswerElement;
+import org.batfish.datamodel.bgp.BgpAggregate;
 import org.batfish.datamodel.bgp.EvpnAddressFamily;
 import org.batfish.datamodel.bgp.Ipv4UnicastAddressFamily;
 import org.batfish.datamodel.bgp.Layer2VniConfig;
@@ -286,6 +289,7 @@ import org.batfish.representation.cisco_nxos.ActionIpAccessListLine;
 import org.batfish.representation.cisco_nxos.AddrGroupIpAddressSpec;
 import org.batfish.representation.cisco_nxos.AddressFamily;
 import org.batfish.representation.cisco_nxos.BgpGlobalConfiguration;
+import org.batfish.representation.cisco_nxos.BgpVrfAddressFamilyAggregateNetworkConfiguration;
 import org.batfish.representation.cisco_nxos.BgpVrfConfiguration;
 import org.batfish.representation.cisco_nxos.BgpVrfIpv4AddressFamilyConfiguration;
 import org.batfish.representation.cisco_nxos.BgpVrfIpv6AddressFamilyConfiguration;
@@ -394,7 +398,6 @@ import org.batfish.representation.cisco_nxos.TcpOptions;
 import org.batfish.representation.cisco_nxos.Track;
 import org.batfish.representation.cisco_nxos.TrackInterface;
 import org.batfish.representation.cisco_nxos.TrackInterface.Mode;
-import org.batfish.representation.cisco_nxos.TrackUnsupported;
 import org.batfish.representation.cisco_nxos.UdpOptions;
 import org.batfish.representation.cisco_nxos.Vlan;
 import org.batfish.representation.cisco_nxos.Vrf;
@@ -1614,7 +1617,7 @@ public final class CiscoNxosGrammarTest {
     String hostname = "nxos_track";
     CiscoNxosConfiguration vc = parseVendorConfig(hostname);
 
-    assertThat(vc.getTracks(), hasKeys(1, 2, 100, 101, 500));
+    assertThat(vc.getTracks(), hasKeys(1, 2, 500));
 
     {
       Track track = vc.getTracks().get(1);
@@ -1639,10 +1642,6 @@ public final class CiscoNxosGrammarTest {
       assertThat(trackInterface.getInterface(), equalTo("loopback1"));
       assertThat(trackInterface.getMode(), equalTo(Mode.IPV6_ROUTING));
     }
-
-    // Should have placeholders for unsupported track types in the VS model
-    assertThat(vc.getTracks().get(100), instanceOf(TrackUnsupported.class));
-    assertThat(vc.getTracks().get(101), instanceOf(TrackUnsupported.class));
   }
 
   @Test
@@ -1657,12 +1656,6 @@ public final class CiscoNxosGrammarTest {
                 "Unsupported interface type: mgmt, expected [ETHERNET, PORT_CHANNEL, LOOPBACK]"),
             hasComment(
                 "Unsupported interface type: Vlan, expected [ETHERNET, PORT_CHANNEL, LOOPBACK]"),
-            allOf(
-                hasComment("This track method is not yet supported and will be ignored."),
-                ParseWarningMatchers.hasText(containsString("ip route 192.0.2.1/32 reachability"))),
-            allOf(
-                hasComment("This track method is not yet supported and will be ignored."),
-                ParseWarningMatchers.hasText(containsString("ip sla 1 reachability"))),
             hasComment("Expected track object-id in range 1-500, but got '0'"),
             hasComment("Expected track object-id in range 1-500, but got '501'"),
             // Undefined references are not accepted by the CLI
@@ -8844,9 +8837,6 @@ public final class CiscoNxosGrammarTest {
     assertThat(ans, hasNumReferrers(filename, CiscoNxosStructureType.TRACK, "1", 2));
     assertThat(ans, hasNumReferrers(filename, CiscoNxosStructureType.TRACK, "2", 1));
     assertThat(ans, hasNumReferrers(filename, CiscoNxosStructureType.TRACK, "3", 0));
-    // Unsupported track methods should still produce correct references
-    assertThat(ans, hasNumReferrers(filename, CiscoNxosStructureType.TRACK, "100", 1));
-    assertThat(ans, hasNumReferrers(filename, CiscoNxosStructureType.TRACK, "101", 1));
 
     assertThat(
         ans,
@@ -8867,4 +8857,176 @@ public final class CiscoNxosGrammarTest {
             "499",
             CiscoNxosStructureUsage.INTERFACE_HSRP_GROUP_TRACK));
   }
+
+  @Test
+  public void testAggregateAddressExtraction() {
+    String hostname = "nxos-aggregate-address";
+    CiscoNxosConfiguration vc = parseVendorConfig(hostname);
+    Map<Prefix, BgpVrfAddressFamilyAggregateNetworkConfiguration> aggs =
+        vc.getBgpGlobalConfiguration()
+            .getVrfs()
+            .get(DEFAULT_VRF_NAME)
+            .getIpv4UnicastAddressFamily()
+            .getAggregateNetworks();
+    assertThat(aggs, aMapWithSize(9));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("1.1.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    null, false, null, false, null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("1.2.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    null, false, "atm1", false, null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("2.1.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    null, true, null, false, null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("2.2.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    "adm", true, null, false, null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("2.3.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    null, true, "atm2", false, null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("3.1.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    null, false, null, true, null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("3.2.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    null, false, null, false, "sm1"))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("3.3.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    null, false, null, true, "sm2"))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("4.0.0.0/16")),
+            equalTo(
+                new BgpVrfAddressFamilyAggregateNetworkConfiguration(
+                    "undefined", false, "undefined", false, "undefined"))));
+  }
+
+  @Test
+  public void testAggregateAddressConversion() throws IOException {
+    String hostname = "nxos-aggregate-address";
+    Configuration c = parseConfig(hostname);
+
+    Map<Prefix, BgpAggregate> aggs = c.getDefaultVrf().getBgpProcess().getAggregates();
+    assertThat(aggs, aMapWithSize(9));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("1.1.0.0/16")),
+            equalTo(BgpAggregate.of(Prefix.parse("1.1.0.0/16"), null, null, null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("1.2.0.0/16")),
+            equalTo(BgpAggregate.of(Prefix.parse("1.2.0.0/16"), null, null, "atm1"))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("2.1.0.0/16")),
+            equalTo(
+                BgpAggregate.of(
+                    Prefix.parse("2.1.0.0/16"),
+                    null,
+                    // TODO: generation policy should incorporate as-set
+                    null,
+                    null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("2.2.0.0/16")),
+            equalTo(
+                BgpAggregate.of(
+                    Prefix.parse("2.2.0.0/16"),
+                    null,
+                    // TODO: generation policy should incorporate as-set and advertise-map
+                    null,
+                    null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("2.3.0.0/16")),
+            equalTo(
+                BgpAggregate.of(
+                    Prefix.parse("2.3.0.0/16"),
+                    null,
+                    // TODO: generation policy should incorporate as-set
+                    null,
+                    "atm2"))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("3.1.0.0/16")),
+            equalTo(
+                BgpAggregate.of(
+                    Prefix.parse("3.1.0.0/16"),
+                    SUMMARY_ONLY_SUPPRESSION_POLICY_NAME,
+                    null,
+                    null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("3.2.0.0/16")),
+            equalTo(
+                BgpAggregate.of(
+                    Prefix.parse("3.2.0.0/16"),
+                    // TODO: suppression policy should incorporate suppress-map
+                    null,
+                    null,
+                    null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("3.3.0.0/16")),
+            equalTo(
+                BgpAggregate.of(
+                    Prefix.parse("3.3.0.0/16"),
+                    // TODO: suppression policy should incorporate suppress-map and ignore
+                    //       summary-only.
+                    SUMMARY_ONLY_SUPPRESSION_POLICY_NAME,
+                    null,
+                    null))));
+    assertThat(
+        aggs,
+        hasEntry(
+            equalTo(Prefix.parse("4.0.0.0/16")),
+            equalTo(
+                BgpAggregate.of(
+                    Prefix.parse("4.0.0.0/16"),
+                    // TODO: verify undefined route-map treated as omitted
+                    null,
+                    null,
+                    null))));
+  }
 }
diff --git a/projects/batfish/src/test/resources/org/batfish/grammar/cisco_nxos/testconfigs/nxos-aggregate-address b/projects/batfish/src/test/resources/org/batfish/grammar/cisco_nxos/testconfigs/nxos-aggregate-address
new file mode 100644
index 0000000000..dec17b36ff
--- /dev/null
+++ b/projects/batfish/src/test/resources/org/batfish/grammar/cisco_nxos/testconfigs/nxos-aggregate-address
@@ -0,0 +1,40 @@
+!RANCID-CONTENT-TYPE: cisco-nx
+!
+hostname nxos-aggregate-address
+!
+route-map atm1 permit 100
+ set metric 50
+!
+route-map adm permit 100
+ match tag 5
+!
+route-map atm2 permit 100
+ set community 1:1
+!
+route-map sm1 permit 100
+ match tag 5
+!
+route-map sm2 permit 100
+ match tag 5
+!
+router bgp 1
+ address-family ipv4 unicast
+   ! no suppression, everything more specific contributes, no inheritance
+   aggregate-address 1.1.0.0/16
+   ! attribute-map should transform aggregate
+   aggregate-address 1.2.0.0/16 attribute-map atm1
+   ! inheritance
+   aggregate-address 2.1.0.0/16 as-set
+   ! inheritance and contributors are limited to routes passing advertise-map,
+   aggregate-address 2.2.0.0/16 as-set advertise-map adm
+   ! attribute-map applies after inheritance
+   aggregate-address 2.3.0.0/16 as-set attribute-map atm2
+   ! suppression everything more specific
+   aggregate-address 3.1.0.0/16 summary-only
+   ! suppress only routes passing suppress-map
+   aggregate-address 3.2.0.0/16 suppress-map sm1
+   ! suppress only routes passing suppress-map, completely ignoring summary-only
+   aggregate-address 3.3.0.0/16 summary-only suppress-map sm2
+   ! undefined route-maps
+   aggregate-address 4.0.0.0/16 advertise-map undefined attribute-map undefined suppress-map undefined
+!
