diff --git a/crates/matrix-sdk-ui/src/timeline/tests/mod.rs b/crates/matrix-sdk-ui/src/timeline/tests/mod.rs
index db408c1f9..82cf9127e 100644
--- a/crates/matrix-sdk-ui/src/timeline/tests/mod.rs
+++ b/crates/matrix-sdk-ui/src/timeline/tests/mod.rs
@@ -30,7 +30,7 @@ use matrix_sdk::{
     config::RequestConfig,
     crypto::OlmMachine,
     deserialized_responses::{EncryptionInfo, TimelineEvent},
-    event_cache::paginator::{PaginableRoom, PaginatorError},
+    paginators::{thread::PaginableThread, PaginableRoom, PaginatorError},
     room::{EventWithContextResponse, Messages, MessagesOptions, PushContext, Relations},
     send_queue::RoomSendQueueUpdate,
     BoxFuture,
@@ -303,13 +303,30 @@ impl PaginableRoom for TestRoomDataProvider {
     }
 }
 
+impl PaginableThread for TestRoomDataProvider {
+    async fn relations(
+        &self,
+        _thread_root: OwnedEventId,
+        _opts: matrix_sdk::room::RelationsOptions,
+    ) -> Result<Relations, matrix_sdk::Error> {
+        unimplemented!();
+    }
+
+    async fn load_event(
+        &self,
+        _event_id: &OwnedEventId,
+    ) -> Result<TimelineEvent, matrix_sdk::Error> {
+        unimplemented!();
+    }
+}
+
 impl PinnedEventsRoom for TestRoomDataProvider {
     fn load_event_with_relations<'a>(
         &'a self,
         _event_id: &'a EventId,
         _request_config: Option<RequestConfig>,
         _related_event_filters: Option<Vec<RelationType>>,
-    ) -> BoxFuture<'a, Result<(TimelineEvent, Vec<TimelineEvent>), PaginatorError>> {
+    ) -> BoxFuture<'a, Result<(TimelineEvent, Vec<TimelineEvent>), matrix_sdk::Error>> {
         unimplemented!();
     }
 
@@ -430,14 +447,6 @@ impl RoomDataProvider for TestRoomDataProvider {
         self.encryption_info.get(session_id).cloned()
     }
 
-    async fn relations(
-        &self,
-        _event_id: OwnedEventId,
-        _opts: matrix_sdk::room::RelationsOptions,
-    ) -> Result<Relations, matrix_sdk::Error> {
-        unimplemented!();
-    }
-
     async fn load_event<'a>(&'a self, _event_id: &'a EventId) -> matrix_sdk::Result<TimelineEvent> {
         unimplemented!();
     }
