diff --git a/packages/server/test/unit/open_project_spec.js b/packages/server/test/unit/open_project_spec.js
index 92824e7720..75671d1f20 100644
--- a/packages/server/test/unit/open_project_spec.js
+++ b/packages/server/test/unit/open_project_spec.js
@@ -1,6 +1,7 @@
 require('../spec_helper')
 
 const path = require('path')
+const os = require('os')
 const chokidar = require('chokidar')
 const browsers = require(`${root}lib/browsers`)
 const ProjectBase = require(`${root}lib/project-base`).ProjectBase
@@ -37,6 +38,18 @@ describe('lib/open_project', () => {
     return openProject.create('/project/root', {}, {})
   })
 
+  context('#create', () => {
+    // @see https://github.com/cypress-io/cypress/issues/18094
+    it('warns on win 32bit', async () => {
+      sinon.stub(os, 'platform').returns('win32')
+      sinon.stub(os, 'arch').returns('ia32')
+      const onWarning = sinon.stub()
+
+      await openProject.create('/root', {}, { onWarning })
+      expect(onWarning.getCall(0).args[0].message).to.include('You are running a 32-bit build')
+    })
+  })
+
   context('#launch', () => {
     beforeEach(async function () {
       await openProject.create('/root', {}, {})
