diff --git a/core/scc/cscc/write_batch_test.go b/core/scc/cscc/write_batch_test.go
new file mode 100644
index 000000000..a48f6939b
--- /dev/null
+++ b/core/scc/cscc/write_batch_test.go
@@ -0,0 +1,23 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package cscc
+
+import (
+	"testing"
+
+	"github.com/hyperledger/fabric/core/scc/cscc/mocks"
+	"github.com/stretchr/testify/require"
+)
+
+func TestWriteBatch(t *testing.T) {
+	mockStub := &mocks.ChaincodeStub{}
+	mockStub.StartWriteBatch()
+	err := mockStub.FinishWriteBatch()
+	require.NoError(t, err)
+	require.Equal(t, 1, mockStub.StartWriteBatchCallCount())
+	require.Equal(t, 1, mockStub.FinishWriteBatchCallCount())
+}
