diff --git a/crates/ruff_linter/resources/test/fixtures/pydocstyle/sections.py b/crates/ruff_linter/resources/test/fixtures/pydocstyle/sections.py
index d04c38a74b..22769903df 100644
--- a/crates/ruff_linter/resources/test/fixtures/pydocstyle/sections.py
+++ b/crates/ruff_linter/resources/test/fixtures/pydocstyle/sections.py
@@ -564,6 +564,17 @@ def titlecase_sub_section_header():
     """
 
 
+def newline_sub_section_header():
+    """Below, `returns:` should be considered a section header.
+
+    Args:
+        arg: Here's a note.
+
+    returns:
+        Here's another note.
+    """
+
+
 def test_method_should_be_correctly_capitalized(parameters: list[str], other_parameters: dict[str, str]): # noqa: D213
     """Test parameters and attributes sections are capitalized correctly.
