diff --git a/hack/tests/e2e-ansible-molecule.sh b/hack/tests/e2e-ansible-molecule.sh
index 20fe4f26..4095ffbf 100755
--- a/hack/tests/e2e-ansible-molecule.sh
+++ b/hack/tests/e2e-ansible-molecule.sh
@@ -65,7 +65,7 @@ header_text "Test Ansible Molecule scenarios"
 pushd "${ROOTDIR}/test/ansible"
 DEST_IMAGE="quay.io/example/ansible-test-operator:v0.0.1"
 sed -i".bak" -E -e 's/(FROM quay.io\/operator-framework\/ansible-operator)(:.*)?/\1:dev/g' build/Dockerfile; rm -f build/Dockerfile.bak
-operator-sdk build "$DEST_IMAGE"
+operator-sdk build "$DEST_IMAGE" --image-build-args="--no-cache"
 load_image_if_kind "$DEST_IMAGE"
 OPERATOR_PULL_POLICY=Never OPERATOR_IMAGE=${DEST_IMAGE} TEST_CLUSTER_PORT=24443 TEST_NAMESPACE=osdk-test molecule test --all
 
diff --git a/pkg/ansible/watches/watches_test.go b/pkg/ansible/watches/watches_test.go
index bb200fa8..3c578db0 100644
--- a/pkg/ansible/watches/watches_test.go
+++ b/pkg/ansible/watches/watches_test.go
@@ -15,10 +15,12 @@
 package watches
 
 import (
+	"fmt"
 	"html/template"
 	"os"
 	"path/filepath"
 	"reflect"
+	"sort"
 	"strconv"
 	"testing"
 	"time"
@@ -302,16 +304,26 @@ func TestLoad(t *testing.T) {
 			ManageStatus: true,
 			Vars:         map[string]interface{}{"sentinel": "reconciling"},
 		},
+		Watch{
+			GroupVersionKind: schema.GroupVersionKind{
+				Version: "v1alpha1",
+				Group:   "app.example.com",
+				Kind:    "AnsibleCollectionEnvTest",
+			},
+			Role:         filepath.Join(cwd, "testdata", "ansible_collections", "nameSpace", "collection", "roles", "someRole"),
+			ManageStatus: true,
+		},
 	}
 
 	testCases := []struct {
-		name                           string
-		path                           string
-		maxWorkers                     int
-		ansibleVerbosity               int
-		expected                       []Watch
-		shouldError                    bool
-		shouldSetAnsibleRolePathEnvVar bool
+		name                                 string
+		path                                 string
+		maxWorkers                           int
+		ansibleVerbosity                     int
+		expected                             []Watch
+		shouldError                          bool
+		shouldSetAnsibleRolePathEnvVar       bool
+		shouldSetAnsibleCollectionPathEnvVar bool
 	}{
 		{
 			name:        "error duplicate GVK",
@@ -379,19 +391,26 @@ func TestLoad(t *testing.T) {
 			shouldError: true,
 		},
 		{
-			name:             "valid watches file",
-			path:             "testdata/valid.yaml",
-			maxWorkers:       1,
-			ansibleVerbosity: 2,
-			expected:         validWatches,
+			name:        "if collection env var is not set and collection is not installed to the default locations, fail",
+			path:        "testdata/invalid_collection.yaml",
+			shouldError: true,
 		},
 		{
-			name:                           "should watches file successfully with ANSIBLE ROLES PATH ENV VAR set",
-			path:                           "testdata/valid.yaml",
-			maxWorkers:                     1,
-			ansibleVerbosity:               2,
-			shouldSetAnsibleRolePathEnvVar: true,
-			expected:                       validWatches,
+			name:                                 "valid watches file",
+			path:                                 "testdata/valid.yaml",
+			maxWorkers:                           1,
+			ansibleVerbosity:                     2,
+			shouldSetAnsibleCollectionPathEnvVar: true,
+			expected:                             validWatches,
+		},
+		{
+			name:                                 "should load file successfully with ANSIBLE ROLES PATH ENV VAR set",
+			path:                                 "testdata/valid.yaml",
+			maxWorkers:                           1,
+			ansibleVerbosity:                     2,
+			shouldSetAnsibleRolePathEnvVar:       true,
+			shouldSetAnsibleCollectionPathEnvVar: true,
+			expected:                             validWatches,
 		},
 	}
 
@@ -409,6 +428,12 @@ func TestLoad(t *testing.T) {
 				os.Setenv("ANSIBLE_ROLES_PATH", anisbleEnvVar)
 				defer os.Unsetenv("ANSIBLE_ROLES_PATH")
 			}
+			if tc.shouldSetAnsibleCollectionPathEnvVar {
+
+				ansibleCollectionPathEnv := filepath.Join(projutil.MustGetwd(), "testdata")
+				os.Setenv("ANSIBLE_COLLECTIONS_PATH", ansibleCollectionPathEnv)
+				defer os.Unsetenv("ANSIBLE_COLLECTIONS_PATH")
+			}
 
 			watchSlice, err := Load(tc.path, tc.maxWorkers, tc.ansibleVerbosity)
 			if err != nil && !tc.shouldError {
@@ -637,80 +662,101 @@ func TestAnsibleVerbosity(t *testing.T) {
 	}
 }
 
-// Test the func getFullRolePath when no envVar is set.
-func TestGetFullRolePath(t *testing.T) {
+// Test the func getPossibleRolePaths.
+func TestGetPossibleRolePaths(t *testing.T) {
 	// Mock default Full Path based in the current directory
 	rolesPath := filepath.Join(projutil.MustGetwd(), "roles")
+	home, _ := os.UserHomeDir()
 
 	type args struct {
-		path                           string
-		shouldSetAnsibleRolePathEnvVar bool
-		env                            string
+		path           string
+		rolesEnv       string
+		collectionsEnv string
 	}
 	tests := []struct {
 		name string
 		args args
-		want string
+		want []string
 	}{
 		{
-			name: "Should work with default values",
+			name: "check the current dir for a role name",
 			args: args{
 				path: "Foo",
 			},
-			want: filepath.Join(rolesPath, "Foo"),
-		},
-		{
-			name: "Should work with full role",
-			args: args{
-				path: filepath.Join(rolesPath, "Foo"),
-			},
-			want: filepath.Join(rolesPath, "Foo"),
+			want: []string{filepath.Join(rolesPath, "Foo")},
 		},
 		{
-			name: "Should work with relative role",
+			name: "check the current dir for a relative path",
 			args: args{
 				path: "relative/Foo",
 			},
-			want: filepath.Join(rolesPath, "relative/Foo"),
+			want: []string{filepath.Join(rolesPath, "relative/Foo")},
 		},
 		{
-			name: "Should work when the full role is informed and the envvar is set",
+			name: "check all paths in ANSIBLE_ROLES_PATH env var",
 			args: args{
-				path:                           filepath.Join(rolesPath, "Foo"),
-				shouldSetAnsibleRolePathEnvVar: true,
+				rolesEnv: "relative:nested/relative:/and/abs",
+				path:     "Foo",
+			},
+			want: []string{
+				filepath.Join(rolesPath, "Foo"),
+				filepath.Join("relative", "Foo"),
+				filepath.Join("relative", "roles", "Foo"),
+				filepath.Join("nested/relative", "Foo"),
+				filepath.Join("nested/relative", "roles", "Foo"),
+				filepath.Join("/and/abs", "Foo"),
+				filepath.Join("/and/abs", "roles", "Foo"),
 			},
-			want: filepath.Join(rolesPath, "Foo"),
 		},
 		{
-			name: "Should return the default full role path based in the current directory when all " +
-				"custom paths are invalids",
+			name: "Check for roles inside default collection locations when given fqcn",
 			args: args{
-				env:                            "invalid/myroles:invalid:mypath/invalid",
-				path:                           "Foo",
-				shouldSetAnsibleRolePathEnvVar: true,
+				path: "myNS.myCol.myRole",
+			},
+			want: []string{
+				filepath.Join(rolesPath, "myNS.myCol.myRole"),
+				filepath.Join("/usr/share/ansible/collections", "ansible_collections", "myNS", "myCol", "roles", "myRole"),
+				filepath.Join(home, ".ansible/collections", "ansible_collections", "myNS", "myCol", "roles", "myRole"),
 			},
-			want: filepath.Join(rolesPath, "Foo"),
 		},
 		{
-			name: "Should return the default full role when the role is not found in the customized one",
+			name: "Check for roles inside ANSIBLE_COLLECTIONS_PATH locations when set and given path is fqcn",
 			args: args{
-				env:                            "customized/myroles",
-				path:                           "Foo",
-				shouldSetAnsibleRolePathEnvVar: true,
+				path:           "myNS.myCol.myRole",
+				collectionsEnv: "/my/collections/",
+			},
+			want: []string{
+				filepath.Join(rolesPath, "myNS.myCol.myRole"),
+				filepath.Join("/my/collections/", "ansible_collections", "myNS", "myCol", "roles", "myRole"),
+				// Note: Defaults are not checked when the env variable is set
 			},
-			want: filepath.Join(rolesPath, "Foo"),
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 
-			if tt.args.shouldSetAnsibleRolePathEnvVar {
-				os.Setenv("ANSIBLE_ROLES_PATH", tt.args.env)
+			if len(tt.args.rolesEnv) > 0 {
+				os.Setenv("ANSIBLE_ROLES_PATH", tt.args.rolesEnv)
 				defer os.Unsetenv("ANSIBLE_ROLES_PATH")
 			}
+			if len(tt.args.collectionsEnv) > 0 {
+				os.Setenv("ANSIBLE_COLLECTIONS_PATH", tt.args.collectionsEnv)
+				defer os.Unsetenv("ANSIBLE_COLLECTIONS_PATH")
+			}
 
-			if got := getFullRolePath(tt.args.path); got != tt.want {
-				t.Errorf("Error to check getFullRolePath() = %v, want %v", got, tt.want)
+			allPathsToCheck := getPossibleRolePaths(tt.args.path)
+			sort.Strings(tt.want)
+			sort.Strings(allPathsToCheck)
+			if !reflect.DeepEqual(allPathsToCheck, tt.want) {
+				t.Errorf("Unexpected paths returned")
+				fmt.Println("Returned:")
+				for i, path := range allPathsToCheck {
+					fmt.Println(i, path)
+				}
+				fmt.Println("Wanted:")
+				for i, path := range tt.want {
+					fmt.Println(i, path)
+				}
 			}
 		})
 	}
diff --git a/test/ansible-memcached/molecule.yml b/test/ansible-memcached/molecule.yml
index 628a524a..cc29e0d2 100644
--- a/test/ansible-memcached/molecule.yml
+++ b/test/ansible-memcached/molecule.yml
@@ -10,7 +10,7 @@ platforms:
 - name: kind-test-local
   groups:
   - k8s
-  image: bsycorp/kind:latest-1.15
+  image: bsycorp/kind:latest-1.16
   privileged: True
   override_command: no
   exposed_ports:
diff --git a/test/ansible/build/Dockerfile b/test/ansible/build/Dockerfile
index 50aaed3a..93ffe548 100644
--- a/test/ansible/build/Dockerfile
+++ b/test/ansible/build/Dockerfile
@@ -7,3 +7,9 @@ COPY roles/ ${HOME}/roles/
 COPY playbooks/ ${HOME}/playbooks/
 COPY inventory/ ${HOME}/inventory/
 COPY plugins/ ${HOME}/plugins/
+COPY fixture_collection/ /tmp/fixture_collection/
+USER root
+RUN chmod -R ug+rwx /tmp/fixture_collection
+USER 1001
+RUN ansible-galaxy collection build /tmp/fixture_collection/ --output-path /tmp/fixture_collection/
+RUN ansible-galaxy collection install /tmp/fixture_collection/operator_sdk-test_fixtures-0.0.0.tar.gz
diff --git a/test/ansible/deploy/crds/test.example.com_collections_crd.yaml b/test/ansible/deploy/crds/test.example.com_collections_crd.yaml
new file mode 100644
index 00000000..2611b51d
--- /dev/null
+++ b/test/ansible/deploy/crds/test.example.com_collections_crd.yaml
@@ -0,0 +1,22 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: collections.test.example.com
+spec:
+  group: test.example.com
+  names:
+    kind: Collection
+    listKind: CollectionList
+    plural: collections
+    singular: collection
+  scope: Namespaced
+  subresources:
+    status: {}
+  validation:
+    openAPIV3Schema:
+      type: object
+      x-kubernetes-preserve-unknown-fields: true
+  versions:
+  - name: v1alpha1
+    served: true
+    storage: true
diff --git a/test/ansible/fixture_collection/galaxy.yml b/test/ansible/fixture_collection/galaxy.yml
new file mode 100644
index 00000000..0f2699d9
--- /dev/null
+++ b/test/ansible/fixture_collection/galaxy.yml
@@ -0,0 +1,6 @@
+namespace: operator_sdk
+name: test_fixtures
+version: 0.0.0
+readme: README.md
+authors:
+- your name <example@domain.com>
diff --git a/test/ansible/fixture_collection/roles/dummy/tasks/main.yml b/test/ansible/fixture_collection/roles/dummy/tasks/main.yml
new file mode 100644
index 00000000..a6aee9ce
--- /dev/null
+++ b/test/ansible/fixture_collection/roles/dummy/tasks/main.yml
@@ -0,0 +1,12 @@
+---
+- name: Create ConfigMap
+  k8s:
+    definition:
+      kind: ConfigMap
+      apiVersion: v1
+      metadata:
+        name: test-this-collection
+        namespace: "{{ meta.namespace }}"
+      data:
+        did_it_work: "indeed"
+    state: present
diff --git a/test/ansible/molecule/cluster/tasks/collections_test.yml b/test/ansible/molecule/cluster/tasks/collections_test.yml
new file mode 100644
index 00000000..79539bb3
--- /dev/null
+++ b/test/ansible/molecule/cluster/tasks/collections_test.yml
@@ -0,0 +1,22 @@
+---
+- name: Create the test.example.com/v1alpha1.Collection
+  k8s:
+    state: present
+    namespace: '{{ namespace }}'
+    definition:
+      apiVersion: test.example.com/v1alpha1
+      kind: Collection
+      metadata:
+        name: collection-test
+    wait: yes
+    wait_timeout: 300
+    wait_condition:
+      type: Running
+      reason: Successful
+      status: "True"
+
+- name: Assert ConfigMap has been created by collection Role
+  assert:
+    that: cm.data.did_it_work == 'indeed'
+  vars:
+    cm: "{{ q('k8s', api_version='v1', kind='ConfigMap', namespace=namespace, resource_name='test-this-collection').0 }}"
diff --git a/test/ansible/molecule/cluster/verify.yml b/test/ansible/molecule/cluster/verify.yml
index 96ae43f6..0e8a073c 100644
--- a/test/ansible/molecule/cluster/verify.yml
+++ b/test/ansible/molecule/cluster/verify.yml
@@ -6,3 +6,4 @@
   tasks:
     - import_tasks: tasks/inventory_test.yml
     - import_tasks: tasks/servicemonitor_test.yml
+    - import_tasks: tasks/collections_test.yml
diff --git a/test/ansible/watches.yaml b/test/ansible/watches.yaml
index 5fe42a28..23fe6961 100644
--- a/test/ansible/watches.yaml
+++ b/test/ansible/watches.yaml
@@ -3,3 +3,8 @@
   group: test.example.com
   kind: Inventory
   playbook: /opt/ansible/playbooks/inventory.yml
+
+- version: v1alpha1
+  group: test.example.com
+  kind: Collection
+  role: operator_sdk.test_fixtures.dummy
