diff --git a/crates/ruff_linter/resources/test/fixtures/ruff/RUF024.py b/crates/ruff_linter/resources/test/fixtures/ruff/RUF024.py
index 3be5e56fc4..486abcdb56 100644
--- a/crates/ruff_linter/resources/test/fixtures/ruff/RUF024.py
+++ b/crates/ruff_linter/resources/test/fixtures/ruff/RUF024.py
@@ -32,3 +32,8 @@ bar.fromkeys(pierogi_fillings, [])
 def bad_dict() -> None:
     dict = MysteryBox()
     dict.fromkeys(pierogi_fillings, [])
+
+
+key = "xy"
+key_0 = "z"
+dict.fromkeys("ABC", list(key))
