diff --git a/pkg/antctl/command_definition_test.go b/pkg/antctl/command_definition_test.go
index e4c74cdc1..65f12b9b1 100644
--- a/pkg/antctl/command_definition_test.go
+++ b/pkg/antctl/command_definition_test.go
@@ -17,337 +17,22 @@ package antctl
 import (
 	"bytes"
 	"encoding/json"
-	"fmt"
 	"io"
 	"reflect"
 	"strings"
 	"testing"
-	"time"
 
 	"github.com/spf13/cobra"
 	"github.com/stretchr/testify/assert"
-	v1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-	"antrea.io/antrea/pkg/agent/apiserver/handlers/agentinfo"
-	"antrea.io/antrea/pkg/agent/apiserver/handlers/memberlist"
-	"antrea.io/antrea/pkg/agent/apiserver/handlers/podinterface"
-	"antrea.io/antrea/pkg/antctl/output"
 	"antrea.io/antrea/pkg/antctl/runtime"
-	"antrea.io/antrea/pkg/antctl/transform/addressgroup"
-	"antrea.io/antrea/pkg/antctl/transform/appliedtogroup"
-	"antrea.io/antrea/pkg/antctl/transform/common"
-	"antrea.io/antrea/pkg/antctl/transform/controllerinfo"
-	"antrea.io/antrea/pkg/antctl/transform/networkpolicy"
 	"antrea.io/antrea/pkg/antctl/transform/version"
-	cpv1beta "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
-	"antrea.io/antrea/pkg/apis/crd/v1beta1"
-	controllernetworkpolicy "antrea.io/antrea/pkg/controller/networkpolicy"
 )
 
 type Foobar struct {
 	Foo string `json:"foo"`
 }
 
-var (
-	AntreaPolicyTierPriority = int32(250)
-	AntreaPolicyPriority     = float64(1.0)
-)
-
-func TestCommandList_tableOutputForGetCommands(t *testing.T) {
-	for _, tc := range []struct {
-		name            string
-		rawResponseData interface{}
-		expected        string
-	}{
-		{
-			name: "StructureData-ControllerInfo-Single",
-			rawResponseData: controllerinfo.Response{
-				Version: "v0.4.0",
-				PodRef: v1.ObjectReference{
-					Kind:      "Pod",
-					Namespace: "kube-system",
-					Name:      "antrea-controller-55b9bcd59f-h9ll4",
-				},
-				NodeRef: v1.ObjectReference{
-					Kind: "Node",
-					Name: "node-control-plane",
-				},
-				ServiceRef: v1.ObjectReference{
-					Kind: "Service",
-					Name: "antrea",
-				},
-				NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
-					NetworkPolicyNum:  1,
-					AddressGroupNum:   1,
-					AppliedToGroupNum: 2,
-				},
-				ConnectedAgentNum: 2,
-				ControllerConditions: []v1beta1.ControllerCondition{
-					{
-						Type:              "ControllerHealthy",
-						Status:            "True",
-						LastHeartbeatTime: metav1.NewTime(time.Now()),
-					},
-				},
-			},
-			expected: `POD                                            NODE               STATUS  NETWORK-POLICIES ADDRESS-GROUPS APPLIED-TO-GROUPS CONNECTED-AGENTS
-kube-system/antrea-controller-55b9bcd59f-h9ll4 node-control-plane Healthy 1                1              2                 2               
-`,
-		},
-		{
-			name: "StructureData-AgentInfo-Single",
-			rawResponseData: agentinfo.AntreaAgentInfoResponse{
-				Version: "v0.4.0",
-				PodRef: v1.ObjectReference{
-					Kind:      "Pod",
-					Namespace: "kube-system",
-					Name:      "antrea-agent-0",
-				},
-				NodeRef: v1.ObjectReference{
-					Kind: "Node",
-					Name: "node-worker",
-				},
-				NodeSubnets: []string{"192.168.1.0/24", "192.168.1.1/24"},
-				OVSInfo: v1beta1.OVSInfo{
-					Version:    "1.0",
-					BridgeName: "br-int",
-					FlowTable: map[string]int32{
-						"0":  5,
-						"10": 7,
-					},
-				},
-				NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
-					NetworkPolicyNum:  1,
-					AddressGroupNum:   1,
-					AppliedToGroupNum: 2,
-				},
-				LocalPodNum: 3,
-				AgentConditions: []v1beta1.AgentCondition{
-					{
-						Type:              "AgentHealthy",
-						Status:            "True",
-						LastHeartbeatTime: metav1.NewTime(time.Now()),
-					},
-				},
-			},
-			expected: `POD                        NODE        STATUS  NODE-SUBNET                   NETWORK-POLICIES ADDRESS-GROUPS APPLIED-TO-GROUPS LOCAL-PODS
-kube-system/antrea-agent-0 node-worker Healthy 192.168.1.0/24,192.168.1.1/24 1                1              2                 3         
-`,
-		},
-		{
-			name:            "StructureData-NonTableOutput-Single",
-			rawResponseData: Foobar{Foo: "foo"},
-			expected: `foo            
-foo            
-`,
-		},
-		{
-			name: "StructureData-NonTableOutput-List",
-			rawResponseData: []Foobar{
-				{Foo: "foo1"},
-				{Foo: "foo2"},
-			},
-			expected: `foo            
-foo1           
-foo2           
-`,
-		},
-		{
-			name: "StructureData-NetworkPolicy-List-HasSummary-RandomFieldOrder",
-			rawResponseData: []networkpolicy.Response{
-				{
-					NetworkPolicy: &cpv1beta.NetworkPolicy{
-						ObjectMeta: metav1.ObjectMeta{
-							Name: "6001549b-ba63-4752-8267-30f52b4332db",
-						},
-						AppliedToGroups: []string{"32ef631b-6817-5a18-86eb-93f4abf0467c", "c4c59cfe-9160-5de5-a85b-01a58d11963e"},
-						Rules: []cpv1beta.NetworkPolicyRule{
-							{
-								Direction: "In",
-								Services:  nil,
-							},
-						},
-						SourceRef: &cpv1beta.NetworkPolicyReference{
-							Type:      cpv1beta.K8sNetworkPolicy,
-							Namespace: "default",
-							Name:      "allow-all",
-							UID:       "6001549b-ba63-4752-8267-30f52b4332db",
-						},
-					},
-				},
-				{
-					NetworkPolicy: &cpv1beta.NetworkPolicy{
-						ObjectMeta: metav1.ObjectMeta{
-							Name: "880db7e8-fc2a-4030-aefe-09afc5f341ad",
-						},
-						TierPriority:    &AntreaPolicyTierPriority,
-						Priority:        &AntreaPolicyPriority,
-						AppliedToGroups: []string{"32ef631b-6817-5a18-86eb-93f4abf0467c"},
-						Rules: []cpv1beta.NetworkPolicyRule{
-							{
-								Direction: "In",
-								Services:  nil,
-							},
-							{
-								Direction: "In",
-								Services:  nil,
-							},
-						},
-						SourceRef: &cpv1beta.NetworkPolicyReference{
-							Type:      cpv1beta.AntreaNetworkPolicy,
-							Namespace: "default",
-							Name:      "allow-all",
-							UID:       "880db7e8-fc2a-4030-aefe-09afc5f341ad",
-						},
-					},
-				},
-			},
-			expected: `NAME                                 APPLIED-TO                                       RULES SOURCE                                TIER-PRIORITY PRIORITY
-6001549b-ba63-4752-8267-30f52b4332db 32ef631b-6817-5a18-86eb-93f4abf0467c + 1 more... 1     K8sNetworkPolicy:default/allow-all    <NONE>        <NONE>  
-880db7e8-fc2a-4030-aefe-09afc5f341ad 32ef631b-6817-5a18-86eb-93f4abf0467c             2     AntreaNetworkPolicy:default/allow-all 250           1       
-`,
-		},
-		{
-			name: "StructureData-AddressGroup-List-HasSummary-HasEmpty",
-			rawResponseData: []addressgroup.Response{
-				{
-					Name: "GroupName1",
-					Pods: []common.GroupMember{
-						{IP: "127.0.0.1"}, {IP: "192.168.0.1"}, {IP: "127.0.0.2"},
-						{IP: "127.0.0.3"}, {IP: "10.0.0.3"}, {IP: "127.0.0.5"}, {IP: "127.0.0.6"},
-					},
-				},
-				{
-					Name: "GroupName2",
-					Pods: []common.GroupMember{},
-				},
-			},
-			expected: `NAME       POD-IPS                                            NODE-IPS
-GroupName1 10.0.0.3,127.0.0.1,127.0.0.2,127.0.0.3 + 2 more... <NONE>  
-GroupName2 <NONE>                                             <NONE>  
-`,
-		},
-		{
-			name: "StructureData-AddressGroup-HasNode",
-			rawResponseData: []addressgroup.Response{
-				{
-					Name: "AddressGroupNameHasNode",
-					Nodes: []common.GroupMember{
-						{IP: "127.0.0.1"}, {IP: "192.168.0.1"}, {IP: "10.176.27.105"}, {IP: "127.0.0.3"},
-					},
-				},
-				{
-					Name: "AddressGroupNameHasPod",
-					Pods: []common.GroupMember{
-						{IP: "127.0.0.1"}, {IP: "192.168.0.1"}, {IP: "127.0.0.2"},
-						{IP: "127.0.0.3"}, {IP: "10.0.0.3"}, {IP: "127.0.0.5"}, {IP: "127.0.0.6"},
-					},
-				},
-				{
-					Name: "AddressGroupNameNone",
-					Pods: []common.GroupMember{},
-				},
-			},
-			expected: `NAME                    POD-IPS                                            NODE-IPS                                     
-AddressGroupNameHasNode <NONE>                                             10.176.27.105,127.0.0.1,127.0.0.3,192.168.0.1
-AddressGroupNameHasPod  10.0.0.3,127.0.0.1,127.0.0.2,127.0.0.3 + 2 more... <NONE>                                       
-AddressGroupNameNone    <NONE>                                             <NONE>                                       
-`,
-		},
-		{
-			name: "StructureData-AppliedToGroup-Single-NoSummary",
-			rawResponseData: appliedtogroup.Response{
-				Name: "GroupName",
-				Pods: []common.GroupMember{
-					{Pod: &cpv1beta.PodReference{
-						Name:      "nginx-6db489d4b7-324rc",
-						Namespace: "PodNamespace",
-					}},
-					{Pod: &cpv1beta.PodReference{
-						Name:      "nginx-6db489d4b7-vgv7v",
-						Namespace: "PodNamespace",
-					}},
-				},
-			},
-			expected: `NAME      PODS                                           
-GroupName PodNamespace/nginx-6db489d4b7-324rc + 1 more...
-`,
-		},
-		{
-			name:            "StructureData-NetworkPolicy-List-EmptyRespCase",
-			rawResponseData: []networkpolicy.Response{},
-			expected:        "\n",
-		},
-		{
-			name: "StructureData-AppliedToGroup-Single-EmptyRespCase",
-			rawResponseData: appliedtogroup.Response{
-				Name: "GroupName",
-				Pods: []common.GroupMember{},
-			},
-			expected: `NAME      PODS  
-GroupName <NONE>
-`,
-		},
-		{
-			name: "StructureData-PodInterface-List",
-			rawResponseData: []podinterface.Response{
-				{
-					PodName:       "nginx-6db489d4b7-vgv7v",
-					PodNamespace:  "default",
-					InterfaceName: "Interface",
-					IPs:           []string{"127.0.0.1"},
-					MAC:           "07-16-76-00-02-86",
-					PortUUID:      "portuuid0",
-					OFPort:        80,
-					ContainerID:   "dve7a2d6c224otm9m0eas8dtwr78",
-				},
-				{
-					PodName:       "nginx-32b489d4b7-vgv7v",
-					PodNamespace:  "default",
-					InterfaceName: "Interface2",
-					IPs:           []string{"127.0.0.2"},
-					MAC:           "07-16-76-00-02-87",
-					PortUUID:      "portuuid1",
-					OFPort:        35572,
-					ContainerID:   "uci2ucsd6dx87dasuk232312csse",
-				},
-			},
-			expected: `NAMESPACE NAME                   INTERFACE-NAME IP        MAC               PORT-UUID OF-PORT CONTAINER-ID
-default   nginx-32b489d4b7-vgv7v Interface2     127.0.0.2 07-16-76-00-02-87 portuuid1 35572   uci2ucsd6dx 
-default   nginx-6db489d4b7-vgv7v Interface      127.0.0.1 07-16-76-00-02-86 portuuid0 80      dve7a2d6c22 
-`,
-		},
-		{
-			name: "StructuredData-Memberlist-State",
-			rawResponseData: []memberlist.Response{
-				{
-					NodeName: "node1",
-					IP:       "192.168.1.2",
-					Status:   "Alive",
-				},
-				{
-					NodeName: "node2",
-					IP:       "192.168.1.3",
-					Status:   "Dead",
-				},
-			},
-			expected: `NODE  IP          STATUS
-node1 192.168.1.2 Alive 
-node2 192.168.1.3 Dead  
-`,
-		},
-	} {
-		t.Run(tc.name, func(t *testing.T) {
-			var outputBuf bytes.Buffer
-			err := output.TableOutputForGetCommands(tc.rawResponseData, &outputBuf)
-			fmt.Println(outputBuf.String())
-			assert.Nil(t, err)
-			assert.Equal(t, tc.expected, outputBuf.String())
-		})
-	}
-}
-
 // TestFormat ensures the formatter and AddonTransform works as expected.
 func TestFormat(t *testing.T) {
 	for _, tc := range []struct {
@@ -908,106 +593,6 @@ func TestGetRequestErrorFallback(t *testing.T) {
 	}
 }
 
-func TestTableOutputForQueryEndpoint(t *testing.T) {
-	policyRef0 := controllernetworkpolicy.PolicyRef{Namespace: "testNamespace", Name: "test-ingress-egress", UID: "uid-1"}
-	policyRef1 := controllernetworkpolicy.PolicyRef{Namespace: "testNamespace", Name: "default-deny-egress", UID: "uid-2"}
-	tc := []struct {
-		name            string
-		rawResponseData interface{}
-		expected        string
-	}{
-		{
-			name: "Pod selected by no policy",
-			rawResponseData: &controllernetworkpolicy.EndpointQueryResponse{
-				Endpoints: []controllernetworkpolicy.Endpoint{
-					{Namespace: "testNamespace", Name: "podA", Policies: []controllernetworkpolicy.Policy{}, Rules: []controllernetworkpolicy.Rule{}},
-				},
-			},
-			expected: `Endpoint testNamespace/podA
-Applied Policies: None
-
-Egress Rules: None
-
-Ingress Rules: None
-
-`,
-		},
-		{
-			name: "Pod selected by 1 policy",
-			rawResponseData: &controllernetworkpolicy.EndpointQueryResponse{
-				Endpoints: []controllernetworkpolicy.Endpoint{
-					{
-						Namespace: "testNamespace",
-						Name:      "podA",
-						Policies:  []controllernetworkpolicy.Policy{{PolicyRef: policyRef0}},
-						Rules: []controllernetworkpolicy.Rule{
-							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionOut, RuleIndex: 0},
-							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionIn, RuleIndex: 0},
-						},
-					},
-				},
-			},
-			expected: `Endpoint testNamespace/podA
-Applied Policies:
-Name                Namespace     UID  
-test-ingress-egress testNamespace uid-1
-
-Egress Rules:
-Name                Namespace     Index UID  
-test-ingress-egress testNamespace 0     uid-1
-
-Ingress Rules:
-Name                Namespace     Index UID  
-test-ingress-egress testNamespace 0     uid-1
-
-`,
-		},
-		{
-			name: "Pod selected by 2 different policies",
-			rawResponseData: &controllernetworkpolicy.EndpointQueryResponse{
-				Endpoints: []controllernetworkpolicy.Endpoint{
-					{
-						Namespace: "testNamespace",
-						Name:      "podA",
-						Policies: []controllernetworkpolicy.Policy{
-							{PolicyRef: policyRef0},
-							{PolicyRef: policyRef1},
-						},
-						Rules: []controllernetworkpolicy.Rule{
-							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionOut, RuleIndex: 0},
-							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionIn, RuleIndex: 0},
-						},
-					},
-				},
-			},
-			expected: `Endpoint testNamespace/podA
-Applied Policies:
-Name                Namespace     UID  
-default-deny-egress testNamespace uid-2
-test-ingress-egress testNamespace uid-1
-
-Egress Rules:
-Name                Namespace     Index UID  
-test-ingress-egress testNamespace 0     uid-1
-
-Ingress Rules:
-Name                Namespace     Index UID  
-test-ingress-egress testNamespace 0     uid-1
-
-`,
-		},
-	}
-	for _, tt := range tc {
-		t.Run(tt.name, func(t *testing.T) {
-			cd := &commandDefinition{}
-			var outputBuf bytes.Buffer
-			err := cd.tableOutputForQueryEndpoint(tt.rawResponseData, &outputBuf)
-			assert.Nil(t, err)
-			assert.Equal(t, tt.expected, outputBuf.String())
-		})
-	}
-}
-
 func TestCollectFlags(t *testing.T) {
 	tc := []struct {
 		name          string
diff --git a/pkg/antctl/output/output_test.go b/pkg/antctl/output/output_test.go
new file mode 100644
index 000000000..839547e32
--- /dev/null
+++ b/pkg/antctl/output/output_test.go
@@ -0,0 +1,443 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package output
+
+import (
+	"bytes"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"antrea.io/antrea/pkg/agent/apiserver/handlers/agentinfo"
+	"antrea.io/antrea/pkg/agent/apiserver/handlers/memberlist"
+	"antrea.io/antrea/pkg/agent/apiserver/handlers/podinterface"
+	"antrea.io/antrea/pkg/antctl/transform/addressgroup"
+	"antrea.io/antrea/pkg/antctl/transform/appliedtogroup"
+	"antrea.io/antrea/pkg/antctl/transform/common"
+	"antrea.io/antrea/pkg/antctl/transform/controllerinfo"
+	"antrea.io/antrea/pkg/antctl/transform/networkpolicy"
+	cpv1beta "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/crd/v1beta1"
+	endpointserver "antrea.io/antrea/pkg/apiserver/handlers/endpoint"
+)
+
+type Foobar struct {
+	Foo string `json:"foo"`
+}
+
+var (
+	AntreaPolicyTierPriority = int32(250)
+	AntreaPolicyPriority     = float64(1.0)
+)
+
+func TestCommandList_tableOutputForGetCommands(t *testing.T) {
+	for _, tc := range []struct {
+		name            string
+		rawResponseData interface{}
+		expected        string
+	}{
+		{
+			name: "StructureData-ControllerInfo-Single",
+			rawResponseData: controllerinfo.Response{
+				Version: "v0.4.0",
+				PodRef: v1.ObjectReference{
+					Kind:      "Pod",
+					Namespace: "kube-system",
+					Name:      "antrea-controller-55b9bcd59f-h9ll4",
+				},
+				NodeRef: v1.ObjectReference{
+					Kind: "Node",
+					Name: "node-control-plane",
+				},
+				ServiceRef: v1.ObjectReference{
+					Kind: "Service",
+					Name: "antrea",
+				},
+				NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+					NetworkPolicyNum:  1,
+					AddressGroupNum:   1,
+					AppliedToGroupNum: 2,
+				},
+				ConnectedAgentNum: 2,
+				ControllerConditions: []v1beta1.ControllerCondition{
+					{
+						Type:              "ControllerHealthy",
+						Status:            "True",
+						LastHeartbeatTime: metav1.NewTime(time.Now()),
+					},
+				},
+			},
+			expected: `POD                                            NODE               STATUS  NETWORK-POLICIES ADDRESS-GROUPS APPLIED-TO-GROUPS CONNECTED-AGENTS
+kube-system/antrea-controller-55b9bcd59f-h9ll4 node-control-plane Healthy 1                1              2                 2               
+`,
+		},
+		{
+			name: "StructureData-AgentInfo-Single",
+			rawResponseData: agentinfo.AntreaAgentInfoResponse{
+				Version: "v0.4.0",
+				PodRef: v1.ObjectReference{
+					Kind:      "Pod",
+					Namespace: "kube-system",
+					Name:      "antrea-agent-0",
+				},
+				NodeRef: v1.ObjectReference{
+					Kind: "Node",
+					Name: "node-worker",
+				},
+				NodeSubnets: []string{"192.168.1.0/24", "192.168.1.1/24"},
+				OVSInfo: v1beta1.OVSInfo{
+					Version:    "1.0",
+					BridgeName: "br-int",
+					FlowTable: map[string]int32{
+						"0":  5,
+						"10": 7,
+					},
+				},
+				NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+					NetworkPolicyNum:  1,
+					AddressGroupNum:   1,
+					AppliedToGroupNum: 2,
+				},
+				LocalPodNum: 3,
+				AgentConditions: []v1beta1.AgentCondition{
+					{
+						Type:              "AgentHealthy",
+						Status:            "True",
+						LastHeartbeatTime: metav1.NewTime(time.Now()),
+					},
+				},
+			},
+			expected: `POD                        NODE        STATUS  NODE-SUBNET                   NETWORK-POLICIES ADDRESS-GROUPS APPLIED-TO-GROUPS LOCAL-PODS
+kube-system/antrea-agent-0 node-worker Healthy 192.168.1.0/24,192.168.1.1/24 1                1              2                 3         
+`,
+		},
+		{
+			name:            "StructureData-NonTableOutput-Single",
+			rawResponseData: Foobar{Foo: "foo"},
+			expected: `foo            
+foo            
+`,
+		},
+		{
+			name: "StructureData-NonTableOutput-List",
+			rawResponseData: []Foobar{
+				{Foo: "foo1"},
+				{Foo: "foo2"},
+			},
+			expected: `foo            
+foo1           
+foo2           
+`,
+		},
+		{
+			name: "StructureData-NetworkPolicy-List-HasSummary-RandomFieldOrder",
+			rawResponseData: []networkpolicy.Response{
+				{
+					NetworkPolicy: &cpv1beta.NetworkPolicy{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "6001549b-ba63-4752-8267-30f52b4332db",
+						},
+						AppliedToGroups: []string{"32ef631b-6817-5a18-86eb-93f4abf0467c", "c4c59cfe-9160-5de5-a85b-01a58d11963e"},
+						Rules: []cpv1beta.NetworkPolicyRule{
+							{
+								Direction: "In",
+								Services:  nil,
+							},
+						},
+						SourceRef: &cpv1beta.NetworkPolicyReference{
+							Type:      cpv1beta.K8sNetworkPolicy,
+							Namespace: "default",
+							Name:      "allow-all",
+							UID:       "6001549b-ba63-4752-8267-30f52b4332db",
+						},
+					},
+				},
+				{
+					NetworkPolicy: &cpv1beta.NetworkPolicy{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "880db7e8-fc2a-4030-aefe-09afc5f341ad",
+						},
+						TierPriority:    &AntreaPolicyTierPriority,
+						Priority:        &AntreaPolicyPriority,
+						AppliedToGroups: []string{"32ef631b-6817-5a18-86eb-93f4abf0467c"},
+						Rules: []cpv1beta.NetworkPolicyRule{
+							{
+								Direction: "In",
+								Services:  nil,
+							},
+							{
+								Direction: "In",
+								Services:  nil,
+							},
+						},
+						SourceRef: &cpv1beta.NetworkPolicyReference{
+							Type:      cpv1beta.AntreaNetworkPolicy,
+							Namespace: "default",
+							Name:      "allow-all",
+							UID:       "880db7e8-fc2a-4030-aefe-09afc5f341ad",
+						},
+					},
+				},
+			},
+			expected: `NAME                                 APPLIED-TO                                       RULES SOURCE                                TIER-PRIORITY PRIORITY
+6001549b-ba63-4752-8267-30f52b4332db 32ef631b-6817-5a18-86eb-93f4abf0467c + 1 more... 1     K8sNetworkPolicy:default/allow-all    <NONE>        <NONE>  
+880db7e8-fc2a-4030-aefe-09afc5f341ad 32ef631b-6817-5a18-86eb-93f4abf0467c             2     AntreaNetworkPolicy:default/allow-all 250           1       
+`,
+		},
+		{
+			name: "StructureData-AddressGroup-List-HasSummary-HasEmpty",
+			rawResponseData: []addressgroup.Response{
+				{
+					Name: "GroupName1",
+					Pods: []common.GroupMember{
+						{IP: "127.0.0.1"}, {IP: "192.168.0.1"}, {IP: "127.0.0.2"},
+						{IP: "127.0.0.3"}, {IP: "10.0.0.3"}, {IP: "127.0.0.5"}, {IP: "127.0.0.6"},
+					},
+				},
+				{
+					Name: "GroupName2",
+					Pods: []common.GroupMember{},
+				},
+			},
+			expected: `NAME       POD-IPS                                            NODE-IPS
+GroupName1 10.0.0.3,127.0.0.1,127.0.0.2,127.0.0.3 + 2 more... <NONE>  
+GroupName2 <NONE>                                             <NONE>  
+`,
+		},
+		{
+			name: "StructureData-AddressGroup-HasNode",
+			rawResponseData: []addressgroup.Response{
+				{
+					Name: "AddressGroupNameHasNode",
+					Nodes: []common.GroupMember{
+						{IP: "127.0.0.1"}, {IP: "192.168.0.1"}, {IP: "10.176.27.105"}, {IP: "127.0.0.3"},
+					},
+				},
+				{
+					Name: "AddressGroupNameHasPod",
+					Pods: []common.GroupMember{
+						{IP: "127.0.0.1"}, {IP: "192.168.0.1"}, {IP: "127.0.0.2"},
+						{IP: "127.0.0.3"}, {IP: "10.0.0.3"}, {IP: "127.0.0.5"}, {IP: "127.0.0.6"},
+					},
+				},
+				{
+					Name: "AddressGroupNameNone",
+					Pods: []common.GroupMember{},
+				},
+			},
+			expected: `NAME                    POD-IPS                                            NODE-IPS                                     
+AddressGroupNameHasNode <NONE>                                             10.176.27.105,127.0.0.1,127.0.0.3,192.168.0.1
+AddressGroupNameHasPod  10.0.0.3,127.0.0.1,127.0.0.2,127.0.0.3 + 2 more... <NONE>                                       
+AddressGroupNameNone    <NONE>                                             <NONE>                                       
+`,
+		},
+		{
+			name: "StructureData-AppliedToGroup-Single-NoSummary",
+			rawResponseData: appliedtogroup.Response{
+				Name: "GroupName",
+				Pods: []common.GroupMember{
+					{Pod: &cpv1beta.PodReference{
+						Name:      "nginx-6db489d4b7-324rc",
+						Namespace: "PodNamespace",
+					}},
+					{Pod: &cpv1beta.PodReference{
+						Name:      "nginx-6db489d4b7-vgv7v",
+						Namespace: "PodNamespace",
+					}},
+				},
+			},
+			expected: `NAME      PODS                                           
+GroupName PodNamespace/nginx-6db489d4b7-324rc + 1 more...
+`,
+		},
+		{
+			name:            "StructureData-NetworkPolicy-List-EmptyRespCase",
+			rawResponseData: []networkpolicy.Response{},
+			expected:        "\n",
+		},
+		{
+			name: "StructureData-AppliedToGroup-Single-EmptyRespCase",
+			rawResponseData: appliedtogroup.Response{
+				Name: "GroupName",
+				Pods: []common.GroupMember{},
+			},
+			expected: `NAME      PODS  
+GroupName <NONE>
+`,
+		},
+		{
+			name: "StructureData-PodInterface-List",
+			rawResponseData: []podinterface.Response{
+				{
+					PodName:       "nginx-6db489d4b7-vgv7v",
+					PodNamespace:  "default",
+					InterfaceName: "Interface",
+					IPs:           []string{"127.0.0.1"},
+					MAC:           "07-16-76-00-02-86",
+					PortUUID:      "portuuid0",
+					OFPort:        80,
+					ContainerID:   "dve7a2d6c224otm9m0eas8dtwr78",
+				},
+				{
+					PodName:       "nginx-32b489d4b7-vgv7v",
+					PodNamespace:  "default",
+					InterfaceName: "Interface2",
+					IPs:           []string{"127.0.0.2"},
+					MAC:           "07-16-76-00-02-87",
+					PortUUID:      "portuuid1",
+					OFPort:        35572,
+					ContainerID:   "uci2ucsd6dx87dasuk232312csse",
+				},
+			},
+			expected: `NAMESPACE NAME                   INTERFACE-NAME IP        MAC               PORT-UUID OF-PORT CONTAINER-ID
+default   nginx-32b489d4b7-vgv7v Interface2     127.0.0.2 07-16-76-00-02-87 portuuid1 35572   uci2ucsd6dx 
+default   nginx-6db489d4b7-vgv7v Interface      127.0.0.1 07-16-76-00-02-86 portuuid0 80      dve7a2d6c22 
+`,
+		},
+		{
+			name: "StructuredData-Memberlist-State",
+			rawResponseData: []memberlist.Response{
+				{
+					NodeName: "node1",
+					IP:       "192.168.1.2",
+					Status:   "Alive",
+				},
+				{
+					NodeName: "node2",
+					IP:       "192.168.1.3",
+					Status:   "Dead",
+				},
+			},
+			expected: `NODE  IP          STATUS
+node1 192.168.1.2 Alive 
+node2 192.168.1.3 Dead  
+`,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			var outputBuf bytes.Buffer
+			err := TableOutputForGetCommands(tc.rawResponseData, &outputBuf)
+			fmt.Println(outputBuf.String())
+			assert.Nil(t, err)
+			assert.Equal(t, tc.expected, outputBuf.String())
+		})
+	}
+}
+
+func TestTableOutputForQueryEndpoint(t *testing.T) {
+	policyRef0 := cpv1beta.NetworkPolicyReference{Namespace: "testNamespace", Name: "test-ingress-egress", UID: "uid-1", Type: cpv1beta.AntreaNetworkPolicy}
+	policyRef1 := cpv1beta.NetworkPolicyReference{Namespace: "testNamespace", Name: "default-deny-egress", UID: "uid-2", Type: cpv1beta.AntreaNetworkPolicy}
+	tc := []struct {
+		name            string
+		rawResponseData interface{}
+		expected        string
+	}{
+		{
+			name: "Pod selected by no policy",
+			rawResponseData: &endpointserver.EndpointQueryResponse{
+				Endpoints: []endpointserver.Endpoint{
+					{Namespace: "testNamespace", Name: "podA", AppliedPolicies: []cpv1beta.NetworkPolicyReference{}, EgressDstRules: []endpointserver.Rule{}, IngressSrcRules: []endpointserver.Rule{}},
+				},
+			},
+			expected: `Endpoint testNamespace/podA
+Applied Policies on Endpoint: None
+
+Egress Rules Referencing Endpoint as Destination: None
+
+Ingress Rules Referencing Endpoint as Source: None
+
+`,
+		},
+		{
+			name: "Pod selected by 1 policy",
+			rawResponseData: &endpointserver.EndpointQueryResponse{
+				Endpoints: []endpointserver.Endpoint{
+					{
+						Namespace:       "testNamespace",
+						Name:            "podA",
+						AppliedPolicies: []cpv1beta.NetworkPolicyReference{policyRef0},
+						EgressDstRules: []endpointserver.Rule{
+							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionOut, RuleIndex: 0},
+						},
+						IngressSrcRules: []endpointserver.Rule{
+							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionIn, RuleIndex: 0},
+						},
+					},
+				},
+			},
+			expected: `Endpoint testNamespace/podA
+Applied Policies on Endpoint:
+Name                Namespace     UID  
+test-ingress-egress testNamespace uid-1
+
+Egress Rules Referencing Endpoint as Destination:
+Name                Namespace     Index UID  
+test-ingress-egress testNamespace 0     uid-1
+
+Ingress Rules Referencing Endpoint as Source:
+Name                Namespace     Index UID  
+test-ingress-egress testNamespace 0     uid-1
+
+`,
+		},
+		{
+			name: "Pod selected by 2 different policies",
+			rawResponseData: &endpointserver.EndpointQueryResponse{
+				Endpoints: []endpointserver.Endpoint{
+					{
+						Namespace: "testNamespace",
+						Name:      "podA",
+						AppliedPolicies: []cpv1beta.NetworkPolicyReference{
+							policyRef0, policyRef1,
+						},
+						EgressDstRules: []endpointserver.Rule{
+							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionOut, RuleIndex: 0},
+						},
+						IngressSrcRules: []endpointserver.Rule{
+							{PolicyRef: policyRef0, Direction: cpv1beta.DirectionIn, RuleIndex: 0},
+						},
+					},
+				},
+			},
+			expected: `Endpoint testNamespace/podA
+Applied Policies on Endpoint:
+Name                Namespace     UID  
+default-deny-egress testNamespace uid-2
+test-ingress-egress testNamespace uid-1
+
+Egress Rules Referencing Endpoint as Destination:
+Name                Namespace     Index UID  
+test-ingress-egress testNamespace 0     uid-1
+
+Ingress Rules Referencing Endpoint as Source:
+Name                Namespace     Index UID  
+test-ingress-egress testNamespace 0     uid-1
+
+`,
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			var outputBuf bytes.Buffer
+			err := TableOutputForQueryEndpoint(tt.rawResponseData, &outputBuf)
+			assert.Nil(t, err)
+			assert.Equal(t, tt.expected, outputBuf.String())
+		})
+	}
+}
diff --git a/pkg/apiserver/handlers/endpoint/handler_test.go b/pkg/apiserver/handlers/endpoint/handler_test.go
index 7b63546b2..bb2cb0039 100644
--- a/pkg/apiserver/handlers/endpoint/handler_test.go
+++ b/pkg/apiserver/handlers/endpoint/handler_test.go
@@ -22,11 +22,11 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"go.uber.org/mock/gomock"
-	v1 "k8s.io/api/admission/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
 
-	"antrea.io/antrea/pkg/controller/networkpolicy"
+	"antrea.io/antrea/pkg/apis/controlplane"
+	"antrea.io/antrea/pkg/apis/controlplane/v1beta2"
 	queriermock "antrea.io/antrea/pkg/controller/networkpolicy/testing"
+	antreatypes "antrea.io/antrea/pkg/controller/types"
 )
 
 type TestCase struct {
@@ -34,53 +34,12 @@ type TestCase struct {
 	handlerRequest string
 	expectedStatus int
 	// expected result written by handler function
-	expectedContent response
+	expectedResponse *EndpointQueryResponse
 
 	// arguments of call to mock
 	argsMock []string
 	// results of call to mock
-	mockQueryResponse response
-}
-
-type response struct {
-	response *networkpolicy.EndpointQueryResponse
-	error    error
-}
-
-var responses = []response{
-	{
-		response: &networkpolicy.EndpointQueryResponse{Endpoints: nil},
-		error:    errors.NewNotFound(v1.Resource("pod"), "pod"),
-	},
-	{
-		response: &networkpolicy.EndpointQueryResponse{Endpoints: []networkpolicy.Endpoint{
-			{
-				Policies: []networkpolicy.Policy{
-					{
-						PolicyRef: networkpolicy.PolicyRef{Name: "policy1"},
-					},
-				},
-			},
-		},
-		},
-		error: nil,
-	},
-	{
-		response: &networkpolicy.EndpointQueryResponse{Endpoints: []networkpolicy.Endpoint{
-			{
-				Policies: []networkpolicy.Policy{
-					{
-						PolicyRef: networkpolicy.PolicyRef{Name: "policy1"},
-					},
-					{
-						PolicyRef: networkpolicy.PolicyRef{Name: "policy2"},
-					},
-				},
-			},
-		},
-		},
-		error: nil,
-	},
+	mockQueryResponse *antreatypes.EndpointNetworkPolicyRules
 }
 
 // TestIncompleteArguments tests how the handler function responds when the user passes in a query command
@@ -116,13 +75,10 @@ func TestInvalidArguments(t *testing.T) {
 	// outline test cases with expected behavior
 	testCases := map[string]TestCase{
 		"Responds with error given no invalid selection": {
-			handlerRequest: "?namespace=namespace&pod=pod",
-			expectedStatus: http.StatusNotFound,
-			argsMock:       []string{namespace, pod},
-			mockQueryResponse: response{
-				response: nil,
-				error:    nil,
-			},
+			handlerRequest:    "?namespace=namespace&pod=pod",
+			expectedStatus:    http.StatusNotFound,
+			argsMock:          []string{namespace, pod},
+			mockQueryResponse: nil,
 		},
 	}
 
@@ -130,7 +86,7 @@ func TestInvalidArguments(t *testing.T) {
 
 }
 
-// TestSinglePolicyResponse tests how the handler function responds when the user passes in a endpoint with a
+// TestSinglePolicyResponse tests how the handler function responds when the user passes in an endpoint with a
 // single policy response
 func TestSinglePolicyResponse(t *testing.T) {
 	mockCtrl := gomock.NewController(t)
@@ -139,22 +95,33 @@ func TestSinglePolicyResponse(t *testing.T) {
 	// outline test cases with expected behavior
 	testCases := map[string]TestCase{
 		"Responds with list of single element": {
-			handlerRequest:  "?namespace=namespace&pod=pod",
-			expectedStatus:  http.StatusOK,
-			expectedContent: responses[1],
-			argsMock:        []string{namespace, pod},
-			mockQueryResponse: response{
-				response: &networkpolicy.EndpointQueryResponse{Endpoints: []networkpolicy.Endpoint{
+			handlerRequest: "?namespace=namespace&pod=pod",
+			expectedStatus: http.StatusOK,
+			expectedResponse: &EndpointQueryResponse{Endpoints: []Endpoint{
+				{
+					AppliedPolicies: []v1beta2.NetworkPolicyReference{
+						{Name: "policy1"},
+					},
+					IngressSrcRules: []Rule{
+						{PolicyRef: v1beta2.NetworkPolicyReference{Name: "policy2"}},
+					},
+				},
+			},
+			},
+			argsMock: []string{namespace, pod},
+			mockQueryResponse: &antreatypes.EndpointNetworkPolicyRules{
+				AppliedPolicies: []*controlplane.NetworkPolicyReference{
+					{Name: "policy1"},
+				},
+				EndpointAsIngressSrcRules: []*antreatypes.RuleInfo{
 					{
-						Policies: []networkpolicy.Policy{
-							{
-								PolicyRef: networkpolicy.PolicyRef{Name: "policy1"},
-							},
+						Policy: &controlplane.NetworkPolicyReference{Name: "policy2"},
+						Index:  0,
+						Rule: &controlplane.NetworkPolicyRule{
+							Direction: controlplane.DirectionIn,
 						},
 					},
 				},
-				},
-				error: nil,
 			},
 		},
 	}
@@ -163,7 +130,7 @@ func TestSinglePolicyResponse(t *testing.T) {
 
 }
 
-// TestMultiPolicyResponse tests how the handler function responds when the user passes in a endpoint with
+// TestMultiPolicyResponse tests how the handler function responds when the user passes in an endpoint with
 // multiple policy responses
 func TestMultiPolicyResponse(t *testing.T) {
 	mockCtrl := gomock.NewController(t)
@@ -172,25 +139,21 @@ func TestMultiPolicyResponse(t *testing.T) {
 	// outline test cases with expected behavior
 	testCases := map[string]TestCase{
 		"Responds with list of single element": {
-			handlerRequest:  "?namespace=namespace&pod=pod",
-			expectedStatus:  http.StatusOK,
-			expectedContent: responses[2],
-			argsMock:        []string{namespace, pod},
-			mockQueryResponse: response{
-				response: &networkpolicy.EndpointQueryResponse{Endpoints: []networkpolicy.Endpoint{
-					{
-						Policies: []networkpolicy.Policy{
-							{
-								PolicyRef: networkpolicy.PolicyRef{Name: "policy1"},
-							},
-							{
-								PolicyRef: networkpolicy.PolicyRef{Name: "policy2"},
-							},
-						},
+			handlerRequest: "?namespace=namespace&pod=pod",
+			expectedStatus: http.StatusOK,
+			expectedResponse: &EndpointQueryResponse{Endpoints: []Endpoint{
+				{
+					AppliedPolicies: []v1beta2.NetworkPolicyReference{
+						{Name: "policy1"}, {Name: "policy2"},
 					},
 				},
+			},
+			},
+			argsMock: []string{namespace, pod},
+			mockQueryResponse: &antreatypes.EndpointNetworkPolicyRules{
+				AppliedPolicies: []*controlplane.NetworkPolicyReference{
+					{Name: "policy1"}, {Name: "policy2"},
 				},
-				error: nil,
 			},
 		},
 	}
@@ -204,7 +167,7 @@ func evaluateTestCases(testCases map[string]TestCase, mockCtrl *gomock.Controlle
 		// create mock querier with expected behavior outlined in testCase
 		mockQuerier := queriermock.NewMockEndpointQuerier(mockCtrl)
 		if tc.expectedStatus != http.StatusBadRequest {
-			mockQuerier.EXPECT().QueryNetworkPolicies(tc.argsMock[0], tc.argsMock[1]).Return(tc.mockQueryResponse.response, tc.mockQueryResponse.error)
+			mockQuerier.EXPECT().QueryNetworkPolicyRules(tc.argsMock[0], tc.argsMock[1]).Return(tc.mockQueryResponse, nil)
 		}
 		// initialize handler with mockQuerier
 		handler := HandleFunc(mockQuerier)
@@ -218,11 +181,17 @@ func evaluateTestCases(testCases map[string]TestCase, mockCtrl *gomock.Controlle
 			return
 		}
 		// check response is expected
-		var received networkpolicy.EndpointQueryResponse
+		var received EndpointQueryResponse
 		err = json.Unmarshal(recorder.Body.Bytes(), &received)
 		assert.Nil(t, err)
-		for i, policy := range tc.expectedContent.response.Endpoints[0].Policies {
-			assert.Equal(t, policy.Name, received.Endpoints[0].Policies[i].Name)
+		for i, policy := range tc.expectedResponse.Endpoints[0].AppliedPolicies {
+			assert.Equal(t, policy.Name, received.Endpoints[0].AppliedPolicies[i].Name)
+		}
+		for i, rule := range tc.expectedResponse.Endpoints[0].IngressSrcRules {
+			assert.Equal(t, rule.PolicyRef.Name, received.Endpoints[0].IngressSrcRules[i].PolicyRef.Name)
+		}
+		for i, rule := range tc.expectedResponse.Endpoints[0].EgressDstRules {
+			assert.Equal(t, rule.PolicyRef.Name, received.Endpoints[0].EgressDstRules[i].PolicyRef.Name)
 		}
 	}
 }
diff --git a/pkg/controller/networkpolicy/endpoint_querier_perf_test.go b/pkg/controller/networkpolicy/endpoint_querier_perf_test.go
index 386344d4b..19911dfdb 100644
--- a/pkg/controller/networkpolicy/endpoint_querier_perf_test.go
+++ b/pkg/controller/networkpolicy/endpoint_querier_perf_test.go
@@ -75,9 +75,9 @@ func testQueryEndpoint(t *testing.T, maxExecutionTime time.Duration, namespaces
 	// track execution time by calling query endpoint 1000 times on random pods
 	for i := 0; i < 1000; i++ {
 		pod, namespace := pods[i].(*v1.Pod).Name, pods[i].(*v1.Pod).Namespace
-		response, err := querier.QueryNetworkPolicies(namespace, pod)
+		response, err := querier.QueryNetworkPolicyRules(namespace, pod)
 		require.Equal(t, err, nil)
-		require.Equal(t, len(response.Endpoints[0].Policies), responseLength)
+		require.Equal(t, len(response.AppliedPolicies), responseLength)
 	}
 	// Stop tracking go routines
 	stopCh <- struct{}{}
diff --git a/pkg/controller/networkpolicy/endpoint_querier_test.go b/pkg/controller/networkpolicy/endpoint_querier_test.go
index c9a0a0a80..e1690b108 100644
--- a/pkg/controller/networkpolicy/endpoint_querier_test.go
+++ b/pkg/controller/networkpolicy/endpoint_querier_test.go
@@ -26,7 +26,8 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 
-	"antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/controlplane"
+	antreatypes "antrea.io/antrea/pkg/controller/types"
 )
 
 // pods represent kubernetes pods for testing proper query results
@@ -79,15 +80,16 @@ var pods = []*corev1.Pod{
 
 // polices represent kubernetes policies for testing proper query results
 //
-// policy 0: select all pods and deny default ingress
-// policy 1: select all pods and deny default egress
+// policy 0: select all matching pods and allow ingress and egress from matching pods
+// policy 1: select all matching pods and deny default egress
+// policy 2: select all matching pods and allow ingress from multiple matching pods
 
 var policies = []*networkingv1.NetworkPolicy{
 	{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "test-ingress-egress",
 			Namespace: "testNamespace",
-			UID:       types.UID("uid-1"),
+			UID:       types.UID("uid-0"),
 		},
 		Spec: networkingv1.NetworkPolicySpec{
 			PodSelector: metav1.LabelSelector{
@@ -125,7 +127,7 @@ var policies = []*networkingv1.NetworkPolicy{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "default-deny-egress",
 			Namespace: "testNamespace",
-			UID:       types.UID("uid-2"),
+			UID:       types.UID("uid-1"),
 		},
 		Spec: networkingv1.NetworkPolicySpec{
 			PodSelector: metav1.LabelSelector{
@@ -140,7 +142,7 @@ var policies = []*networkingv1.NetworkPolicy{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "test-multiple-ingress-rules",
 			Namespace: "testNamespace",
-			UID:       types.UID("uid-3"),
+			UID:       types.UID("uid-2"),
 		},
 		Spec: networkingv1.NetworkPolicySpec{
 			PodSelector: metav1.LabelSelector{
@@ -212,117 +214,118 @@ func makeControllerAndEndpointQuerier(objects ...runtime.Object) *endpointQuerie
 	return querier
 }
 
-func TestEndpointQuery(t *testing.T) {
-	policyRef0 := PolicyRef{policies[0].Namespace, policies[0].Name, policies[0].UID}
-	policyRef1 := PolicyRef{policies[1].Namespace, policies[1].Name, policies[1].UID}
-	policyRef2 := PolicyRef{policies[2].Namespace, policies[2].Name, policies[2].UID}
+func TestQueryNetworkPolicyRules(t *testing.T) {
+	policyRef := controlplane.NetworkPolicyReference{Type: controlplane.K8sNetworkPolicy, Namespace: policies[0].Namespace, Name: policies[0].Name, UID: policies[0].UID}
+	policyRef1 := controlplane.NetworkPolicyReference{Type: controlplane.K8sNetworkPolicy, Namespace: policies[1].Namespace, Name: policies[1].Name, UID: policies[1].UID}
+	policyRef2 := controlplane.NetworkPolicyReference{Type: controlplane.K8sNetworkPolicy, Namespace: policies[2].Namespace, Name: policies[2].Name, UID: policies[2].UID}
+	ns, podA := "testNamespace", "podA"
 
 	testCases := []struct {
 		name             string
 		objs             []runtime.Object
 		podNamespace     string
 		podName          string
-		expectedResponse *EndpointQueryResponse
+		expectedResponse *antreatypes.EndpointNetworkPolicyRules
 	}{
 		{
-			"InvalidSelector", // provided Namespace / Name does not match any Pod
-			[]runtime.Object{},
-			"non-existing-namespace",
-			"non-existing-pod",
-			nil,
+			name:         "No matching pod",
+			objs:         []runtime.Object{},
+			podNamespace: "non-existing-namespace",
+			podName:      "non-existing-pod",
 		},
 		{
-			"NoPolicy", // Pod is not selected by any policy
-			[]runtime.Object{namespaces[0], pods[0]},
-			"testNamespace",
-			"podA",
-			&EndpointQueryResponse{
-				[]Endpoint{
-					{Namespace: "testNamespace", Name: "podA", Policies: []Policy{}, Rules: []Rule{}},
-				},
-			},
+			name:             "Empty response",
+			objs:             []runtime.Object{namespaces[0], pods[0]},
+			podNamespace:     ns,
+			podName:          podA,
+			expectedResponse: &antreatypes.EndpointNetworkPolicyRules{Namespace: ns, Name: podA},
 		},
 		{
-			"SingleAppliedIngressEgressPolicy", // Pod is selected by a single policy
-			[]runtime.Object{namespaces[0], pods[0], policies[0]},
-			"testNamespace",
-			"podA",
-			&EndpointQueryResponse{
-				[]Endpoint{
-					{
-						Namespace: "testNamespace",
-						Name:      "podA",
-						Policies:  []Policy{{policyRef0}},
-						Rules: []Rule{
-							{policyRef0, v1beta2.DirectionOut, 0},
-							{policyRef0, v1beta2.DirectionIn, 0},
-						},
-					},
+			name:    "Default namespace",
+			objs:    []runtime.Object{namespaces[0], pods[0]},
+			podName: podA,
+		},
+		{
+			name:         "Single KNP applied with ingress and egress rules",
+			objs:         []runtime.Object{namespaces[0], pods[0], policies[0]},
+			podNamespace: ns,
+			podName:      podA,
+			expectedResponse: &antreatypes.EndpointNetworkPolicyRules{
+				Namespace:       ns,
+				Name:            podA,
+				AppliedPolicies: []*controlplane.NetworkPolicyReference{&policyRef},
+				EndpointAsIngressSrcRules: []*antreatypes.RuleInfo{
+					{Policy: &policyRef, Rule: &controlplane.NetworkPolicyRule{Direction: controlplane.DirectionIn}},
+				},
+				EndpointAsEgressDstRules: []*antreatypes.RuleInfo{
+					{Policy: &policyRef, Rule: &controlplane.NetworkPolicyRule{Direction: controlplane.DirectionOut}},
 				},
 			},
 		},
 		{
-			"MultiplePolicy", // Pod is selected by different policies
-			[]runtime.Object{namespaces[0], pods[0], policies[0], policies[1]},
-			"testNamespace",
-			"podA",
-			&EndpointQueryResponse{
-				[]Endpoint{
-					{
-						Namespace: "testNamespace",
-						Name:      "podA",
-						Policies: []Policy{
-							{policyRef0},
-							{policyRef1},
-						},
-						Rules: []Rule{
-							{policyRef0, v1beta2.DirectionOut, 0},
-							{policyRef0, v1beta2.DirectionIn, 0},
-						},
-					},
+			name:         "Multiple KNP applied", // Pod is selected by different policies
+			objs:         []runtime.Object{namespaces[0], pods[0], policies[0], policies[1]},
+			podNamespace: ns,
+			podName:      podA,
+			expectedResponse: &antreatypes.EndpointNetworkPolicyRules{
+				Namespace:       ns,
+				Name:            podA,
+				AppliedPolicies: []*controlplane.NetworkPolicyReference{&policyRef, &policyRef1},
+				EndpointAsIngressSrcRules: []*antreatypes.RuleInfo{
+					{Policy: &policyRef, Rule: &controlplane.NetworkPolicyRule{Direction: controlplane.DirectionIn}},
+				},
+				EndpointAsEgressDstRules: []*antreatypes.RuleInfo{
+					{Policy: &policyRef, Rule: &controlplane.NetworkPolicyRule{Direction: controlplane.DirectionOut}},
 				},
 			},
 		},
 		{
-			"MultipleRule", // Pod is selected by policy with multiple rules
-			[]runtime.Object{namespaces[0], pods[0], policies[2]},
-			"testNamespace",
-			"podA",
-			&EndpointQueryResponse{
-				[]Endpoint{
-					{
-						Namespace: "testNamespace",
-						Name:      "podA",
-						Policies: []Policy{
-							{policyRef2},
-						},
-						Rules: []Rule{
-							{policyRef2, v1beta2.DirectionIn, 1},
-						},
-					},
+			name:         "Single KNP applied with multiple ingress rules", // Pod is selected by policy with multiple rules
+			objs:         []runtime.Object{namespaces[0], pods[0], policies[2]},
+			podNamespace: ns,
+			podName:      podA,
+			expectedResponse: &antreatypes.EndpointNetworkPolicyRules{
+				Namespace:       ns,
+				Name:            podA,
+				AppliedPolicies: []*controlplane.NetworkPolicyReference{&policyRef2},
+				EndpointAsIngressSrcRules: []*antreatypes.RuleInfo{
+					{Policy: &policyRef2, Index: 1, Rule: &controlplane.NetworkPolicyRule{Direction: controlplane.DirectionIn}},
 				},
 			},
 		},
 	}
 
+	evaluateResponse := func(expectedRules, responseRules []*antreatypes.RuleInfo) {
+		assert.Equal(t, len(expectedRules), len(responseRules))
+		for idx := range expectedRules {
+			assert.EqualValues(t, expectedRules[idx].Rule.Direction, responseRules[idx].Rule.Direction)
+			assert.Equal(t, expectedRules[idx].Index, responseRules[idx].Index)
+			assert.Equal(t, expectedRules[idx].Policy, responseRules[idx].Policy)
+		}
+		return
+	}
+
 	for _, tc := range testCases {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
 			t.Parallel()
-
 			endpointQuerier := makeControllerAndEndpointQuerier(tc.objs...)
-			response, err := endpointQuerier.QueryNetworkPolicies(tc.podNamespace, tc.podName)
+			response, err := endpointQuerier.QueryNetworkPolicyRules(tc.podNamespace, tc.podName)
 			require.NoErrorf(t, err, "Expected QueryNetworkPolicies to succeed")
 			if tc.expectedResponse == nil {
-				assert.Nil(t, response, "Expected nil response from QueryNetworkPolicies")
+				assert.Nil(t, response, "Expected nil response from QueryNetworkPolicyRules")
 			} else {
-				assert.Len(t, response.Endpoints, 1, "QueryNetworkPolicies should only return responses with a single endpoint")
-				expectedEndpoint := &tc.expectedResponse.Endpoints[0]
-				endpoint := &response.Endpoints[0]
-				assert.Equal(t, expectedEndpoint.Namespace, endpoint.Namespace)
-				assert.Equal(t, expectedEndpoint.Name, endpoint.Name)
-				assert.ElementsMatch(t, expectedEndpoint.Rules, endpoint.Rules)
-				assert.ElementsMatch(t, expectedEndpoint.Policies, endpoint.Policies)
+				assert.Equal(t, tc.expectedResponse.Namespace, response.Namespace)
+				assert.Equal(t, tc.expectedResponse.Name, response.Name)
+				assert.Equal(t, len(tc.expectedResponse.AppliedPolicies), len(response.AppliedPolicies))
+				var expectedPolicies, responsePolicies []*controlplane.NetworkPolicyReference
+				for idx, expected := range tc.expectedResponse.AppliedPolicies {
+					expectedPolicies = append(expectedPolicies, expected)
+					responsePolicies = append(responsePolicies, response.AppliedPolicies[idx])
+				}
+				assert.ElementsMatch(t, expectedPolicies, responsePolicies)
+				evaluateResponse(tc.expectedResponse.EndpointAsIngressSrcRules, response.EndpointAsIngressSrcRules)
+				evaluateResponse(tc.expectedResponse.EndpointAsEgressDstRules, response.EndpointAsEgressDstRules)
 			}
 		})
 	}
