diff --git a/snips_nlu/tests/test_crf_slot_filler.py b/snips_nlu/tests/test_crf_slot_filler.py
index 0e99cfd..5b5e7bb 100644
--- a/snips_nlu/tests/test_crf_slot_filler.py
+++ b/snips_nlu/tests/test_crf_slot_filler.py
@@ -9,7 +9,6 @@ from mock import MagicMock
 from snips_nlu.constants import (
     DATA, END, ENTITY, ENTITY_KIND, LANGUAGE_EN, RES_MATCH_RANGE, SLOT_NAME,
     SNIPS_DATETIME, START, TEXT, VALUE)
-from snips_nlu.dataset import validate_and_format_dataset
 from snips_nlu.pipeline.configs import CRFSlotFillerConfig
 from snips_nlu.preprocessing import Token, tokenize
 from snips_nlu.result import unresolved_slot
@@ -23,12 +22,13 @@ from snips_nlu.slot_filler.feature_factory import (
     IsDigitFactory, NgramFactory, ShapeNgramFactory)
 from snips_nlu.tests.utils import (
     BEVERAGE_DATASET, FixtureTest, SAMPLE_DATASET, TEST_PATH, WEATHER_DATASET)
+from snips_nlu.utils import NotTrained
 
 
 class TestCRFSlotFiller(FixtureTest):
     def test_should_get_slots(self):
         # Given
-        dataset = validate_and_format_dataset(BEVERAGE_DATASET)
+        dataset = BEVERAGE_DATASET
         config = CRFSlotFillerConfig(random_seed=42)
         intent = "MakeTea"
         slot_filler = CRFSlotFiller(config)
@@ -47,7 +47,7 @@ class TestCRFSlotFiller(FixtureTest):
 
     def test_should_get_builtin_slots(self):
         # Given
-        dataset = validate_and_format_dataset(WEATHER_DATASET)
+        dataset = WEATHER_DATASET
         config = CRFSlotFillerConfig(random_seed=42)
         intent = "SearchWeatherForecast"
         slot_filler = CRFSlotFiller(config)
@@ -69,9 +69,74 @@ class TestCRFSlotFiller(FixtureTest):
         ]
         self.assertListEqual(expected_slots, slots)
 
+    def test_should_not_use_crf_when_dataset_with_no_slots(self):
+        # Given
+        dataset = {
+            "language": "en",
+            "intents": {
+                "intent1": {
+                    "utterances": [
+                        {
+                            "data": [
+                                {
+                                    "text": "This is an utterance without "
+                                            "slots"
+                                }
+                            ]
+                        }
+                    ]
+                }
+            },
+            "entities": {}
+        }
+        slot_filler = CRFSlotFiller()
+        mock_compute_features = MagicMock()
+        slot_filler.compute_features = mock_compute_features
+
+        # When
+        slot_filler.fit(dataset, "intent1")
+        slots = slot_filler.get_slots("This is an utterance without slots")
+
+        # Then
+        mock_compute_features.assert_not_called()
+        self.assertListEqual([], slots)
+
+    def test_should_compute_sequence_probability_when_no_slots(self):
+        # Given
+        dataset = {
+            "language": "en",
+            "intents": {
+                "intent1": {
+                    "utterances": [
+                        {
+                            "data": [
+                                {
+                                    "text": "This is an utterance without "
+                                            "slots"
+                                }
+                            ]
+                        }
+                    ]
+                }
+            },
+            "entities": {}
+        }
+        slot_filler = CRFSlotFiller().fit(dataset, "intent1")
+        tokens = tokenize("hello world foo bar", "en")
+
+        # When
+        res1 = slot_filler.get_sequence_probability(
+            tokens, ["O", "O", "O", "O"])
+        res2 = slot_filler.get_sequence_probability(
+            tokens, ["O", "O", "B-location", "O"])
+
+        # Then
+        self.assertEqual(1.0, res1)
+        self.assertEqual(0.0, res2)
+
     def test_should_parse_naughty_strings(self):
         # Given
-        dataset = validate_and_format_dataset(SAMPLE_DATASET)
+        dataset = SAMPLE_DATASET
         naughty_strings_path = TEST_PATH / "resources" / "naughty_strings.txt"
         with naughty_strings_path.open(encoding='utf8') as f:
             naughty_strings = [line.strip("\n") for line in f.readlines()]
@@ -84,6 +149,31 @@ class TestCRFSlotFiller(FixtureTest):
             with self.fail_if_exception("Naughty string crashes"):
                 slot_filler.get_slots(s)
 
+    def test_should_not_get_slots_when_not_fitted(self):
+        # Given
+        slot_filler = CRFSlotFiller()
+
+        # When / Then
+        self.assertFalse(slot_filler.fitted)
+        with self.assertRaises(NotTrained):
+            slot_filler.get_slots("foobar")
+
+    def test_should_not_get_sequence_probability_when_not_fitted(self):
+        # Given
+        slot_filler = CRFSlotFiller()
+
+        # When / Then
+        with self.assertRaises(NotTrained):
+            slot_filler.get_sequence_probability(tokens=[], labels=[])
+
+    def test_should_not_log_weights_when_not_fitted(self):
+        # Given
+        slot_filler = CRFSlotFiller()
+
+        # When / Then
+        with self.assertRaises(NotTrained):
+            slot_filler.log_weights()
+
     def test_should_fit_with_naughty_strings_no_tags(self):
         # Given
         naughty_strings_path = TEST_PATH / "resources" / "naughty_strings.txt"
@@ -138,8 +228,6 @@ class TestCRFSlotFiller(FixtureTest):
             "snips_nlu_version": "0.0.1"
         }
 
-        naughty_dataset = validate_and_format_dataset(naughty_dataset)
-
         # Then
         with self.fail_if_exception("Naughty string make NLU crash"):
             slot_filler = CRFSlotFiller()
@@ -158,7 +246,7 @@ class TestCRFSlotFiller(FixtureTest):
 
     def test_should_get_slots_after_deserialization(self):
         # Given
-        dataset = validate_and_format_dataset(BEVERAGE_DATASET)
+        dataset = BEVERAGE_DATASET
         config = CRFSlotFillerConfig(random_seed=42)
         intent = "MakeTea"
         slot_filler = CRFSlotFiller(config)
@@ -293,7 +381,7 @@ class TestCRFSlotFiller(FixtureTest):
         config = CRFSlotFillerConfig(
             tagging_scheme=TaggingScheme.BILOU,
             feature_factory_configs=features_factories)
-        dataset = validate_and_format_dataset(SAMPLE_DATASET)
+        dataset = SAMPLE_DATASET
 
         slot_filler = CRFSlotFiller(config)
         intent = "dummy_intent_1"
@@ -411,6 +499,86 @@ class TestCRFSlotFiller(FixtureTest):
         crf_path = Path(slot_filler.crf_model.modelfile.name)
         self.assertFileContent(crf_path, "foo bar")
 
+    def test_should_be_serializable_when_fitted_without_slots(self):
+        # Given
+        features_factories = [
+            {
+                "factory_name": ShapeNgramFactory.name,
+                "args": {"n": 1},
+                "offsets": [0]
+            },
+            {
+                "factory_name": IsDigitFactory.name,
+                "args": {},
+                "offsets": [-1, 0]
+            }
+        ]
+        config = CRFSlotFillerConfig(
+            tagging_scheme=TaggingScheme.BILOU,
+            feature_factory_configs=features_factories)
+        dataset = {
+            "language": "en",
+            "intents": {
+                "intent1": {
+                    "utterances": [
+                        {
+                            "data": [
+                                {
+                                    "text": "This is an utterance without "
+                                            "slots"
+                                }
+                            ]
+                        }
+                    ]
+                }
+            },
+            "entities": {}
+        }
+
+        slot_filler = CRFSlotFiller(config)
+        slot_filler.fit(dataset, intent="intent1")
+
+        # When
+        slot_filler.persist(self.tmp_file_path)
+
+        # Then
+        metadata_path = self.tmp_file_path / "metadata.json"
+        self.assertJsonContent(metadata_path, {"unit_name": "crf_slot_filler"})
+        self.assertIsNone(slot_filler.crf_model)
+
+    def test_should_be_deserializable_when_fitted_without_slots(self):
+        # Given
+        dataset = {
+            "language": "en",
+            "intents": {
+                "intent1": {
+                    "utterances": [
+                        {
+                            "data": [
+                                {
+                                    "text": "This is an utterance without "
+                                            "slots"
+                                }
+                            ]
+                        }
+                    ]
+                }
+            },
+            "entities": {}
+        }
+
+        slot_filler = CRFSlotFiller()
+        slot_filler.fit(dataset, intent="intent1")
+        slot_filler.persist(self.tmp_file_path)
+        loaded_slot_filler = CRFSlotFiller.from_path(self.tmp_file_path)
+
+        # When
+        slots = loaded_slot_filler.get_slots(
+            "This is an utterance without slots")
+
+        # Then
+        self.assertListEqual([], slots)
+
     def test_should_be_serializable_into_bytearray(self):
         # Given
         dataset = BEVERAGE_DATASET
@@ -448,7 +616,7 @@ class TestCRFSlotFiller(FixtureTest):
         slot_filler = CRFSlotFiller(slot_filler_config)
 
         tokens = tokenize("foo hello world bar", LANGUAGE_EN)
-        dataset = validate_and_format_dataset(SAMPLE_DATASET)
+        dataset = SAMPLE_DATASET
         slot_filler.fit(dataset, intent="dummy_intent_1")
 
         # When
diff --git a/snips_nlu/tests/test_deterministic_intent_parser.py b/snips_nlu/tests/test_deterministic_intent_parser.py
index 57149c6..4b7602b 100644
--- a/snips_nlu/tests/test_deterministic_intent_parser.py
+++ b/snips_nlu/tests/test_deterministic_intent_parser.py
@@ -17,6 +17,7 @@ from snips_nlu.pipeline.configs import DeterministicIntentParserConfig
 from snips_nlu.result import intent_classification_result, unresolved_slot
 from snips_nlu.tests.utils import FixtureTest, SAMPLE_DATASET, TEST_PATH, \
     BEVERAGE_DATASET
+from snips_nlu.utils import NotTrained
 
 
 class TestDeterministicIntentParser(FixtureTest):
@@ -215,6 +216,15 @@ class TestDeterministicIntentParser(FixtureTest):
         self.assertEqual(intent_name_1, res_1[RES_INTENT][RES_INTENT_NAME])
         self.assertEqual(intent_name_2, res_2[RES_INTENT][RES_INTENT_NAME])
 
+    def test_should_not_parse_when_not_fitted(self):
+        # Given
+        parser = DeterministicIntentParser()
+
+        # When / Then
+        self.assertFalse(parser.fitted)
+        with self.assertRaises(NotTrained):
+            parser.parse("foobar")
+
     def test_should_get_intent_after_deserialization(self):
         # Given
         dataset = validate_and_format_dataset(self.slots_dataset)
diff --git a/snips_nlu/tests/test_log_reg_intent_classifier.py b/snips_nlu/tests/test_log_reg_intent_classifier.py
index a859d5f..ad2867c 100644
--- a/snips_nlu/tests/test_log_reg_intent_classifier.py
+++ b/snips_nlu/tests/test_log_reg_intent_classifier.py
@@ -20,9 +20,9 @@ from snips_nlu.pipeline.configs import (
     IntentClassifierDataAugmentationConfig, LogRegIntentClassifierConfig)
 from snips_nlu.tests.utils import (
     BEVERAGE_DATASET, FixtureTest, SAMPLE_DATASET, get_empty_dataset)
+from snips_nlu.utils import NotTrained
 
 
-# pylint: disable=W0613
 def get_mocked_augment_utterances(dataset, intent_name, language,
                                   min_utterances, capitalization_ratio,
                                   add_builtin_entities_examples,
@@ -30,9 +30,6 @@ def get_mocked_augment_utterances(dataset, intent_name, language,
     return dataset[INTENTS][intent_name][UTTERANCES]
 
 
-# pylint: enable=W0613
-
-
 class TestLogRegIntentClassifier(FixtureTest):
     def test_intent_classifier_should_get_intent(self):
         # Given
@@ -69,6 +66,15 @@ class TestLogRegIntentClassifier(FixtureTest):
         self.assertEqual("MakeCoffee", res2[RES_INTENT_NAME])
         self.assertEqual(None, res3)
 
+    def test_should_not_get_intent_when_not_fitted(self):
+        # Given
+        intent_classifier = LogRegIntentClassifier()
+
+        # When / Then
+        self.assertFalse(intent_classifier.fitted)
+        with self.assertRaises(NotTrained):
+            intent_classifier.get_intent("foobar")
+
     def test_should_get_none_if_empty_dataset(self):
         # Given
         dataset = validate_and_format_dataset(get_empty_dataset(LANGUAGE_EN))
diff --git a/snips_nlu/tests/test_nlu_engine.py b/snips_nlu/tests/test_nlu_engine.py
index 27d7879..0ca9d0f 100644
--- a/snips_nlu/tests/test_nlu_engine.py
+++ b/snips_nlu/tests/test_nlu_engine.py
@@ -24,7 +24,7 @@ from snips_nlu.result import (
     resolved_slot, unresolved_slot)
 from snips_nlu.tests.utils import (
     BEVERAGE_DATASET, FixtureTest, SAMPLE_DATASET, get_empty_dataset)
-from snips_nlu.utils import json_string
+from snips_nlu.utils import json_string, NotTrained
 
 
 class TestSnipsNLUEngine(FixtureTest):
@@ -244,6 +244,15 @@ class TestSnipsNLUEngine(FixtureTest):
         # Then
         self.assertEqual(empty_result("hello world"), result)
 
+    def test_should_not_get_slots_when_not_fitted(self):
+        # Given
+        engine = SnipsNLUEngine()
+
+        # When / Then
+        self.assertFalse(engine.fitted)
+        with self.assertRaises(NotTrained):
+            engine.parse("foobar")
+
     def test_should_be_serializable_into_zip(self):
         # Given
         register_processing_unit(TestIntentParser1)
diff --git a/snips_nlu/tests/test_probabilistic_intent_parser.py b/snips_nlu/tests/test_probabilistic_intent_parser.py
index 4843559..f2c3ccb 100644
--- a/snips_nlu/tests/test_probabilistic_intent_parser.py
+++ b/snips_nlu/tests/test_probabilistic_intent_parser.py
@@ -16,7 +16,7 @@ from snips_nlu.pipeline.units_registry import register_processing_unit, \
     reset_processing_units
 from snips_nlu.slot_filler import CRFSlotFiller, SlotFiller
 from snips_nlu.tests.utils import BEVERAGE_DATASET, FixtureTest
-from snips_nlu.utils import json_string
+from snips_nlu.utils import json_string, NotTrained
 
 
 class TestProbabilisticIntentParser(FixtureTest):
@@ -76,6 +76,15 @@ class TestProbabilisticIntentParser(FixtureTest):
             parser.fit(BEVERAGE_DATASET, force_retrain=False)
             self.assertEqual(1, mock_fit.call_count)
 
+    def test_should_not_parse_when_not_fitted(self):
+        # Given
+        parser = ProbabilisticIntentParser()
+
+        # When / Then
+        self.assertFalse(parser.fitted)
+        with self.assertRaises(NotTrained):
+            parser.parse("foobar")
+
     def test_should_be_serializable_before_fitting(self):
         # Given
         parser = ProbabilisticIntentParser()
diff --git a/snips_nlu/tests/utils.py b/snips_nlu/tests/utils.py
index 3d3db30..dba8c11 100644
--- a/snips_nlu/tests/utils.py
+++ b/snips_nlu/tests/utils.py
@@ -12,7 +12,7 @@ from unittest import TestCase
 from snips_nlu_ontology import get_all_languages
 
 from snips_nlu.resources import load_resources
-from snips_nlu.utils import json_string
+from snips_nlu.utils import json_string, unicode_string
 
 TEST_PATH = Path(__file__).parent
 SAMPLE_DATASET_PATH = TEST_PATH / "resources" / "sample_dataset.json"
@@ -57,9 +57,8 @@ class SnipsTest(TestCase):
 
     @staticmethod
     def writeFileContent(path, content):
-        content = bytes(content, encoding="utf8")
         with path.open(mode="w") as f:
-            f.write(content.decode("utf8"))
+            f.write(unicode_string(content))
 
 
 class FixtureTest(SnipsTest):
