diff --git a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExecutionFailureHandlerTest.java b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExecutionFailureHandlerTest.java
index 2b250af3e71..a9d3beada8a 100644
--- a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExecutionFailureHandlerTest.java
+++ b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExecutionFailureHandlerTest.java
@@ -39,6 +39,7 @@ import java.util.Collections;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Collectors;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -57,6 +58,8 @@ class ExecutionFailureHandlerTest {
 
     private TestFailoverStrategy failoverStrategy;
 
+    private AtomicBoolean isNewAttempt;
+
     private TestRestartBackoffTimeStrategy backoffTimeStrategy;
 
     private ExecutionFailureHandler executionFailureHandler;
@@ -71,7 +74,9 @@ class ExecutionFailureHandlerTest {
 
         failoverStrategy = new TestFailoverStrategy();
         testingFailureEnricher = new TestingFailureEnricher();
-        backoffTimeStrategy = new TestRestartBackoffTimeStrategy(true, RESTART_DELAY_MS);
+        isNewAttempt = new AtomicBoolean(true);
+        backoffTimeStrategy =
+                new TestRestartBackoffTimeStrategy(true, RESTART_DELAY_MS, isNewAttempt::get);
         executionFailureHandler =
                 new ExecutionFailureHandler(
                         schedulingTopology,
@@ -158,6 +163,8 @@ class ExecutionFailureHandlerTest {
         final Throwable error =
                 new Exception(new SuppressRestartsException(new Exception("test failure")));
         final long timestamp = System.currentTimeMillis();
+
+        isNewAttempt.set(false);
         final FailureHandlingResult result =
                 executionFailureHandler.getFailureHandlingResult(execution, error, timestamp);
 
@@ -171,6 +178,10 @@ class ExecutionFailureHandlerTest {
         assertThat(result.getFailureLabels().get())
                 .isEqualTo(testingFailureEnricher.getFailureLabels());
         assertThat(result.getTimestamp()).isEqualTo(timestamp);
+        assertThat(result.isRootCause())
+                .as(
+                        "A NonRecoverableFailure should be new attempt even if RestartBackoffTimeStrategy consider it's not new attempt.")
+                .isTrue();
 
         assertThatThrownBy(result::getVerticesToRestart)
                 .as("getVerticesToRestart is not allowed when restarting is suppressed")
@@ -183,6 +194,59 @@ class ExecutionFailureHandlerTest {
         assertThat(executionFailureHandler.getNumberOfRestarts()).isZero();
     }
 
+    @Test
+    void testNewAttemptAndNumberOfRestarts() throws Exception {
+        final Set<ExecutionVertexID> tasksToRestart =
+                Collections.singleton(new ExecutionVertexID(new JobVertexID(), 0));
+        failoverStrategy.setTasksToRestart(tasksToRestart);
+
+        Execution execution =
+                FailureHandlingResultTest.createExecution(EXECUTOR_RESOURCE.getExecutor());
+        final Throwable error = new Exception("expected test failure");
+
+        testHandlingRootException(execution, error);
+
+        isNewAttempt.set(false);
+        testHandlingConcurrentException(execution, error);
+        testHandlingConcurrentException(execution, error);
+
+        isNewAttempt.set(true);
+        testHandlingRootException(execution, error);
+        testHandlingRootException(execution, error);
+
+        isNewAttempt.set(false);
+        testHandlingConcurrentException(execution, error);
+        testHandlingConcurrentException(execution, error);
+    }
+
+    private void testHandlingRootException(Execution execution, Throwable error) {
+        final long originalNumberOfRestarts = executionFailureHandler.getNumberOfRestarts();
+        FailureHandlingResult result =
+                executionFailureHandler.getFailureHandlingResult(
+                        execution, error, System.currentTimeMillis());
+        assertThat(result.isRootCause())
+                .as(
+                        "The FailureHandlingResult should be the root cause if exception is new attempt.")
+                .isTrue();
+        assertThat(executionFailureHandler.getNumberOfRestarts())
+                .as("The numberOfRestarts should be increased when it's a root exception.")
+                .isEqualTo(originalNumberOfRestarts + 1);
+    }
+
+    private void testHandlingConcurrentException(Execution execution, Throwable error) {
+        final long originalNumberOfRestarts = executionFailureHandler.getNumberOfRestarts();
+        FailureHandlingResult result =
+                executionFailureHandler.getFailureHandlingResult(
+                        execution, error, System.currentTimeMillis());
+        assertThat(result.isRootCause())
+                .as(
+                        "The FailureHandlingResult shouldn't be the root cause if exception isn't new attempt.")
+                .isFalse();
+        assertThat(executionFailureHandler.getNumberOfRestarts())
+                .as("The numberOfRestarts shouldn't be increased when it isn't a root exception.")
+                .isEqualTo(originalNumberOfRestarts);
+    }
+
     /** Tests the check for unrecoverable error. */
     @Test
     void testUnrecoverableErrorCheck() {
diff --git a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExponentialDelayRestartBackoffTimeStrategyTest.java b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExponentialDelayRestartBackoffTimeStrategyTest.java
index cfe37bec94d..eeee69dcf74 100644
--- a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExponentialDelayRestartBackoffTimeStrategyTest.java
+++ b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/ExponentialDelayRestartBackoffTimeStrategyTest.java
@@ -49,7 +49,7 @@ class ExponentialDelayRestartBackoffTimeStrategyTest {
 
         for (int i = 0; i <= maxAttempts; i++) {
             assertThat(restartStrategy.canRestart()).isTrue();
-            restartStrategy.notifyFailure(failure);
+            assertThat(restartStrategy.notifyFailure(failure)).isTrue();
             clock.advanceTime(Duration.ofMillis(maxBackoffMS + 1));
         }
         assertThat(restartStrategy.canRestart()).isFalse();
@@ -110,19 +110,19 @@ class ExponentialDelayRestartBackoffTimeStrategyTest {
                         0.25,
                         Integer.MAX_VALUE);
 
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
 
         clock.advanceTime(
                 resetBackoffThresholdMS + restartStrategy.getBackoffTime() - 1,
                 TimeUnit.MILLISECONDS);
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.getBackoffTime())
                 .as("Backoff should be increased")
                 .isEqualTo(2L);
 
         clock.advanceTime(
                 resetBackoffThresholdMS + restartStrategy.getBackoffTime(), TimeUnit.MILLISECONDS);
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.getBackoffTime())
                 .as("Backoff should be reset")
                 .isEqualTo(initialBackoffMS);
@@ -146,15 +146,15 @@ class ExponentialDelayRestartBackoffTimeStrategyTest {
                         jitterFactor,
                         10);
 
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isEqualTo(4L); // 4
         clock.advanceTime(Duration.ofMillis(maxBackoffMS + 1));
 
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isEqualTo(9L); // 4 * 2.3
         clock.advanceTime(Duration.ofMillis(maxBackoffMS + 1));
 
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isEqualTo(21L); // 4 * 2.3 * 2.3
         clock.advanceTime(Duration.ofMillis(maxBackoffMS + 1));
     }
@@ -223,7 +223,7 @@ class ExponentialDelayRestartBackoffTimeStrategyTest {
                     RestartBackoffTimeStrategy restartStrategy = factory.create();
                     for (int i = 0; i < failureCount; i++) {
                         clock.advanceTime(Duration.ofMillis(advanceMsEachFailure));
-                        restartStrategy.notifyFailure(failure);
+                        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
                     }
                     return restartStrategy.getBackoffTime();
                 },
@@ -250,31 +250,31 @@ class ExponentialDelayRestartBackoffTimeStrategyTest {
                         Integer.MAX_VALUE);
 
         // ensure initial data
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.canRestart()).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isEqualTo(initialBackoffMS);
 
         // ensure backoff time is initial after the first failure
         clock.advanceTime(resetBackoffThresholdMS + 1, TimeUnit.MILLISECONDS);
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.canRestart()).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isEqualTo(initialBackoffMS);
 
         // ensure backoff increases until threshold is reached
         clock.advanceTime(4, TimeUnit.MILLISECONDS);
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.canRestart()).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isEqualTo(2L);
 
         // ensure backoff is reset after threshold is reached
         clock.advanceTime(resetBackoffThresholdMS + 9 + 1, TimeUnit.MILLISECONDS);
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.canRestart()).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isOne();
         clock.advanceTime(Duration.ofMillis(maxBackoffMS + 1));
 
         // ensure backoff still increases
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.canRestart()).isTrue();
         assertThat(restartStrategy.getBackoffTime()).isEqualTo(2L);
     }
@@ -313,7 +313,7 @@ class ExponentialDelayRestartBackoffTimeStrategyTest {
 
         // After advance time it's a new round, and it reaches the maxAttempts.
         clock.advanceTime(1, TimeUnit.MILLISECONDS);
-        restartStrategy.notifyFailure(failure);
+        assertThat(restartStrategy.notifyFailure(failure)).isTrue();
         assertThat(restartStrategy.canRestart()).isFalse();
     }
 
@@ -360,9 +360,15 @@ class ExponentialDelayRestartBackoffTimeStrategyTest {
             ManualClock clock,
             long expectedBackoffMS,
             ExponentialDelayRestartBackoffTimeStrategy restartStrategy) {
+        boolean expectedNewAttempt = true;
         for (int advanceMs = 0; advanceMs < expectedBackoffMS; advanceMs++) {
             for (int i = 0; i < 10; i++) {
-                restartStrategy.notifyFailure(failure);
+                assertThat(restartStrategy.notifyFailure(failure)).isEqualTo(expectedNewAttempt);
+                if (expectedNewAttempt) {
+                    // Only the first one is new attempt, all rest of failures aren't new attempt.
+                    expectedNewAttempt = false;
+                }
+
                 assertThat(restartStrategy.canRestart()).isTrue();
                 assertThat(restartStrategy.getBackoffTime())
                         .isEqualTo(expectedBackoffMS - advanceMs);
diff --git a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/FailureHandlingResultTest.java b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/FailureHandlingResultTest.java
index 55f20890e6a..a590c99809f 100644
--- a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/FailureHandlingResultTest.java
+++ b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/FailureHandlingResultTest.java
@@ -24,8 +24,9 @@ import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
 import org.apache.flink.testutils.TestingUtils;
 import org.apache.flink.testutils.executor.TestExecutorExtension;
 
-import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
 
 import java.util.Collections;
 import java.util.HashSet;
@@ -47,8 +48,9 @@ class FailureHandlingResultTest {
             TestingUtils.defaultExecutorExtension();
 
     /** Tests normal FailureHandlingResult. */
-    @Test
-    void testNormalFailureHandlingResult() throws Exception {
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void testNormalFailureHandlingResult(boolean isNewAttempt) throws Exception {
 
         // create a normal FailureHandlingResult
         Execution execution = createExecution(EXECUTOR_RESOURCE.getExecutor());
@@ -63,7 +65,14 @@ class FailureHandlingResultTest {
                 CompletableFuture.completedFuture(Collections.singletonMap("key", "value"));
         FailureHandlingResult result =
                 FailureHandlingResult.restartable(
-                        execution, error, timestamp, failureLabels, tasks, delay, false);
+                        execution,
+                        error,
+                        timestamp,
+                        failureLabels,
+                        tasks,
+                        delay,
+                        false,
+                        isNewAttempt);
 
         assertThat(result.canRestart()).isTrue();
         assertThat(delay).isEqualTo(result.getRestartDelayMS());
@@ -73,24 +82,29 @@ class FailureHandlingResultTest {
         assertThat(result.getTimestamp()).isEqualTo(timestamp);
         assertThat(result.getFailedExecution()).isPresent();
         assertThat(result.getFailedExecution().get()).isSameAs(execution);
+        assertThat(result.isRootCause()).isEqualTo(isNewAttempt);
     }
 
     /** Tests FailureHandlingResult which suppresses restarts. */
-    @Test
-    void testRestartingSuppressedFailureHandlingResultWithNoCausingExecutionVertexId() {
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void testRestartingSuppressedFailureHandlingResultWithNoCausingExecutionVertexId(
+            boolean isNewAttempt) {
         // create a FailureHandlingResult with error
         final Throwable error = new Exception("test error");
         final long timestamp = System.currentTimeMillis();
         final CompletableFuture<Map<String, String>> failureLabels =
                 CompletableFuture.completedFuture(Collections.singletonMap("key", "value"));
         FailureHandlingResult result =
-                FailureHandlingResult.unrecoverable(null, error, timestamp, failureLabels, false);
+                FailureHandlingResult.unrecoverable(
+                        null, error, timestamp, failureLabels, false, isNewAttempt);
 
         assertThat(result.canRestart()).isFalse();
         assertThat(result.getError()).isSameAs(error);
         assertThat(result.getTimestamp()).isEqualTo(timestamp);
         assertThat(result.getFailureLabels()).isEqualTo(failureLabels);
         assertThat(result.getFailedExecution()).isNotPresent();
+        assertThat(result.isRootCause()).isEqualTo(isNewAttempt);
 
         assertThatThrownBy(result::getVerticesToRestart)
                 .as("getVerticesToRestart is not allowed when restarting is suppressed")
diff --git a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/TestRestartBackoffTimeStrategy.java b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/TestRestartBackoffTimeStrategy.java
index 0307891fd9e..1dd725f8e19 100644
--- a/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/TestRestartBackoffTimeStrategy.java
+++ b/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/TestRestartBackoffTimeStrategy.java
@@ -19,6 +19,8 @@
 
 package org.apache.flink.runtime.executiongraph.failover;
 
+import java.util.function.Supplier;
+
 /** A RestartBackoffTimeStrategy implementation for tests. */
 public class TestRestartBackoffTimeStrategy implements RestartBackoffTimeStrategy {
 
@@ -26,9 +28,17 @@ public class TestRestartBackoffTimeStrategy implements RestartBackoffTimeStrateg
 
     private long backoffTime;
 
+    private Supplier<Boolean> isNewAttempt;
+
     public TestRestartBackoffTimeStrategy(boolean canRestart, long backoffTime) {
+        this(canRestart, backoffTime, () -> true);
+    }
+
+    public TestRestartBackoffTimeStrategy(
+            boolean canRestart, long backoffTime, Supplier<Boolean> isNewAttempt) {
         this.canRestart = canRestart;
         this.backoffTime = backoffTime;
+        this.isNewAttempt = isNewAttempt;
     }
 
     @Override
@@ -42,8 +52,9 @@ public class TestRestartBackoffTimeStrategy implements RestartBackoffTimeStrateg
     }
 
     @Override
-    public void notifyFailure(Throwable cause) {
+    public boolean notifyFailure(Throwable cause) {
         // ignore
+        return isNewAttempt.get();
     }
 
     public void setCanRestart(final boolean canRestart) {
@@ -53,4 +64,8 @@ public class TestRestartBackoffTimeStrategy implements RestartBackoffTimeStrateg
     public void setBackoffTime(final long backoffTime) {
         this.backoffTime = backoffTime;
     }
+
+    public void setIsNewAttempt(Supplier<Boolean> isNewAttempt) {
+        this.isNewAttempt = isNewAttempt;
+    }
 }
diff --git a/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java b/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java
index 46c56281492..24c530c53c6 100644
--- a/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java
+++ b/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java
@@ -128,6 +128,7 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
@@ -1440,6 +1441,9 @@ public class DefaultSchedulerTest {
 
     @Test
     void testExceptionHistoryConcurrentRestart() throws Exception {
+        AtomicBoolean isNewAttempt = new AtomicBoolean(true);
+        testRestartBackoffTimeStrategy.setIsNewAttempt(isNewAttempt::get);
+
         final JobGraph jobGraph = singleJobVertexJobGraph(2);
 
         final TaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
@@ -1476,9 +1480,9 @@ public class DefaultSchedulerTest {
                         exception0);
 
         // multi-ExecutionVertex failure
+        isNewAttempt.set(false);
         final RuntimeException exception1 = new RuntimeException("failure #1");
-        failoverStrategyFactory.setTasksToRestart(
-                executionVertex1.getID(), executionVertex0.getID());
+        failoverStrategyFactory.setTasksToRestart(executionVertex1.getID());
         final long updateStateTriggeringRestartTimestamp1 =
                 initiateFailure(
                         scheduler,
@@ -1491,7 +1495,7 @@ public class DefaultSchedulerTest {
 
         delayExecutor.triggerNonPeriodicScheduledTasks();
 
-        assertThat(scheduler.getExceptionHistory()).hasSize(2);
+        assertThat(scheduler.getExceptionHistory()).hasSize(1);
         final Iterator<RootExceptionHistoryEntry> actualExceptionHistory =
                 scheduler.getExceptionHistory().iterator();
 
@@ -1513,17 +1517,6 @@ public class DefaultSchedulerTest {
                                         updateStateTriggeringRestartTimestamp1,
                                         executionVertex1.getTaskNameWithSubtaskIndex(),
                                         executionVertex1.getCurrentAssignedResourceLocation()));
-
-        final RootExceptionHistoryEntry entry1 = actualExceptionHistory.next();
-        assertThat(
-                        ExceptionHistoryEntryTestingUtils.matchesFailure(
-                                entry1,
-                                exception1,
-                                updateStateTriggeringRestartTimestamp1,
-                                executionVertex1.getTaskNameWithSubtaskIndex(),
-                                executionVertex1.getCurrentAssignedResourceLocation()))
-                .isTrue();
-        assertThat(entry1.getConcurrentExceptions()).isEmpty();
     }
 
     @Test
diff --git a/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java b/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java
index 69ad4095ef1..21ed10c0357 100644
--- a/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java
+++ b/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshotTest.java
@@ -96,7 +96,8 @@ class FailureHandlingResultSnapshotTest {
                                 .map(ExecutionVertex::getID)
                                 .collect(Collectors.toSet()),
                         0L,
-                        false);
+                        false,
+                        true);
 
         assertThatThrownBy(
                         () ->
@@ -124,7 +125,8 @@ class FailureHandlingResultSnapshotTest {
                                 .map(ExecutionVertex::getID)
                                 .collect(Collectors.toSet()),
                         0L,
-                        false);
+                        false,
+                        true);
 
         // FailedExecution with failure labels
         assertThat(failureHandlingResult.getFailureLabels().get())
@@ -145,6 +147,7 @@ class FailureHandlingResultSnapshotTest {
         assertThat(testInstance.getRootCauseExecution()).isPresent();
         assertThat(testInstance.getRootCauseExecution().get())
                 .isSameAs(rootCauseExecutionVertex.getCurrentExecutionAttempt());
+        assertThat(testInstance.isRootCause()).isTrue();
     }
 
     @Test
@@ -170,7 +173,8 @@ class FailureHandlingResultSnapshotTest {
                                 .map(ExecutionVertex::getID)
                                 .collect(Collectors.toSet()),
                         0L,
-                        false);
+                        false,
+                        true);
 
         final FailureHandlingResultSnapshot testInstance =
                 FailureHandlingResultSnapshot.create(
@@ -183,6 +187,7 @@ class FailureHandlingResultSnapshotTest {
                 .isSameAs(rootCauseExecutionVertex.getCurrentExecutionAttempt());
         assertThat(testInstance.getConcurrentlyFailedExecution())
                 .containsExactly(otherFailedExecutionVertex.getCurrentExecutionAttempt());
+        assertThat(testInstance.isRootCause()).isTrue();
     }
 
     @Test
@@ -196,7 +201,8 @@ class FailureHandlingResultSnapshotTest {
                                         new RuntimeException("Expected exception"),
                                         System.currentTimeMillis(),
                                         FailureEnricherUtils.EMPTY_FAILURE_LABELS,
-                                        Collections.singleton(rootCauseExecution)))
+                                        Collections.singleton(rootCauseExecution),
+                                        true))
                 .isInstanceOf(IllegalArgumentException.class);
     }
 
@@ -227,6 +233,7 @@ class FailureHandlingResultSnapshotTest {
                                 .map(ExecutionVertex::getID)
                                 .collect(Collectors.toSet()),
                         0L,
+                        true,
                         true);
 
         // FailedExecution with failure labels
@@ -244,6 +251,7 @@ class FailureHandlingResultSnapshotTest {
                 .containsExactlyInAnyOrder(
                         failedExecutionVertex0.getCurrentExecutionAttempt(),
                         failedExecutionVertex1.getCurrentExecutionAttempt());
+        assertThat(testInstance.isRootCause()).isTrue();
     }
 
     private Collection<Execution> getCurrentExecutions(ExecutionVertexID executionVertexId) {
diff --git a/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntryTest.java b/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntryTest.java
index 76539d21143..61c1a156cf2 100644
--- a/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntryTest.java
+++ b/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntryTest.java
@@ -81,10 +81,11 @@ class RootExceptionHistoryEntryTest {
         final CompletableFuture<Map<String, String>> rootFailureLabels =
                 CompletableFuture.completedFuture(Collections.singletonMap("key", "value"));
 
-        final Throwable concurrentException = new IllegalStateException("Expected other failure");
-        final ExecutionVertex concurrentlyFailedExecutionVertex = extractExecutionVertex(1);
-        final long concurrentExceptionTimestamp =
-                triggerFailure(concurrentlyFailedExecutionVertex, concurrentException);
+        final Throwable concurrentException1 = new IllegalStateException("Expected other failure1");
+        final ExecutionVertex concurrentlyFailedExecutionVertex1 = extractExecutionVertex(1);
+        Predicate<ExceptionHistoryEntry> exception1Predicate =
+                triggerFailureAndCreateEntryMatcher(
+                        concurrentException1, concurrentlyFailedExecutionVertex1);
 
         final FailureHandlingResultSnapshot snapshot =
                 new FailureHandlingResultSnapshot(
@@ -93,7 +94,8 @@ class RootExceptionHistoryEntryTest {
                         rootTimestamp,
                         rootFailureLabels,
                         Collections.singleton(
-                                concurrentlyFailedExecutionVertex.getCurrentExecutionAttempt()));
+                                concurrentlyFailedExecutionVertex1.getCurrentExecutionAttempt()),
+                        true);
         final RootExceptionHistoryEntry actualEntry =
                 RootExceptionHistoryEntry.fromFailureHandlingResultSnapshot(snapshot);
 
@@ -105,15 +107,24 @@ class RootExceptionHistoryEntryTest {
                                 rootFailureLabels.get(),
                                 rootExecutionVertex.getTaskNameWithSubtaskIndex(),
                                 rootExecutionVertex.getCurrentAssignedResourceLocation()));
+
+        assertThat(actualEntry.getConcurrentExceptions()).hasSize(1).allMatch(exception1Predicate);
+
+        // Test for addConcurrentExceptions
+        final Throwable concurrentException2 = new IllegalStateException("Expected other failure2");
+        final ExecutionVertex concurrentlyFailedExecutionVertex2 = extractExecutionVertex(2);
+        Predicate<ExceptionHistoryEntry> exception2Predicate =
+                triggerFailureAndCreateEntryMatcher(
+                        concurrentException2, concurrentlyFailedExecutionVertex2);
+
+        actualEntry.addConcurrentExceptions(
+                concurrentlyFailedExecutionVertex2.getCurrentExecutions());
         assertThat(actualEntry.getConcurrentExceptions())
-                .hasSize(1)
+                .hasSize(2)
                 .allMatch(
-                        ExceptionHistoryEntryMatcher.matchesFailure(
-                                concurrentException,
-                                concurrentExceptionTimestamp,
-                                concurrentlyFailedExecutionVertex.getTaskNameWithSubtaskIndex(),
-                                concurrentlyFailedExecutionVertex
-                                        .getCurrentAssignedResourceLocation()));
+                        exceptionHistoryEntry ->
+                                exception1Predicate.test(exceptionHistoryEntry)
+                                        || exception2Predicate.test(exceptionHistoryEntry));
     }
 
     @Test
@@ -121,14 +132,16 @@ class RootExceptionHistoryEntryTest {
         final Throwable concurrentException0 =
                 new RuntimeException("Expected concurrent failure #0");
         final ExecutionVertex concurrentlyFailedExecutionVertex0 = extractExecutionVertex(0);
-        final long concurrentExceptionTimestamp0 =
-                triggerFailure(concurrentlyFailedExecutionVertex0, concurrentException0);
+        final Predicate<ExceptionHistoryEntry> exception0Predicate =
+                triggerFailureAndCreateEntryMatcher(
+                        concurrentException0, concurrentlyFailedExecutionVertex0);
 
         final Throwable concurrentException1 =
                 new IllegalStateException("Expected concurrent failure #1");
         final ExecutionVertex concurrentlyFailedExecutionVertex1 = extractExecutionVertex(1);
-        final long concurrentExceptionTimestamp1 =
-                triggerFailure(concurrentlyFailedExecutionVertex1, concurrentException1);
+        final Predicate<ExceptionHistoryEntry> exception1Predicate =
+                triggerFailureAndCreateEntryMatcher(
+                        concurrentException1, concurrentlyFailedExecutionVertex1);
 
         final Throwable rootCause = new Exception("Expected root failure");
         final long rootTimestamp = System.currentTimeMillis();
@@ -150,18 +163,6 @@ class RootExceptionHistoryEntryTest {
                         ExceptionHistoryEntryMatcher.matchesGlobalFailure(
                                 rootCause, rootTimestamp, rootFailureLabels.get()));
 
-        final Predicate<ExceptionHistoryEntry> exception0Predicate =
-                ExceptionHistoryEntryMatcher.matchesFailure(
-                        concurrentException0,
-                        concurrentExceptionTimestamp0,
-                        concurrentlyFailedExecutionVertex0.getTaskNameWithSubtaskIndex(),
-                        concurrentlyFailedExecutionVertex0.getCurrentAssignedResourceLocation());
-        final Predicate<ExceptionHistoryEntry> exception1Predicate =
-                ExceptionHistoryEntryMatcher.matchesFailure(
-                        concurrentException1,
-                        concurrentExceptionTimestamp1,
-                        concurrentlyFailedExecutionVertex1.getTaskNameWithSubtaskIndex(),
-                        concurrentlyFailedExecutionVertex1.getCurrentAssignedResourceLocation());
         assertThat(actualEntry.getConcurrentExceptions())
                 .allMatch(
                         exceptionHistoryEntry ->
@@ -169,6 +170,17 @@ class RootExceptionHistoryEntryTest {
                                         || exception1Predicate.test(exceptionHistoryEntry));
     }
 
+    private Predicate<ExceptionHistoryEntry> triggerFailureAndCreateEntryMatcher(
+            Throwable concurrentException0, ExecutionVertex concurrentlyFailedExecutionVertex0) {
+        final long concurrentExceptionTimestamp0 =
+                triggerFailure(concurrentlyFailedExecutionVertex0, concurrentException0);
+        return ExceptionHistoryEntryMatcher.matchesFailure(
+                concurrentException0,
+                concurrentExceptionTimestamp0,
+                concurrentlyFailedExecutionVertex0.getTaskNameWithSubtaskIndex(),
+                concurrentlyFailedExecutionVertex0.getCurrentAssignedResourceLocation());
+    }
+
     private long triggerFailure(ExecutionVertex executionVertex, Throwable throwable) {
         executionGraph.updateState(
                 new TaskExecutionStateTransition(
