diff --git a/crates/matrix-sdk/src/authentication/oidc/tests.rs b/crates/matrix-sdk/src/authentication/oauth/tests.rs
similarity index 77%
rename from crates/matrix-sdk/src/authentication/oidc/tests.rs
rename to crates/matrix-sdk/src/authentication/oauth/tests.rs
index 3e2fb6ca0..d756c82f1 100644
--- a/crates/matrix-sdk/src/authentication/oidc/tests.rs
+++ b/crates/matrix-sdk/src/authentication/oauth/tests.rs
@@ -20,12 +20,12 @@ use wiremock::{
 
 use super::{
     registrations::OidcRegistrations, AuthorizationCode, AuthorizationError, AuthorizationResponse,
-    Oidc, OidcAuthorizationData, OidcError, RedirectUriQueryParseError,
+    OAuth, OAuthAuthorizationData, OAuthError, RedirectUriQueryParseError,
 };
 use crate::{
-    authentication::oidc::{
-        error::{AuthorizationCodeErrorResponseType, OauthClientRegistrationError},
-        AccountManagementActionFull, AuthorizationValidationData, OauthAuthorizationCodeError,
+    authentication::oauth::{
+        error::{AuthorizationCodeErrorResponseType, OAuthClientRegistrationError},
+        AccountManagementActionFull, AuthorizationValidationData, OAuthAuthorizationCodeError,
     },
     test_utils::{
         client::{
@@ -38,9 +38,9 @@ use crate::{
     Client, Error, SessionChange,
 };
 
-const REDIRECT_URI_STRING: &str = "http://127.0.0.1:6778/oidc/callback";
+const REDIRECT_URI_STRING: &str = "http://127.0.0.1:6778/oauth/callback";
 
-async fn mock_environment() -> anyhow::Result<(Oidc, MatrixMockServer, Url, OidcRegistrations)> {
+async fn mock_environment() -> anyhow::Result<(OAuth, MatrixMockServer, Url, OidcRegistrations)> {
     let server = MatrixMockServer::new().await;
     server.mock_who_am_i().ok().named("whoami").mount().await;
 
@@ -52,17 +52,18 @@ async fn mock_environment() -> anyhow::Result<(Oidc, MatrixMockServer, Url, Oidc
     let client = server.client_builder().unlogged().build().await;
     let client_metadata = mock_client_metadata();
 
-    let registrations_path = tempdir().unwrap().path().join("oidc").join("registrations.json");
+    let registrations_path =
+        tempdir().unwrap().path().join("matrix-sdk-oauth").join("registrations.json");
     let registrations =
         OidcRegistrations::new(&registrations_path, client_metadata, HashMap::new()).unwrap();
 
-    Ok((client.oidc(), server, mock_redirect_uri(), registrations))
+    Ok((client.oauth(), server, mock_redirect_uri(), registrations))
 }
 
 /// Check the URL in the given authorization data.
 async fn check_authorization_url(
-    authorization_data: &OidcAuthorizationData,
-    oidc: &Oidc,
+    authorization_data: &OAuthAuthorizationData,
+    oauth: &OAuth,
     issuer: &Url,
     device_id: Option<&DeviceId>,
     expected_prompt: Option<&str>,
@@ -70,7 +71,7 @@ async fn check_authorization_url(
 ) {
     tracing::debug!("authorization data URL = {}", authorization_data.url);
 
-    let data = oidc.data().unwrap();
+    let data = oauth.data().unwrap();
     let authorization_data_guard = data.authorization_data.lock().await;
     let validation_data =
         authorization_data_guard.get(&authorization_data.state).expect("missing validation data");
@@ -152,25 +153,27 @@ async fn check_authorization_url(
 #[async_test]
 async fn test_high_level_login() -> anyhow::Result<()> {
     // Given a fresh environment.
-    let (oidc, _server, mut redirect_uri, registrations) = mock_environment().await.unwrap();
-    assert!(oidc.issuer().is_none());
-    assert!(oidc.client_id().is_none());
+    let (oauth, _server, mut redirect_uri, registrations) = mock_environment().await.unwrap();
+    assert!(oauth.issuer().is_none());
+    assert!(oauth.client_id().is_none());
 
     // When getting the OIDC login URL.
-    let authorization_data =
-        oidc.url_for_oidc(registrations, redirect_uri.clone(), Some(Prompt::Create)).await.unwrap();
+    let authorization_data = oauth
+        .url_for_oidc(registrations, redirect_uri.clone(), Some(Prompt::Create))
+        .await
+        .unwrap();
 
     // Then the client should be configured correctly.
-    assert_let!(Some(issuer) = oidc.issuer());
-    assert!(oidc.client_id().is_some());
+    assert_let!(Some(issuer) = oauth.issuer());
+    assert!(oauth.client_id().is_some());
 
-    check_authorization_url(&authorization_data, &oidc, issuer, None, Some("create"), None).await;
+    check_authorization_url(&authorization_data, &oauth, issuer, None, Some("create"), None).await;
 
     // When completing the login with a valid callback.
     redirect_uri.set_query(Some(&format!("code=42&state={}", authorization_data.state.secret())));
 
     // Then the login should succeed.
-    oidc.login_with_oidc_callback(&authorization_data, redirect_uri).await?;
+    oauth.login_with_oidc_callback(&authorization_data, redirect_uri).await?;
 
     Ok(())
 }
@@ -178,14 +181,14 @@ async fn test_high_level_login() -> anyhow::Result<()> {
 #[async_test]
 async fn test_high_level_login_cancellation() -> anyhow::Result<()> {
     // Given a client ready to complete login.
-    let (oidc, _server, mut redirect_uri, registrations) = mock_environment().await.unwrap();
+    let (oauth, _server, mut redirect_uri, registrations) = mock_environment().await.unwrap();
     let authorization_data =
-        oidc.url_for_oidc(registrations, redirect_uri.clone(), None).await.unwrap();
+        oauth.url_for_oidc(registrations, redirect_uri.clone(), None).await.unwrap();
 
-    assert_let!(Some(issuer) = oidc.issuer());
-    assert!(oidc.client_id().is_some());
+    assert_let!(Some(issuer) = oauth.issuer());
+    assert!(oauth.client_id().is_some());
 
-    check_authorization_url(&authorization_data, &oidc, issuer, None, None, None).await;
+    check_authorization_url(&authorization_data, &oauth, issuer, None, None, None).await;
 
     // When completing login with a cancellation callback.
     redirect_uri.set_query(Some(&format!(
@@ -193,12 +196,13 @@ async fn test_high_level_login_cancellation() -> anyhow::Result<()> {
         authorization_data.state.secret()
     )));
 
-    let error = oidc.login_with_oidc_callback(&authorization_data, redirect_uri).await.unwrap_err();
+    let error =
+        oauth.login_with_oidc_callback(&authorization_data, redirect_uri).await.unwrap_err();
 
     // Then a cancellation error should be thrown.
     assert_matches!(
         error,
-        Error::Oidc(OidcError::AuthorizationCode(OauthAuthorizationCodeError::Cancelled))
+        Error::OAuth(OAuthError::AuthorizationCode(OAuthAuthorizationCodeError::Cancelled))
     );
 
     Ok(())
@@ -207,24 +211,25 @@ async fn test_high_level_login_cancellation() -> anyhow::Result<()> {
 #[async_test]
 async fn test_high_level_login_invalid_state() -> anyhow::Result<()> {
     // Given a client ready to complete login.
-    let (oidc, _server, mut redirect_uri, registrations) = mock_environment().await.unwrap();
+    let (oauth, _server, mut redirect_uri, registrations) = mock_environment().await.unwrap();
     let authorization_data =
-        oidc.url_for_oidc(registrations, redirect_uri.clone(), None).await.unwrap();
+        oauth.url_for_oidc(registrations, redirect_uri.clone(), None).await.unwrap();
 
-    assert_let!(Some(issuer) = oidc.issuer());
-    assert!(oidc.client_id().is_some());
+    assert_let!(Some(issuer) = oauth.issuer());
+    assert!(oauth.client_id().is_some());
 
-    check_authorization_url(&authorization_data, &oidc, issuer, None, None, None).await;
+    check_authorization_url(&authorization_data, &oauth, issuer, None, None, None).await;
 
     // When completing login with an old/tampered state.
     redirect_uri.set_query(Some("code=42&state=imposter_alert"));
 
-    let error = oidc.login_with_oidc_callback(&authorization_data, redirect_uri).await.unwrap_err();
+    let error =
+        oauth.login_with_oidc_callback(&authorization_data, redirect_uri).await.unwrap_err();
 
     // Then the login should fail by flagging the invalid state.
     assert_matches!(
         error,
-        Error::Oidc(OidcError::AuthorizationCode(OauthAuthorizationCodeError::InvalidState))
+        Error::OAuth(OAuthError::AuthorizationCode(OAuthAuthorizationCodeError::InvalidState))
     );
 
     Ok(())
@@ -239,7 +244,7 @@ async fn test_login_url() -> anyhow::Result<()> {
     oauth_server.mock_server_metadata().ok().expect(1..).mount().await;
 
     let client = server.client_builder().registered_with_oauth(server.server().uri()).build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
 
     let device_id = owned_device_id!("D3V1C31D"); // yo this is 1999 speaking
 
@@ -248,19 +253,19 @@ async fn test_login_url() -> anyhow::Result<()> {
 
     // No extra parameters.
     let authorization_data =
-        oidc.login(redirect_uri.clone(), Some(device_id.clone()))?.build().await?;
-    check_authorization_url(&authorization_data, &oidc, &issuer, Some(&device_id), None, None)
+        oauth.login(redirect_uri.clone(), Some(device_id.clone()))?.build().await?;
+    check_authorization_url(&authorization_data, &oauth, &issuer, Some(&device_id), None, None)
         .await;
 
     // With prompt parameter.
-    let authorization_data = oidc
+    let authorization_data = oauth
         .login(redirect_uri.clone(), Some(device_id.clone()))?
         .prompt(vec![Prompt::Create])
         .build()
         .await?;
     check_authorization_url(
         &authorization_data,
-        &oidc,
+        &oauth,
         &issuer,
         Some(&device_id),
         Some("create"),
@@ -269,14 +274,14 @@ async fn test_login_url() -> anyhow::Result<()> {
     .await;
 
     // With user_id_hint parameter.
-    let authorization_data = oidc
+    let authorization_data = oauth
         .login(redirect_uri.clone(), Some(device_id.clone()))?
         .user_id_hint(user_id!("@joe:example.org"))
         .build()
         .await?;
     check_authorization_url(
         &authorization_data,
-        &oidc,
+        &oauth,
         &issuer,
         Some(&device_id),
         None,
@@ -326,10 +331,10 @@ async fn test_finish_authorization() -> anyhow::Result<()> {
     oauth_server.mock_token().ok().expect(1).named("token").mount().await;
 
     let client = server.client_builder().registered_with_oauth(server.server().uri()).build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
 
     // If the state is missing, then any attempt to finish authorizing will fail.
-    let res = oidc
+    let res = oauth
         .finish_authorization(AuthorizationCode {
             code: "42".to_owned(),
             state: CsrfToken::new("none".to_owned()),
@@ -338,7 +343,7 @@ async fn test_finish_authorization() -> anyhow::Result<()> {
 
     assert_matches!(
         res,
-        Err(OidcError::AuthorizationCode(OauthAuthorizationCodeError::InvalidState))
+        Err(OAuthError::AuthorizationCode(OAuthAuthorizationCodeError::InvalidState))
     );
     assert!(client.session_tokens().is_none());
 
@@ -352,13 +357,13 @@ async fn test_finish_authorization() -> anyhow::Result<()> {
     };
 
     {
-        let data = oidc.data().context("missing data")?;
+        let data = oauth.data().context("missing data")?;
         let prev = data.authorization_data.lock().await.insert(state.clone(), auth_validation_data);
         assert!(prev.is_none());
     }
 
     // Finishing the authorization for another state won't work.
-    let res = oidc
+    let res = oauth
         .finish_authorization(AuthorizationCode {
             code: "1337".to_owned(),
             state: CsrfToken::new("none".to_owned()),
@@ -367,40 +372,41 @@ async fn test_finish_authorization() -> anyhow::Result<()> {
 
     assert_matches!(
         res,
-        Err(OidcError::AuthorizationCode(OauthAuthorizationCodeError::InvalidState))
+        Err(OAuthError::AuthorizationCode(OAuthAuthorizationCodeError::InvalidState))
     );
     assert!(client.session_tokens().is_none());
-    assert!(oidc.data().unwrap().authorization_data.lock().await.get(&state).is_some());
+    assert!(oauth.data().unwrap().authorization_data.lock().await.get(&state).is_some());
 
     // Finishing the authorization for the expected state will work.
-    oidc.finish_authorization(AuthorizationCode { code: "1337".to_owned(), state: state.clone() })
+    oauth
+        .finish_authorization(AuthorizationCode { code: "1337".to_owned(), state: state.clone() })
         .await?;
 
     assert!(client.session_tokens().is_some());
-    assert!(oidc.data().unwrap().authorization_data.lock().await.get(&state).is_none());
+    assert!(oauth.data().unwrap().authorization_data.lock().await.get(&state).is_none());
 
     Ok(())
 }
 
 #[async_test]
-async fn test_oidc_session() -> anyhow::Result<()> {
+async fn test_oauth_session() -> anyhow::Result<()> {
     let client = MockClientBuilder::new("https://example.org".to_owned()).unlogged().build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
 
     let tokens = mock_session_tokens_with_refresh();
-    let issuer = "https://oidc.example.com/issuer";
+    let issuer = "https://oauth.example.com/issuer";
     let session = mock_session(tokens.clone(), issuer);
-    oidc.restore_session(session.clone()).await?;
+    oauth.restore_session(session.clone()).await?;
 
     // Test a few extra getters.
     assert_eq!(client.session_tokens().unwrap(), tokens);
 
-    let user_session = oidc.user_session().unwrap();
+    let user_session = oauth.user_session().unwrap();
     assert_eq!(user_session.meta, session.user.meta);
     assert_eq!(user_session.tokens, tokens);
     assert_eq!(user_session.issuer.as_str(), issuer);
 
-    let full_session = oidc.full_session().unwrap();
+    let full_session = oauth.full_session().unwrap();
 
     assert_eq!(full_session.client_id.as_str(), "test_client_id");
     assert_eq!(full_session.user.meta, session.user.meta);
@@ -436,15 +442,15 @@ async fn test_insecure_clients() -> anyhow::Result<()> {
             .build()
             .await?,
     ] {
-        let oidc = client.oidc();
+        let oauth = client.oauth();
 
         // Restore the previous session so we have an existing set of refresh tokens.
-        oidc.restore_session(mock_session(prev_tokens.clone(), &server_url)).await?;
+        oauth.restore_session(mock_session(prev_tokens.clone(), &server_url)).await?;
 
         let mut session_changes = client.subscribe_to_session_changes();
 
         // A refresh in insecure mode should work Just Fine.
-        oidc.refresh_access_token().await?;
+        oauth.refresh_access_token().await?;
 
         assert_eq!(client.session_tokens().unwrap(), next_tokens);
 
@@ -469,7 +475,7 @@ async fn test_register_client() {
     let server = MatrixMockServer::new().await;
     let oauth_server = server.oauth();
     let client = server.client_builder().unlogged().build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
     let client_metadata = mock_client_metadata();
 
     // Server doesn't support registration, it fails.
@@ -481,10 +487,10 @@ async fn test_register_client() {
         .mount()
         .await;
 
-    let result = oidc.register_client(&client_metadata).await;
+    let result = oauth.register_client(&client_metadata).await;
     assert_matches!(
         result,
-        Err(OidcError::ClientRegistration(OauthClientRegistrationError::NotSupported))
+        Err(OAuthError::ClientRegistration(OAuthClientRegistrationError::NotSupported))
     );
 
     server.verify_and_reset().await;
@@ -499,10 +505,10 @@ async fn test_register_client() {
         .await;
     oauth_server.mock_registration().ok().expect(1).named("registration").mount().await;
 
-    let response = oidc.register_client(&client_metadata).await.unwrap();
+    let response = oauth.register_client(&client_metadata).await.unwrap();
     assert_eq!(response.client_id.as_str(), "test_client_id");
 
-    let auth_data = oidc.data().unwrap();
+    let auth_data = oauth.data().unwrap();
     // There is a difference of ending slash between the strings so we parse them
     // with `Url` which will normalize that.
     assert_eq!(auth_data.issuer, Url::parse(&server.server().uri()).unwrap());
@@ -517,23 +523,23 @@ async fn test_management_url_cache() {
     oauth_server.mock_server_metadata().ok().expect(1).mount().await;
 
     let client = server.client_builder().logged_in_with_oauth(server.server().uri()).build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
 
     // The cache should not contain the entry.
-    assert!(!client.inner.caches.provider_metadata.lock().await.contains("PROVIDER_METADATA"));
+    assert!(!client.inner.caches.server_metadata.lock().await.contains("SERVER_METADATA"));
 
-    let management_url = oidc
+    let management_url = oauth
         .account_management_url(Some(AccountManagementActionFull::Profile))
         .await
         .expect("We should be able to fetch the account management url");
 
     assert!(management_url.is_some());
 
-    // Check that the provider metadata has been inserted into the cache.
-    assert!(client.inner.caches.provider_metadata.lock().await.contains("PROVIDER_METADATA"));
+    // Check that the server metadata has been inserted into the cache.
+    assert!(client.inner.caches.server_metadata.lock().await.contains("SERVER_METADATA"));
 
     // Another parameter doesn't make another request for the metadata.
-    let management_url = oidc
+    let management_url = oauth
         .account_management_url(Some(AccountManagementActionFull::SessionsList))
         .await
         .expect("We should be able to fetch the account management url");
@@ -542,14 +548,14 @@ async fn test_management_url_cache() {
 }
 
 #[async_test]
-async fn test_provider_metadata() {
+async fn test_server_metadata() {
     let server = MatrixMockServer::new().await;
     let client = server.client_builder().unlogged().build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
     let issuer = server.server().uri();
 
     // The endpoint is not mocked so it is not supported.
-    let error = oidc.provider_metadata().await.unwrap_err();
+    let error = oauth.server_metadata().await.unwrap_err();
     assert!(error.is_not_supported());
 
     // Mock the `GET /auth_issuer` fallback endpoint.
@@ -568,11 +574,11 @@ async fn test_provider_metadata() {
         .named("openid-configuration")
         .mount(server.server())
         .await;
-    oidc.provider_metadata().await.unwrap();
+    oauth.server_metadata().await.unwrap();
 
     // Mock the `GET /auth_metadata` endpoint.
     let oauth_server = server.oauth();
     oauth_server.mock_server_metadata().ok().expect(1).named("auth_metadata").mount().await;
 
-    oidc.provider_metadata().await.unwrap();
+    oauth.server_metadata().await.unwrap();
 }
diff --git a/crates/matrix-sdk/tests/integration/encryption/cross_signing.rs b/crates/matrix-sdk/tests/integration/encryption/cross_signing.rs
index 825088a00..d6e17df82 100644
--- a/crates/matrix-sdk/tests/integration/encryption/cross_signing.rs
+++ b/crates/matrix-sdk/tests/integration/encryption/cross_signing.rs
@@ -111,7 +111,7 @@ async fn test_reset_legacy_auth() {
 
 #[cfg(feature = "experimental-oidc")]
 #[async_test]
-async fn test_reset_oidc() {
+async fn test_reset_oauth() {
     use assert_matches2::assert_let;
     use matrix_sdk::{encryption::CrossSigningResetAuthType, test_utils::mocks::MatrixMockServer};
     use similar_asserts::assert_eq;
@@ -163,9 +163,9 @@ async fn test_reset_oidc() {
         .unwrap()
         .expect("We should have received a reset handle");
 
-    assert_let!(CrossSigningResetAuthType::Oidc(oidc_info) = reset_handle.auth_type());
+    assert_let!(CrossSigningResetAuthType::OAuth(oauth_info) = reset_handle.auth_type());
     assert_eq!(
-        oidc_info.approval_url.as_str(),
+        oauth_info.approval_url.as_str(),
         format!("{}/account/?action=org.matrix.cross_signing_reset", server.server().uri())
     );
 
diff --git a/crates/matrix-sdk/tests/integration/refresh_token.rs b/crates/matrix-sdk/tests/integration/refresh_token.rs
index 58c251287..55f615b90 100644
--- a/crates/matrix-sdk/tests/integration/refresh_token.rs
+++ b/crates/matrix-sdk/tests/integration/refresh_token.rs
@@ -581,9 +581,10 @@ async fn test_oauth_refresh_token_handled_success() {
 
     let issuer = server.server().uri();
     let client = server.client_builder().unlogged().handle_refresh_tokens().build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
 
-    oidc.restore_session(mock_session(mock_prev_session_tokens_with_refresh(), issuer))
+    oauth
+        .restore_session(mock_session(mock_prev_session_tokens_with_refresh(), issuer))
         .await
         .unwrap();
 
@@ -608,7 +609,7 @@ async fn test_oauth_refresh_token_handled_success() {
 #[async_test]
 async fn test_oauth_refresh_token_handled_failure() {
     use matrix_sdk::{
-        authentication::oidc::OidcError,
+        authentication::oauth::OAuthError,
         test_utils::{
             client::{mock_prev_session_tokens_with_refresh, oauth::mock_session},
             mocks::MatrixMockServer,
@@ -633,9 +634,10 @@ async fn test_oauth_refresh_token_handled_failure() {
 
     let issuer = server.server().uri();
     let client = server.client_builder().unlogged().handle_refresh_tokens().build().await;
-    let oidc = client.oidc();
+    let oauth = client.oauth();
 
-    oidc.restore_session(mock_session(mock_prev_session_tokens_with_refresh(), issuer))
+    oauth
+        .restore_session(mock_session(mock_prev_session_tokens_with_refresh(), issuer))
         .await
         .unwrap();
 
@@ -644,13 +646,13 @@ async fn test_oauth_refresh_token_handled_failure() {
     // The request fails with a refresh token error.
     let res = client.whoami().await;
     assert_let!(
-        Err(HttpError::RefreshToken(RefreshTokenError::Oidc(oidc_err))) = res,
-        "The request should fail with a refresh token error from the OIDC API"
+        Err(HttpError::RefreshToken(RefreshTokenError::OAuth(oauth_err))) = res,
+        "The request should fail with a refresh token error from the OAuth 2.0 API"
     );
     assert_matches!(
-        *oidc_err,
-        OidcError::RefreshToken(_),
-        "The OIDC error should be a refresh token error"
+        *oauth_err,
+        OAuthError::RefreshToken(_),
+        "The OAuth 2.0 error should be a refresh token error"
     );
 
     // We get notified once that the token is invalid.
@@ -685,8 +687,9 @@ async fn test_oauth_handle_refresh_tokens() {
     let issuer = server.server().uri();
     let client = server.client_builder().unlogged().handle_refresh_tokens().build().await;
 
-    let oidc = client.oidc();
-    oidc.restore_session(mock_session(mock_prev_session_tokens_with_refresh(), issuer))
+    let oauth = client.oauth();
+    oauth
+        .restore_session(mock_session(mock_prev_session_tokens_with_refresh(), issuer))
         .await
         .unwrap();
 
@@ -709,7 +712,7 @@ async fn test_oauth_handle_refresh_tokens() {
     // Refresh token successfully.
     oauth_server.mock_token().ok().mock_once().named("refresh_token").mount().await;
 
-    oidc.refresh_access_token().await.unwrap();
+    oauth.refresh_access_token().await.unwrap();
 
     // The tokens were updated.
     assert_eq!(
