diff --git a/mountpoint-s3-fs/tests/mock_s3_tests.rs b/mountpoint-s3-fs/tests/mock_s3_tests.rs
index c19d6012..83f999d6 100644
--- a/mountpoint-s3-fs/tests/mock_s3_tests.rs
+++ b/mountpoint-s3-fs/tests/mock_s3_tests.rs
@@ -1,7 +1,6 @@
 use std::num::NonZeroUsize;
 
 use common::make_test_filesystem_with_client;
-use httpmock::{Method, MockServer, Then};
 use mountpoint_s3_client::S3CrtClient;
 use mountpoint_s3_client::config::{
     AddressingStyle, Allocator, EndpointConfig, S3ClientAuthConfig, S3ClientConfig, Uri,
@@ -9,6 +8,8 @@ use mountpoint_s3_client::config::{
 use mountpoint_s3_client::error_metadata::ClientErrorMetadata;
 use mountpoint_s3_fs::fs::error_metadata::{ErrorMetadata, MOUNTPOINT_ERROR_CLIENT};
 use mountpoint_s3_fs::fs::{FUSE_ROOT_INODE, OpenFlags, ToErrno};
+use wiremock::matchers::{method, path, query_param};
+use wiremock::{Mock, MockServer, ResponseTemplate};
 
 use mountpoint_s3_fs::S3Filesystem;
 use mountpoint_s3_fs::memory::PagedPool;
@@ -32,8 +33,8 @@ async fn test_lookup_throttled_mock(head_object_throttled: bool, list_object_thr
     </Error>
     "#
     .to_string();
-    let head_object_503_headers = [];
-    let head_object_ok_headers = [
+    let head_object_503_headers = vec![];
+    let head_object_ok_headers = vec![
         ("ETag", "71a5b8dcb22444f1b2b899dedc1e4122"),
         ("Date", "Thu, 12 Jan 2023 00:04:21 GMT"),
         ("Last-Modified", "Tue, 10 Jan 2023 23:39:32 GMT"),
@@ -43,7 +44,7 @@ async fn test_lookup_throttled_mock(head_object_throttled: bool, list_object_thr
         ("Content-Length", "1024"),
     ];
 
-    let (fs, server) = create_fs_with_mock_s3(bucket);
+    let (fs, server) = create_fs_with_mock_s3(bucket).await;
 
     // lookup will issue 2 S3 calls, let's mock them
     // those are the *assumed* S3 responses, based on: https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
@@ -52,22 +53,25 @@ async fn test_lookup_throttled_mock(head_object_throttled: bool, list_object_thr
     } else {
         (200, list_empty_response(bucket, key))
     };
-    let (head_object_http_code, head_object_headers): (u16, &[(&str, &str)]) = if head_object_throttled {
-        (503, &head_object_503_headers)
+    let (head_object_http_code, head_object_headers) = if head_object_throttled {
+        (503, head_object_503_headers)
     } else {
-        (200, &head_object_ok_headers)
+        (200, head_object_ok_headers)
     };
-    server.mock(|when, then| {
-        when.method(Method::GET)
-            .path(format!("/{bucket}/"))
-            .query_param("list-type", "2")
-            .query_param("prefix", format!("{key}/"));
-        then.status(list_http_code).body(list_response);
-    });
-    server.mock(|when, then| {
-        when.method(Method::HEAD).path(format!("/{bucket}/{key}"));
-        set_response_headers(then.status(head_object_http_code), head_object_headers);
-    });
+    Mock::given(method("GET"))
+        .and(path(format!("/{bucket}/")))
+        .and(query_param("list-type", "2"))
+        .and(query_param("prefix", format!("{key}/")))
+        .respond_with(ResponseTemplate::new(list_http_code).set_body_string(list_response))
+        .mount(&server)
+        .await;
+
+    let head_response = ResponseTemplate::new(head_object_http_code).append_headers(head_object_headers);
+    Mock::given(method("HEAD"))
+        .and(path(format!("/{bucket}/{key}")))
+        .respond_with(head_response)
+        .mount(&server)
+        .await;
 
     // perform a lookup
     let err = fs
@@ -100,18 +104,19 @@ async fn test_lookup_throttled_mock(head_object_throttled: bool, list_object_thr
 async fn test_lookup_unhandled_error_mock() {
     let bucket = "bucket";
     let key = "unhandled";
-    let (fs, server) = create_fs_with_mock_s3(bucket);
-    server.mock(|when, then| {
-        when.method(Method::GET)
-            .path(format!("/{bucket}/"))
-            .query_param("list-type", "2")
-            .query_param("prefix", format!("{key}/"));
-        then.status(200).body(list_empty_response(bucket, key));
-    });
-    server.mock(|when, then| {
-        when.method(Method::HEAD).path(format!("/{bucket}/{key}"));
-        then.status(409); // let's say S3 adds 409 response code for HeadObject in future
-    });
+    let (fs, server) = create_fs_with_mock_s3(bucket).await;
+    Mock::given(method("GET"))
+        .and(path(format!("/{bucket}/")))
+        .and(query_param("list-type", "2"))
+        .and(query_param("prefix", format!("{key}/")))
+        .respond_with(ResponseTemplate::new(200).set_body_string(list_empty_response(bucket, key)))
+        .mount(&server)
+        .await;
+    Mock::given(method("HEAD"))
+        .and(path(format!("/{bucket}/{key}")))
+        .respond_with(ResponseTemplate::new(409)) // let's say S3 adds 409 response code for HeadObject in future
+        .mount(&server)
+        .await;
     // perform a lookup
     let err = fs
         .lookup(FUSE_ROOT_INODE, key.as_ref())
@@ -148,24 +153,29 @@ async fn test_read_unhandled_error_mock() {
     ];
     let get_object_error_resp = r#"<?xml version="1.0" encoding="UTF-8"?><Error><Code>NotARealError</Code><Message>This error is made up.</Message><RequestId>CM0R497NB0WAQ977</RequestId><HostId>w1TqUKGaIuNAIgzqm/L2azuzgEBINxTngWPbV1iH2IvpLsVCCTKHJTh4HsGp4JnggHqVkA+KN1MGqHDw1+WEuA==</HostId></Error>"#;
 
-    let (fs, server) = create_fs_with_mock_s3(bucket);
+    let (fs, server) = create_fs_with_mock_s3(bucket).await;
 
     // set up a mock for a successful lookup but a failed read
-    server.mock(|when, then| {
-        when.method(Method::GET)
-            .path(format!("/{bucket}/"))
-            .query_param("list-type", "2")
-            .query_param("prefix", format!("{key}/"));
-        then.status(200).body(list_empty_response(bucket, key));
-    });
-    server.mock(|when, then| {
-        when.method(Method::HEAD).path(format!("/{bucket}/{key}"));
-        set_response_headers(then.status(200), &head_object_ok_headers);
-    });
-    server.mock(|when, then| {
-        when.method(Method::GET).path(format!("/{bucket}/{key}"));
-        then.status(418).body(get_object_error_resp);
-    });
+    Mock::given(method("GET"))
+        .and(path(format!("/{bucket}/")))
+        .and(query_param("list-type", "2"))
+        .and(query_param("prefix", format!("{key}/")))
+        .respond_with(ResponseTemplate::new(200).set_body_string(list_empty_response(bucket, key)))
+        .mount(&server)
+        .await;
+
+    let head_response = ResponseTemplate::new(200).append_headers(head_object_ok_headers);
+    Mock::given(method("HEAD"))
+        .and(path(format!("/{bucket}/{key}")))
+        .respond_with(head_response)
+        .mount(&server)
+        .await;
+
+    Mock::given(method("GET"))
+        .and(path(format!("/{bucket}/{key}")))
+        .respond_with(ResponseTemplate::new(418).set_body_string(get_object_error_resp))
+        .mount(&server)
+        .await;
 
     // perform a read
     let entry = fs
@@ -197,9 +207,9 @@ async fn test_read_unhandled_error_mock() {
     );
 }
 
-fn create_fs_with_mock_s3(bucket: &str) -> (S3Filesystem<S3CrtClient>, MockServer) {
-    let server = MockServer::start();
-    let endpoint = format!("http://{}", server.address());
+async fn create_fs_with_mock_s3(bucket: &str) -> (S3Filesystem<S3CrtClient>, MockServer) {
+    let server = MockServer::start().await;
+    let endpoint = server.uri();
     let endpoint = Uri::new_from_str(&Allocator::default(), endpoint).expect("must be a valid uri");
     let endpoint_config = EndpointConfig::new("PLACEHOLDER")
         .addressing_style(AddressingStyle::Path) // mock server responds to path style requests only
@@ -231,10 +241,3 @@ fn list_empty_response(bucket: &str, prefix: &str) -> String {
     "#;
     template.replace("__PREFIX__", prefix).replace("__BUCKET__", bucket)
 }
-
-fn set_response_headers(mut then: Then, headers: &[(&str, &str)]) -> Then {
-    for (key, value) in headers.iter() {
-        then = then.header(*key, *value);
-    }
-    then
-}
