diff --git a/linkerd/app/gateway/src/http/tests.rs b/linkerd/app/gateway/src/http/tests.rs
index 3d8a5044..5c90917b 100644
--- a/linkerd/app/gateway/src/http/tests.rs
+++ b/linkerd/app/gateway/src/http/tests.rs
@@ -4,7 +4,6 @@ use linkerd_app_core::{
     svc::{NewService, ServiceExt},
     tls,
     trace::test::trace_init,
-    transport::ServerAddr,
     Error, NameAddr,
 };
 use linkerd_app_inbound::GatewayLoop;
@@ -53,7 +52,7 @@ async fn upgraded_request_remains_relative_form() {
 
     impl svc::Param<OrigDstAddr> for Target {
         fn param(&self) -> OrigDstAddr {
-            OrigDstAddr(Self::dst_addr())
+            OrigDstAddr(([10, 10, 10, 10], 4143).into())
         }
     }
 
@@ -133,8 +132,7 @@ async fn upgraded_request_remains_relative_form() {
                     }]))]),
                 },
             };
-            let (policy, tx) =
-                inbound::policy::AllowPolicy::for_test(ServerAddr(Self::dst_addr()), policy);
+            let (policy, tx) = inbound::policy::AllowPolicy::for_test(self.param(), policy);
             tokio::spawn(async move {
                 tx.closed().await;
             });
@@ -142,12 +140,6 @@ async fn upgraded_request_remains_relative_form() {
         }
     }
 
-    impl Target {
-        fn dst_addr() -> std::net::SocketAddr {
-            ([10, 10, 10, 10], 4143).into()
-        }
-    }
-
     let (inner, mut handle) =
         mock::pair::<http::Request<http::BoxBody>, http::Response<http::BoxBody>>();
     handle.allow(1);
diff --git a/linkerd/app/inbound/src/accept/tests.rs b/linkerd/app/inbound/src/accept/tests.rs
deleted file mode 100644
index b24599a9..00000000
--- a/linkerd/app/inbound/src/accept/tests.rs
+++ /dev/null
@@ -1,140 +0,0 @@
-use super::*;
-use crate::{policy::Store, test_util};
-use futures::future;
-use linkerd_app_core::{
-    svc::{NewService, ServiceExt},
-    Error,
-};
-use linkerd_proxy_server_policy::{Authentication, Authorization, Meta, ServerPolicy};
-use std::sync::Arc;
-use tokio::{sync::watch, time};
-
-const PROXY_PORT: u16 = 999;
-
-#[tokio::test(flavor = "current_thread")]
-async fn fixed_policy() {
-    let (io, _) = io::duplex(1);
-    let policies = Store::fixed_for_test(std::iter::once((1000, allow_policy())));
-
-    inbound()
-        .with_stack(new_ok())
-        .push_accept(
-            PROXY_PORT,
-            policies,
-            new_panic("direct stack must not be built"),
-        )
-        .into_inner()
-        .new_service(Target(1000))
-        .oneshot(io)
-        .await
-        .expect("should succeed");
-}
-
-#[tokio::test(flavor = "current_thread")]
-async fn discovers() {
-    let (policies, mut disco) = Store::for_test(std::iter::empty());
-
-    let (svc, mut svc_handle) = tower_test::mock::pair::<io::DuplexStream, ()>();
-    let stack = inbound()
-        .with_stack(move |Accept { orig_dst_addr, .. }| {
-            assert_eq!(orig_dst_addr.port(), 1000);
-            svc.clone()
-        })
-        .push_accept(
-            PROXY_PORT,
-            policies,
-            new_panic("direct stack must not be built"),
-        )
-        .into_inner();
-
-    // Start processing a connection.
-    let (io, _) = io::duplex(1);
-    let mut fut = stack.new_service(Target(1000)).oneshot(io);
-
-    // Ensure that a discovery fires for the port.
-    tokio::select! {
-        biased;
-        _ = &mut fut => panic!("unexpected response"),
-        _ = time::sleep(time::Duration::from_millis(100)) => panic!("timed out"),
-        reqrsp = disco.next_request() => {
-            let (port, respond) = reqrsp.expect("should receive request");
-            assert_eq!(port, 1000);
-            // Satisfy the policy.
-            let (tx, rx) = watch::channel(allow_policy());
-            tokio::spawn(async move { tx.closed().await; });
-            respond.send_response(tonic::Response::new(rx));
-        }
-    };
-
-    // Ensure that the connection is routed to the inner service.
-    tokio::select! {
-        biased;
-        _ = &mut fut => panic!("unexpected response"),
-        _ = time::sleep(time::Duration::from_millis(100)) => panic!("timed out"),
-         reqrsp = svc_handle.next_request() => {
-            let (_io, respond) = reqrsp.expect("should receive request");
-            respond.send_response(());
-            fut.await.expect("should succeed");
-        }
-    }
-}
-
-#[tokio::test(flavor = "current_thread")]
-async fn skips_discovery_for_direct() {
-    let (io, _) = io::duplex(1);
-    let policies = Store::fixed_for_test(std::iter::empty());
-    inbound()
-        .with_stack(new_panic("detect stack must not be built"))
-        .push_accept(PROXY_PORT, policies, new_ok())
-        .into_inner()
-        .new_service(Target(PROXY_PORT))
-        .oneshot(io)
-        .await
-        .expect("should succeed");
-}
-
-fn allow_policy() -> ServerPolicy {
-    ServerPolicy {
-        protocol: linkerd_proxy_server_policy::Protocol::Opaque(Arc::new([Authorization {
-            authentication: Authentication::Unauthenticated,
-            networks: vec![Default::default()],
-            meta: Arc::new(Meta::Resource {
-                group: "policy.linkerd.io".into(),
-                kind: "serverauthorization".into(),
-                name: "testsaz".into(),
-            }),
-        }])),
-        meta: Arc::new(Meta::Resource {
-            group: "policy.linkerd.io".into(),
-            kind: "server".into(),
-            name: "testallow".into(),
-        }),
-    }
-}
-
-fn inbound() -> Inbound<()> {
-    Inbound::new(test_util::default_config(), test_util::runtime().0)
-}
-
-fn new_panic<T>(msg: &'static str) -> svc::ArcNewTcp<T, io::DuplexStream> {
-    svc::ArcNewService::new(move |_| panic!("{msg}"))
-}
-
-fn new_ok<T>() -> svc::ArcNewTcp<T, io::DuplexStream> {
-    svc::ArcNewService::new(|_| svc::BoxService::new(svc::mk(|_| future::ok::<(), Error>(()))))
-}
-
-#[derive(Clone, Debug)]
-struct Target(u16);
-
-impl svc::Param<OrigDstAddr> for Target {
-    fn param(&self) -> OrigDstAddr {
-        OrigDstAddr(([192, 0, 2, 2], self.0).into())
-    }
-}
-
-impl svc::Param<Remote<ClientAddr>> for Target {
-    fn param(&self) -> Remote<ClientAddr> {
-        Remote(ClientAddr(([192, 0, 2, 3], 54321).into()))
-    }
-}
diff --git a/linkerd/app/inbound/src/detect/tests.rs b/linkerd/app/inbound/src/detect/tests.rs
index 93223b7c..d7f01502 100644
--- a/linkerd/app/inbound/src/detect/tests.rs
+++ b/linkerd/app/inbound/src/detect/tests.rs
@@ -27,7 +27,7 @@ fn authzs() -> Arc<[Authorization]> {
 
 fn allow(protocol: Protocol) -> AllowPolicy {
     let (allow, _tx) = AllowPolicy::for_test(
-        ServerAddr(orig_dst_addr().into()),
+        orig_dst_addr(),
         ServerPolicy {
             protocol,
             meta: Arc::new(Meta::Resource {
diff --git a/linkerd/app/inbound/src/http/tests.rs b/linkerd/app/inbound/src/http/tests.rs
index 27c4e829..cf55c435 100644
--- a/linkerd/app/inbound/src/http/tests.rs
+++ b/linkerd/app/inbound/src/http/tests.rs
@@ -578,15 +578,11 @@ impl Target {
     fn addr() -> SocketAddr {
         ([127, 0, 0, 1], 80).into()
     }
-
-    fn dst_addr() -> SocketAddr {
-        ([192, 0, 2, 2], 80).into()
-    }
 }
 
 impl svc::Param<OrigDstAddr> for Target {
     fn param(&self) -> OrigDstAddr {
-        OrigDstAddr(Self::dst_addr())
+        OrigDstAddr(([192, 0, 2, 2], 80).into())
     }
 }
 
@@ -626,7 +622,7 @@ impl svc::Param<policy::AllowPolicy> for Target {
             }),
         }]);
         let (policy, _) = policy::AllowPolicy::for_test(
-            ServerAddr(Self::dst_addr()),
+            self.param(),
             policy::ServerPolicy {
                 protocol: policy::Protocol::Http1(Arc::new([
                     linkerd_proxy_server_policy::http::default(authorizations),
diff --git a/linkerd/app/inbound/src/policy/http/tests.rs b/linkerd/app/inbound/src/policy/http/tests.rs
index 4543a5ee..ef711af0 100644
--- a/linkerd/app/inbound/src/policy/http/tests.rs
+++ b/linkerd/app/inbound/src/policy/http/tests.rs
@@ -1,12 +1,12 @@
 use super::*;
 use crate::policy::{Authentication, Authorization, Meta, Protocol, ServerPolicy};
-use linkerd_app_core::{svc::Service, transport::ServerAddr, Infallible};
+use linkerd_app_core::{svc::Service, Infallible};
 use std::sync::Arc;
 
 macro_rules! conn {
     ($client:expr, $dst:expr) => {{
         ConnectionMeta {
-            dst: ServerAddr(($dst, 8080).into()),
+            dst: OrigDstAddr(($dst, 8080).into()),
             client: Remote(ClientAddr(($client, 30120).into())),
             tls: tls::ConditionalServerTls::Some(tls::ServerTls::Established {
                 client_id: Some("foo.bar.bah".parse().unwrap()),
diff --git a/linkerd/app/inbound/src/policy/tcp/tests.rs b/linkerd/app/inbound/src/policy/tcp/tests.rs
index 4c636aa0..40b58c80 100644
--- a/linkerd/app/inbound/src/policy/tcp/tests.rs
+++ b/linkerd/app/inbound/src/policy/tcp/tests.rs
@@ -1,5 +1,6 @@
 use super::*;
 use crate::policy::*;
+use linkerd_app_core::{proxy::http, Error};
 use linkerd_proxy_server_policy::{
     authz::Suffix, Authentication, Authorization, Protocol, ServerPolicy,
 };
@@ -31,12 +32,12 @@ async fn unauthenticated_allowed() {
     };
 
     let tls = tls::ConditionalServerTls::None(tls::NoServerTls::NoClientHello);
-    let permitted = check_authorized(&policy, server_addr(), client_addr(), &tls)
+    let permitted = check_authorized(&policy, orig_dst_addr(), client_addr(), &tls)
         .expect("unauthenticated connection must be permitted");
     assert_eq!(
         permitted,
         ServerPermit {
-            dst: server_addr(),
+            dst: orig_dst_addr(),
             protocol: policy.protocol,
             labels: ServerAuthzLabels {
                 authz: Arc::new(Meta::Resource {
@@ -83,12 +84,12 @@ async fn authenticated_identity() {
         client_id: Some(client_id()),
         negotiated_protocol: None,
     });
-    let permitted = check_authorized(&policy, server_addr(), client_addr(), &tls)
+    let permitted = check_authorized(&policy, orig_dst_addr(), client_addr(), &tls)
         .expect("unauthenticated connection must be permitted");
     assert_eq!(
         permitted,
         ServerPermit {
-            dst: server_addr(),
+            dst: orig_dst_addr(),
             protocol: policy.protocol.clone(),
             labels: ServerAuthzLabels {
                 authz: Arc::new(Meta::Resource {
@@ -113,7 +114,7 @@ async fn authenticated_identity() {
         )),
         negotiated_protocol: None,
     });
-    check_authorized(&policy, server_addr(), client_addr(), &tls)
+    check_authorized(&policy, orig_dst_addr(), client_addr(), &tls)
         .expect_err("policy must require a client identity");
 }
 
@@ -147,10 +148,10 @@ async fn authenticated_suffix() {
         negotiated_protocol: None,
     });
     assert_eq!(
-        check_authorized(&policy, server_addr(), client_addr(), &tls)
+        check_authorized(&policy, orig_dst_addr(), client_addr(), &tls)
             .expect("unauthenticated connection must be permitted"),
         ServerPermit {
-            dst: server_addr(),
+            dst: orig_dst_addr(),
             protocol: policy.protocol.clone(),
             labels: ServerAuthzLabels {
                 authz: Arc::new(Meta::Resource {
@@ -175,7 +176,7 @@ async fn authenticated_suffix() {
         ),
         negotiated_protocol: None,
     });
-    check_authorized(&policy, server_addr(), client_addr(), &tls)
+    check_authorized(&policy, orig_dst_addr(), client_addr(), &tls)
         .expect_err("policy must require a client identity");
 }
 
@@ -206,10 +207,10 @@ async fn tls_unauthenticated() {
         negotiated_protocol: None,
     });
     assert_eq!(
-        check_authorized(&policy, server_addr(), client_addr(), &tls)
+        check_authorized(&policy, orig_dst_addr(), client_addr(), &tls)
             .expect("unauthenticated connection must be permitted"),
         ServerPermit {
-            dst: server_addr(),
+            dst: orig_dst_addr(),
             protocol: policy.protocol.clone(),
             labels: ServerAuthzLabels {
                 authz: Arc::new(Meta::Resource {
@@ -231,7 +232,7 @@ async fn tls_unauthenticated() {
             .parse()
             .unwrap(),
     });
-    check_authorized(&policy, server_addr(), client_addr(), &tls)
+    check_authorized(&policy, orig_dst_addr(), client_addr(), &tls)
         .expect_err("policy must require a TLS termination identity");
 }
 
@@ -245,6 +246,32 @@ fn client_addr() -> Remote<ClientAddr> {
     Remote(ClientAddr(([192, 0, 2, 3], 54321).into()))
 }
 
-fn server_addr() -> ServerAddr {
-    ServerAddr(([192, 0, 2, 2], 1000).into())
+fn orig_dst_addr() -> OrigDstAddr {
+    OrigDstAddr(([192, 0, 2, 2], 1000).into())
+}
+
+impl tonic::client::GrpcService<tonic::body::BoxBody> for MockSvc {
+    type ResponseBody = linkerd_app_core::control::RspBody;
+    type Error = Error;
+    type Future = futures::future::Pending<Result<http::Response<Self::ResponseBody>, Self::Error>>;
+
+    fn poll_ready(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Error>> {
+        unreachable!()
+    }
+
+    fn call(&mut self, _req: http::Request<tonic::body::BoxBody>) -> Self::Future {
+        unreachable!()
+    }
+}
+
+impl Store<MockSvc> {
+    pub(crate) fn for_test(
+        default: impl Into<DefaultPolicy>,
+        ports: impl IntoIterator<Item = (u16, ServerPolicy)>,
+    ) -> Self {
+        Self::spawn_fixed(default.into(), std::time::Duration::MAX, ports)
+    }
 }
diff --git a/linkerd/app/integration/src/tests/transparency.rs b/linkerd/app/integration/src/tests/transparency.rs
index 881996a1..781966b8 100644
--- a/linkerd/app/integration/src/tests/transparency.rs
+++ b/linkerd/app/integration/src/tests/transparency.rs
@@ -221,16 +221,9 @@ async fn test_server_speaks_first(env: TestEnv) {
         .run()
         .await;
 
-    let policy = controller::policy().with_inbound_default(policy::all_unauthenticated());
-    let policy_tx = policy.inbound_tx(srv.addr.port());
-    // disable protocol detection on this port.
-    policy_tx.send(policy::opaque_unauthenticated());
-
-    let policy = policy.run().await;
-
     let proxy = proxy::new()
+        .disable_inbound_ports_protocol_detection(vec![srv.addr.port()])
         .inbound(srv)
-        .policy(policy)
         .run_with_test_env(env)
         .await;
 
@@ -1231,43 +1224,74 @@ mod proxy_to_proxy {
                 self.out_proxy.join_servers(),
             };
         }
-    }
 
-    http1_tests! { proxy: |srv: server::Listening| async move {
-        let ctrl = controller::new();
-        let srv_addr = srv.addr;
-        let dst = format!("transparency.test.svc.cluster.local:{}", srv_addr.port());
-        let _profile_in = ctrl.profile_tx_default(&dst, "transparency.test.svc.cluster.local");
-        let ctrl = ctrl
-            .run()
-            .instrument(tracing::info_span!("ctrl", "inbound"))
-            .await;
-        let inbound = proxy::new().controller(ctrl).inbound(srv).run().await;
+        async fn mk(srv: server::Listening) -> Self {
+            let ctrl = controller::new();
+            let srv_addr = srv.addr;
+            let dst = format!("transparency.test.svc.cluster.local:{}", srv_addr.port());
+            let _profile_in = ctrl.profile_tx_default(&dst, "transparency.test.svc.cluster.local");
+            let ctrl = ctrl
+                .run()
+                .instrument(tracing::info_span!("ctrl", "inbound"))
+                .await;
+            let inbound = proxy::new().controller(ctrl).inbound(srv).run().await;
 
-        let ctrl = controller::new();
-        let _profile_out = ctrl.profile_tx_default(srv_addr, "transparency.test.svc.cluster.local");
-        let dst = ctrl.destination_tx(dst);
-        dst.send_h2_hinted(inbound.inbound);
+            let ctrl = controller::new();
+            let _profile_out =
+                ctrl.profile_tx_default(srv_addr, "transparency.test.svc.cluster.local");
+            let dst = ctrl.destination_tx(dst);
+            dst.send_h2_hinted(inbound.inbound);
 
-        let ctrl = ctrl
-            .run()
-            .instrument(tracing::info_span!("ctrl", "outbound"))
-            .await;
-        let outbound = proxy::new()
-            .controller(ctrl)
-            .outbound_ip(srv_addr)
-            .run().await;
-
-        let addr = outbound.outbound;
-        ProxyToProxy {
-            _dst: dst,
-            _profile_in,
-            _profile_out,
-            in_proxy: inbound,
-            out_proxy: outbound,
-            inbound: addr,
+            let ctrl = ctrl
+                .run()
+                .instrument(tracing::info_span!("ctrl", "outbound"))
+                .await;
+            let outbound = proxy::new()
+                .controller(ctrl)
+                .outbound_ip(srv_addr)
+                .run()
+                .await;
+
+            let addr = outbound.outbound;
+            ProxyToProxy {
+                _dst: dst,
+                _profile_in,
+                _profile_out,
+                in_proxy: inbound,
+                out_proxy: outbound,
+                inbound: addr,
+            }
         }
-    } }
+    }
+
+    http1_tests! { proxy: ProxyToProxy::mk }
+
+    /// Reproduces https://github.com/linkerd/linkerd2/issues/10090 --- an error
+    /// is logged by the inbound proxy's Hyper server due to the presence of a
+    /// `connection: close` header on the HTTP response, which is invalid in HTTP/2.
+    ///
+    /// Unfortunately, this test cannot actually *panic* when this error is
+    /// logged, but the test still exercises the behavior.
+    #[tokio::test]
+    async fn http1_inbound_timeout() {
+        let _trace = trace_init();
+
+        // server will always time out
+        let srv = server::http1()
+            .delay_listen(futures::future::pending())
+            .await;
+        let proxies = ProxyToProxy::mk(srv).await;
+        let client = client::http1(
+            proxies.out_proxy.outbound,
+            "transparency.test.svc.cluster.local",
+        );
+        let res = client.request(client.request_builder("/")).await.unwrap();
+        // tracing::info!(res);
+        assert_eq!(res.status(), http::StatusCode::BAD_GATEWAY);
+
+        // ensure panics from the server are propagated
+        proxies.join_servers().await;
+    }
 }
 
 #[tokio::test]
