diff --git a/internal/jsonpath/jsonpath_test.go b/internal/jsonpath/jsonpath_test.go
index 074db1259..b16abdccb 100644
--- a/internal/jsonpath/jsonpath_test.go
+++ b/internal/jsonpath/jsonpath_test.go
@@ -25,7 +25,18 @@ import (
 func TestJSONPathEval(t *testing.T) {
 	jsonStr := []byte(`{
 		"test": [1, 2, 3],
-		"obj": {"key": "value"}
+		"obj": {"key": "value"},
+		"counts": {
+			"A": { "stored": 10 },
+			"B": { "stored": 5, "uncommited": 5 }
+		},
+		"array": [
+			{"A": "value1"},
+			{"B": "value2"}
+		],
+		"info": {
+			"stored": 100
+		}
 	}`)
 
 	tests := []struct {
@@ -53,7 +64,7 @@ func TestJSONPathEval(t *testing.T) {
 			json:     jsonStr,
 			name:     "access map length",
 			path:     "$.length()",
-			expected: 2,
+			expected: 5,
 			wantErr:  false,
 		},
 		{
@@ -90,6 +101,48 @@ func TestJSONPathEval(t *testing.T) {
 			expected: "value",
 			wantErr:  false,
 		},
+		{
+			json:     jsonStr,
+			name:     "* for map",
+			path:     ".obj.*",
+			expected: []any{"value"},
+			wantErr:  false,
+		},
+		{
+			json:     jsonStr,
+			name:     "*.key",
+			path:     ".counts.*.stored",
+			expected: []any{10.0, 5.0},
+			wantErr:  false,
+		},
+		{
+			json:     jsonStr,
+			name:     "*.*",
+			path:     ".counts.*.*",
+			expected: []any{10.0, 5.0, 5.0},
+			wantErr:  false,
+		},
+		{
+			json:     jsonStr,
+			name:     "*.* for array",
+			path:     ".array.*.*",
+			expected: []any{"value1", "value2"},
+			wantErr:  false,
+		},
+		{
+			json:     jsonStr,
+			name:     "array index",
+			path:     ".test.0",
+			expected: 1.0,
+			wantErr:  false,
+		},
+		{
+			json:     jsonStr,
+			name:     "sum",
+			path:     ".info.sum()",
+			expected: 100.0,
+			wantErr:  false,
+		},
 	}
 
 	for _, tt := range tests {
diff --git a/tests/v2/e2e/assets/index_correction.yaml b/tests/v2/e2e/assets/index_correction.yaml
new file mode 100644
index 000000000..f3312d86b
--- /dev/null
+++ b/tests/v2/e2e/assets/index_correction.yaml
@@ -0,0 +1,228 @@
+#
+# Copyright (C) 2019-2025 vdaas.org vald team <vald@vdaas.org>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+time_zone: UTC
+logging:
+  format: raw
+  level: info
+  logger: glg
+dataset:
+  name: _E2E_DATASET_PATH_
+kubernetes:
+  kube_config: $HOME/.kube/config
+  port_forward:
+    enabled: true
+    local_port: 8082
+    namespace: _E2E_TARGET_NAMESPACE_
+    service_name: _E2E_TARGET_NAME_
+    target_port: 8081
+target:
+  addrs:
+    - 127.0.0.1:8082
+  health_check_duration: "1s"
+  connection_pool:
+    enable_dns_resolver: true
+    enable_rebalance: true
+    old_conn_close_duration: 2m
+    rebalance_duration: 30m
+    size: 3
+  backoff:
+    backoff_factor: 1.1
+    backoff_time_limit: 5s
+    enable_error_log: false
+    initial_duration: 5ms
+    jitter_limit: 100ms
+    maximum_duration: 5s
+    retry_count: 100
+  call_option:
+    content_subtype: ""
+    max_recv_msg_size: 0
+    max_retry_rpc_buffer_size: 0
+    max_send_msg_size: 0
+    wait_for_ready: true
+  dial_option:
+    authority: ""
+    backoff_base_delay: 1s
+    backoff_jitter: 0.2
+    backoff_max_delay: 120s
+    backoff_multiplier: 1.6
+    disable_retry: false
+    enable_backoff: true
+    idle_timeout: 1h
+    initial_connection_window_size: 2097152
+    initial_window_size: 1048576
+    insecure: true
+    interceptors: []
+    keepalive:
+      permit_without_stream: false
+      time: ""
+      timeout: 30s
+    max_call_attempts: 0
+    max_header_list_size: 0
+    max_msg_size: 0
+    min_connection_timeout: 20s
+    net:
+      dialer:
+        dual_stack_enabled: true
+        keepalive: ""
+        timeout: ""
+      dns:
+        cache_enabled: true
+        cache_expiration: 1h
+        refresh_duration: 30m
+      network: tcp
+      socket_option:
+        ip_recover_destination_addr: false
+        ip_transparent: false
+        reuse_addr: true
+        reuse_port: true
+        tcp_cork: false
+        tcp_defer_accept: false
+        tcp_fast_open: false
+        tcp_no_delay: false
+        tcp_quick_ack: false
+      tls:
+        ca: /path/to/ca
+        cert: /path/to/cert
+        enabled: false
+        insecure_skip_verify: false
+        key: /path/to/key
+    read_buffer_size: 0
+    shared_write_buffer: false
+    timeout: ""
+    user_agent: Vald-gRPC
+    write_buffer_size: 0
+  tls:
+    ca: /path/to/ca
+    cert: /path/to/cert
+    enabled: false
+    insecure_skip_verify: false
+    key: /path/to/key
+metadata:
+  key1: sample metadata value1
+  key2: sample metadata value2
+  key3: sample metadata value3
+metadata_string: key4=value4,key5=value5
+strategies:
+  - concurrency: 1
+    name: check Index Property
+    operations:
+      - name: IndexProperty
+        executions:
+          - mode: unary
+            name: IndexProperty
+            type: index_property
+            wait: 3s
+  - concurrency: 1
+    name: Initial Insert and Wait
+    operations:
+      - name: Insert Operation
+        executions:
+          - name: Flush
+            mode: unary
+            type: flush
+          - mode: unary
+            name: IndexInfo
+            type: index_info
+            expect:
+              - value: {}
+          - name: Insert
+            type: insert
+            mode: unary
+            parallelism: 10
+            num: 1000
+            qps: 10
+            wait: 2m
+          - mode: unary
+            name: IndexInfo
+            type: index_info
+            expect:
+              - status_code: ok
+                path: $.*
+                value: 3000
+  - concurrency: 1
+    name: Index Correction -> Check the number of Replica
+    operations:
+      - name: IndexCorrection
+        executions:
+          - name: IndexCorrection
+            type: kubernetes
+            mode: other
+            kubernetes:
+              kind: job
+              namespace: default
+              name: vald-index-correction
+              action: create
+          - name: Wait
+            type: kubernetes
+            mode: other
+            kubernetes:
+              kind: pod
+              namespace: default
+              label_selector: "app=vald-index-correction"
+              action: wait
+              status: completed
+          - mode: unary
+            name: IndexInfo
+            type: index_info
+            expect:
+              - status_code: ok
+                path: $.*
+                value: 3000
+  - concurrency: 1
+    name: Kill an agent -> Index Correction -> Check the number of Replica
+    operations:
+      - name: IndexCorrection
+        executions:
+          - name: KillAgent
+            type: kubernetes
+            mode: other
+            kubernetes:
+              kind: pod
+              namespace: default
+              name: vald-agent-0
+              action: delete
+          - mode: unary
+            name: IndexInfo
+            type: index_info
+            expect:
+              - status_code: ok
+                path: $.*
+                op: lt
+                value: 3000
+          - name: IndexCorrection
+            type: kubernetes
+            mode: other
+            kubernetes:
+              kind: job
+              namespace: default
+              name: vald-index-correction
+              action: create
+          - name: Wait
+            type: kubernetes
+            mode: other
+            kubernetes:
+              kind: pod
+              namespace: default
+              label_selector: "app=vald-index-correction"
+              status: completed
+              action: wait
+          - mode: unary
+            name: IndexInfo
+            type: index_info
+            expect:
+              - status_code: ok
+                path: $.*
+                value: 3000
diff --git a/tests/v2/e2e/crud/grpc_test.go b/tests/v2/e2e/crud/grpc_test.go
index d1522b212..954652867 100644
--- a/tests/v2/e2e/crud/grpc_test.go
+++ b/tests/v2/e2e/crud/grpc_test.go
@@ -125,7 +125,7 @@ func handleGRPCWithStatusCode(
 			case config.Gt, config.Ge, config.Lt, config.Le:
 				a, b, ok := compare(val, expect.Value)
 				if !ok {
-					errs = append(errs, commonErr)
+					errs = append(errs, fmt.Errorf("assert_%v failed, cannot compare values, JSONPath: %s, expected: %v actual: %v", expect.Op, expect.Path, expect.Value, val))
 					continue
 				}
 				switch op {
diff --git a/tests/v2/e2e/crud/kubernetes_test.go b/tests/v2/e2e/crud/kubernetes_test.go
index cf9b740e8..fa4cb1a5e 100644
--- a/tests/v2/e2e/crud/kubernetes_test.go
+++ b/tests/v2/e2e/crud/kubernetes_test.go
@@ -21,6 +21,8 @@ package crud
 
 import (
 	"context"
+	"fmt"
+	"os"
 	"testing"
 
 	"github.com/vdaas/vald/internal/log"
@@ -133,12 +135,21 @@ func (r *runner) processKubernetes(t *testing.T, ctx context.Context, plan *conf
 		if !ok {
 			t.Errorf("failed to wait for %s: %v", plan.Kubernetes.Kind, err)
 		}
+	case config.KubernetesActionCreate:
+		if plan.Kubernetes.Kind == config.Job {
+			cronJob := kubernetes.CronJob(r.k8s, plan.Kubernetes.Namespace)
+			_, err = cronJob.CreateJob(ctx, plan.Kubernetes.Name, kubernetes.EmptyGetOptions, kubernetes.EmptyCreateOptions)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "failed to create job from cronjob: %v", err)
+				t.Errorf("failed to create job from cronjob: %v\n", err)
+			}
+			return
+		}
+		t.Errorf("kubernetes action create is only supported for creating job from cronjob")
 	case config.KubernetesActionExec:
 		t.Errorf("kubernetes action %s is not supported yet", plan.Kubernetes.Action)
 	case config.KubernetesActionApply:
 		t.Errorf("kubernetes action %s is not supported yet", plan.Kubernetes.Action)
-	case config.KubernetesActionCreate:
-		t.Errorf("kubernetes action %s is not supported yet", plan.Kubernetes.Action)
 	case config.KubernetesActionPatch:
 		t.Errorf("kubernetes action %s is not supported yet", plan.Kubernetes.Action)
 	case config.KubernetesActionScale:
diff --git a/tests/v2/e2e/kubernetes/client.go b/tests/v2/e2e/kubernetes/client.go
index 498fdfb85..7b5336361 100644
--- a/tests/v2/e2e/kubernetes/client.go
+++ b/tests/v2/e2e/kubernetes/client.go
@@ -16,7 +16,6 @@
 // limitations under the License.
 //
 
-// package kubernetes provides kubernetes client functions
 package kubernetes
 
 import (
diff --git a/tests/v2/e2e/kubernetes/doc.go b/tests/v2/e2e/kubernetes/doc.go
new file mode 100644
index 000000000..664a6eeff
--- /dev/null
+++ b/tests/v2/e2e/kubernetes/doc.go
@@ -0,0 +1,20 @@
+//go:build e2e
+
+//
+// Copyright (C) 2019-2025 vdaas.org vald team <vald@vdaas.org>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// package kubernetes provides kubernetes client functions
+package kubernetes
diff --git a/tests/v2/e2e/kubernetes/resources.go b/tests/v2/e2e/kubernetes/resources.go
index 540be8aea..b26297f2d 100644
--- a/tests/v2/e2e/kubernetes/resources.go
+++ b/tests/v2/e2e/kubernetes/resources.go
@@ -16,11 +16,12 @@
 // limitations under the License.
 //
 
-// package kubernetes provides kubernetes e2e tests
 package kubernetes
 
 import (
 	"context"
+	"fmt"
+	"time"
 
 	"github.com/vdaas/vald/internal/errors"
 	"github.com/vdaas/vald/internal/sync"
@@ -115,6 +116,12 @@ type WorkloadControllerResourceClient[T Object, L ObjectList, C NamedObject] int
 	WorkloadResourceClient[T, L, C]
 	PodTemplateInterface[T]
 	PodAnnotationInterface[T]
+	CreateJob(
+		ctx context.Context,
+		from string,
+		gopts metav1.GetOptions,
+		copts metav1.CreateOptions,
+	) (*batchv1.Job, error)
 }
 
 type (
@@ -614,3 +621,28 @@ func (b *baseClient[T, L, C]) SetPodAnnotations(
 	}
 	return b.Update(ctx, obj, uopts)
 }
+
+func (b *baseClient[T, L, C]) CreateJob(
+	ctx context.Context, from string, gopts metav1.GetOptions, copts metav1.CreateOptions,
+) (*batchv1.Job, error) {
+	cobj, err := b.Get(ctx, from, gopts)
+	if err != nil {
+		return nil, err
+	}
+	tmpl, err := b.GetPodTemplate(cobj)
+	if err != nil {
+		return nil, err
+	}
+	if tmpl == nil {
+		return nil, errors.ErrPodTemplateNotFound
+	}
+	return Job(b.GetClient(), b.GetNamespace()).Create(ctx, &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", from, time.Now().Format("2006-01-02-15-04-05")),
+			Namespace: b.GetNamespace(),
+		},
+		Spec: batchv1.JobSpec{
+			Template: *tmpl,
+		},
+	}, copts)
+}
diff --git a/tests/v2/e2e/kubernetes/rollout.go b/tests/v2/e2e/kubernetes/rollout.go
index 8d685edd7..ef12c17d4 100644
--- a/tests/v2/e2e/kubernetes/rollout.go
+++ b/tests/v2/e2e/kubernetes/rollout.go
@@ -16,7 +16,6 @@
 // limitations under the License.
 //
 
-// package kubernetes provides kubernetes e2e tests
 package kubernetes
 
 import (
diff --git a/tests/v2/e2e/kubernetes/status.go b/tests/v2/e2e/kubernetes/status.go
index ab1aa1a6b..21ac8d410 100644
--- a/tests/v2/e2e/kubernetes/status.go
+++ b/tests/v2/e2e/kubernetes/status.go
@@ -16,7 +16,6 @@
 // limitations under the License.
 //
 
-// package kubernetes provides kubernetes e2e tests
 package kubernetes
 
 import (
@@ -149,7 +148,7 @@ func WaitForStatus[T Object, L ObjectList, C NamedObject, I ResourceInterface[T,
 			items, err := extractItems[T](l)
 			// if no resources found yet, keep polling
 			if len(items) == 0 {
-				continue
+				return false, errors.New("No resource matched for labelSelector: " + labelSelector)
 			}
 			if err != nil {
 				return false, errors.Wrap(err, "failed to extract items")
