diff --git a/cmd/mdatagen/internal/loader_test.go b/cmd/mdatagen/internal/loader_test.go
index ec2912b4a..27c24655f 100644
--- a/cmd/mdatagen/internal/loader_test.go
+++ b/cmd/mdatagen/internal/loader_test.go
@@ -460,7 +460,7 @@ func TestLoadMetadata(t *testing.T) {
 								Since: "1.5.0",
 								Note:  "This metric will be removed in favor of batch_send_trigger_size",
 							},
-							Unit: strPtr("{times}"),
+							Unit: strPtr("{time}"),
 							Sum: &Sum{
 								MetricValueType: MetricValueType{pmetric.NumberDataPointValueTypeInt},
 								Mono:            Mono{Monotonic: true},
@@ -500,7 +500,7 @@ func TestLoadMetadata(t *testing.T) {
 								Description:           "This metric is optional and therefore not initialized in NewTelemetryBuilder.",
 								ExtendedDocumentation: "For example this metric only exists if feature A is enabled.",
 							},
-							Unit:     strPtr("{items}"),
+							Unit:     strPtr("{item}"),
 							Optional: true,
 							Gauge: &Gauge{
 								MetricValueType: MetricValueType{
@@ -515,7 +515,7 @@ func TestLoadMetadata(t *testing.T) {
 								Description: "Queue capacity - sync gauge example.",
 								Stability:   component.StabilityLevelDevelopment,
 							},
-							Unit: strPtr("{items}"),
+							Unit: strPtr("{item}"),
 							Gauge: &Gauge{
 								MetricValueType: MetricValueType{
 									ValueType: pmetric.NumberDataPointValueTypeInt,
diff --git a/cmd/mdatagen/internal/samplereceiver/internal/metadatatest/generated_telemetrytest.go b/cmd/mdatagen/internal/samplereceiver/internal/metadatatest/generated_telemetrytest.go
index 81261758c..12c6c4ceb 100644
--- a/cmd/mdatagen/internal/samplereceiver/internal/metadatatest/generated_telemetrytest.go
+++ b/cmd/mdatagen/internal/samplereceiver/internal/metadatatest/generated_telemetrytest.go
@@ -26,7 +26,7 @@ func AssertEqualBatchSizeTriggerSend(t *testing.T, tt *componenttest.Telemetry,
 	want := metricdata.Metrics{
 		Name:        "otelcol_batch_size_trigger_send",
 		Description: "Number of times the batch was sent due to a size trigger [Deprecated]",
-		Unit:        "{times}",
+		Unit:        "{time}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -58,7 +58,7 @@ func AssertEqualQueueCapacity(t *testing.T, tt *componenttest.Telemetry, dps []m
 	want := metricdata.Metrics{
 		Name:        "otelcol_queue_capacity",
 		Description: "Queue capacity - sync gauge example. [Development]",
-		Unit:        "{items}",
+		Unit:        "{item}",
 		Data: metricdata.Gauge[int64]{
 			DataPoints: dps,
 		},
@@ -72,7 +72,7 @@ func AssertEqualQueueLength(t *testing.T, tt *componenttest.Telemetry, dps []met
 	want := metricdata.Metrics{
 		Name:        "otelcol_queue_length",
 		Description: "This metric is optional and therefore not initialized in NewTelemetryBuilder. [Alpha]",
-		Unit:        "{items}",
+		Unit:        "{item}",
 		Data: metricdata.Gauge[int64]{
 			DataPoints: dps,
 		},
diff --git a/exporter/exporterhelper/internal/metadatatest/generated_telemetrytest.go b/exporter/exporterhelper/internal/metadatatest/generated_telemetrytest.go
index 46bc8152a..2c501440d 100644
--- a/exporter/exporterhelper/internal/metadatatest/generated_telemetrytest.go
+++ b/exporter/exporterhelper/internal/metadatatest/generated_telemetrytest.go
@@ -16,7 +16,7 @@ func AssertEqualExporterEnqueueFailedLogRecords(t *testing.T, tt *componenttest.
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_enqueue_failed_log_records",
 		Description: "Number of log records failed to be added to the sending queue. [Alpha]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -32,7 +32,7 @@ func AssertEqualExporterEnqueueFailedMetricPoints(t *testing.T, tt *componenttes
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_enqueue_failed_metric_points",
 		Description: "Number of metric points failed to be added to the sending queue. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -48,7 +48,7 @@ func AssertEqualExporterEnqueueFailedProfileSamples(t *testing.T, tt *componentt
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_enqueue_failed_profile_samples",
 		Description: "Number of profile samples failed to be added to the sending queue. [Development]",
-		Unit:        "{samples}",
+		Unit:        "{sample}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -64,7 +64,7 @@ func AssertEqualExporterEnqueueFailedSpans(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_enqueue_failed_spans",
 		Description: "Number of spans failed to be added to the sending queue. [Alpha]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -80,7 +80,7 @@ func AssertEqualExporterQueueBatchSendSize(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_queue_batch_send_size",
 		Description: "Number of units in the batch [Development]",
-		Unit:        "{units}",
+		Unit:        "{unit}",
 		Data: metricdata.Histogram[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			DataPoints:  dps,
@@ -110,7 +110,7 @@ func AssertEqualExporterQueueCapacity(t *testing.T, tt *componenttest.Telemetry,
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_queue_capacity",
 		Description: "Fixed capacity of the retry queue (in batches). [Alpha]",
-		Unit:        "{batches}",
+		Unit:        "{batch}",
 		Data: metricdata.Gauge[int64]{
 			DataPoints: dps,
 		},
@@ -124,7 +124,7 @@ func AssertEqualExporterQueueSize(t *testing.T, tt *componenttest.Telemetry, dps
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_queue_size",
 		Description: "Current size of the retry queue (in batches). [Alpha]",
-		Unit:        "{batches}",
+		Unit:        "{batch}",
 		Data: metricdata.Gauge[int64]{
 			DataPoints: dps,
 		},
@@ -138,7 +138,7 @@ func AssertEqualExporterSendFailedLogRecords(t *testing.T, tt *componenttest.Tel
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_send_failed_log_records",
 		Description: "Number of log records in failed attempts to send to destination. At detailed telemetry level, includes attributes: error.type (semantic convention), error.permanent. [Alpha]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -154,7 +154,7 @@ func AssertEqualExporterSendFailedMetricPoints(t *testing.T, tt *componenttest.T
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_send_failed_metric_points",
 		Description: "Number of metric points in failed attempts to send to destination. At detailed telemetry level, includes attributes: error.type (semantic convention), error.permanent. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -170,7 +170,7 @@ func AssertEqualExporterSendFailedProfileSamples(t *testing.T, tt *componenttest
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_send_failed_profile_samples",
 		Description: "Number of profile samples in failed attempts to send to destination. At detailed telemetry level, includes attributes: error.type (semantic convention), error.permanent. [Development]",
-		Unit:        "{samples}",
+		Unit:        "{sample}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -186,7 +186,7 @@ func AssertEqualExporterSendFailedSpans(t *testing.T, tt *componenttest.Telemetr
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_send_failed_spans",
 		Description: "Number of spans in failed attempts to send to destination. At detailed telemetry level, includes attributes: error.type (semantic convention), error.permanent. [Alpha]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -202,7 +202,7 @@ func AssertEqualExporterSentLogRecords(t *testing.T, tt *componenttest.Telemetry
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_sent_log_records",
 		Description: "Number of log record successfully sent to destination. [Alpha]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -218,7 +218,7 @@ func AssertEqualExporterSentMetricPoints(t *testing.T, tt *componenttest.Telemet
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_sent_metric_points",
 		Description: "Number of metric points successfully sent to destination. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -234,7 +234,7 @@ func AssertEqualExporterSentProfileSamples(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_sent_profile_samples",
 		Description: "Number of profile samples successfully sent to destination. [Development]",
-		Unit:        "{samples}",
+		Unit:        "{sample}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -250,7 +250,7 @@ func AssertEqualExporterSentSpans(t *testing.T, tt *componenttest.Telemetry, dps
 	want := metricdata.Metrics{
 		Name:        "otelcol_exporter_sent_spans",
 		Description: "Number of spans successfully sent to destination. [Alpha]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
diff --git a/processor/batchprocessor/internal/metadatatest/generated_telemetrytest.go b/processor/batchprocessor/internal/metadatatest/generated_telemetrytest.go
index 925d5ec4e..e31007e54 100644
--- a/processor/batchprocessor/internal/metadatatest/generated_telemetrytest.go
+++ b/processor/batchprocessor/internal/metadatatest/generated_telemetrytest.go
@@ -26,7 +26,7 @@ func AssertEqualProcessorBatchBatchSendSize(t *testing.T, tt *componenttest.Tele
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_batch_batch_send_size",
 		Description: "Number of units in the batch [Development]",
-		Unit:        "{units}",
+		Unit:        "{unit}",
 		Data: metricdata.Histogram[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			DataPoints:  dps,
@@ -56,7 +56,7 @@ func AssertEqualProcessorBatchBatchSizeTriggerSend(t *testing.T, tt *componentte
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_batch_batch_size_trigger_send",
 		Description: "Number of times the batch was sent due to a size trigger [Development]",
-		Unit:        "{times}",
+		Unit:        "{time}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -72,7 +72,7 @@ func AssertEqualProcessorBatchMetadataCardinality(t *testing.T, tt *componenttes
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_batch_metadata_cardinality",
 		Description: "Number of distinct metadata value combinations being processed [Development]",
-		Unit:        "{combinations}",
+		Unit:        "{combination}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: false,
@@ -88,7 +88,7 @@ func AssertEqualProcessorBatchTimeoutTriggerSend(t *testing.T, tt *componenttest
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_batch_timeout_trigger_send",
 		Description: "Number of times the batch was sent due to a timeout trigger [Development]",
-		Unit:        "{times}",
+		Unit:        "{time}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
diff --git a/processor/memorylimiterprocessor/internal/metadatatest/generated_telemetrytest.go b/processor/memorylimiterprocessor/internal/metadatatest/generated_telemetrytest.go
index c4a34915d..2b9584a11 100644
--- a/processor/memorylimiterprocessor/internal/metadatatest/generated_telemetrytest.go
+++ b/processor/memorylimiterprocessor/internal/metadatatest/generated_telemetrytest.go
@@ -26,7 +26,7 @@ func AssertEqualProcessorAcceptedLogRecords(t *testing.T, tt *componenttest.Tele
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_accepted_log_records",
 		Description: "Number of log records successfully pushed into the next component in the pipeline. [Deprecated]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -42,7 +42,7 @@ func AssertEqualProcessorAcceptedMetricPoints(t *testing.T, tt *componenttest.Te
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_accepted_metric_points",
 		Description: "Number of metric points successfully pushed into the next component in the pipeline. [Deprecated]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -58,7 +58,7 @@ func AssertEqualProcessorAcceptedSpans(t *testing.T, tt *componenttest.Telemetry
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_accepted_spans",
 		Description: "Number of spans successfully pushed into the next component in the pipeline. [Deprecated]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -74,7 +74,7 @@ func AssertEqualProcessorRefusedLogRecords(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_refused_log_records",
 		Description: "Number of log records that were rejected by the next component in the pipeline. [Deprecated]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -90,7 +90,7 @@ func AssertEqualProcessorRefusedMetricPoints(t *testing.T, tt *componenttest.Tel
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_refused_metric_points",
 		Description: "Number of metric points that were rejected by the next component in the pipeline. [Deprecated]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -106,7 +106,7 @@ func AssertEqualProcessorRefusedSpans(t *testing.T, tt *componenttest.Telemetry,
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_refused_spans",
 		Description: "Number of spans that were rejected by the next component in the pipeline. [Deprecated]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
diff --git a/processor/processorhelper/internal/metadatatest/generated_telemetrytest.go b/processor/processorhelper/internal/metadatatest/generated_telemetrytest.go
index cf5012f40..614b93dfe 100644
--- a/processor/processorhelper/internal/metadatatest/generated_telemetrytest.go
+++ b/processor/processorhelper/internal/metadatatest/generated_telemetrytest.go
@@ -16,7 +16,7 @@ func AssertEqualProcessorIncomingItems(t *testing.T, tt *componenttest.Telemetry
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_incoming_items",
 		Description: "Number of items passed to the processor. [Alpha]",
-		Unit:        "{items}",
+		Unit:        "{item}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -47,7 +47,7 @@ func AssertEqualProcessorOutgoingItems(t *testing.T, tt *componenttest.Telemetry
 	want := metricdata.Metrics{
 		Name:        "otelcol_processor_outgoing_items",
 		Description: "Number of items emitted from the processor. [Alpha]",
-		Unit:        "{items}",
+		Unit:        "{item}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
diff --git a/receiver/otlpreceiver/otlp_test.go b/receiver/otlpreceiver/otlp_test.go
index 59c8bf30a..f80cbb30f 100644
--- a/receiver/otlpreceiver/otlp_test.go
+++ b/receiver/otlpreceiver/otlp_test.go
@@ -1329,7 +1329,7 @@ func assertReceiverTraces(t *testing.T, tt *componenttest.Telemetry, id componen
 		metricdata.Metrics{
 			Name:        "otelcol_receiver_failed_spans",
 			Description: "The number of spans that failed to be processed by the receiver due to internal errors. [Alpha]",
-			Unit:        "{spans}",
+			Unit:        "{span}",
 			Data: metricdata.Sum[int64]{
 				Temporality: metricdata.CumulativeTemporality,
 				IsMonotonic: true,
@@ -1350,7 +1350,7 @@ func assertReceiverTraces(t *testing.T, tt *componenttest.Telemetry, id componen
 		metricdata.Metrics{
 			Name:        "otelcol_receiver_accepted_spans",
 			Description: "Number of spans successfully pushed into the pipeline. [Alpha]",
-			Unit:        "{spans}",
+			Unit:        "{span}",
 			Data: metricdata.Sum[int64]{
 				Temporality: metricdata.CumulativeTemporality,
 				IsMonotonic: true,
@@ -1371,7 +1371,7 @@ func assertReceiverTraces(t *testing.T, tt *componenttest.Telemetry, id componen
 		metricdata.Metrics{
 			Name:        "otelcol_receiver_refused_spans",
 			Description: "Number of spans that could not be pushed into the pipeline. [Alpha]",
-			Unit:        "{spans}",
+			Unit:        "{span}",
 			Data: metricdata.Sum[int64]{
 				Temporality: metricdata.CumulativeTemporality,
 				IsMonotonic: true,
@@ -1416,7 +1416,7 @@ func assertReceiverTraces(t *testing.T, tt *componenttest.Telemetry, id componen
 			metricdata.Metrics{
 				Name:        "otelcol_receiver_requests",
 				Description: "The number of requests performed.",
-				Unit:        "{requests}",
+				Unit:        "{request}",
 				Data: metricdata.Sum[int64]{
 					Temporality: metricdata.CumulativeTemporality,
 					IsMonotonic: true,
@@ -1448,7 +1448,7 @@ func assertReceiverMetrics(t *testing.T, tt *componenttest.Telemetry, id compone
 		metricdata.Metrics{
 			Name:        "otelcol_receiver_failed_metric_points",
 			Description: "The number of metric points that failed to be processed by the receiver due to internal errors. [Alpha]",
-			Unit:        "{datapoints}",
+			Unit:        "{datapoint}",
 			Data: metricdata.Sum[int64]{
 				Temporality: metricdata.CumulativeTemporality,
 				IsMonotonic: true,
@@ -1469,7 +1469,7 @@ func assertReceiverMetrics(t *testing.T, tt *componenttest.Telemetry, id compone
 		metricdata.Metrics{
 			Name:        "otelcol_receiver_accepted_metric_points",
 			Description: "Number of metric points successfully pushed into the pipeline. [Alpha]",
-			Unit:        "{datapoints}",
+			Unit:        "{datapoint}",
 			Data: metricdata.Sum[int64]{
 				Temporality: metricdata.CumulativeTemporality,
 				IsMonotonic: true,
@@ -1490,7 +1490,7 @@ func assertReceiverMetrics(t *testing.T, tt *componenttest.Telemetry, id compone
 		metricdata.Metrics{
 			Name:        "otelcol_receiver_refused_metric_points",
 			Description: "Number of metric points that could not be pushed into the pipeline. [Alpha]",
-			Unit:        "{datapoints}",
+			Unit:        "{datapoint}",
 			Data: metricdata.Sum[int64]{
 				Temporality: metricdata.CumulativeTemporality,
 				IsMonotonic: true,
@@ -1535,7 +1535,7 @@ func assertReceiverMetrics(t *testing.T, tt *componenttest.Telemetry, id compone
 			metricdata.Metrics{
 				Name:        "otelcol_receiver_requests",
 				Description: "The number of requests performed.",
-				Unit:        "{requests}",
+				Unit:        "{request}",
 				Data: metricdata.Sum[int64]{
 					Temporality: metricdata.CumulativeTemporality,
 					IsMonotonic: true,
diff --git a/receiver/receiverhelper/internal/metadatatest/generated_telemetrytest.go b/receiver/receiverhelper/internal/metadatatest/generated_telemetrytest.go
index 00818ffa0..cd435bd19 100644
--- a/receiver/receiverhelper/internal/metadatatest/generated_telemetrytest.go
+++ b/receiver/receiverhelper/internal/metadatatest/generated_telemetrytest.go
@@ -16,7 +16,7 @@ func AssertEqualReceiverAcceptedLogRecords(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_accepted_log_records",
 		Description: "Number of log records successfully pushed into the pipeline. [Alpha]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -32,7 +32,7 @@ func AssertEqualReceiverAcceptedMetricPoints(t *testing.T, tt *componenttest.Tel
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_accepted_metric_points",
 		Description: "Number of metric points successfully pushed into the pipeline. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -48,7 +48,7 @@ func AssertEqualReceiverAcceptedProfileSamples(t *testing.T, tt *componenttest.T
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_accepted_profile_samples",
 		Description: "Number of profile samples successfully pushed into the pipeline. [Alpha]",
-		Unit:        "{samples}",
+		Unit:        "{sample}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -64,7 +64,7 @@ func AssertEqualReceiverAcceptedSpans(t *testing.T, tt *componenttest.Telemetry,
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_accepted_spans",
 		Description: "Number of spans successfully pushed into the pipeline. [Alpha]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -80,7 +80,7 @@ func AssertEqualReceiverFailedLogRecords(t *testing.T, tt *componenttest.Telemet
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_failed_log_records",
 		Description: "The number of log records that failed to be processed by the receiver due to internal errors. [Alpha]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -96,7 +96,7 @@ func AssertEqualReceiverFailedMetricPoints(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_failed_metric_points",
 		Description: "The number of metric points that failed to be processed by the receiver due to internal errors. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -112,7 +112,7 @@ func AssertEqualReceiverFailedProfileSamples(t *testing.T, tt *componenttest.Tel
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_failed_profile_samples",
 		Description: "The number of profile samples that failed to be processed by the receiver due to internal errors. [Alpha]",
-		Unit:        "{samples}",
+		Unit:        "{sample}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -128,7 +128,7 @@ func AssertEqualReceiverFailedSpans(t *testing.T, tt *componenttest.Telemetry, d
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_failed_spans",
 		Description: "The number of spans that failed to be processed by the receiver due to internal errors. [Alpha]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -144,7 +144,7 @@ func AssertEqualReceiverRefusedLogRecords(t *testing.T, tt *componenttest.Teleme
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_refused_log_records",
 		Description: "Number of log records that could not be pushed into the pipeline. [Alpha]",
-		Unit:        "{records}",
+		Unit:        "{record}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -160,7 +160,7 @@ func AssertEqualReceiverRefusedMetricPoints(t *testing.T, tt *componenttest.Tele
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_refused_metric_points",
 		Description: "Number of metric points that could not be pushed into the pipeline. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -176,7 +176,7 @@ func AssertEqualReceiverRefusedProfileSamples(t *testing.T, tt *componenttest.Te
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_refused_profile_samples",
 		Description: "Number of profile samples that could not be pushed into the pipeline. [Alpha]",
-		Unit:        "{samples}",
+		Unit:        "{sample}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -192,7 +192,7 @@ func AssertEqualReceiverRefusedSpans(t *testing.T, tt *componenttest.Telemetry,
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_refused_spans",
 		Description: "Number of spans that could not be pushed into the pipeline. [Alpha]",
-		Unit:        "{spans}",
+		Unit:        "{span}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -208,7 +208,7 @@ func AssertEqualReceiverRequests(t *testing.T, tt *componenttest.Telemetry, dps
 	want := metricdata.Metrics{
 		Name:        "otelcol_receiver_requests",
 		Description: "The number of requests performed. [Alpha]",
-		Unit:        "{requests}",
+		Unit:        "{request}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
diff --git a/scraper/scraperhelper/internal/metadatatest/generated_telemetrytest.go b/scraper/scraperhelper/internal/metadatatest/generated_telemetrytest.go
index 8e86843fd..d4da93971 100644
--- a/scraper/scraperhelper/internal/metadatatest/generated_telemetrytest.go
+++ b/scraper/scraperhelper/internal/metadatatest/generated_telemetrytest.go
@@ -16,7 +16,7 @@ func AssertEqualScraperErroredLogRecords(t *testing.T, tt *componenttest.Telemet
 	want := metricdata.Metrics{
 		Name:        "otelcol_scraper_errored_log_records",
 		Description: "Number of log records that were unable to be scraped. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -32,7 +32,7 @@ func AssertEqualScraperErroredMetricPoints(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_scraper_errored_metric_points",
 		Description: "Number of metric points that were unable to be scraped. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -48,7 +48,7 @@ func AssertEqualScraperScrapedLogRecords(t *testing.T, tt *componenttest.Telemet
 	want := metricdata.Metrics{
 		Name:        "otelcol_scraper_scraped_log_records",
 		Description: "Number of log records successfully scraped. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -64,7 +64,7 @@ func AssertEqualScraperScrapedMetricPoints(t *testing.T, tt *componenttest.Telem
 	want := metricdata.Metrics{
 		Name:        "otelcol_scraper_scraped_metric_points",
 		Description: "Number of metric points successfully scraped. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
diff --git a/scraper/scraperhelper/xscraperhelper/internal/metadatatest/generated_telemetrytest.go b/scraper/scraperhelper/xscraperhelper/internal/metadatatest/generated_telemetrytest.go
index 55e46c580..dd5f3aca2 100644
--- a/scraper/scraperhelper/xscraperhelper/internal/metadatatest/generated_telemetrytest.go
+++ b/scraper/scraperhelper/xscraperhelper/internal/metadatatest/generated_telemetrytest.go
@@ -16,7 +16,7 @@ func AssertEqualScraperErroredProfileRecords(t *testing.T, tt *componenttest.Tel
 	want := metricdata.Metrics{
 		Name:        "otelcol_scraper_errored_profile_records",
 		Description: "Number of profile records that were unable to be scraped. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
@@ -32,7 +32,7 @@ func AssertEqualScraperScrapedProfileRecords(t *testing.T, tt *componenttest.Tel
 	want := metricdata.Metrics{
 		Name:        "otelcol_scraper_scraped_profile_records",
 		Description: "Number of profile records successfully scraped. [Alpha]",
-		Unit:        "{datapoints}",
+		Unit:        "{datapoint}",
 		Data: metricdata.Sum[int64]{
 			Temporality: metricdata.CumulativeTemporality,
 			IsMonotonic: true,
