diff --git a/internal/cache/metadata/type_cache_test.go b/internal/cache/metadata/type_cache_test.go
new file mode 100644
index 000000000..6becf4c5f
--- /dev/null
+++ b/internal/cache/metadata/type_cache_test.go
@@ -0,0 +1,216 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	. "github.com/jacobsa/ogletest"
+)
+
+const (
+	TTL                 time.Duration = time.Millisecond
+	TypeCacheMaxEntries               = 5
+)
+
+var (
+	now               time.Time = time.Now()
+	expiration        time.Time = now.Add(TTL)
+	beforeExpiration  time.Time = expiration.Add(-time.Nanosecond)
+	afterExpiration   time.Time = expiration.Add(time.Nanosecond)
+	now2              time.Time = now.Add(TTL / 2)
+	expiration2       time.Time = now2.Add(TTL)
+	beforeExpiration2 time.Time = expiration2.Add(-time.Nanosecond)
+)
+
+func TestTypeCache(t *testing.T) { RunTests(t) }
+
+////////////////////////////////////////////////////////////////////////
+// Boilerplate
+////////////////////////////////////////////////////////////////////////
+
+type TypeCacheTest struct {
+	cache TypeCache
+	ttl   time.Duration
+}
+
+type ZeroMaxEntriesTypeCacheTest struct {
+	cache TypeCache
+	ttl   time.Duration
+}
+
+type ZeroTtlTypeCacheTest struct {
+	cache TypeCache
+}
+
+func init() {
+	RegisterTestSuite(&TypeCacheTest{})
+	RegisterTestSuite(&ZeroMaxEntriesTypeCacheTest{})
+	RegisterTestSuite(&ZeroTtlTypeCacheTest{})
+}
+
+func (t *TypeCacheTest) SetUp(ti *TestInfo) {
+	t.ttl = TTL
+	t.cache = createNewTypeCache(TypeCacheMaxEntries, t.ttl)
+}
+
+func (t *ZeroMaxEntriesTypeCacheTest) SetUp(ti *TestInfo) {
+	t.ttl = TTL
+	t.cache = createNewTypeCache(0, t.ttl)
+}
+
+func (t *ZeroTtlTypeCacheTest) SetUp(ti *TestInfo) {
+	t.cache = createNewTypeCache(TypeCacheMaxEntries, 0)
+}
+
+////////////////////////////////////////////////////////////////////////
+// Helpers
+////////////////////////////////////////////////////////////////////////
+
+func createNewTypeCache(maxEntries int, ttl time.Duration) *typeCache {
+	tc := NewTypeCache(maxEntries, ttl)
+
+	AssertNe(nil, tc)
+	AssertNe(nil, tc.(*typeCache))
+
+	return tc.(*typeCache)
+}
+
+////////////////////////////////////////////////////////////////////////
+// Tests for regulat TypeCache - TypeCacheTest
+////////////////////////////////////////////////////////////////////////
+
+func (t *TypeCacheTest) TestNewTypeCache() {
+	input := []struct {
+		maxEntries         int
+		ttl                time.Duration
+		entriesShouldBeNil bool
+	}{
+		{
+			maxEntries:         0,
+			ttl:                time.Second,
+			entriesShouldBeNil: true,
+		},
+		{
+			maxEntries:         1,
+			ttl:                0,
+			entriesShouldBeNil: true,
+		},
+		{
+			maxEntries: -1,
+			ttl:        time.Second,
+		},
+		{
+			maxEntries: 1,
+			ttl:        time.Second,
+		}}
+
+	for _, input := range input {
+		tc := createNewTypeCache(input.maxEntries, input.ttl)
+
+		AssertEq(input.entriesShouldBeNil, tc.entries == nil)
+	}
+}
+
+func (t *TypeCacheTest) TestGetFromEmptyTypeCache() {
+	ExpectEq(UnknownType, t.cache.Get(now, "abc"))
+}
+
+func (t *TypeCacheTest) TestGetUninsertedEntry() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+
+	ExpectEq(UnknownType, t.cache.Get(beforeExpiration, "abc"))
+}
+
+func (t *TypeCacheTest) TestGetOverwrittenEntry() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+	t.cache.Insert(now, "abcd", ExplicitDirType)
+
+	ExpectEq(ExplicitDirType, t.cache.Get(beforeExpiration, "abcd"))
+}
+
+func (t *TypeCacheTest) TestGetBeforeTtlExpiration() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+
+	ExpectEq(RegularFileType, t.cache.Get(beforeExpiration, "abcd"))
+}
+
+func (t *TypeCacheTest) TestGetAfterTtlExpiration() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+
+	ExpectEq(UnknownType, t.cache.Get(afterExpiration, "abcd"))
+}
+
+func (t *TypeCacheTest) TestGetAfterMaxEntriesExpiration() {
+	entriesToBeInserted := TypeCacheMaxEntries + 1
+
+	for i := 0; i < entriesToBeInserted; i++ {
+		t.cache.Insert(now, fmt.Sprint(i), RegularFileType)
+	}
+
+	// Verify that Get works, by accessing the last entry inserted.
+	ExpectEq(RegularFileType, t.cache.Get(beforeExpiration, fmt.Sprint(entriesToBeInserted-1)))
+
+	// The first inserted entry should have been evicted by all the later insertions.
+	ExpectEq(UnknownType, t.cache.Get(beforeExpiration, fmt.Sprint(0)))
+
+	// The second entry should not have been evicted
+	ExpectEq(RegularFileType, t.cache.Get(beforeExpiration, fmt.Sprint(1)))
+}
+
+func (t *TypeCacheTest) TestGetErasedEntry() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+	t.cache.Erase("abcd")
+
+	ExpectEq(UnknownType, t.cache.Get(beforeExpiration, "abcd"))
+}
+
+func (t *TypeCacheTest) TestGetReinsertedEntry() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+	t.cache.Erase("abcd")
+	t.cache.Insert(now2, "abcd", ExplicitDirType)
+
+	ExpectEq(ExplicitDirType, t.cache.Get(beforeExpiration2, "abcd"))
+}
+
+////////////////////////////////////////////////////////////////////////
+// Tests for ZeroMaxCountTypeCacheTest
+////////////////////////////////////////////////////////////////////////
+
+func (t *ZeroMaxEntriesTypeCacheTest) TestGetFromEmptyTypeCache() {
+	ExpectEq(UnknownType, t.cache.Get(now, "abc"))
+}
+
+func (t *ZeroMaxEntriesTypeCacheTest) TestGetInsertedEntry() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+
+	ExpectEq(UnknownType, t.cache.Get(beforeExpiration, "abcd"))
+}
+
+////////////////////////////////////////////////////////////////////////
+// Tests for ZeroTtlTypeCacheTest
+////////////////////////////////////////////////////////////////////////
+
+func (t *ZeroTtlTypeCacheTest) TestGetFromEmptyTypeCache() {
+	ExpectEq(UnknownType, t.cache.Get(now, "abc"))
+}
+
+func (t *ZeroTtlTypeCacheTest) TestGetInsertedEntry() {
+	t.cache.Insert(now, "abcd", RegularFileType)
+
+	ExpectEq(UnknownType, t.cache.Get(beforeExpiration, "abcd"))
+}
diff --git a/internal/config/yaml_parser_test.go b/internal/config/yaml_parser_test.go
index b73b3e5c6..ed1fb8ba7 100644
--- a/internal/config/yaml_parser_test.go
+++ b/internal/config/yaml_parser_test.go
@@ -111,6 +111,7 @@ func (t *YamlParserTest) TestReadConfigFile_ValidConfig() {
 
 	// metadata-cache config
 	ExpectEq(5, mountConfig.MetadataCacheConfig.TtlInSeconds)
+	ExpectEq(1, mountConfig.MetadataCacheConfig.TypeCacheMaxEntries)
 }
 
 func (t *YamlParserTest) TestReadConfigFile_InvalidLogConfig() {
@@ -165,3 +166,18 @@ func (t *YamlParserTest) TestReadConfigFile_MetatadaCacheConfig_TtlTooHigh() {
 	AssertNe(nil, err)
 	AssertThat(err, oglematchers.Error(oglematchers.HasSubstr(MetadataCacheTtlSecsTooHighError)))
 }
+
+func (t *YamlParserTest) TestReadConfigFile_MetatadaCacheConfig_InvalidTypeCacheMaxEntries() {
+	_, err := ParseConfigFile("testdata/metadata_cache_config_invalid_type-cache-max-entries.yaml")
+
+	AssertNe(nil, err)
+	AssertThat(err, oglematchers.Error(oglematchers.HasSubstr(TypeCacheMaxEntriesInvalidValueError)))
+}
+
+func (t *YamlParserTest) TestReadConfigFile_MetatadaCacheConfig_TypeCacheMaxEntriesNotSet() {
+	mountConfig, err := ParseConfigFile("testdata/metadata_cache_config_type-cache-max-entries_unset.yaml")
+
+	AssertEq(nil, err)
+	AssertNe(nil, mountConfig)
+	AssertEq(DefaultTypeCacheMaxEntries, mountConfig.MetadataCacheConfig.TypeCacheMaxEntries)
+}
diff --git a/internal/fs/handle/dir_handle_test.go b/internal/fs/handle/dir_handle_test.go
index 9a84e6db0..8f69bdddc 100644
--- a/internal/fs/handle/dir_handle_test.go
+++ b/internal/fs/handle/dir_handle_test.go
@@ -77,7 +77,8 @@ func (t *DirHandleTest) resetDirHandle() {
 		0,     // typeCacheTTL
 		&t.bucket,
 		&t.clock,
-		&t.clock)
+		&t.clock,
+		0)
 
 	t.dh = NewDirHandle(
 		dirInode,
diff --git a/internal/fs/inode/base_dir_test.go b/internal/fs/inode/base_dir_test.go
index 34b8ae6fc..ba545d1e6 100644
--- a/internal/fs/inode/base_dir_test.go
+++ b/internal/fs/inode/base_dir_test.go
@@ -20,6 +20,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/metadata"
 	"github.com/googlecloudplatform/gcsfuse/internal/storage/fake"
 	"golang.org/x/net/context"
 
@@ -171,7 +172,7 @@ func (t *BaseDirTest) LookUpChild_BucketFound() {
 	ExpectEq("bucketA/", result.FullName.LocalName())
 	ExpectEq("", result.FullName.GcsObjectName())
 	ExpectEq(nil, result.Object)
-	ExpectEq(ImplicitDirType, result.Type())
+	ExpectEq(metadata.ImplicitDirType, result.Type())
 
 	result, err = t.in.LookUpChild(t.ctx, "bucketB")
 
@@ -183,7 +184,7 @@ func (t *BaseDirTest) LookUpChild_BucketFound() {
 	ExpectEq("bucketB/", result.FullName.LocalName())
 	ExpectEq("", result.FullName.GcsObjectName())
 	ExpectEq(nil, result.Object)
-	ExpectEq(ImplicitDirType, result.Type())
+	ExpectEq(metadata.ImplicitDirType, result.Type())
 }
 
 func (t *BaseDirTest) LookUpChild_BucketCached() {
diff --git a/internal/fs/inode/core_test.go b/internal/fs/inode/core_test.go
index eaff8d05e..1e03f3139 100644
--- a/internal/fs/inode/core_test.go
+++ b/internal/fs/inode/core_test.go
@@ -18,6 +18,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/metadata"
 	"github.com/googlecloudplatform/gcsfuse/internal/fs/inode"
 	"github.com/googlecloudplatform/gcsfuse/internal/gcsx"
 	"github.com/googlecloudplatform/gcsfuse/internal/storage/fake"
@@ -68,7 +69,7 @@ func (t *CoreTest) File() {
 		Object:   o,
 	}
 	ExpectTrue(c.Exists())
-	ExpectEq(inode.RegularFileType, c.Type())
+	ExpectEq(metadata.RegularFileType, c.Type())
 }
 
 func (t *CoreTest) LocalFile() {
@@ -80,7 +81,7 @@ func (t *CoreTest) LocalFile() {
 		Local:    true,
 	}
 	ExpectTrue(c.Exists())
-	ExpectEq(inode.RegularFileType, c.Type())
+	ExpectEq(metadata.RegularFileType, c.Type())
 }
 
 func (t *CoreTest) ExplicitDir() {
@@ -94,7 +95,7 @@ func (t *CoreTest) ExplicitDir() {
 		Object:   o,
 	}
 	ExpectTrue(c.Exists())
-	ExpectEq(inode.ExplicitDirType, c.Type())
+	ExpectEq(metadata.ExplicitDirType, c.Type())
 }
 
 func (t *CoreTest) ImplicitDir() {
@@ -105,7 +106,7 @@ func (t *CoreTest) ImplicitDir() {
 		Object:   nil,
 	}
 	ExpectTrue(c.Exists())
-	ExpectEq(inode.ImplicitDirType, c.Type())
+	ExpectEq(metadata.ImplicitDirType, c.Type())
 }
 
 func (t *CoreTest) BucketRootDir() {
@@ -115,13 +116,13 @@ func (t *CoreTest) BucketRootDir() {
 		Object:   nil,
 	}
 	ExpectTrue(c.Exists())
-	ExpectEq(inode.ImplicitDirType, c.Type())
+	ExpectEq(metadata.ImplicitDirType, c.Type())
 }
 
 func (t *CoreTest) Nonexistent() {
 	var c *inode.Core
 	ExpectFalse(c.Exists())
-	ExpectEq(inode.UnknownType, c.Type())
+	ExpectEq(metadata.UnknownType, c.Type())
 }
 
 func (t *CoreTest) SanityCheck() {
diff --git a/internal/fs/inode/dir_test.go b/internal/fs/inode/dir_test.go
index d54f8fe10..637d60644 100644
--- a/internal/fs/inode/dir_test.go
+++ b/internal/fs/inode/dir_test.go
@@ -22,6 +22,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/metadata"
+	"github.com/googlecloudplatform/gcsfuse/internal/config"
 	"github.com/googlecloudplatform/gcsfuse/internal/contentcache"
 	"github.com/googlecloudplatform/gcsfuse/internal/storage/fake"
 	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
@@ -53,6 +55,7 @@ type DirTest struct {
 	clock  timeutil.SimulatedClock
 
 	in DirInode
+	tc metadata.TypeCache
 }
 
 var _ SetUpInterface = &DirTest{}
@@ -86,7 +89,11 @@ func (p DirentSlice) Len() int           { return len(p) }
 func (p DirentSlice) Less(i, j int) bool { return p[i].Name < p[j].Name }
 func (p DirentSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
 
-func (t *DirTest) resetInode(implicitDirs bool, enableNonexistentTypeCache bool) {
+func (t *DirTest) resetInode(implicitDirs, enableNonexistentTypeCache bool) {
+	t.resetInodeWithTypeCacheConfigs(implicitDirs, enableNonexistentTypeCache, config.DefaultTypeCacheMaxEntries, typeCacheTTL)
+}
+
+func (t *DirTest) resetInodeWithTypeCacheConfigs(implicitDirs, enableNonexistentTypeCache bool, typeCacheMaxEntries int, typeCacheTTL time.Duration) {
 	if t.in != nil {
 		t.in.Unlock()
 	}
@@ -104,11 +111,21 @@ func (t *DirTest) resetInode(implicitDirs bool, enableNonexistentTypeCache bool)
 		typeCacheTTL,
 		&t.bucket,
 		&t.clock,
-		&t.clock)
+		&t.clock,
+		typeCacheMaxEntries)
+
+	d := t.in.(*dirInode)
+	AssertNe(nil, d)
+	t.tc = d.cache
+	AssertNe(nil, t.tc)
 
 	t.in.Lock()
 }
 
+func (t *DirTest) getTypeFromCache(name string) metadata.Type {
+	return t.tc.Get(t.in.(*dirInode).cacheClock.Now(), name)
+}
+
 // Read all of the entries and sort them by name.
 func (t *DirTest) readAllEntries() (entries []fuseutil.Dirent, err error) {
 	tok := ""
@@ -194,10 +211,13 @@ func (t *DirTest) Attributes() {
 }
 
 func (t *DirTest) LookUpChild_NonExistent() {
-	result, err := t.in.LookUpChild(t.ctx, "qux")
+	const name = "qux"
+
+	result, err := t.in.LookUpChild(t.ctx, name)
 
 	AssertEq(nil, err)
 	AssertEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
 }
 
 func (t *DirTest) LookUpChild_FileOnly() {
@@ -215,6 +235,7 @@ func (t *DirTest) LookUpChild_FileOnly() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
 
 	ExpectEq(objName, result.FullName.GcsObjectName())
 	ExpectEq(objName, result.Object.Name)
@@ -225,6 +246,7 @@ func (t *DirTest) LookUpChild_FileOnly() {
 	result, err = t.in.LookUpChild(t.ctx, name+ConflictingFileNameSuffix)
 	AssertEq(nil, err)
 	ExpectEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
 }
 
 func (t *DirTest) LookUpChild_DirOnly() {
@@ -242,6 +264,7 @@ func (t *DirTest) LookUpChild_DirOnly() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(objName, result.FullName.GcsObjectName())
 	ExpectEq(objName, result.Object.Name)
@@ -252,6 +275,7 @@ func (t *DirTest) LookUpChild_DirOnly() {
 	result, err = t.in.LookUpChild(t.ctx, name+ConflictingFileNameSuffix)
 	AssertEq(nil, err)
 	ExpectEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
 }
 
 func (t *DirTest) LookUpChild_ImplicitDirOnly_Disabled() {
@@ -267,11 +291,13 @@ func (t *DirTest) LookUpChild_ImplicitDirOnly_Disabled() {
 	result, err := t.in.LookUpChild(t.ctx, name)
 	AssertEq(nil, err)
 	ExpectEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
 
 	// Ditto with a conflict marker.
 	result, err = t.in.LookUpChild(t.ctx, name+ConflictingFileNameSuffix)
 	AssertEq(nil, err)
 	ExpectEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
 }
 
 func (t *DirTest) LookUpChild_ImplicitDirOnly_Enabled() {
@@ -293,14 +319,16 @@ func (t *DirTest) LookUpChild_ImplicitDirOnly_Enabled() {
 
 	AssertEq(nil, err)
 	ExpectEq(nil, result.Object)
+	ExpectEq(metadata.ImplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(objName, result.FullName.GcsObjectName())
-	ExpectEq(ImplicitDirType, result.Type())
+	ExpectEq(metadata.ImplicitDirType, result.Type())
 
 	// A conflict marker should not work.
 	result, err = t.in.LookUpChild(t.ctx, name+ConflictingFileNameSuffix)
 	AssertEq(nil, err)
 	ExpectEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
 }
 
 func (t *DirTest) LookUpChild_FileAndDir() {
@@ -322,6 +350,7 @@ func (t *DirTest) LookUpChild_FileAndDir() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(dirObjName, result.FullName.GcsObjectName())
 	ExpectEq(dirObjName, result.Object.Name)
@@ -333,6 +362,7 @@ func (t *DirTest) LookUpChild_FileAndDir() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	AssertEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
 
 	ExpectEq(fileObjName, result.FullName.GcsObjectName())
 	ExpectEq(fileObjName, result.Object.Name)
@@ -363,6 +393,14 @@ func (t *DirTest) LookUpChild_SymlinkAndDir() {
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
 
+	// The following check should have been for metadata.SymlinkType,
+	// because of the t.setSymlinkTarget call above, but it is not.
+	// This is so because the above symlink is
+	// created as a regular directory object on GCS,
+	// and is read back the same to gcsfuse and is this stored in type-cache
+	// also as a directory.
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
+
 	ExpectEq(dirObjName, result.FullName.GcsObjectName())
 	ExpectEq(dirObjName, result.Object.Name)
 	ExpectEq(dirObj.Generation, result.Object.Generation)
@@ -373,6 +411,7 @@ func (t *DirTest) LookUpChild_SymlinkAndDir() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
 
 	ExpectEq(linkObjName, result.FullName.GcsObjectName())
 	ExpectEq(linkObjName, result.Object.Name)
@@ -404,6 +443,8 @@ func (t *DirTest) LookUpChild_FileAndDirAndImplicitDir_Disabled() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(path.Join(dirInodeName, name)))
 
 	ExpectEq(dirObjName, result.FullName.GcsObjectName())
 	ExpectEq(dirObjName, result.Object.Name)
@@ -449,6 +490,8 @@ func (t *DirTest) LookUpChild_FileAndDirAndImplicitDir_Enabled() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(path.Join(dirInodeName, name)))
 
 	ExpectEq(dirObjName, result.FullName.GcsObjectName())
 	ExpectEq(dirObjName, result.Object.Name)
@@ -460,6 +503,7 @@ func (t *DirTest) LookUpChild_FileAndDirAndImplicitDir_Enabled() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
 
 	ExpectEq(fileObjName, result.FullName.GcsObjectName())
 	ExpectEq(fileObjName, result.Object.Name)
@@ -483,6 +527,7 @@ func (t *DirTest) LookUpChild_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
 
 	ExpectEq(fileObjName, result.Object.Name)
 
@@ -496,6 +541,7 @@ func (t *DirTest) LookUpChild_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
 
 	ExpectEq(fileObjName, result.Object.Name)
 
@@ -506,6 +552,7 @@ func (t *DirTest) LookUpChild_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(dirObjName, result.Object.Name)
 }
@@ -532,6 +579,7 @@ func (t *DirTest) LookUpChild_NonExistentTypeCache_ImplicitDirsDisabled() {
 
 	AssertEq(nil, err)
 	AssertEq(nil, result)
+	ExpectEq(metadata.NonexistentType, t.getTypeFromCache(name))
 
 	// But after the TTL expires, the behavior should flip.
 	t.clock.AdvanceTime(typeCacheTTL + time.Millisecond)
@@ -541,6 +589,7 @@ func (t *DirTest) LookUpChild_NonExistentTypeCache_ImplicitDirsDisabled() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(objName, result.FullName.GcsObjectName())
 	ExpectEq(objName, result.Object.Name)
@@ -560,6 +609,7 @@ func (t *DirTest) LookUpChild_NonExistentTypeCache_ImplicitDirsEnabled() {
 
 	AssertEq(nil, err)
 	AssertEq(nil, result)
+	ExpectEq(metadata.NonexistentType, t.getTypeFromCache(name))
 
 	// Create an object that implicitly defines the directory.
 	otherObjName := path.Join(objName, "asdf")
@@ -571,6 +621,7 @@ func (t *DirTest) LookUpChild_NonExistentTypeCache_ImplicitDirsEnabled() {
 
 	AssertEq(nil, err)
 	AssertEq(nil, result)
+	ExpectEq(metadata.NonexistentType, t.getTypeFromCache(name))
 
 	// But after the TTL expires, the behavior should flip.
 	t.clock.AdvanceTime(typeCacheTTL + time.Millisecond)
@@ -582,12 +633,79 @@ func (t *DirTest) LookUpChild_NonExistentTypeCache_ImplicitDirsEnabled() {
 	ExpectEq(nil, result.Object)
 
 	ExpectEq(objName, result.FullName.GcsObjectName())
-	ExpectEq(ImplicitDirType, result.Type())
+	ExpectEq(metadata.ImplicitDirType, result.Type())
 
 	// A conflict marker should not work.
 	result, err = t.in.LookUpChild(t.ctx, name+ConflictingFileNameSuffix)
 	AssertEq(nil, err)
 	ExpectEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name+ConflictingFileNameSuffix))
+}
+
+func (t *DirTest) LookUpChild_TypeCacheEnabled() {
+	inputs := []struct {
+		typeCacheMaxEntries int
+		typeCacheTTL        time.Duration
+	}{{
+		typeCacheMaxEntries: config.DefaultTypeCacheMaxEntries,
+		typeCacheTTL:        time.Second,
+	}, {
+		typeCacheMaxEntries: -1,
+		typeCacheTTL:        time.Second,
+	}}
+
+	for _, input := range inputs {
+		t.resetInodeWithTypeCacheConfigs(true, true, input.typeCacheMaxEntries, input.typeCacheTTL)
+
+		const name = "qux"
+		objName := path.Join(dirInodeName, name)
+
+		// Create a backing object.
+		o, err := storageutil.CreateObject(t.ctx, t.bucket, objName, []byte("taco"))
+
+		AssertEq(nil, err)
+		AssertNe(nil, o)
+
+		// Look up nonexistent object, return nil
+		result, err := t.in.LookUpChild(t.ctx, name)
+
+		AssertEq(nil, err)
+		AssertNe(nil, result)
+		ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
+	}
+}
+
+func (t *DirTest) LookUpChild_TypeCacheDisabled() {
+	inputs := []struct {
+		typeCacheMaxEntries int
+		typeCacheTTL        time.Duration
+	}{{
+		typeCacheMaxEntries: 0,
+		typeCacheTTL:        time.Second,
+	}, {
+		typeCacheMaxEntries: config.DefaultTypeCacheMaxEntries,
+		typeCacheTTL:        0,
+	}}
+
+	for _, input := range inputs {
+		t.resetInodeWithTypeCacheConfigs(true, true, input.typeCacheMaxEntries, input.typeCacheTTL)
+
+		const name = "qux"
+		objName := path.Join(dirInodeName, name)
+
+		// Create a backing object.
+		o, err := storageutil.CreateObject(t.ctx, t.bucket, objName, []byte("taco"))
+
+		AssertEq(nil, err)
+		AssertNe(nil, o)
+
+		// Look up nonexistent object, return nil
+		result, err := t.in.LookUpChild(t.ctx, name)
+
+		AssertEq(nil, err)
+		AssertNe(nil, result)
+		ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
+	}
 }
 
 func (t *DirTest) ReadDescendants_Empty() {
@@ -660,18 +778,22 @@ func (t *DirTest) ReadEntries_NonEmpty_ImplicitDirsDisabled() {
 	entry = entries[0]
 	ExpectEq("backed_dir_empty", entry.Name)
 	ExpectEq(fuseutil.DT_Directory, entry.Type)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache("backed_dir_empty"))
 
 	entry = entries[1]
 	ExpectEq("backed_dir_nonempty", entry.Name)
 	ExpectEq(fuseutil.DT_Directory, entry.Type)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache("backed_dir_nonempty"))
 
 	entry = entries[2]
 	ExpectEq("file", entry.Name)
 	ExpectEq(fuseutil.DT_File, entry.Type)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache("file"))
 
 	entry = entries[3]
 	ExpectEq("symlink", entry.Name)
 	ExpectEq(fuseutil.DT_Link, entry.Type)
+	ExpectEq(metadata.SymlinkType, t.getTypeFromCache("symlink"))
 }
 
 func (t *DirTest) ReadEntries_NonEmpty_ImplicitDirsEnabled() {
@@ -707,22 +829,27 @@ func (t *DirTest) ReadEntries_NonEmpty_ImplicitDirsEnabled() {
 	entry = entries[0]
 	ExpectEq("backed_dir_empty", entry.Name)
 	ExpectEq(fuseutil.DT_Directory, entry.Type)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache("backed_dir_empty"))
 
 	entry = entries[1]
 	ExpectEq("backed_dir_nonempty", entry.Name)
 	ExpectEq(fuseutil.DT_Directory, entry.Type)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache("backed_dir_nonempty"))
 
 	entry = entries[2]
 	ExpectEq("file", entry.Name)
 	ExpectEq(fuseutil.DT_File, entry.Type)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache("file"))
 
 	entry = entries[3]
 	ExpectEq("implicit_dir", entry.Name)
 	ExpectEq(fuseutil.DT_Directory, entry.Type)
+	ExpectEq(metadata.ImplicitDirType, t.getTypeFromCache("implicit_dir"))
 
 	entry = entries[4]
 	ExpectEq("symlink", entry.Name)
 	ExpectEq(fuseutil.DT_Link, entry.Type)
+	ExpectEq(metadata.SymlinkType, t.getTypeFromCache("symlink"))
 }
 
 func (t *DirTest) ReadEntries_TypeCaching() {
@@ -739,6 +866,7 @@ func (t *DirTest) ReadEntries_TypeCaching() {
 	// Read the directory, priming the type cache.
 	_, err = t.readAllEntries()
 	AssertEq(nil, err)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
 
 	// Create a backing object for a directory.
 	_, err = storageutil.CreateObject(t.ctx, t.bucket, dirObjName, []byte("taco"))
@@ -750,6 +878,7 @@ func (t *DirTest) ReadEntries_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
 
 	ExpectEq(fileObjName, result.Object.Name)
 
@@ -760,6 +889,7 @@ func (t *DirTest) ReadEntries_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(dirObjName, result.Object.Name)
 }
@@ -778,6 +908,7 @@ func (t *DirTest) CreateChildFile_DoesntExist() {
 	ExpectEq(result.FullName.GcsObjectName(), result.Object.Name)
 	ExpectEq(objName, result.Object.Name)
 	ExpectFalse(IsSymlink(result.Object))
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
 
 	ExpectEq(1, len(result.Object.Metadata))
 	ExpectEq(
@@ -799,6 +930,8 @@ func (t *DirTest) CreateChildFile_Exists() {
 	_, err = t.in.CreateChildFile(t.ctx, name)
 	ExpectThat(err, Error(HasSubstr("Precondition")))
 	ExpectThat(err, Error(HasSubstr("exists")))
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
 }
 
 func (t *DirTest) CreateChildFile_TypeCaching() {
@@ -822,6 +955,7 @@ func (t *DirTest) CreateChildFile_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache(name))
 
 	ExpectEq(fileObjName, result.Object.Name)
 
@@ -832,6 +966,7 @@ func (t *DirTest) CreateChildFile_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(dirObjName, result.Object.Name)
 }
@@ -856,6 +991,7 @@ func (t *DirTest) CloneToChildFile_SourceDoesntExist() {
 	_, err = t.in.CloneToChildFile(t.ctx, path.Base(dstName), src)
 	var notFoundErr *gcs.NotFoundError
 	ExpectTrue(errors.As(err, &notFoundErr))
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(dstName))
 }
 
 func (t *DirTest) CloneToChildFile_DestinationDoesntExist() {
@@ -871,6 +1007,7 @@ func (t *DirTest) CloneToChildFile_DestinationDoesntExist() {
 	AssertEq(nil, err)
 	AssertNe(nil, result)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache("qux"))
 
 	ExpectEq(t.bucket.Name(), result.Bucket.Name())
 	ExpectEq(result.FullName.GcsObjectName(), result.Object.Name)
@@ -881,6 +1018,7 @@ func (t *DirTest) CloneToChildFile_DestinationDoesntExist() {
 	contents, err := storageutil.ReadObject(t.ctx, t.bucket, dstName)
 	AssertEq(nil, err)
 	ExpectEq("taco", string(contents))
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache("qux"))
 }
 
 func (t *DirTest) CloneToChildFile_DestinationExists() {
@@ -900,6 +1038,7 @@ func (t *DirTest) CloneToChildFile_DestinationExists() {
 	AssertEq(nil, err)
 	AssertNe(nil, result)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache("qux"))
 
 	ExpectEq(t.bucket.Name(), result.Bucket.Name())
 	ExpectEq(result.FullName.GcsObjectName(), result.Object.Name)
@@ -911,6 +1050,7 @@ func (t *DirTest) CloneToChildFile_DestinationExists() {
 	contents, err := storageutil.ReadObject(t.ctx, t.bucket, dstName)
 	AssertEq(nil, err)
 	ExpectEq("taco", string(contents))
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache("qux"))
 }
 
 func (t *DirTest) CloneToChildFile_TypeCaching() {
@@ -938,6 +1078,7 @@ func (t *DirTest) CloneToChildFile_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.RegularFileType, t.getTypeFromCache("qux"))
 
 	ExpectEq(dstName, result.Object.Name)
 
@@ -948,6 +1089,7 @@ func (t *DirTest) CloneToChildFile_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache("qux"))
 
 	ExpectEq(dirObjName, result.Object.Name)
 }
@@ -962,6 +1104,7 @@ func (t *DirTest) CreateChildSymlink_DoesntExist() {
 	AssertEq(nil, err)
 	AssertNe(nil, result)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.SymlinkType, t.getTypeFromCache(name))
 
 	ExpectEq(t.bucket.Name(), result.Bucket.Name())
 	ExpectEq(result.FullName.GcsObjectName(), result.Object.Name)
@@ -984,6 +1127,7 @@ func (t *DirTest) CreateChildSymlink_Exists() {
 	_, err = t.in.CreateChildSymlink(t.ctx, name, target)
 	ExpectThat(err, Error(HasSubstr("Precondition")))
 	ExpectThat(err, Error(HasSubstr("exists")))
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
 }
 
 func (t *DirTest) CreateChildSymlink_TypeCaching() {
@@ -1008,6 +1152,7 @@ func (t *DirTest) CreateChildSymlink_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.SymlinkType, t.getTypeFromCache(name))
 
 	ExpectEq(linkObjName, result.Object.Name)
 
@@ -1018,6 +1163,7 @@ func (t *DirTest) CreateChildSymlink_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(dirObjName, result.Object.Name)
 }
@@ -1031,6 +1177,7 @@ func (t *DirTest) CreateChildDir_DoesntExist() {
 	AssertEq(nil, err)
 	AssertNe(nil, result)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(t.bucket.Name(), result.Bucket.Name())
 	ExpectEq(result.FullName.GcsObjectName(), result.Object.Name)
@@ -1052,6 +1199,7 @@ func (t *DirTest) CreateChildDir_Exists() {
 	_, err = t.in.CreateChildDir(t.ctx, name)
 	ExpectThat(err, Error(HasSubstr("Precondition")))
 	ExpectThat(err, Error(HasSubstr("exists")))
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
 }
 
 func (t *DirTest) DeleteChildFile_DoesntExist() {
@@ -1059,6 +1207,7 @@ func (t *DirTest) DeleteChildFile_DoesntExist() {
 
 	err := t.in.DeleteChildFile(t.ctx, name, 0, nil)
 	ExpectEq(nil, err)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
 }
 
 func (t *DirTest) DeleteChildFile_WrongGeneration() {
@@ -1176,6 +1325,7 @@ func (t *DirTest) DeleteChildFile_TypeCaching() {
 
 	AssertEq(nil, err)
 	AssertNe(nil, result.Object)
+	ExpectEq(metadata.ExplicitDirType, t.getTypeFromCache(name))
 
 	ExpectEq(dirObjName, result.Object.Name)
 }
@@ -1228,6 +1378,7 @@ func (t *DirTest) CreateLocalChildFile_ShouldnotCreateObjectInGCS() {
 	result, err := t.in.LookUpChild(t.ctx, name)
 	AssertEq(nil, err)
 	AssertEq(nil, result)
+	ExpectEq(metadata.UnknownType, t.getTypeFromCache(name))
 }
 
 func (t *DirTest) LocalFileEntriesEmpty() {
diff --git a/main_test.go b/main_test.go
index 15b244234..9c3867a15 100644
--- a/main_test.go
+++ b/main_test.go
@@ -94,7 +94,7 @@ func (t *MainTest) TestStringifyShouldReturnAllFlagsPassedInMountConfigAsMarshal
 
 	actual := util.Stringify(mountConfig)
 
-	expected := "{\"CreateEmptyFile\":false,\"Severity\":\"TRACE\",\"Format\":\"\",\"FilePath\":\"\\\"path\\\"to\\\"file\\\"\",\"LogRotateConfig\":{\"MaxFileSizeMB\":2,\"BackupFileCount\":2,\"Compress\":true},\"MaxSizeInMB\":0,\"CacheFileForRangeRead\":false,\"CacheLocation\":\"\",\"TtlInSeconds\":0}"
+	expected := "{\"CreateEmptyFile\":false,\"Severity\":\"TRACE\",\"Format\":\"\",\"FilePath\":\"\\\"path\\\"to\\\"file\\\"\",\"LogRotateConfig\":{\"MaxFileSizeMB\":2,\"BackupFileCount\":2,\"Compress\":true},\"MaxSizeInMB\":0,\"CacheFileForRangeRead\":false,\"CacheLocation\":\"\",\"TtlInSeconds\":0,\"TypeCacheMaxEntries\":0}"
 	AssertEq(expected, actual)
 }
