diff --git a/flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSinkITCase.java b/flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSinkITCase.java
index 493a5df51ad..be8453b53b5 100644
--- a/flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSinkITCase.java
+++ b/flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSinkITCase.java
@@ -151,12 +151,12 @@ public class HiveTableSinkITCase {
         }
     }
 
-    @Test(timeout = 120000)
+    @Test
     public void testDefaultSerPartStreamingWrite() throws Exception {
         testStreamingWrite(true, false, "textfile", this::checkSuccessFiles);
     }
 
-    @Test(timeout = 120000)
+    @Test
     public void testPartStreamingWrite() throws Exception {
         testStreamingWrite(true, false, "parquet", this::checkSuccessFiles);
         // disable vector orc writer test for hive 2.x due to dependency conflict
@@ -165,7 +165,7 @@ public class HiveTableSinkITCase {
         }
     }
 
-    @Test(timeout = 120000)
+    @Test
     public void testNonPartStreamingWrite() throws Exception {
         testStreamingWrite(false, false, "parquet", (p) -> {});
         // disable vector orc writer test for hive 2.x due to dependency conflict
@@ -174,7 +174,7 @@ public class HiveTableSinkITCase {
         }
     }
 
-    @Test(timeout = 120000)
+    @Test
     public void testPartStreamingMrWrite() throws Exception {
         testStreamingWrite(true, true, "parquet", this::checkSuccessFiles);
         // doesn't support writer 2.0 orc table
@@ -183,7 +183,7 @@ public class HiveTableSinkITCase {
         }
     }
 
-    @Test(timeout = 120000)
+    @Test
     public void testNonPartStreamingMrWrite() throws Exception {
         testStreamingWrite(false, true, "parquet", (p) -> {});
         // doesn't support writer 2.0 orc table
@@ -192,7 +192,7 @@ public class HiveTableSinkITCase {
         }
     }
 
-    @Test(timeout = 120000)
+    @Test
     public void testStreamingAppend() throws Exception {
         testStreamingWrite(
                 false,
@@ -232,7 +232,7 @@ public class HiveTableSinkITCase {
                 });
     }
 
-    @Test(timeout = 120000)
+    @Test
     public void testStreamingSinkWithTimestampLtzWatermark() throws Exception {
         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
         env.setParallelism(1);
diff --git a/flink-core/src/test/java/org/apache/flink/api/common/eventtime/WatermarkOutputMultiplexerTest.java b/flink-core/src/test/java/org/apache/flink/api/common/eventtime/WatermarkOutputMultiplexerTest.java
index 362caf7be73..61ce8d5e307 100644
--- a/flink-core/src/test/java/org/apache/flink/api/common/eventtime/WatermarkOutputMultiplexerTest.java
+++ b/flink-core/src/test/java/org/apache/flink/api/common/eventtime/WatermarkOutputMultiplexerTest.java
@@ -23,6 +23,7 @@ import org.junit.Test;
 import java.util.UUID;
 
 import static org.apache.flink.api.common.eventtime.WatermarkMatchers.watermark;
+import static org.hamcrest.CoreMatchers.equalTo;
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.CoreMatchers.nullValue;
 import static org.junit.Assert.assertEquals;
@@ -139,6 +140,9 @@ public class WatermarkOutputMultiplexerTest {
      * This test makes sure that we don't output any update if there are zero outputs. Due to how
      * aggregation of deferred updates in the KafkaConsumer works we had a bug there that caused a
      * Long.MAX_VALUE watermark to be emitted in case of zero partitions.
+     *
+     * <p>Additionally it verifies that the combined output is not IDLE during the initial phase
+     * when there are no splits assigned and the combined watermark is at its initial value.
      */
     @Test
     public void noCombinedDeferredUpdateWhenWeHaveZeroOutputs() {
@@ -149,6 +153,7 @@ public class WatermarkOutputMultiplexerTest {
         multiplexer.onPeriodicEmit();
 
         assertThat(underlyingWatermarkOutput.lastWatermark(), is(nullValue()));
+        assertThat(underlyingWatermarkOutput.isIdle(), is(false));
     }
 
     @Test
@@ -376,6 +381,24 @@ public class WatermarkOutputMultiplexerTest {
         assertFalse(unregistered);
     }
 
+    @Test
+    public void testEmittingIdleAfterAllSplitsRemoved() {
+        final TestingWatermarkOutput underlyingWatermarkOutput = createTestingWatermarkOutput();
+        final WatermarkOutputMultiplexer multiplexer =
+                new WatermarkOutputMultiplexer(underlyingWatermarkOutput);
+
+        Watermark emittedWatermark = new Watermark(1);
+        final String id = UUID.randomUUID().toString();
+        multiplexer.registerNewOutput(id);
+        WatermarkOutput immediateOutput = multiplexer.getImmediateOutput(id);
+        immediateOutput.emitWatermark(emittedWatermark);
+        multiplexer.unregisterOutput(id);
+
+        multiplexer.onPeriodicEmit();
+        assertThat(underlyingWatermarkOutput.lastWatermark(), equalTo(emittedWatermark));
+        assertThat(underlyingWatermarkOutput.isIdle(), equalTo(true));
+    }
+
     /**
      * Convenience method so we don't have to go through the output ID dance when we only want an
      * immediate output for a given output ID.
diff --git a/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskTest.java b/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskTest.java
index 4075ba68796..6f5e8ebcece 100644
--- a/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskTest.java
+++ b/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskTest.java
@@ -626,12 +626,8 @@ public class MultipleInputStreamTaskTest {
             // make source active once again, emit a watermark and go idle again.
             addSourceRecords(testHarness, 1, initialTime + 10);
 
-            // FLIP-27 sources do not emit active status on new records, we wrap a record with
-            // ACTIVE/IDLE sequence
-            expectedOutput.add(StreamStatus.ACTIVE);
             expectedOutput.add(
                     new StreamRecord<>("" + (initialTime + 10), TimestampAssigner.NO_TIMESTAMP));
-            expectedOutput.add(StreamStatus.IDLE);
             expectedOutput.add(StreamStatus.ACTIVE); // activate source on new watermark
             expectedOutput.add(new Watermark(initialTime + 10)); // forward W from source
             expectedOutput.add(StreamStatus.IDLE); // go idle after reading all records
diff --git a/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java b/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java
index 858f0229535..c845e1db732 100644
--- a/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java
+++ b/flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java
@@ -264,6 +264,138 @@ public class OneInputStreamTaskTest extends TestLogger {
         assertEquals(2, resultElements.size());
     }
 
+    /**
+     * This test verifies that watermarks are not forwarded when the task is idle. It also verifies
+     * that when task is idle, watermarks generated in the middle of chains are also blocked and
+     * never forwarded.
+     *
+     * <p>The tested chain will be: (HEAD: normal operator) --> (watermark generating operator) -->
+     * (normal operator). The operators will throw an exception and fail the test if either of them
+     * were forwarded watermarks when the task is idle.
+     */
+    @Test
+    public void testWatermarksNotForwardedWithinChainWhenIdle() throws Exception {
+
+        final OneInputStreamTaskTestHarness<String, String> testHarness =
+                new OneInputStreamTaskTestHarness<>(
+                        OneInputStreamTask::new,
+                        1,
+                        1,
+                        BasicTypeInfo.STRING_TYPE_INFO,
+                        BasicTypeInfo.STRING_TYPE_INFO);
+
+        TriggerableFailOnWatermarkTestOperator headOperator =
+                new TriggerableFailOnWatermarkTestOperator();
+        WatermarkGeneratingTestOperator watermarkOperator = new WatermarkGeneratingTestOperator();
+        TriggerableFailOnWatermarkTestOperator tailOperator =
+                new TriggerableFailOnWatermarkTestOperator();
+
+        testHarness
+                .setupOperatorChain(new OperatorID(42L, 42L), headOperator)
+                .chain(new OperatorID(4711L, 42L), watermarkOperator, StringSerializer.INSTANCE)
+                .chain(new OperatorID(123L, 123L), tailOperator, StringSerializer.INSTANCE)
+                .finish();
+
+        // --------------------- begin test ---------------------
+
+        ConcurrentLinkedQueue<Object> expectedOutput = new ConcurrentLinkedQueue<>();
+
+        testHarness.invoke();
+        testHarness.waitForTaskRunning();
+
+        // the task starts as active, so all generated watermarks should be forwarded
+        testHarness.processElement(
+                new StreamRecord<>(
+                        TriggerableFailOnWatermarkTestOperator.EXPECT_FORWARDED_WATERMARKS_MARKER));
+
+        testHarness.processElement(new StreamRecord<>("10"), 0, 0);
+
+        // this watermark will be forwarded since the task is currently active,
+        // but should not be in the final output because it should be blocked by the watermark
+        // generator in the chain
+        testHarness.processElement(new Watermark(15));
+
+        testHarness.processElement(new StreamRecord<>("20"), 0, 0);
+        testHarness.processElement(new StreamRecord<>("30"), 0, 0);
+
+        testHarness.waitForInputProcessing();
+
+        expectedOutput.add(
+                new StreamRecord<>(
+                        TriggerableFailOnWatermarkTestOperator.EXPECT_FORWARDED_WATERMARKS_MARKER));
+        expectedOutput.add(new StreamRecord<>("10"));
+        expectedOutput.add(new Watermark(10));
+        expectedOutput.add(new StreamRecord<>("20"));
+        expectedOutput.add(new Watermark(20));
+        expectedOutput.add(new StreamRecord<>("30"));
+        expectedOutput.add(new Watermark(30));
+        TestHarnessUtil.assertOutputEquals(
+                "Output was not correct.", expectedOutput, testHarness.getOutput());
+
+        // now, toggle the task to be idle, and let the watermark generator produce some watermarks
+        testHarness.processElement(StreamStatus.IDLE);
+
+        // after this, the operators will throw an exception if they are forwarded watermarks
+        // anywhere in the chain
+        testHarness.processElement(
+                new StreamRecord<>(
+                        TriggerableFailOnWatermarkTestOperator.NO_FORWARDED_WATERMARKS_MARKER));
+
+        // NOTE: normally, tasks will not have records to process while idle;
+        // we're doing this here only to mimic watermark generating in operators
+        testHarness.processElement(new StreamRecord<>("40"), 0, 0);
+        testHarness.processElement(new StreamRecord<>("50"), 0, 0);
+        testHarness.processElement(new StreamRecord<>("60"), 0, 0);
+        testHarness.processElement(
+                new Watermark(
+                        65)); // the test will fail if any of the operators were forwarded this
+        testHarness.waitForInputProcessing();
+
+        // the 40 - 60 watermarks should not be forwarded, only the stream status toggle element and
+        // records
+        expectedOutput.add(StreamStatus.IDLE);
+        expectedOutput.add(
+                new StreamRecord<>(
+                        TriggerableFailOnWatermarkTestOperator.NO_FORWARDED_WATERMARKS_MARKER));
+        expectedOutput.add(new StreamRecord<>("40"));
+        expectedOutput.add(new StreamRecord<>("50"));
+        expectedOutput.add(new StreamRecord<>("60"));
+        TestHarnessUtil.assertOutputEquals(
+                "Output was not correct.", expectedOutput, testHarness.getOutput());
+
+        // re-toggle the task to be active and see if new watermarks are correctly forwarded again
+        testHarness.processElement(StreamStatus.ACTIVE);
+        testHarness.processElement(
+                new StreamRecord<>(
+                        TriggerableFailOnWatermarkTestOperator.EXPECT_FORWARDED_WATERMARKS_MARKER));
+
+        testHarness.processElement(new StreamRecord<>("70"), 0, 0);
+        testHarness.processElement(new StreamRecord<>("80"), 0, 0);
+        testHarness.processElement(new StreamRecord<>("90"), 0, 0);
+        testHarness.waitForInputProcessing();
+
+        expectedOutput.add(StreamStatus.ACTIVE);
+        expectedOutput.add(
+                new StreamRecord<>(
+                        TriggerableFailOnWatermarkTestOperator.EXPECT_FORWARDED_WATERMARKS_MARKER));
+        expectedOutput.add(new StreamRecord<>("70"));
+        expectedOutput.add(new Watermark(70));
+        expectedOutput.add(new StreamRecord<>("80"));
+        expectedOutput.add(new Watermark(80));
+        expectedOutput.add(new StreamRecord<>("90"));
+        expectedOutput.add(new Watermark(90));
+        TestHarnessUtil.assertOutputEquals(
+                "Output was not correct.", expectedOutput, testHarness.getOutput());
+
+        testHarness.endInput();
+
+        testHarness.waitForTaskCompletion();
+
+        List<String> resultElements =
+                TestHarnessUtil.getRawElementsFromOutput(testHarness.getOutput());
+        assertEquals(12, resultElements.size());
+    }
+
     /** This test verifies that checkpoint barriers are correctly forwarded. */
     @Test
     public void testCheckpointBarriers() throws Exception {
@@ -999,4 +1131,81 @@ public class OneInputStreamTaskTest extends TestLogger {
             return value;
         }
     }
+
+    /** A {@link TriggerableFailOnWatermarkTestOperator} that generates watermarks. */
+    private static class WatermarkGeneratingTestOperator
+            extends TriggerableFailOnWatermarkTestOperator {
+
+        private static final long serialVersionUID = -5064871833244157221L;
+
+        private long lastWatermark;
+
+        @Override
+        protected void handleElement(StreamRecord<String> element) {
+            long timestamp = Long.valueOf(element.getValue());
+            if (timestamp > lastWatermark) {
+                output.emitWatermark(new Watermark(timestamp));
+                lastWatermark = timestamp;
+            }
+        }
+
+        @Override
+        protected void handleWatermark(Watermark mark) {
+            if (mark.equals(Watermark.MAX_WATERMARK)) {
+                output.emitWatermark(mark);
+                lastWatermark = Long.MAX_VALUE;
+            }
+        }
+    }
+
+    /**
+     * An operator that can be triggered whether or not to expect watermarks forwarded to it,
+     * toggled by letting it process special trigger marker records.
+     *
+     * <p>If it receives a watermark when it's not expecting one, it'll throw an exception and fail.
+     */
+    private static class TriggerableFailOnWatermarkTestOperator
+            extends AbstractStreamOperator<String>
+            implements OneInputStreamOperator<String, String> {
+
+        private static final long serialVersionUID = 2048954179291813243L;
+
+        public static final String EXPECT_FORWARDED_WATERMARKS_MARKER = "EXPECT_WATERMARKS";
+        public static final String NO_FORWARDED_WATERMARKS_MARKER = "NO_WATERMARKS";
+
+        protected boolean expectForwardedWatermarks;
+
+        @Override
+        public void processElement(StreamRecord<String> element) throws Exception {
+            output.collect(element);
+
+            if (element.getValue().equals(EXPECT_FORWARDED_WATERMARKS_MARKER)) {
+                this.expectForwardedWatermarks = true;
+            } else if (element.getValue().equals(NO_FORWARDED_WATERMARKS_MARKER)) {
+                this.expectForwardedWatermarks = false;
+            } else {
+                handleElement(element);
+            }
+        }
+
+        @Override
+        public void processWatermark(Watermark mark) throws Exception {
+            if (!expectForwardedWatermarks) {
+                throw new Exception(
+                        "Received a "
+                                + mark
+                                + ", but this operator should not be forwarded watermarks.");
+            } else {
+                handleWatermark(mark);
+            }
+        }
+
+        protected void handleElement(StreamRecord<String> element) {
+            // do nothing
+        }
+
+        protected void handleWatermark(Watermark mark) {
+            output.emitWatermark(mark);
+        }
+    }
 }
diff --git a/flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperatorTest.java b/flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperatorTest.java
index 97dc443ad62..fcc81377fd3 100644
--- a/flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperatorTest.java
+++ b/flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperatorTest.java
@@ -61,14 +61,10 @@ public class WatermarkAssignerOperatorTest extends WatermarkAssignerOperatorTest
         ConcurrentLinkedQueue<Object> output = testHarness.getOutput();
         List<Object> expectedOutput = new ArrayList<>();
 
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(1L)));
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(2L)));
         testHarness.processWatermark(new Watermark(2)); // this watermark should be ignored
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(3L)));
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(4L)));
 
         // trigger watermark emit
@@ -82,13 +78,9 @@ public class WatermarkAssignerOperatorTest extends WatermarkAssignerOperatorTest
 
         expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(4L)));
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(5L)));
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(6L)));
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(7L)));
-        expectedOutput.add(StreamStatus.ACTIVE);
         testHarness.processElement(new StreamRecord<>(GenericRowData.of(8L)));
 
         testHarness.setProcessingTime(1060);
