diff --git a/lib/components/Alert/Alert.test.js b/lib/components/Alert/Alert.test.tsx
similarity index 77%
rename from lib/components/Alert/Alert.test.js
rename to lib/components/Alert/Alert.test.tsx
index 4299a810a0d..0622cdaed2c 100644
--- a/lib/components/Alert/Alert.test.js
+++ b/lib/components/Alert/Alert.test.tsx
@@ -1,15 +1,16 @@
 import React from 'react';
 import { Alert } from '../';
+import { AlertProps } from './Alert';
 import { render } from 'react-testing-library';
 import forEachTheme from '../../test/utils/forEachTheme';
 
 describe('Alert', () => {
   forEachTheme(({ ThemeProvider }) => {
     describe('Tones', () => {
-      const tones = ['critical', 'info'];
+      const tones: Array<AlertProps['tone']> = ['critical', 'info', undefined];
 
       tones.forEach(tone => {
-        test(tone, () => {
+        test(tone || 'default', () => {
           const { container } = render(
             <ThemeProvider>
               <Alert tone={tone}>{tone}</Alert>
diff --git a/lib/components/Box/Box.test.js b/lib/components/Box/Box.test.tsx
similarity index 95%
rename from lib/components/Box/Box.test.js
rename to lib/components/Box/Box.test.tsx
index 3b576f7272f..d2574df0acc 100644
--- a/lib/components/Box/Box.test.js
+++ b/lib/components/Box/Box.test.tsx
@@ -40,9 +40,11 @@ describe('Box', () => {
     });
 
     test('Unknown values', () => {
+      const UnsafeBox: any = Box;
+
       const { container } = render(
         <ThemeProvider>
-          <Box
+          <UnsafeBox
             paddingTop="UNKNOWN"
             paddingBottom="UNKNOWN"
             paddingLeft="UNKNOWN"
@@ -56,7 +58,7 @@ describe('Box', () => {
             backgroundColor="UNKNOWN"
           >
             Children
-          </Box>
+          </UnsafeBox>
         </ThemeProvider>
       );
       expect(container.firstChild).toMatchSnapshot();
diff --git a/lib/components/Bullet/Bullet.test.js b/lib/components/Bullet/Bullet.test.tsx
similarity index 100%
rename from lib/components/Bullet/Bullet.test.js
rename to lib/components/Bullet/Bullet.test.tsx
diff --git a/lib/components/BulletList/BulletList.test.js b/lib/components/BulletList/BulletList.test.tsx
similarity index 100%
rename from lib/components/BulletList/BulletList.test.js
rename to lib/components/BulletList/BulletList.test.tsx
diff --git a/lib/components/Card/Card.test.js b/lib/components/Card/Card.test.tsx
similarity index 100%
rename from lib/components/Card/Card.test.js
rename to lib/components/Card/Card.test.tsx
diff --git a/lib/components/Checkbox/Checkbox.test.js b/lib/components/Checkbox/Checkbox.test.tsx
similarity index 98%
rename from lib/components/Checkbox/Checkbox.test.js
rename to lib/components/Checkbox/Checkbox.test.tsx
index 4db577c248f..faf0391f8bd 100644
--- a/lib/components/Checkbox/Checkbox.test.js
+++ b/lib/components/Checkbox/Checkbox.test.tsx
@@ -3,7 +3,7 @@ import { Checkbox } from '..';
 import { render } from 'react-testing-library';
 import forEachTheme from '../../test/utils/forEachTheme';
 
-const noop = () => {};
+const noop = () => undefined;
 
 describe('Checkbox', () => {
   forEachTheme(({ ThemeProvider }) => {
diff --git a/lib/components/ChecklistCard/ChecklistCard.test.js b/lib/components/ChecklistCard/ChecklistCard.test.tsx
similarity index 96%
rename from lib/components/ChecklistCard/ChecklistCard.test.js
rename to lib/components/ChecklistCard/ChecklistCard.test.tsx
index bdf8d13d253..f11dd5b2c5c 100644
--- a/lib/components/ChecklistCard/ChecklistCard.test.js
+++ b/lib/components/ChecklistCard/ChecklistCard.test.tsx
@@ -3,7 +3,7 @@ import { ChecklistCard, Checkbox } from '..';
 import { render } from 'react-testing-library';
 import forEachTheme from '../../test/utils/forEachTheme';
 
-const noop = () => {};
+const noop = () => undefined;
 
 describe('ChecklistCard', () => {
   forEachTheme(({ ThemeProvider }) => {
diff --git a/lib/components/Divider/Divider.test.js b/lib/components/Divider/Divider.test.tsx
similarity index 100%
rename from lib/components/Divider/Divider.test.js
rename to lib/components/Divider/Divider.test.tsx
diff --git a/lib/components/FieldLabel/FieldLabel.test.js b/lib/components/FieldLabel/FieldLabel.test.tsx
similarity index 100%
rename from lib/components/FieldLabel/FieldLabel.test.js
rename to lib/components/FieldLabel/FieldLabel.test.tsx
diff --git a/lib/components/FieldMessage/FieldMessage.test.js b/lib/components/FieldMessage/FieldMessage.test.tsx
similarity index 100%
rename from lib/components/FieldMessage/FieldMessage.test.js
rename to lib/components/FieldMessage/FieldMessage.test.tsx
diff --git a/lib/components/Radio/Radio.test.js b/lib/components/Radio/Radio.test.tsx
similarity index 98%
rename from lib/components/Radio/Radio.test.js
rename to lib/components/Radio/Radio.test.tsx
index daf0f51d02e..82b5a80c6e4 100644
--- a/lib/components/Radio/Radio.test.js
+++ b/lib/components/Radio/Radio.test.tsx
@@ -3,7 +3,7 @@ import { Radio } from '..';
 import { render } from 'react-testing-library';
 import forEachTheme from '../../test/utils/forEachTheme';
 
-const noop = () => {};
+const noop = () => undefined;
 
 describe('Radio', () => {
   forEachTheme(({ ThemeProvider }) => {
diff --git a/lib/components/Reset/Reset.test.js b/lib/components/Reset/Reset.test.tsx
similarity index 89%
rename from lib/components/Reset/Reset.test.js
rename to lib/components/Reset/Reset.test.tsx
index cd297340a92..db8e968a3db 100644
--- a/lib/components/Reset/Reset.test.js
+++ b/lib/components/Reset/Reset.test.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import React, { AllHTMLAttributes } from 'react';
 import { Reset } from '..';
 import { render } from 'react-testing-library';
 import forEachTheme from '../../test/utils/forEachTheme';
@@ -24,7 +24,7 @@ describe('Reset', () => {
     });
 
     test('Component', () => {
-      function CustomComponent(props) {
+      function CustomComponent(props: AllHTMLAttributes<HTMLElement>) {
         return <span data-custom="component" {...props} />;
       }
 
diff --git a/lib/components/Strong/Strong.test.js b/lib/components/Strong/Strong.test.tsx
similarity index 100%
rename from lib/components/Strong/Strong.test.js
rename to lib/components/Strong/Strong.test.tsx
diff --git a/lib/components/Text/Text.test.js b/lib/components/Text/Text.test.tsx
similarity index 100%
rename from lib/components/Text/Text.test.js
rename to lib/components/Text/Text.test.tsx
diff --git a/lib/components/icons/ChevronIcon/ChevronIcon.test.js b/lib/components/icons/ChevronIcon/ChevronIcon.test.tsx
similarity index 100%
rename from lib/components/icons/ChevronIcon/ChevronIcon.test.js
rename to lib/components/icons/ChevronIcon/ChevronIcon.test.tsx
diff --git a/lib/components/icons/ErrorIcon/ErrorIcon.test.js b/lib/components/icons/ErrorIcon/ErrorIcon.test.tsx
similarity index 100%
rename from lib/components/icons/ErrorIcon/ErrorIcon.test.js
rename to lib/components/icons/ErrorIcon/ErrorIcon.test.tsx
diff --git a/lib/components/icons/InfoIcon/InfoIcon.test.js b/lib/components/icons/InfoIcon/InfoIcon.test.tsx
similarity index 100%
rename from lib/components/icons/InfoIcon/InfoIcon.test.js
rename to lib/components/icons/InfoIcon/InfoIcon.test.tsx
diff --git a/lib/components/icons/TickCircleIcon/TickCircleIcon.test.js b/lib/components/icons/TickCircleIcon/TickCircleIcon.test.tsx
similarity index 100%
rename from lib/components/icons/TickCircleIcon/TickCircleIcon.test.js
rename to lib/components/icons/TickCircleIcon/TickCircleIcon.test.tsx
diff --git a/lib/components/icons/TickIcon/TickIcon.test.js b/lib/components/icons/TickIcon/TickIcon.test.tsx
similarity index 100%
rename from lib/components/icons/TickIcon/TickIcon.test.js
rename to lib/components/icons/TickIcon/TickIcon.test.tsx
diff --git a/lib/test/utils/forEachTheme.js b/lib/test/utils/forEachTheme.js
deleted file mode 100644
index 4f1c088b3de..00000000000
--- a/lib/test/utils/forEachTheme.js
+++ /dev/null
@@ -1,15 +0,0 @@
-import React from 'react';
-import each from 'lodash/each';
-import * as themes from '../../themes';
-import { ThemeProvider } from '../../components';
-
-export default callback => {
-  each(themes, theme => {
-    describe(`Theme: ${theme.name}`, () => {
-      callback({
-        theme,
-        ThemeProvider: props => <ThemeProvider theme={theme} {...props} />
-      });
-    });
-  });
-};
diff --git a/lib/test/utils/forEachTheme.tsx b/lib/test/utils/forEachTheme.tsx
new file mode 100644
index 00000000000..2c8bd2b2347
--- /dev/null
+++ b/lib/test/utils/forEachTheme.tsx
@@ -0,0 +1,26 @@
+import React, { FunctionComponent } from 'react';
+import each from 'lodash/each';
+import { Omit } from 'utility-types';
+import * as themes from '../../themes';
+import { Theme } from '../../themes/theme';
+import ThemeProvider, {
+  ThemeProviderProps
+} from '../../components/ThemeProvider/ThemeProvider';
+
+type ForEachThemeCallback = (
+  args: {
+    theme: Theme;
+    ThemeProvider: FunctionComponent<Omit<ThemeProviderProps, 'theme'>>;
+  }
+) => void;
+
+export default (callback: ForEachThemeCallback) => {
+  each(themes, theme => {
+    describe(`Theme: ${theme.name}`, () => {
+      callback({
+        theme,
+        ThemeProvider: props => <ThemeProvider theme={theme} {...props} />
+      });
+    });
+  });
+};
diff --git a/lib/themes/themes.test.js b/lib/themes/themes.test.js
deleted file mode 100644
index 79ca828e4cc..00000000000
--- a/lib/themes/themes.test.js
+++ /dev/null
@@ -1,16 +0,0 @@
-/* eslint-disable import/namespace */
-import * as themes from './';
-import deepKeys from 'deep-keys';
-const themeNames = Object.keys(themes);
-
-const contract = deepKeys(themes[themeNames[0]]);
-
-describe('Themes', () => {
-  describe('Consistent API', () => {
-    themeNames.forEach(themeName => {
-      test(themeName, () => {
-        expect(deepKeys(themes[themeName])).toEqual(contract);
-      });
-    });
-  });
-});
diff --git a/lib/themes/utils/toUnit.test.js b/lib/themes/utils/toUnit.test.ts
similarity index 100%
rename from lib/themes/utils/toUnit.test.js
rename to lib/themes/utils/toUnit.test.ts
