diff --git a/pkg/machinery/config/types/v1alpha1/v1alpha1_stability_test.go b/pkg/machinery/config/types/v1alpha1/v1alpha1_stability_test.go
index f593abd21..3bab495dd 100644
--- a/pkg/machinery/config/types/v1alpha1/v1alpha1_stability_test.go
+++ b/pkg/machinery/config/types/v1alpha1/v1alpha1_stability_test.go
@@ -40,6 +40,7 @@ func TestConfigEncodingStability(t *testing.T) {
 		config.TalosVersion1_5,
 		config.TalosVersion1_6,
 		config.TalosVersion1_7,
+		config.TalosVersion1_8,
 	}
 
 	currentVersion := ensure.Value(semver.ParseTolerant(gendata.VersionTag))
