diff --git a/exporter/datadogexporter/internal/model/attributes/azure/azure_test.go b/exporter/datadogexporter/internal/model/attributes/azure/azure_test.go
index 6771e48673..ca866505e3 100644
--- a/exporter/datadogexporter/internal/model/attributes/azure/azure_test.go
+++ b/exporter/datadogexporter/internal/model/attributes/azure/azure_test.go
@@ -18,6 +18,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/pdata/pcommon"
 	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/internal/model/internal/testutils"
@@ -38,22 +39,55 @@ var (
 )
 
 func TestHostInfoFromAttributes(t *testing.T) {
-	hostInfo := HostInfoFromAttributes(testAttrs)
-	require.NotNil(t, hostInfo)
-	assert.ElementsMatch(t, hostInfo.HostAliases, []string{testVMID})
+	tests := []struct {
+		name       string
+		attrs      pcommon.Map
+		usePreview bool
 
-	emptyHostInfo := HostInfoFromAttributes(testEmpty)
-	require.NotNil(t, emptyHostInfo)
-	assert.ElementsMatch(t, emptyHostInfo.HostAliases, []string{})
-}
+		ok       bool
+		hostname string
+		aliases  []string
+	}{
+		{
+			name:  "all attributes",
+			attrs: testAttrs,
 
-func TestHostnameFromAttributes(t *testing.T) {
-	hostname, ok := HostnameFromAttributes(testAttrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testHostname)
+			ok:       true,
+			hostname: testHostname,
+			aliases:  []string{testVMID},
+		},
+		{
+			name:  "empty",
+			attrs: testEmpty,
+		},
+		{
+			name:       "all attributes with preview",
+			attrs:      testAttrs,
+			usePreview: true,
 
-	_, ok = HostnameFromAttributes(testEmpty)
-	assert.False(t, ok)
+			ok:       true,
+			hostname: testVMID,
+		},
+		{
+			name:       "empty with preview",
+			attrs:      testEmpty,
+			usePreview: true,
+		},
+	}
+
+	for _, testInstance := range tests {
+		t.Run(testInstance.name, func(t *testing.T) {
+			hostInfo := HostInfoFromAttributes(testInstance.attrs, testInstance.usePreview)
+			require.NotNil(t, hostInfo)
+			assert.ElementsMatch(t, testInstance.aliases, hostInfo.HostAliases)
+			hostname, ok := HostnameFromAttributes(testInstance.attrs, testInstance.usePreview)
+
+			assert.Equal(t, testInstance.ok, ok)
+			if testInstance.ok || ok {
+				assert.Equal(t, testInstance.hostname, hostname)
+			}
+		})
+	}
 }
 
 func TestClusterNameFromAttributes(t *testing.T) {
diff --git a/exporter/datadogexporter/internal/model/attributes/ec2/ec2_test.go b/exporter/datadogexporter/internal/model/attributes/ec2/ec2_test.go
index 70bcd5de2d..af1fcdf13e 100644
--- a/exporter/datadogexporter/internal/model/attributes/ec2/ec2_test.go
+++ b/exporter/datadogexporter/internal/model/attributes/ec2/ec2_test.go
@@ -44,7 +44,11 @@ func TestHostnameFromAttributes(t *testing.T) {
 		conventions.AttributeHostID:        testInstanceID,
 		conventions.AttributeHostName:      testIP,
 	})
-	hostname, ok := HostnameFromAttributes(attrs)
+	hostname, ok := HostnameFromAttributes(attrs, false)
+	assert.True(t, ok)
+	assert.Equal(t, hostname, testInstanceID)
+
+	hostname, ok = HostnameFromAttributes(attrs, true)
 	assert.True(t, ok)
 	assert.Equal(t, hostname, testInstanceID)
 }
diff --git a/exporter/datadogexporter/internal/model/attributes/gcp/gcp_test.go b/exporter/datadogexporter/internal/model/attributes/gcp/gcp_test.go
index 30f9d6633f..fc7520626c 100644
--- a/exporter/datadogexporter/internal/model/attributes/gcp/gcp_test.go
+++ b/exporter/datadogexporter/internal/model/attributes/gcp/gcp_test.go
@@ -18,6 +18,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/pdata/pcommon"
 	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/internal/model/internal/testutils"
@@ -33,23 +34,8 @@ const (
 	testBadHostname   = "badhostname"
 )
 
-func TestHostnameFromAttributes(t *testing.T) {
-	attrs := testutils.NewAttributeMap(map[string]string{
-		conventions.AttributeCloudProvider: conventions.AttributeCloudProviderGCP,
-		conventions.AttributeHostID:        testHostID,
-		conventions.AttributeHostName:      testHostname,
-	})
-	hostname, ok := HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testHostname)
-
-	attrs = testutils.NewAttributeMap(map[string]string{})
-	_, ok = HostnameFromAttributes(attrs)
-	assert.False(t, ok)
-}
-
-func TestHostInfoFromAttributes(t *testing.T) {
-	attrs := testutils.NewAttributeMap(map[string]string{
+var (
+	testFullMap = testutils.NewAttributeMap(map[string]string{
 		conventions.AttributeCloudProvider:         conventions.AttributeCloudProviderGCP,
 		conventions.AttributeHostID:                testHostID,
 		conventions.AttributeHostName:              testHostname,
@@ -57,14 +43,8 @@ func TestHostInfoFromAttributes(t *testing.T) {
 		conventions.AttributeHostType:              testHostType,
 		conventions.AttributeCloudAccountID:        testCloudAccount,
 	})
-	hostInfo := HostInfoFromAttributes(attrs)
-	assert.ElementsMatch(t, hostInfo.HostAliases, []string{fmt.Sprintf("%s.%s", testShortHostname, testCloudAccount)})
-	assert.ElementsMatch(t, hostInfo.GCPTags,
-		[]string{"instance-id:hostID", "zone:zone", "instance-type:machineType", "project:projectID"})
-}
 
-func TestHostInfoFromAttributesBadHostname(t *testing.T) {
-	attrs := testutils.NewAttributeMap(map[string]string{
+	testFullBadMap = testutils.NewAttributeMap(map[string]string{
 		conventions.AttributeCloudProvider:         conventions.AttributeCloudProviderGCP,
 		conventions.AttributeHostID:                testHostID,
 		conventions.AttributeHostName:              testBadHostname,
@@ -72,8 +52,75 @@ func TestHostInfoFromAttributesBadHostname(t *testing.T) {
 		conventions.AttributeHostType:              testHostType,
 		conventions.AttributeCloudAccountID:        testCloudAccount,
 	})
-	hostInfo := HostInfoFromAttributes(attrs)
-	assert.ElementsMatch(t, hostInfo.HostAliases, []string{fmt.Sprintf("%s.%s", testBadHostname, testCloudAccount)})
-	assert.ElementsMatch(t, hostInfo.GCPTags,
-		[]string{"instance-id:hostID", "zone:zone", "instance-type:machineType", "project:projectID"})
+
+	testGCPIntegrationHostname    = fmt.Sprintf("%s.%s", testShortHostname, testCloudAccount)
+	testGCPIntegrationBadHostname = fmt.Sprintf("%s.%s", testBadHostname, testCloudAccount)
+)
+
+func TestInfoFromAttributes(t *testing.T) {
+	tags := []string{"instance-id:hostID", "zone:zone", "instance-type:machineType", "project:projectID"}
+	tests := []struct {
+		name       string
+		attrs      pcommon.Map
+		usePreview bool
+
+		ok          bool
+		hostname    string
+		hostAliases []string
+		gcpTags     []string
+	}{
+		{
+			name:        "no preview",
+			attrs:       testFullMap,
+			ok:          true,
+			hostname:    testHostname,
+			hostAliases: []string{testGCPIntegrationHostname},
+			gcpTags:     tags,
+		},
+		{
+			name:  "no hostname, no preview",
+			attrs: testutils.NewAttributeMap(map[string]string{}),
+		},
+		{
+			name:       "preview",
+			attrs:      testFullMap,
+			usePreview: true,
+			ok:         true,
+			hostname:   testGCPIntegrationHostname,
+			gcpTags:    tags,
+		},
+		{
+			name:        "bad hostname, no preview",
+			attrs:       testFullBadMap,
+			ok:          true,
+			hostname:    testBadHostname,
+			hostAliases: []string{testGCPIntegrationBadHostname},
+			gcpTags:     tags,
+		},
+		{
+			name:       "bad hostname, preview",
+			attrs:      testFullBadMap,
+			usePreview: true,
+			ok:         true,
+			hostname:   testGCPIntegrationBadHostname,
+			gcpTags:    tags,
+		},
+		{
+			name:       "no hostname, preview",
+			attrs:      testutils.NewAttributeMap(map[string]string{}),
+			usePreview: true,
+		},
+	}
+
+	for _, testInstance := range tests {
+		t.Run(testInstance.name, func(t *testing.T) {
+			hostname, ok := HostnameFromAttributes(testInstance.attrs, testInstance.usePreview)
+			assert.Equal(t, testInstance.ok, ok)
+			assert.Equal(t, testInstance.hostname, hostname)
+
+			hostInfo := HostInfoFromAttributes(testInstance.attrs, testInstance.usePreview)
+			assert.ElementsMatch(t, testInstance.hostAliases, hostInfo.HostAliases)
+			assert.ElementsMatch(t, testInstance.gcpTags, hostInfo.GCPTags)
+		})
+	}
 }
diff --git a/exporter/datadogexporter/internal/model/attributes/hostname_test.go b/exporter/datadogexporter/internal/model/attributes/hostname_test.go
index 410d389082..4e22b73717 100644
--- a/exporter/datadogexporter/internal/model/attributes/hostname_test.go
+++ b/exporter/datadogexporter/internal/model/attributes/hostname_test.go
@@ -18,6 +18,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/pdata/pcommon"
 	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/internal/model/attributes/azure"
@@ -25,100 +26,160 @@ import (
 )
 
 const (
-	testHostID      = "example-host-id"
-	testHostName    = "example-host-name"
-	testContainerID = "example-container-id"
-	testClusterName = "clusterName"
-	testNodeName    = "nodeName"
-	testCustomName  = "example-custom-host-name"
+	testHostID                 = "example-host-id"
+	testHostName               = "example-host-name"
+	testContainerID            = "example-container-id"
+	testClusterName            = "clusterName"
+	testNodeName               = "nodeName"
+	testCustomName             = "example-custom-host-name"
+	testCloudAccount           = "projectID"
+	testGCPHostname            = testHostName + ".c." + testCloudAccount + ".internal"
+	testGCPIntegrationHostname = testHostName + "." + testCloudAccount
 )
 
 func TestHostnameFromAttributes(t *testing.T) {
-	// Custom hostname
-	attrs := testutils.NewAttributeMap(map[string]string{
-		AttributeDatadogHostname:            testCustomName,
-		AttributeK8sNodeName:                testNodeName,
-		conventions.AttributeK8SClusterName: testClusterName,
-		conventions.AttributeContainerID:    testContainerID,
-		conventions.AttributeHostID:         testHostID,
-		conventions.AttributeHostName:       testHostName,
-	})
-	hostname, ok := HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testCustomName)
-
-	// Container ID
-	attrs = testutils.NewAttributeMap(map[string]string{
-		conventions.AttributeContainerID: testContainerID,
-	})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testContainerID)
-
-	// AWS cloud provider means relying on the EC2 function
-	attrs = testutils.NewAttributeMap(map[string]string{
-		conventions.AttributeCloudProvider: conventions.AttributeCloudProviderAWS,
-		conventions.AttributeHostID:        testHostID,
-		conventions.AttributeHostName:      testHostName,
-	})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testHostName)
-
-	// AWS cloud provider means relying on the EC2 function
-	attrs = testutils.NewAttributeMap(map[string]string{
-		conventions.AttributeCloudProvider:      conventions.AttributeCloudProviderAWS,
-		conventions.AttributeCloudPlatform:      conventions.AttributeCloudPlatformAWSECS,
-		conventions.AttributeAWSECSTaskARN:      "example-task-ARN",
-		conventions.AttributeAWSECSTaskFamily:   "example-task-family",
-		conventions.AttributeAWSECSTaskRevision: "example-task-revision",
-		conventions.AttributeAWSECSLaunchtype:   conventions.AttributeAWSECSLaunchtypeFargate,
-	})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Empty(t, hostname)
-
-	// GCP cloud provider means relying on the GCP function
-	attrs = testutils.NewAttributeMap(map[string]string{
-		conventions.AttributeCloudProvider: conventions.AttributeCloudProviderGCP,
-		conventions.AttributeHostID:        testHostID,
-		conventions.AttributeHostName:      testHostName,
-	})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testHostName)
-
-	// Azure cloud provider means relying on the Azure function
-	attrs = testutils.NewAttributeMap(map[string]string{
-		conventions.AttributeCloudProvider: conventions.AttributeCloudProviderAzure,
-		conventions.AttributeHostID:        testHostID,
-		conventions.AttributeHostName:      testHostName,
-	})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testHostName)
-
-	// Host Id takes preference
-	attrs = testutils.NewAttributeMap(map[string]string{
-		conventions.AttributeHostID:   testHostID,
-		conventions.AttributeHostName: testHostName,
-	})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.True(t, ok)
-	assert.Equal(t, hostname, testHostID)
-
-	// No labels means no hostname
-	attrs = testutils.NewAttributeMap(map[string]string{})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.False(t, ok)
-	assert.Empty(t, hostname)
-
-	attrs = testutils.NewAttributeMap(map[string]string{
-		AttributeDatadogHostname: "127.0.0.1",
-	})
-	hostname, ok = HostnameFromAttributes(attrs)
-	assert.False(t, ok)
-	assert.Empty(t, hostname)
+	tests := []struct {
+		name       string
+		attrs      pcommon.Map
+		usePreview bool
+
+		ok       bool
+		hostname string
+	}{
+		{
+			name: "custom hostname",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				AttributeDatadogHostname:            testCustomName,
+				AttributeK8sNodeName:                testNodeName,
+				conventions.AttributeK8SClusterName: testClusterName,
+				conventions.AttributeContainerID:    testContainerID,
+				conventions.AttributeHostID:         testHostID,
+				conventions.AttributeHostName:       testHostName,
+			}),
+			ok:       true,
+			hostname: testCustomName,
+		},
+		{
+			name: "container ID",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeContainerID: testContainerID,
+			}),
+			ok:       true,
+			hostname: testContainerID,
+		},
+		{
+			name: "container ID, preview",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeContainerID: testContainerID,
+			}),
+			usePreview: true,
+		},
+		{
+			name: "AWS EC2",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeCloudProvider: conventions.AttributeCloudProviderAWS,
+				conventions.AttributeHostID:        testHostID,
+				conventions.AttributeHostName:      testHostName,
+			}),
+			ok:       true,
+			hostname: testHostName,
+		},
+		{
+			name: "AWS EC2, preview",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeCloudProvider: conventions.AttributeCloudProviderAWS,
+				conventions.AttributeHostID:        testHostID,
+				conventions.AttributeHostName:      testHostName,
+			}),
+			ok:         true,
+			hostname:   testHostID,
+			usePreview: true,
+		},
+		{
+			name: "ECS Fargate",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeCloudProvider:      conventions.AttributeCloudProviderAWS,
+				conventions.AttributeCloudPlatform:      conventions.AttributeCloudPlatformAWSECS,
+				conventions.AttributeAWSECSTaskARN:      "example-task-ARN",
+				conventions.AttributeAWSECSTaskFamily:   "example-task-family",
+				conventions.AttributeAWSECSTaskRevision: "example-task-revision",
+				conventions.AttributeAWSECSLaunchtype:   conventions.AttributeAWSECSLaunchtypeFargate,
+			}),
+			ok:       true,
+			hostname: "",
+		},
+		{
+			name: "GCP",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeCloudProvider: conventions.AttributeCloudProviderGCP,
+				conventions.AttributeHostID:        testHostID,
+				conventions.AttributeHostName:      testGCPHostname,
+			}),
+			ok:       true,
+			hostname: testGCPHostname,
+		},
+		{
+			name: "GCP, preview",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeCloudProvider:  conventions.AttributeCloudProviderGCP,
+				conventions.AttributeHostID:         testHostID,
+				conventions.AttributeHostName:       testGCPHostname,
+				conventions.AttributeCloudAccountID: testCloudAccount,
+			}),
+			usePreview: true,
+			ok:         true,
+			hostname:   testGCPIntegrationHostname,
+		},
+		{
+			name: "azure",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeCloudProvider: conventions.AttributeCloudProviderAzure,
+				conventions.AttributeHostID:        testHostID,
+				conventions.AttributeHostName:      testHostName,
+			}),
+			ok:       true,
+			hostname: testHostName,
+		},
+		{
+			name: "azure, preview",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeCloudProvider: conventions.AttributeCloudProviderAzure,
+				conventions.AttributeHostID:        testHostID,
+				conventions.AttributeHostName:      testHostName,
+			}),
+			usePreview: true,
+			ok:         true,
+			hostname:   testHostID,
+		},
+		{
+			name: "host id v. hostname",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				conventions.AttributeHostID:   testHostID,
+				conventions.AttributeHostName: testHostName,
+			}),
+			ok:       true,
+			hostname: testHostID,
+		},
+		{
+			name:  "no hostname",
+			attrs: testutils.NewAttributeMap(map[string]string{}),
+		},
+		{
+			name: "localhost",
+			attrs: testutils.NewAttributeMap(map[string]string{
+				AttributeDatadogHostname: "127.0.0.1",
+			}),
+		},
+	}
+
+	for _, testInstance := range tests {
+		t.Run(testInstance.name, func(t *testing.T) {
+			hostname, ok := HostnameFromAttributes(testInstance.attrs, testInstance.usePreview)
+			assert.Equal(t, testInstance.ok, ok)
+			assert.Equal(t, testInstance.hostname, hostname)
+		})
+
+	}
 }
 
 func TestGetClusterName(t *testing.T) {
@@ -163,7 +224,7 @@ func TestHostnameKubernetes(t *testing.T) {
 		conventions.AttributeHostID:         testHostID,
 		conventions.AttributeHostName:       testHostName,
 	})
-	hostname, ok := HostnameFromAttributes(attrs)
+	hostname, ok := HostnameFromAttributes(attrs, false)
 	assert.True(t, ok)
 	assert.Equal(t, hostname, "nodeName-clusterName")
 
@@ -174,7 +235,7 @@ func TestHostnameKubernetes(t *testing.T) {
 		conventions.AttributeHostID:      testHostID,
 		conventions.AttributeHostName:    testHostName,
 	})
-	hostname, ok = HostnameFromAttributes(attrs)
+	hostname, ok = HostnameFromAttributes(attrs, false)
 	assert.True(t, ok)
 	assert.Equal(t, hostname, "nodeName")
 
@@ -185,7 +246,7 @@ func TestHostnameKubernetes(t *testing.T) {
 		conventions.AttributeHostID:         testHostID,
 		conventions.AttributeHostName:       testHostName,
 	})
-	hostname, ok = HostnameFromAttributes(attrs)
+	hostname, ok = HostnameFromAttributes(attrs, false)
 	assert.True(t, ok)
 	// cluster name gets ignored, fallback to next option
 	assert.Equal(t, hostname, testHostID)
