diff --git a/hudi-common/src/test/java/org/apache/hudi/common/util/collection/TestExternalSpillableMap.java b/hudi-common/src/test/java/org/apache/hudi/common/util/collection/TestExternalSpillableMap.java
index 2b651ecfd9c4..ed64ee002878 100644
--- a/hudi-common/src/test/java/org/apache/hudi/common/util/collection/TestExternalSpillableMap.java
+++ b/hudi-common/src/test/java/org/apache/hudi/common/util/collection/TestExternalSpillableMap.java
@@ -43,6 +43,7 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.TestMethodOrder;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.EnumSource;
 import org.junit.jupiter.params.provider.MethodSource;
 
 import java.io.IOException;
@@ -81,33 +82,34 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
   public void simpleInsertTest(ExternalSpillableMap.DiskMapType diskMapType, boolean isCompressionEnabled) throws IOException, URISyntaxException {
     Schema schema = HoodieAvroUtils.addMetadataFields(SchemaTestUtil.getSimpleSchema());
 
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
         new ExternalSpillableMap<>(16L, basePath, new DefaultSizeEstimator(),
-            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled); // 16B
+            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) { // 16B
 
-    SchemaTestUtil testUtil = new SchemaTestUtil();
-    List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
-    List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
-    assert (recordKeys.size() == 100);
-
-    // Test iterator
-    Iterator<HoodieRecord<? extends HoodieRecordPayload>> itr = records.iterator();
-    int cntSize = 0;
-    while (itr.hasNext()) {
-      HoodieRecord<? extends HoodieRecordPayload> rec = itr.next();
-      cntSize++;
-      assert recordKeys.contains(rec.getRecordKey());
-    }
-    assertEquals(recordKeys.size(), cntSize);
-
-    // Test value stream
-    List<HoodieRecord<? extends HoodieRecordPayload>> values = records.valueStream().collect(Collectors.toList());
-    cntSize = 0;
-    for (HoodieRecord value : values) {
-      assert recordKeys.contains(value.getRecordKey());
-      cntSize++;
+      SchemaTestUtil testUtil = new SchemaTestUtil();
+      List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
+      List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
+      assert (recordKeys.size() == 100);
+
+      // Test iterator
+      Iterator<HoodieRecord<? extends HoodieRecordPayload>> itr = records.iterator();
+      int cntSize = 0;
+      while (itr.hasNext()) {
+        HoodieRecord<? extends HoodieRecordPayload> rec = itr.next();
+        cntSize++;
+        assert recordKeys.contains(rec.getRecordKey());
+      }
+      assertEquals(recordKeys.size(), cntSize);
+
+      // Test value stream
+      List<HoodieRecord<? extends HoodieRecordPayload>> values = records.valueStream().collect(Collectors.toList());
+      cntSize = 0;
+      for (HoodieRecord value : values) {
+        assert recordKeys.contains(value.getRecordKey());
+        cntSize++;
+      }
+      assertEquals(recordKeys.size(), cntSize);
     }
-    assertEquals(recordKeys.size(), cntSize);
   }
 
   @ParameterizedTest
@@ -115,38 +117,39 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
   public void testSimpleUpsert(ExternalSpillableMap.DiskMapType diskMapType, boolean isCompressionEnabled) throws IOException, URISyntaxException {
     Schema schema = HoodieAvroUtils.addMetadataFields(SchemaTestUtil.getSimpleSchema());
 
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
         new ExternalSpillableMap<>(16L, basePath, new DefaultSizeEstimator(),
-            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled); // 16B
+            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) { // 16B
 
-    SchemaTestUtil testUtil = new SchemaTestUtil();
-    List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
-    List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
-    assert (recordKeys.size() == 100);
-    Iterator<HoodieRecord<? extends HoodieRecordPayload>> itr = records.iterator();
-    while (itr.hasNext()) {
-      HoodieRecord<? extends HoodieRecordPayload> rec = itr.next();
-      assert recordKeys.contains(rec.getRecordKey());
-    }
+      SchemaTestUtil testUtil = new SchemaTestUtil();
+      List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
+      List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
+      assert (recordKeys.size() == 100);
+      Iterator<HoodieRecord<? extends HoodieRecordPayload>> itr = records.iterator();
+      while (itr.hasNext()) {
+        HoodieRecord<? extends HoodieRecordPayload> rec = itr.next();
+        assert recordKeys.contains(rec.getRecordKey());
+      }
 
-    List<IndexedRecord> updatedRecords = SchemaTestUtil.updateHoodieTestRecords(recordKeys,
-        testUtil.generateHoodieTestRecords(0, 100), InProcessTimeGenerator.createNewInstantTime());
+      List<IndexedRecord> updatedRecords = SchemaTestUtil.updateHoodieTestRecords(recordKeys,
+          testUtil.generateHoodieTestRecords(0, 100), InProcessTimeGenerator.createNewInstantTime());
 
-    // update records already inserted
-    SpillableMapTestUtils.upsertRecords(updatedRecords, records);
+      // update records already inserted
+      SpillableMapTestUtils.upsertRecords(updatedRecords, records);
 
-    // make sure we have records spilled to disk
-    assertTrue(records.getDiskBasedMapNumEntries() > 0);
+      // make sure we have records spilled to disk
+      assertTrue(records.getDiskBasedMapNumEntries() > 0);
 
-    // iterate over the updated records and compare the value from Map
-    updatedRecords.forEach(record -> {
-      HoodieRecord rec = records.get(((GenericRecord) record).get(HoodieRecord.RECORD_KEY_METADATA_FIELD));
-      try {
-        assertEquals(((HoodieAvroRecord) rec).getData().getInsertValue(schema).get(), record);
-      } catch (IOException io) {
-        throw new UncheckedIOException(io);
-      }
-    });
+      // iterate over the updated records and compare the value from Map
+      updatedRecords.forEach(record -> {
+        HoodieRecord rec = records.get(((GenericRecord) record).get(HoodieRecord.RECORD_KEY_METADATA_FIELD));
+        try {
+          assertEquals(((HoodieAvroRecord) rec).getData().getInsertValue(schema).get(), record);
+        } catch (IOException io) {
+          throw new UncheckedIOException(io);
+        }
+      });
+    }
   }
 
   @ParameterizedTest
@@ -156,56 +159,57 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
     Schema schema = HoodieAvroUtils.addMetadataFields(SchemaTestUtil.getSimpleSchema());
     String payloadClazz = HoodieAvroPayload.class.getName();
 
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
         new ExternalSpillableMap<>(16L, basePath, new DefaultSizeEstimator(),
-            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled); // 16B
+            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) { // 16B
 
-    SchemaTestUtil testUtil = new SchemaTestUtil();
-    List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
-    // insert a bunch of records so that values spill to disk too
-    List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
-    IndexedRecord inMemoryRecord = iRecords.get(0);
-    String ikey = ((GenericRecord) inMemoryRecord).get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString();
-    String iPartitionPath = ((GenericRecord) inMemoryRecord).get(HoodieRecord.PARTITION_PATH_METADATA_FIELD).toString();
-    HoodieRecord inMemoryHoodieRecord = new HoodieAvroRecord<>(new HoodieKey(ikey, iPartitionPath),
-        new HoodieAvroPayload(Option.of((GenericRecord) inMemoryRecord)));
-
-    IndexedRecord onDiskRecord = iRecords.get(99);
-    String dkey = ((GenericRecord) onDiskRecord).get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString();
-    String dPartitionPath = ((GenericRecord) onDiskRecord).get(HoodieRecord.PARTITION_PATH_METADATA_FIELD).toString();
-    HoodieRecord onDiskHoodieRecord = new HoodieAvroRecord<>(new HoodieKey(dkey, dPartitionPath),
-        new HoodieAvroPayload(Option.of((GenericRecord) onDiskRecord)));
-    // assert size
-    assert records.size() == 100;
-    // get should return the same HoodieKey, same location and same value
-    assert inMemoryHoodieRecord.getKey().equals(records.get(ikey).getKey());
-    assert onDiskHoodieRecord.getKey().equals(records.get(dkey).getKey());
-    // compare the member variables of HoodieRecord not set by the constructor
-    assert records.get(ikey).getCurrentLocation().getFileId().equals(SpillableMapTestUtils.DUMMY_FILE_ID);
-    assert records.get(ikey).getCurrentLocation().getInstantTime().equals(SpillableMapTestUtils.DUMMY_COMMIT_TIME);
-
-    // test contains
-    assertTrue(records.containsKey(ikey));
-    assertTrue(records.containsKey(dkey));
-
-    // test isEmpty
-    assertFalse(records.isEmpty());
-
-    // test containsAll
-    assertTrue(records.keySet().containsAll(recordKeys));
-
-    // remove (from inMemory and onDisk)
-    HoodieRecord removedRecord = records.remove(ikey);
-    assertTrue(removedRecord != null);
-    assertFalse(records.containsKey(ikey));
-
-    removedRecord = records.remove(dkey);
-    assertTrue(removedRecord != null);
-    assertFalse(records.containsKey(dkey));
-
-    // test clear
-    records.clear();
-    assertTrue(records.size() == 0);
+      SchemaTestUtil testUtil = new SchemaTestUtil();
+      List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
+      // insert a bunch of records so that values spill to disk too
+      List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
+      IndexedRecord inMemoryRecord = iRecords.get(0);
+      String ikey = ((GenericRecord) inMemoryRecord).get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString();
+      String iPartitionPath = ((GenericRecord) inMemoryRecord).get(HoodieRecord.PARTITION_PATH_METADATA_FIELD).toString();
+      HoodieRecord inMemoryHoodieRecord = new HoodieAvroRecord<>(new HoodieKey(ikey, iPartitionPath),
+          new HoodieAvroPayload(Option.of((GenericRecord) inMemoryRecord)));
+
+      IndexedRecord onDiskRecord = iRecords.get(99);
+      String dkey = ((GenericRecord) onDiskRecord).get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString();
+      String dPartitionPath = ((GenericRecord) onDiskRecord).get(HoodieRecord.PARTITION_PATH_METADATA_FIELD).toString();
+      HoodieRecord onDiskHoodieRecord = new HoodieAvroRecord<>(new HoodieKey(dkey, dPartitionPath),
+          new HoodieAvroPayload(Option.of((GenericRecord) onDiskRecord)));
+      // assert size
+      assert records.size() == 100;
+      // get should return the same HoodieKey, same location and same value
+      assert inMemoryHoodieRecord.getKey().equals(records.get(ikey).getKey());
+      assert onDiskHoodieRecord.getKey().equals(records.get(dkey).getKey());
+      // compare the member variables of HoodieRecord not set by the constructor
+      assert records.get(ikey).getCurrentLocation().getFileId().equals(SpillableMapTestUtils.DUMMY_FILE_ID);
+      assert records.get(ikey).getCurrentLocation().getInstantTime().equals(SpillableMapTestUtils.DUMMY_COMMIT_TIME);
+
+      // test contains
+      assertTrue(records.containsKey(ikey));
+      assertTrue(records.containsKey(dkey));
+
+      // test isEmpty
+      assertFalse(records.isEmpty());
+
+      // test containsAll
+      assertTrue(records.keySet().containsAll(recordKeys));
+
+      // remove (from inMemory and onDisk)
+      HoodieRecord removedRecord = records.remove(ikey);
+      assertTrue(removedRecord != null);
+      assertFalse(records.containsKey(ikey));
+
+      removedRecord = records.remove(dkey);
+      assertTrue(removedRecord != null);
+      assertFalse(records.containsKey(dkey));
+
+      // test clear
+      records.clear();
+      assertTrue(records.size() == 0);
+    }
   }
 
   @ParameterizedTest
@@ -213,20 +217,21 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
   public void simpleTestWithException(ExternalSpillableMap.DiskMapType diskMapType, boolean isCompressionEnabled) throws IOException, URISyntaxException {
     Schema schema = HoodieAvroUtils.addMetadataFields(SchemaTestUtil.getSimpleSchema());
 
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records = new ExternalSpillableMap<>(16L,
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records = new ExternalSpillableMap<>(16L,
         failureOutputPath, new DefaultSizeEstimator(),
-        new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled); // 16B
+        new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) { // 16B
 
-    SchemaTestUtil testUtil = new SchemaTestUtil();
-    List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
-    List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
-    assert (recordKeys.size() == 100);
-    Iterator<HoodieRecord<? extends HoodieRecordPayload>> itr = records.iterator();
-    assertThrows(IOException.class, () -> {
-      while (itr.hasNext()) {
-        throw new IOException("Testing failures...");
-      }
-    });
+      SchemaTestUtil testUtil = new SchemaTestUtil();
+      List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
+      List<String> recordKeys = SpillableMapTestUtils.upsertRecords(iRecords, records);
+      assert (recordKeys.size() == 100);
+      Iterator<HoodieRecord<? extends HoodieRecordPayload>> itr = records.iterator();
+      assertThrows(IOException.class, () -> {
+        while (itr.hasNext()) {
+          throw new IOException("Testing failures...");
+        }
+      });
+    }
   }
 
   @ParameterizedTest
@@ -237,52 +242,53 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
 
     Schema schema = HoodieAvroUtils.addMetadataFields(SchemaTestUtil.getSimpleSchema());
 
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
         new ExternalSpillableMap<>(16L, basePath, new DefaultSizeEstimator(),
-            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled); // 16B
+            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) { // 16B
+
+      List<String> recordKeys = new ArrayList<>();
+      // Ensure we spill to disk
+      while (records.getDiskBasedMapNumEntries() < 1) {
+        SchemaTestUtil testUtil = new SchemaTestUtil();
+        List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
+        recordKeys.addAll(SpillableMapTestUtils.upsertRecords(iRecords, records));
+      }
 
-    List<String> recordKeys = new ArrayList<>();
-    // Ensure we spill to disk
-    while (records.getDiskBasedMapNumEntries() < 1) {
-      SchemaTestUtil testUtil = new SchemaTestUtil();
-      List<IndexedRecord> iRecords = testUtil.generateHoodieTestRecords(0, 100);
-      recordKeys.addAll(SpillableMapTestUtils.upsertRecords(iRecords, records));
+      // Get a record from the in-Memory map
+      String key = recordKeys.get(0);
+      HoodieAvroRecord record = (HoodieAvroRecord) records.get(key);
+      List<IndexedRecord> recordsToUpdate = new ArrayList<>();
+      recordsToUpdate.add((IndexedRecord) record.getData().getInsertValue(schema).get());
+
+      String newCommitTime = InProcessTimeGenerator.createNewInstantTime();
+      List<String> keysToBeUpdated = new ArrayList<>();
+      keysToBeUpdated.add(key);
+      // Update the instantTime for this record
+      List<IndexedRecord> updatedRecords =
+          SchemaTestUtil.updateHoodieTestRecords(keysToBeUpdated, recordsToUpdate, newCommitTime);
+      // Upsert this updated record
+      SpillableMapTestUtils.upsertRecords(updatedRecords, records);
+      GenericRecord gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
+      // The record returned for this key should have the updated commitTime
+      assert newCommitTime.contentEquals(gRecord.get(HoodieRecord.COMMIT_TIME_METADATA_FIELD).toString());
+
+      // Get a record from the disk based map
+      key = recordKeys.get(recordKeys.size() - 1);
+      record = (HoodieAvroRecord) records.get(key);
+      recordsToUpdate = new ArrayList<>();
+      recordsToUpdate.add((IndexedRecord) record.getData().getInsertValue(schema).get());
+
+      newCommitTime = InProcessTimeGenerator.createNewInstantTime();
+      keysToBeUpdated = new ArrayList<>();
+      keysToBeUpdated.add(key);
+      // Update the commitTime for this record
+      updatedRecords = SchemaTestUtil.updateHoodieTestRecords(keysToBeUpdated, recordsToUpdate, newCommitTime);
+      // Upsert this updated record
+      SpillableMapTestUtils.upsertRecords(updatedRecords, records);
+      gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
+      // The record returned for this key should have the updated instantTime
+      assert newCommitTime.contentEquals(gRecord.get(HoodieRecord.COMMIT_TIME_METADATA_FIELD).toString());
     }
-
-    // Get a record from the in-Memory map
-    String key = recordKeys.get(0);
-    HoodieAvroRecord record = (HoodieAvroRecord) records.get(key);
-    List<IndexedRecord> recordsToUpdate = new ArrayList<>();
-    recordsToUpdate.add((IndexedRecord) record.getData().getInsertValue(schema).get());
-
-    String newCommitTime = InProcessTimeGenerator.createNewInstantTime();
-    List<String> keysToBeUpdated = new ArrayList<>();
-    keysToBeUpdated.add(key);
-    // Update the instantTime for this record
-    List<IndexedRecord> updatedRecords =
-        SchemaTestUtil.updateHoodieTestRecords(keysToBeUpdated, recordsToUpdate, newCommitTime);
-    // Upsert this updated record
-    SpillableMapTestUtils.upsertRecords(updatedRecords, records);
-    GenericRecord gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
-    // The record returned for this key should have the updated commitTime
-    assert newCommitTime.contentEquals(gRecord.get(HoodieRecord.COMMIT_TIME_METADATA_FIELD).toString());
-
-    // Get a record from the disk based map
-    key = recordKeys.get(recordKeys.size() - 1);
-    record = (HoodieAvroRecord) records.get(key);
-    recordsToUpdate = new ArrayList<>();
-    recordsToUpdate.add((IndexedRecord) record.getData().getInsertValue(schema).get());
-
-    newCommitTime = InProcessTimeGenerator.createNewInstantTime();
-    keysToBeUpdated = new ArrayList<>();
-    keysToBeUpdated.add(key);
-    // Update the commitTime for this record
-    updatedRecords = SchemaTestUtil.updateHoodieTestRecords(keysToBeUpdated, recordsToUpdate, newCommitTime);
-    // Upsert this updated record
-    SpillableMapTestUtils.upsertRecords(updatedRecords, records);
-    gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
-    // The record returned for this key should have the updated instantTime
-    assert newCommitTime.contentEquals(gRecord.get(HoodieRecord.COMMIT_TIME_METADATA_FIELD).toString());
   }
 
   @ParameterizedTest
@@ -293,65 +299,66 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
 
     Schema schema = SchemaTestUtil.getSimpleSchema();
 
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
         new ExternalSpillableMap<>(16L, basePath, new DefaultSizeEstimator(),
-            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled); // 16B
+            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) { // 16B
 
-    SchemaTestUtil testUtil = new SchemaTestUtil();
-    List<String> recordKeys = new ArrayList<>();
+      SchemaTestUtil testUtil = new SchemaTestUtil();
+      List<String> recordKeys = new ArrayList<>();
+
+      // Ensure we spill to disk
+      while (records.getDiskBasedMapNumEntries() < 1) {
+        List<HoodieRecord> hoodieRecords = testUtil.generateHoodieTestRecordsWithoutHoodieMetadata(0, 100);
+        hoodieRecords.stream().forEach(r -> {
+          records.put(r.getRecordKey(), r);
+          recordKeys.add(r.getRecordKey());
+        });
+      }
 
-    // Ensure we spill to disk
-    while (records.getDiskBasedMapNumEntries() < 1) {
-      List<HoodieRecord> hoodieRecords = testUtil.generateHoodieTestRecordsWithoutHoodieMetadata(0, 100);
-      hoodieRecords.stream().forEach(r -> {
+      // Get a record from the in-Memory map
+      String key = recordKeys.get(0);
+      HoodieRecord record = records.get(key);
+      // Get the field we want to update
+      String fieldName = schema.getFields().stream().filter(field -> field.schema().getType() == Schema.Type.STRING)
+          .findAny().get().name();
+      // Use a new value to update this field
+      String newValue = "update1";
+      List<HoodieRecord> recordsToUpdate = new ArrayList<>();
+      recordsToUpdate.add(record);
+
+      List<HoodieRecord> updatedRecords =
+          SchemaTestUtil.updateHoodieTestRecordsWithoutHoodieMetadata(recordsToUpdate, schema, fieldName, newValue);
+
+      // Upsert this updated record
+      updatedRecords.forEach(r -> {
         records.put(r.getRecordKey(), r);
-        recordKeys.add(r.getRecordKey());
       });
+      GenericRecord gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
+      // The record returned for this key should have the updated value for the field name
+      assertEquals(gRecord.get(fieldName).toString(), newValue);
+
+      // Get a record from the disk based map
+      key = recordKeys.get(recordKeys.size() - 1);
+      record = records.get(key);
+      // Get the field we want to update
+      fieldName = schema.getFields().stream().filter(field -> field.schema().getType() == Schema.Type.STRING).findAny()
+          .get().name();
+      // Use a new value to update this field
+      newValue = "update2";
+      recordsToUpdate = new ArrayList<>();
+      recordsToUpdate.add(record);
+
+      updatedRecords =
+          SchemaTestUtil.updateHoodieTestRecordsWithoutHoodieMetadata(recordsToUpdate, schema, fieldName, newValue);
+
+      // Upsert this updated record
+      updatedRecords.forEach(r -> {
+        records.put(r.getRecordKey(), r);
+      });
+      gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
+      // The record returned for this key should have the updated value for the field name
+      assertEquals(gRecord.get(fieldName).toString(), newValue);
     }
-
-    // Get a record from the in-Memory map
-    String key = recordKeys.get(0);
-    HoodieRecord record = records.get(key);
-    // Get the field we want to update
-    String fieldName = schema.getFields().stream().filter(field -> field.schema().getType() == Schema.Type.STRING)
-        .findAny().get().name();
-    // Use a new value to update this field
-    String newValue = "update1";
-    List<HoodieRecord> recordsToUpdate = new ArrayList<>();
-    recordsToUpdate.add(record);
-
-    List<HoodieRecord> updatedRecords =
-        SchemaTestUtil.updateHoodieTestRecordsWithoutHoodieMetadata(recordsToUpdate, schema, fieldName, newValue);
-
-    // Upsert this updated record
-    updatedRecords.forEach(r -> {
-      records.put(r.getRecordKey(), r);
-    });
-    GenericRecord gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
-    // The record returned for this key should have the updated value for the field name
-    assertEquals(gRecord.get(fieldName).toString(), newValue);
-
-    // Get a record from the disk based map
-    key = recordKeys.get(recordKeys.size() - 1);
-    record = records.get(key);
-    // Get the field we want to update
-    fieldName = schema.getFields().stream().filter(field -> field.schema().getType() == Schema.Type.STRING).findAny()
-        .get().name();
-    // Use a new value to update this field
-    newValue = "update2";
-    recordsToUpdate = new ArrayList<>();
-    recordsToUpdate.add(record);
-
-    updatedRecords =
-        SchemaTestUtil.updateHoodieTestRecordsWithoutHoodieMetadata(recordsToUpdate, schema, fieldName, newValue);
-
-    // Upsert this updated record
-    updatedRecords.forEach(r -> {
-      records.put(r.getRecordKey(), r);
-    });
-    gRecord = (GenericRecord) records.get(key).getData().getInsertValue(schema).get();
-    // The record returned for this key should have the updated value for the field name
-    assertEquals(gRecord.get(fieldName).toString(), newValue);
   }
 
   @Test
@@ -360,29 +367,30 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
     final boolean isCompressionEnabled = false;
     final Schema schema = SchemaTestUtil.getSimpleSchema();
 
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
         new ExternalSpillableMap<>(16L, basePath, new DefaultSizeEstimator(),
-            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled);
+            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) {
 
-    List<String> recordKeys = new ArrayList<>();
-    SchemaTestUtil testUtil = new SchemaTestUtil();
+      List<String> recordKeys = new ArrayList<>();
+      SchemaTestUtil testUtil = new SchemaTestUtil();
 
-    // Put a single record. Payload size estimation happens as part of this initial put.
-    HoodieRecord seedRecord = testUtil.generateHoodieTestRecordsWithoutHoodieMetadata(0, 1).get(0);
-    records.put(seedRecord.getRecordKey(), seedRecord);
-
-    // Remove the key immediately to make the map empty again.
-    records.remove(seedRecord.getRecordKey());
-
-    // Verify payload size re-estimation does not throw exception
-    SchemaTestUtil testUtilx = new SchemaTestUtil();
-    List<HoodieRecord> hoodieRecords = testUtil.generateHoodieTestRecordsWithoutHoodieMetadata(0, 250);
-    hoodieRecords.stream().forEach(hoodieRecord -> {
-      assertDoesNotThrow(() -> {
-        records.put(hoodieRecord.getRecordKey(), hoodieRecord);
-      }, "ExternalSpillableMap put() should not throw exception!");
-      recordKeys.add(hoodieRecord.getRecordKey());
-    });
+      // Put a single record. Payload size estimation happens as part of this initial put.
+      HoodieRecord seedRecord = testUtil.generateHoodieTestRecordsWithoutHoodieMetadata(0, 1).get(0);
+      records.put(seedRecord.getRecordKey(), seedRecord);
+
+      // Remove the key immediately to make the map empty again.
+      records.remove(seedRecord.getRecordKey());
+
+      // Verify payload size re-estimation does not throw exception
+      SchemaTestUtil testUtilx = new SchemaTestUtil();
+      List<HoodieRecord> hoodieRecords = testUtil.generateHoodieTestRecordsWithoutHoodieMetadata(0, 250);
+      hoodieRecords.stream().forEach(hoodieRecord -> {
+        assertDoesNotThrow(() -> {
+          records.put(hoodieRecord.getRecordKey(), hoodieRecord);
+        }, "ExternalSpillableMap put() should not throw exception!");
+        recordKeys.add(hoodieRecord.getRecordKey());
+      });
+    }
   }
 
   @ParameterizedTest
@@ -408,28 +416,48 @@ public class TestExternalSpillableMap extends HoodieCommonTestHarness {
     // Estimate the first record size and calculate the total memory size that the in-memory map can only contain 100 records.
     long estimatedPayloadSize = keyEstimator.sizeEstimate(key) + valEstimator.sizeEstimate(record);
     long totalEstimatedSizeWith100Records = (long) ((estimatedPayloadSize * 100) / 0.8);
-    ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
         new ExternalSpillableMap<>(totalEstimatedSizeWith100Records, basePath, new DefaultSizeEstimator(),
-            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled);
-
-    // Insert 100 records and then in-memory map will contain 100 records.
-    SpillableMapTestUtils.upsertRecords(iRecords, records);
-
-    // Generate one record and it will be spilled to disk
-    List<IndexedRecord> singleRecord = testUtil.generateHoodieTestRecords(0, 1);
-    List<String> singleRecordKey = SpillableMapTestUtils.upsertRecords(singleRecord, records);
-
-    // Get the field we want to update
-    String fieldName = schema.getFields().stream().filter(field -> field.schema().getType() == Schema.Type.STRING).findAny()
-        .get().name();
-    HoodieRecord hoodieRecord = records.get(singleRecordKey.get(0));
-    // Use a new value to update this field, the estimate size of this record will be less than the first record.
-    String newValue = "";
-    HoodieRecord updatedRecord =
-        SchemaTestUtil.updateHoodieTestRecordsWithoutHoodieMetadata(Arrays.asList(hoodieRecord), schema, fieldName, newValue).get(0);
-    records.put(updatedRecord.getRecordKey(), updatedRecord);
-
-    assertEquals(records.size(), 101);
+            new HoodieRecordSizeEstimator(schema), diskMapType, isCompressionEnabled)) {
+
+      // Insert 100 records and then in-memory map will contain 100 records.
+      SpillableMapTestUtils.upsertRecords(iRecords, records);
+
+      // Generate one record and it will be spilled to disk
+      List<IndexedRecord> singleRecord = testUtil.generateHoodieTestRecords(0, 1);
+      List<String> singleRecordKey = SpillableMapTestUtils.upsertRecords(singleRecord, records);
+
+      // Get the field we want to update
+      String fieldName = schema.getFields().stream().filter(field -> field.schema().getType() == Schema.Type.STRING).findAny()
+          .get().name();
+      HoodieRecord hoodieRecord = records.get(singleRecordKey.get(0));
+      // Use a new value to update this field, the estimate size of this record will be less than the first record.
+      String newValue = "";
+      HoodieRecord updatedRecord =
+          SchemaTestUtil.updateHoodieTestRecordsWithoutHoodieMetadata(Arrays.asList(hoodieRecord), schema, fieldName, newValue).get(0);
+      records.put(updatedRecord.getRecordKey(), updatedRecord);
+
+      assertEquals(records.size(), 101);
+    }
+  }
+
+  @ParameterizedTest
+  @EnumSource(value = ExternalSpillableMap.DiskMapType.class)
+  void assertEmptyMapOperations(ExternalSpillableMap.DiskMapType diskMapType) throws IOException {
+    // validate that operations on an empty map work as expected
+    try (ExternalSpillableMap<String, HoodieRecord<? extends HoodieRecordPayload>> records =
+             new ExternalSpillableMap<>(10, basePath, new DefaultSizeEstimator(),
+                 new DefaultSizeEstimator<>(), diskMapType, false)) {
+      assertTrue(records.isEmpty());
+      assertFalse(records.containsKey("key"));
+      assertFalse(records.containsValue("value"));
+      assertTrue(records.keySet().isEmpty());
+      assertTrue(records.values().isEmpty());
+      assertTrue(records.entrySet().isEmpty());
+      assertEquals(0, records.valueStream().count());
+      assertEquals(0, records.size());
+      assertFalse(records.iterator().hasNext());
+    }
   }
 
   private static Stream<Arguments> testArguments() {
