diff --git a/linera-witty-macros/src/unit_tests/wit_load.rs b/linera-witty-macros/src/unit_tests/wit_load.rs
index d4613f50a..fd68525e4 100644
--- a/linera-witty-macros/src/unit_tests/wit_load.rs
+++ b/linera-witty-macros/src/unit_tests/wit_load.rs
@@ -262,3 +262,295 @@ fn enum_type() {
 
     assert_eq!(output.to_string(), expected.to_string());
 }
+
+/// Check the generated code for the body of the implementation of `WitLoad` for a named struct
+/// with some ignored fields.
+#[test]
+fn named_struct_with_skipped_fields() {
+    let input: ItemStruct = parse_quote! {
+        struct Type {
+            #[witty(skip)]
+            ignored1: u8,
+            first: u8,
+            #[witty(skip)]
+            ignored2: i128,
+            #[witty(skip)]
+            ignored3: String,
+            second: CustomType,
+            #[witty(skip)]
+            ignored4: Vec<()>,
+        }
+    };
+    let output = derive_for_struct(&input.fields);
+
+    let expected = quote! {
+        fn load<Instance>(
+            memory: &linera_witty::Memory<'_, Instance>,
+            mut location: linera_witty::GuestPointer,
+        ) -> Result<Self, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let linera_witty::hlist_pat![first, second] =
+                <linera_witty::HList![u8, CustomType] as linera_witty::WitLoad>::load(
+                    memory,
+                    location
+                )?;
+
+            let ignored1 = Default::default();
+            let ignored2 = Default::default();
+            let ignored3 = Default::default();
+            let ignored4 = Default::default();
+
+            Ok(Self {
+                ignored1,
+                first,
+                ignored2,
+                ignored3,
+                second,
+                ignored4
+            })
+        }
+
+        fn lift_from<Instance>(
+            flat_layout: <Self::Layout as linera_witty::Layout>::Flat,
+            memory: &linera_witty::Memory<'_, Instance>,
+        ) -> Result<Self, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let linera_witty::hlist_pat![first, second] =
+                <linera_witty::HList![u8, CustomType] as linera_witty::WitLoad>::lift_from(
+                    flat_layout,
+                    memory
+                )?;
+
+            let ignored1 = Default::default();
+            let ignored2 = Default::default();
+            let ignored3 = Default::default();
+            let ignored4 = Default::default();
+
+            Ok(Self {
+                ignored1,
+                first,
+                ignored2,
+                ignored3,
+                second,
+                ignored4
+            })
+        }
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
+
+/// Check the generated code for the body of the implementation of `WitLoad` for a tuple struct
+/// with some ignored fields.
+#[test]
+fn tuple_struct_with_skipped_fields() {
+    let input: ItemStruct = parse_quote! {
+        struct Type(
+            #[witty(skip)]
+            PhantomData<T>,
+            String,
+            Vec<CustomType>,
+            #[witty(skip)]
+            bool,
+            i64,
+        );
+    };
+    let output = derive_for_struct(&input.fields);
+
+    let expected = quote! {
+        fn load<Instance>(
+            memory: &linera_witty::Memory<'_, Instance>,
+            mut location: linera_witty::GuestPointer,
+        ) -> Result<Self, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let linera_witty::hlist_pat![field1, field2, field4] = <linera_witty::HList![
+                String,
+                Vec<CustomType>,
+                i64
+            ] as linera_witty::WitLoad>::load(memory, location)?;
+
+            let field0 = Default::default();
+            let field3 = Default::default();
+
+            Ok(Self(field0, field1, field2, field3, field4))
+        }
+
+        fn lift_from<Instance>(
+            flat_layout: <Self::Layout as linera_witty::Layout>::Flat,
+            memory: &linera_witty::Memory<'_, Instance>,
+        ) -> Result<Self, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let linera_witty::hlist_pat![field1, field2, field4] = <linera_witty::HList![
+                String,
+                Vec<CustomType>,
+                i64
+            ] as linera_witty::WitLoad>::lift_from(flat_layout, memory)?;
+
+            let field0 = Default::default();
+            let field3 = Default::default();
+
+            Ok(Self(field0, field1, field2, field3, field4))
+        }
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
+
+/// Check the generated code for the body of the implementation of `WitType` for an enum with some
+/// ignored fields.
+#[test]
+fn enum_type_with_skipped_fields() {
+    let input: ItemEnum = parse_quote! {
+        enum Enum {
+            Empty,
+            Tuple(i8, CustomType, #[witty(skip)] u128),
+            Struct {
+                first: (),
+                #[witty(skip)]
+                ignored1: (u8, u16),
+                #[witty(skip)]
+                ignored2: String,
+                second: String,
+                #[witty(skip)]
+                ignored3: Option<String>,
+            },
+        }
+    };
+    let output = derive_for_enum(&input.ident, input.variants.iter());
+
+    let expected = quote! {
+        fn load<Instance>(
+            memory: &linera_witty::Memory<'_, Instance>,
+            mut location: linera_witty::GuestPointer,
+        ) -> Result<Self, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let discriminant = <u8 as linera_witty::WitLoad>::load(memory, location,)?;
+            location = location
+                .after::<u8>()
+                .after_padding_for::<linera_witty::HList![]>()
+                .after_padding_for::<linera_witty::HList![i8, CustomType]>()
+                .after_padding_for::<linera_witty::HList![(), String]>();
+
+            match discriminant {
+                0 => {
+                    let linera_witty::hlist_pat![] =
+                        <linera_witty::HList![] as linera_witty::WitLoad>::load(memory, location)?;
+
+                    Ok(Enum::Empty)
+                }
+                1 => {
+                    let linera_witty::hlist_pat![field0, field1] =
+                        <linera_witty::HList![i8, CustomType] as linera_witty::WitLoad>::load(
+                            memory,
+                            location
+                        )?;
+
+                    let field2 = Default::default();
+
+                    Ok(Enum::Tuple(field0, field1, field2))
+                }
+                2 => {
+                    let linera_witty::hlist_pat![first, second] =
+                        <linera_witty::HList![(), String] as linera_witty::WitLoad>::load(
+                            memory,
+                            location
+                        )?;
+
+                    let ignored1 = Default::default();
+                    let ignored2 = Default::default();
+                    let ignored3 = Default::default();
+
+                    Ok(Enum::Struct {
+                        first,
+                        ignored1,
+                        ignored2,
+                        second,
+                        ignored3
+                    })
+                }
+                _ => Err(linera_witty::RuntimeError::InvalidVariant),
+            }
+        }
+
+        fn lift_from<Instance>(
+            linera_witty::hlist_pat![discriminant_flat_type, ...flat_layout]:
+                <Self::Layout as linera_witty::Layout>::Flat,
+            memory: &linera_witty::Memory<'_, Instance>,
+        ) -> Result<Self, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let discriminant = <u8 as linera_witty::WitLoad>::lift_from(
+                linera_witty::hlist![discriminant_flat_type],
+                memory,
+            )?;
+
+            match discriminant {
+                0 => {
+                    let linera_witty::hlist_pat![] =
+                        <linera_witty::HList![] as linera_witty::WitLoad>::lift_from(
+                            linera_witty::JoinFlatLayouts::from_joined(flat_layout),
+                            memory,
+                        )?;
+
+                    Ok(Enum::Empty)
+                }
+                1 => {
+                    let linera_witty::hlist_pat![field0, field1] =
+                        <linera_witty::HList![i8, CustomType] as linera_witty::WitLoad>::lift_from(
+                            linera_witty::JoinFlatLayouts::from_joined(flat_layout),
+                            memory,
+                        )?;
+
+                    let field2 = Default::default();
+
+                    Ok(Enum::Tuple(field0, field1, field2))
+                }
+                2 => {
+                    let linera_witty::hlist_pat![first, second] =
+                        <linera_witty::HList![(), String] as linera_witty::WitLoad>::lift_from(
+                            linera_witty::JoinFlatLayouts::from_joined(flat_layout),
+                            memory,
+                        )?;
+
+                    let ignored1 = Default::default();
+                    let ignored2 = Default::default();
+                    let ignored3 = Default::default();
+
+                    Ok(Enum::Struct {
+                        first,
+                        ignored1,
+                        ignored2,
+                        second,
+                        ignored3
+                    })
+                }
+                _ => Err(linera_witty::RuntimeError::InvalidVariant),
+            }
+        }
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
diff --git a/linera-witty-macros/src/unit_tests/wit_store.rs b/linera-witty-macros/src/unit_tests/wit_store.rs
index f443067ea..c5e5d03c6 100644
--- a/linera-witty-macros/src/unit_tests/wit_store.rs
+++ b/linera-witty-macros/src/unit_tests/wit_store.rs
@@ -69,7 +69,7 @@ fn named_struct() {
             <Instance::Runtime as linera_witty::Runtime>::Memory:
                 linera_witty::RuntimeMemory<Instance>,
         {
-            let Self { first, second } = self;
+            let Self { first, second, } = self;
             linera_witty::hlist![first, second].store(memory, location)
         }
 
@@ -82,7 +82,7 @@ fn named_struct() {
             <Instance::Runtime as linera_witty::Runtime>::Memory:
                 linera_witty::RuntimeMemory<Instance>,
         {
-            let Self { first, second } = self;
+            let Self { first, second, } = self;
             linera_witty::hlist![first, second].lower(memory)
         }
     };
@@ -177,7 +177,7 @@ fn enum_type() {
 
                     linera_witty::hlist![field0, field1].store(memory, location)
                 }
-                Enum::Struct { first, second } => {
+                Enum::Struct { first, second, } => {
                     2_u8.store(memory, location)?;
                     location = location
                         .after::<u8>()
@@ -218,7 +218,219 @@ fn enum_type() {
 
                     Ok(flat_layout)
                 }
-                Enum::Struct { first, second } => {
+                Enum::Struct { first, second, } => {
+                    let variant_flat_layout = linera_witty::hlist![first, second].lower(memory)?;
+                    let flat_layout: <Self::Layout as linera_witty::Layout>::Flat =
+                        linera_witty::JoinFlatLayouts::into_joined(
+                            2_u8.lower(memory)? + variant_flat_layout,
+                        );
+
+                    Ok(flat_layout)
+                }
+            }
+        }
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
+
+/// Check the generated code for the body of the implementation of `WitStore` for a named struct
+/// with some ignored fields.
+#[test]
+fn named_struct_with_skipped_fields() {
+    let input: ItemStruct = parse_quote! {
+        struct Type {
+            #[witty(skip)]
+            ignored1: u8,
+            first: u8,
+            #[witty(skip)]
+            ignored2: i128,
+            #[witty(skip)]
+            ignored3: String,
+            second: CustomType,
+            #[witty(skip)]
+            ignored4: Vec<()>,
+        }
+    };
+    let output = derive_for_struct(&input.fields);
+
+    let expected = quote! {
+        fn store<Instance>(
+            &self,
+            memory: &mut linera_witty::Memory<'_, Instance>,
+            mut location: linera_witty::GuestPointer,
+        ) -> Result<(), linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let Self { first, second, .. } = self;
+            linera_witty::hlist![first, second].store(memory, location)
+        }
+
+        fn lower<Instance>(
+            &self,
+            memory: &mut linera_witty::Memory<'_, Instance>,
+        ) -> Result<<Self::Layout as linera_witty::Layout>::Flat, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let Self { first, second, .. } = self;
+            linera_witty::hlist![first, second].lower(memory)
+        }
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
+
+/// Check the generated code for the body of the implementation of `WitStore` for a tuple struct
+/// with some ignored fields.
+#[test]
+fn tuple_struct_with_skipped_fields() {
+    let input: ItemStruct = parse_quote! {
+        struct Type(
+            #[witty(skip)]
+            PhantomData<T>,
+            String,
+            Vec<CustomType>,
+            #[witty(skip)]
+            bool,
+            i64,
+        );
+    };
+    let output = derive_for_struct(&input.fields);
+
+    let expected = quote! {
+        fn store<Instance>(
+            &self,
+            memory: &mut linera_witty::Memory<'_, Instance>,
+            mut location: linera_witty::GuestPointer,
+        ) -> Result<(), linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let Self(_, field1, field2, _, field4) = self;
+            linera_witty::hlist![field1, field2, field4].store(memory, location)
+        }
+
+        fn lower<Instance>(
+            &self,
+            memory: &mut linera_witty::Memory<'_, Instance>,
+        ) -> Result<<Self::Layout as linera_witty::Layout>::Flat, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            let Self(_, field1, field2, _, field4) = self;
+            linera_witty::hlist![field1, field2, field4].lower(memory)
+        }
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
+
+/// Check the generated code for the body of the implementation of `WitStore` for a enum with some
+/// ignored fields.
+#[test]
+fn enum_type_with_skipped_fields() {
+    let input: ItemEnum = parse_quote! {
+        enum Enum {
+            Empty,
+            Tuple(i8, CustomType, #[witty(skip)] u128),
+            Struct {
+                first: (),
+                #[witty(skip)]
+                ignored1: (u8, u16),
+                #[witty(skip)]
+                ignored2: String,
+                second: String,
+                #[witty(skip)]
+                ignored3: Option<String>,
+            },
+        }
+    };
+    let output = derive_for_enum(&input.ident, input.variants.iter());
+
+    let expected = quote! {
+        fn store<Instance>(
+            &self,
+            memory: &mut linera_witty::Memory<'_, Instance>,
+            mut location: linera_witty::GuestPointer,
+        ) -> Result<(), linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            match self {
+                Enum::Empty => {
+                    0_u8.store(memory, location)?;
+                    location = location
+                        .after::<u8>()
+                        .after_padding_for::<linera_witty::HList![]>()
+                        .after_padding_for::<linera_witty::HList![i8, CustomType]>()
+                        .after_padding_for::<linera_witty::HList![(), String]>();
+
+                    linera_witty::hlist![].store(memory, location)
+                }
+                Enum::Tuple(field0, field1, _) => {
+                    1_u8.store(memory, location)?;
+                    location = location
+                        .after::<u8>()
+                        .after_padding_for::<linera_witty::HList![]>()
+                        .after_padding_for::<linera_witty::HList![i8, CustomType]>()
+                        .after_padding_for::<linera_witty::HList![(), String]>();
+
+                    linera_witty::hlist![field0, field1].store(memory, location)
+                }
+                Enum::Struct { first, second, .. } => {
+                    2_u8.store(memory, location)?;
+                    location = location
+                        .after::<u8>()
+                        .after_padding_for::<linera_witty::HList![]>()
+                        .after_padding_for::<linera_witty::HList![i8, CustomType]>()
+                        .after_padding_for::<linera_witty::HList![(), String]>();
+
+                    linera_witty::hlist![first, second].store(memory, location)
+                }
+            }
+        }
+
+        fn lower<Instance>(
+            &self,
+            memory: &mut linera_witty::Memory<'_, Instance>,
+        ) -> Result<<Self::Layout as linera_witty::Layout>::Flat, linera_witty::RuntimeError>
+        where
+            Instance: linera_witty::InstanceWithMemory,
+            <Instance::Runtime as linera_witty::Runtime>::Memory:
+                linera_witty::RuntimeMemory<Instance>,
+        {
+            match self {
+                Enum::Empty => {
+                    let variant_flat_layout = linera_witty::hlist![].lower(memory)?;
+                    let flat_layout: <Self::Layout as linera_witty::Layout>::Flat =
+                        linera_witty::JoinFlatLayouts::into_joined(
+                            0_u8.lower(memory)? + variant_flat_layout,
+                        );
+
+                    Ok(flat_layout)
+                }
+                Enum::Tuple(field0, field1, _) => {
+                    let variant_flat_layout = linera_witty::hlist![field0, field1].lower(memory)?;
+                    let flat_layout: <Self::Layout as linera_witty::Layout>::Flat =
+                        linera_witty::JoinFlatLayouts::into_joined(
+                            1_u8.lower(memory)? + variant_flat_layout,
+                        );
+
+                    Ok(flat_layout)
+                }
+                Enum::Struct { first, second, .. } => {
                     let variant_flat_layout = linera_witty::hlist![first, second].lower(memory)?;
                     let flat_layout: <Self::Layout as linera_witty::Layout>::Flat =
                         linera_witty::JoinFlatLayouts::into_joined(
diff --git a/linera-witty-macros/src/unit_tests/wit_type.rs b/linera-witty-macros/src/unit_tests/wit_type.rs
index d984895f6..829710bfa 100644
--- a/linera-witty-macros/src/unit_tests/wit_type.rs
+++ b/linera-witty-macros/src/unit_tests/wit_type.rs
@@ -130,3 +130,135 @@ fn enum_type() {
 
     assert_eq!(output.to_string(), expected.to_string());
 }
+
+/// Check the generated code for the body of the implementation of `WitLoad` for a named struct
+/// with some ignored fields.
+#[test]
+fn named_struct_with_skipped_fields() {
+    let input: ItemStruct = parse_quote! {
+        struct Type {
+            #[witty(skip)]
+            ignored1: u8,
+            first: u8,
+            #[witty(skip)]
+            ignored2: i128,
+            #[witty(skip)]
+            ignored3: String,
+            second: CustomType,
+            #[witty(skip)]
+            ignored4: Vec<()>,
+        }
+    };
+    let output = derive_for_struct(&input.fields);
+
+    let expected = quote! {
+        const SIZE: u32 = <linera_witty::HList![u8, CustomType] as linera_witty::WitType>::SIZE;
+
+        type Layout = <linera_witty::HList![u8, CustomType] as linera_witty::WitType>::Layout;
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
+
+/// Check the generated code for the body of the implementation of `WitLoad` for a tuple struct
+/// with some ignored fields.
+#[test]
+fn tuple_struct_with_skipped_fields() {
+    let input: ItemStruct = parse_quote! {
+        struct Type(
+            #[witty(skip)]
+            PhantomData<T>,
+            String,
+            Vec<CustomType>,
+            #[witty(skip)]
+            bool,
+            i64,
+        );
+    };
+    let output = derive_for_struct(&input.fields);
+
+    let expected = quote! {
+        const SIZE: u32 =
+            <linera_witty::HList![String, Vec<CustomType>, i64] as linera_witty::WitType>::SIZE;
+
+        type Layout =
+            <linera_witty::HList![String, Vec<CustomType>, i64] as linera_witty::WitType>::Layout;
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
+
+/// Check the generated code for the body of the implementation of `WitType` for an enum
+/// with some ignored fields.
+#[test]
+fn enum_type_with_skipped_fields() {
+    let input: ItemEnum = parse_quote! {
+        enum Enum {
+            Empty,
+            Tuple(i8, CustomType, #[witty(skip)] u128),
+            Struct {
+                first: (),
+                #[witty(skip)]
+                ignored1: (u8, u16),
+                #[witty(skip)]
+                ignored2: String,
+                second: String,
+                #[witty(skip)]
+                ignored3: Option<String>,
+            },
+        }
+    };
+    let output = derive_for_enum(&input.ident, input.variants.iter());
+
+    let expected = quote! {
+        const SIZE: u32 = {
+            let discriminant_size = std::mem::size_of::<u8>() as u32;
+            let mut size = discriminant_size;
+            let mut variants_alignment = <
+                < <linera_witty::HList![] as linera_witty::WitType>::Layout as linera_witty::Merge<
+                    < <linera_witty::HList![i8, CustomType] as linera_witty::WitType>::Layout
+                    as linera_witty::Merge<
+                        <linera_witty::HList![(), String] as linera_witty::WitType>::Layout
+                    >>::Output
+                >>::Output
+            as linera_witty::Layout>::ALIGNMENT;
+            let padding = (-(size as i32) & (variants_alignment as i32 - 1)) as u32;
+
+            let variant_size = discriminant_size
+                + padding
+                + <linera_witty::HList![] as linera_witty::WitType>::SIZE;
+
+            if variant_size > size {
+                size = variant_size;
+            }
+
+            let variant_size = discriminant_size
+                + padding
+                + <linera_witty::HList![i8, CustomType] as linera_witty::WitType>::SIZE;
+
+            if variant_size > size {
+                size = variant_size;
+            }
+
+            let variant_size = discriminant_size
+                + padding
+                + <linera_witty::HList![(), String] as linera_witty::WitType>::SIZE;
+
+            if variant_size > size {
+                size = variant_size;
+            }
+
+            size
+        };
+
+        type Layout = linera_witty::HCons<u8,
+            < <linera_witty::HList![] as linera_witty::WitType>::Layout as linera_witty::Merge<
+                < <linera_witty::HList![i8, CustomType] as linera_witty::WitType>::Layout
+                as linera_witty::Merge<
+                    <linera_witty::HList![(), String] as linera_witty::WitType>::Layout
+                >>::Output
+            >>::Output>;
+    };
+
+    assert_eq!(output.to_string(), expected.to_string());
+}
