diff --git a/detox/src/client/Client.test.js b/detox/src/client/Client.test.js
index 7bf32c681..f8ddc0b36 100644
--- a/detox/src/client/Client.test.js
+++ b/detox/src/client/Client.test.js
@@ -1,24 +1,25 @@
 const tempfile = require('tempfile');
 const actions = require('./actions/actions');
-const config = require('../configuration/configurations.mock').validOneDeviceAndSession.session;
 const invoke = require('../invoke');
 const sleep = require('../utils/sleep');
+const { validOneDeviceAndSession } = require('../configuration/configurations.mock');
 
 describe('Client', () => {
-  let argparse;
   let WebSocket;
   let Client;
   let client;
   let bunyan;
   let log;
+  let sessionConfig;
 
   beforeEach(() => {
+    sessionConfig = {
+      ...validOneDeviceAndSession.session,
+    };
+
     jest.mock('../utils/logger');
     WebSocket = jest.mock('./AsyncWebSocket');
 
-    jest.mock('../utils/argparse');
-    argparse = require('../utils/argparse');
-
     jest.mock('bunyan')
     bunyan = require('bunyan');
 
@@ -169,7 +170,7 @@ describe('Client', () => {
   });
 
   it(`cleanup() - if not connected should do nothing`, async () => {
-    client = new Client(config);
+    client = new Client(sessionConfig);
     client.ws.send.mockReturnValueOnce(response("cleanupDone", {}, 1));
     await client.cleanup();
 
@@ -215,7 +216,7 @@ describe('Client', () => {
     });
 
     it(`execute() - fast invocation should not trigger "slowInvocationStatus"`, async () => {
-      argparse.getArgValue.mockReturnValue(2); // set debug-slow-invocations
+      sessionConfig.debugSynchronization = 2;
       await connect();
       await executeWithSlowInvocation(1);
       expect(client.ws.send).toHaveBeenLastCalledWith({"params": {"args": ["test"], "method": "matcherForAccessibilityLabel:", "target": {"type": "Class", "value": "GREYMatchers"}}, "type": "invoke"}, undefined);
@@ -223,7 +224,7 @@ describe('Client', () => {
     });
 
     it(`execute() - slow invocation should trigger "slowInvocationStatus:`, async () => {
-      argparse.getArgValue.mockReturnValue(2); // set debug-slow-invocations
+      sessionConfig.debugSynchronization = 2;
       await connect();
       await executeWithSlowInvocation(3);
       expect(client.ws.send).toHaveBeenLastCalledWith({"params": {}, "type": "currentStatus"}, undefined);
@@ -231,7 +232,7 @@ describe('Client', () => {
     });
 
     it(`execute() - slow invocation should do nothing if ws was closed`, async () => {
-      argparse.getArgValue.mockReturnValue(2); // set debug-slow-invocations
+      sessionConfig.debugSynchronization = 2;
       await connect();
       client.ws.isOpen.mockReturnValue(false);
       await executeWithSlowInvocation(4);
@@ -242,7 +243,7 @@ describe('Client', () => {
     // NOTE: this test prevents geometrical progression of currentStatus calls
     // when the device gets really busy. Otherwise, we can get 1000 pending currentStatus calls.
     it(`slowInvocationStatus() - should not schedule currentStatus thrice`, async () => {
-      argparse.getArgValue.mockReturnValue(2); // set debug-slow-invocations
+      sessionConfig.debugSynchronization = 2;
       await connect();
 
       jest.spyOn(client, 'slowInvocationStatus');
@@ -440,7 +441,7 @@ describe('Client', () => {
   });
 
   async function connect() {
-    client = new Client(config);
+    client = new Client(sessionConfig);
     client.ws.send.mockReturnValueOnce(response("loginSuccess", {}, 1));
     await client.connect();
     client.ws.isOpen.mockReturnValue(true);
diff --git a/detox/src/configuration/composeSessionConfig.test.js b/detox/src/configuration/composeSessionConfig.test.js
index b440051a3..eff14c342 100644
--- a/detox/src/configuration/composeSessionConfig.test.js
+++ b/detox/src/configuration/composeSessionConfig.test.js
@@ -2,70 +2,201 @@ const DetoxConfigErrorBuilder = require('../errors/DetoxConfigErrorBuilder');
 
 describe('composeSessionConfig', () => {
   let composeSessionConfig;
-  let detoxConfig, deviceConfig;
+  let cliConfig, detoxConfig, deviceConfig;
   /** @type {DetoxConfigErrorBuilder} */
   let errorBuilder;
 
   beforeEach(() => {
     composeSessionConfig = require('./composeSessionConfig');
     errorBuilder = new DetoxConfigErrorBuilder();
+    cliConfig = {};
     detoxConfig = {};
     deviceConfig = {};
   });
 
   const compose = () => composeSessionConfig({
+    cliConfig,
     detoxConfig,
     deviceConfig,
     errorBuilder,
   });
 
   it('should generate a default config', async () => {
-    const sessionConfig = await compose();
-
-    expect(sessionConfig).toMatchObject({
+    expect(await compose()).toEqual({
       autoStart: true,
-      server: expect.stringMatching(/^ws:.*localhost:/),
-      sessionId: expect.stringMatching(/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i),
+      debugSynchronization: false,
+      server: expect.any(String),
+      sessionId: expect.any(String),
+    });
+  });
+
+  describe('sessionId', function() {
+    describe('by default', () => {
+      it('should be autogenerated GUID', async () => {
+        expect(await compose()).toMatchObject({
+          sessionId: expect.stringMatching(/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i),
+        });
+      });
+    });
+
+    it('should pass validations', async () => {
+      detoxConfig.session = { sessionId: 1234 };
+      await expect(compose()).rejects.toThrowError(errorBuilder.invalidSessionIdProperty());
+
+      detoxConfig.session = { sessionId: '' };
+      await expect(compose()).rejects.toThrowError(errorBuilder.invalidSessionIdProperty());
+    });
+
+    describe('when defined in global config', () => {
+      beforeEach(() => {
+        detoxConfig.session = { sessionId: 'someSessionId' };
+      });
+
+      it('should use the specified value', async () => {
+        expect((await compose()).sessionId).toBe('someSessionId');
+      });
+
+      describe('and in device config', () => {
+        beforeEach(() => {
+          deviceConfig.session = { sessionId: 'otherSessionId' };
+        });
+
+        it('should use the specified value', async () => {
+          expect((await compose()).sessionId).toBe('otherSessionId');
+        });
+      });
+    });
+  });
+
+  describe('server', function() {
+    describe('by default', () => {
+      it('should be autogenerated', async () => {
+        expect(await compose()).toMatchObject({
+          server: expect.stringMatching(/^ws:.*localhost:/),
+        });
+      });
+    });
+
+    it('should pass validations', async () => {
+      detoxConfig.session = { server: 1234 };
+      await expect(compose()).rejects.toThrowError(errorBuilder.invalidServerProperty());
+
+      detoxConfig.session = { server: 'http://invalid-protocol.com' };
+      await expect(compose()).rejects.toThrowError(errorBuilder.invalidServerProperty());
+    });
+
+    describe('when defined in global config', () => {
+      beforeEach(() => {
+        detoxConfig.session = { server: 'ws://myserver:1100' };
+      });
+
+      it('should use the specified value', async () => {
+        expect(await compose()).toMatchObject({
+          server: 'ws://myserver:1100',
+        });
+      });
+
+      describe('and in device config', () => {
+        beforeEach(() => {
+          deviceConfig.session = { server: 'ws://otherserver:1100' };
+        });
+
+        it('should use the specified value', async () => {
+          expect(await compose()).toMatchObject({
+            server: 'ws://otherserver:1100',
+          });
+        });
+      });
     });
   });
 
-  describe('if detoxConfig.session is defined', function() {
-    beforeEach(() => {
-      detoxConfig.session = {
-        server: 'ws://localhost:9999',
-        sessionId: 'someSessionId',
-      };
-    })
+  describe('autoStart', function() {
+    describe('by default', () => {
+      it('should be true', async () => {
+        expect(await compose()).toMatchObject({ autoStart: true });
+      });
+    });
+
+    describe('when autoStart is explicitly false', function() {
+      beforeEach(() => {
+        detoxConfig.session = { autoStart: false };
+      });
+
+      it('should override the value', async () => {
+        expect(await compose()).toMatchObject({ autoStart: false });
+      });
+    });
+
+    describe('when server is defined', () => {
+      beforeEach(() => {
+        detoxConfig.session = { server: 'ws://localhost:1100' };
+      });
+
+      it('should be false', async () => {
+        expect(await compose()).toMatchObject({ autoStart: false });
+      });
 
-    it('should return detoxConfig.session', async () => {
-      expect(await compose()).toEqual({
-        server: 'ws://localhost:9999',
-        sessionId: 'someSessionId',
+      describe('when autoStart is explicitly true', function() {
+        beforeEach(() => {
+          detoxConfig.session.autoStart = true;
+        });
+
+        it('should override the value', async () => {
+          expect(await compose()).toMatchObject({ autoStart: true });
+        });
       });
     });
+  });
 
-    test(`providing empty server config should throw`, () => {
-      delete detoxConfig.session.server;
-      expect(compose()).rejects.toThrowError(errorBuilder.missingServerProperty());
+  describe('debugSynchronization', function () {
+    describe('by default', () => {
+      it('should be false', async () => {
+        expect(await compose()).toMatchObject({
+          debugSynchronization: false,
+        });
+      });
     });
 
-    test(`providing server config with no session should throw`, () => {
-      delete detoxConfig.session.sessionId;
-      expect(compose()).rejects.toThrowError(errorBuilder.missingSessionIdProperty());
+    it('should pass validations', async () => {
+      detoxConfig.session = { debugSynchronization: -1 };
+      await expect(compose()).rejects.toThrowError(errorBuilder.invalidDebugSynchronizationProperty());
+
+      detoxConfig.session = { debugSynchronization: '3000' };
+      await expect(compose()).rejects.toThrowError(errorBuilder.invalidDebugSynchronizationProperty());
     });
 
-    describe('if deviceConfig.session is defined', function() {
+    describe('when defined in global config', () => {
       beforeEach(() => {
-        detoxConfig.session = {
-          server: 'ws://localhost:1111',
-          sessionId: 'anotherSession',
-        };
+        detoxConfig.session = { debugSynchronization: 10000 };
       });
 
-      it('should return deviceConfig.session instead of detoxConfig.session', async () => {
-        expect(await compose()).toEqual({
-          server: 'ws://localhost:1111',
-          sessionId: 'anotherSession',
+      it('should use that value', async () => {
+        expect(await compose()).toMatchObject({
+          debugSynchronization: 10000,
+        });
+      });
+
+      describe('and in device config', () => {
+        beforeEach(() => {
+          deviceConfig.session = { debugSynchronization: 20000 };
+        });
+
+        it('should use that value', async () => {
+          expect(await compose()).toMatchObject({
+            debugSynchronization: 20000,
+          });
+        });
+
+        describe('and in CLI config', () => {
+          beforeEach(() => {
+            cliConfig.debugSynchronization = 3000;
+          });
+
+          it('should use that value', async () => {
+            expect(await compose()).toMatchObject({
+              debugSynchronization: 3000,
+            });
+          });
         });
       });
     });
diff --git a/detox/src/errors/DetoxConfigErrorBuilder.test.js b/detox/src/errors/DetoxConfigErrorBuilder.test.js
index c6ad942e6..d8daa9530 100644
--- a/detox/src/errors/DetoxConfigErrorBuilder.test.js
+++ b/detox/src/errors/DetoxConfigErrorBuilder.test.js
@@ -257,13 +257,13 @@ describe('DetoxConfigErrorBuilder', () => {
     });
   });
 
-  describe('.missingServerProperty', () => {
+  describe('.invalidServerProperty', () => {
     beforeEach(() => {
-      build = () => builder.missingServerProperty();
+      build = () => builder.invalidServerProperty();
       builder.setConfigurationName('android.release');
       builder.setDetoxConfig({
         session: {
-          sessionId: '234589798234',
+          server: 'localhost',
         },
         configurations: {
           'android.release': {
@@ -286,9 +286,9 @@ describe('DetoxConfigErrorBuilder', () => {
     });
   });
 
-  describe('.missingSessionId', () => {
+  describe('.invalidSessionId', () => {
     beforeEach(() => {
-      build = () => builder.missingSessionIdProperty();
+      build = () => builder.invalidSessionIdProperty();
       builder.setConfigurationName('android.release');
       builder.setDetoxConfig({
         configurations: {
@@ -298,7 +298,48 @@ describe('DetoxConfigErrorBuilder', () => {
               avdName: 'Pixel_2_API_29',
             },
             session: {
-              server: 'ws://localhost:12837',
+              sessionId: 234589798234,
+            },
+          }
+        }
+      });
+    });
+
+    it('should create a generic error, if the config location is not known', () => {
+      expect(build()).toMatchSnapshot();
+    });
+
+    it('should create an error with a hint, if the config location is known', () => {
+      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
+      expect(build()).toMatchSnapshot();
+    });
+
+    it('should point to global session if there is one', () => {
+      builder.setDetoxConfig({
+        session: {
+          server: 'ws://localhost:12837',
+        },
+        configurations: {},
+      })
+
+      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
+      expect(build()).toMatchSnapshot();
+    });
+  });
+
+  describe('.invalidDebugSynchronizationProperty', () => {
+    beforeEach(() => {
+      build = () => builder.invalidDebugSynchronizationProperty();
+      builder.setConfigurationName('android.release');
+      builder.setDetoxConfig({
+        configurations: {
+          'android.release': {
+            type: 'android.emulator',
+            device: {
+              avdName: 'Pixel_2_API_29',
+            },
+            session: {
+              debugSynchronization: '3000',
             },
           }
         }
diff --git a/detox/src/utils/isValidWebsocketURL.test.js b/detox/src/utils/isValidWebsocketURL.test.js
new file mode 100644
index 000000000..6e545246d
--- /dev/null
+++ b/detox/src/utils/isValidWebsocketURL.test.js
@@ -0,0 +1,14 @@
+const isValidWebsocketURL = require('./isValidWebsocketURL');
+
+describe('isValidWebsocketURL', () => {
+  it.each([
+    [true, 'ws://localhost'],
+    [true, 'wss://localhost:3428'],
+    [false, 'http://localhost:3428'],
+    [false, '3428'],
+    [false, 3428],
+    [false, null],
+  ])('should return %j for %s', (expected, url) => {
+    expect(isValidWebsocketURL(url)).toBe(expected);
+  });
+});
diff --git a/detox/test/package.json b/detox/test/package.json
index 5527640f6..1135d9d72 100644
--- a/detox/test/package.json
+++ b/detox/test/package.json
@@ -11,7 +11,7 @@
     "start": "react-native start",
     "packager": "react-native start",
     "detox-server": "detox run-server",
-    "e2e:ios": "detox test -c ios.sim.release --debug-synchronization",
+    "e2e:ios": "detox test -c ios.sim.release",
     "e2e:android": "detox test -c android.emu.release",
     "e2e:android-debug": "detox test -c android.emu.debug",
     "e2e:ios-ci": "npm run e2e:ios -- --workers 3 --retries 1",
@@ -93,6 +93,9 @@
         "type": "ios.simulator",
         "device": {
           "type": "iPhone 11 Pro"
+        },
+        "session": {
+          "debugSynchronization": 3000
         }
       },
       "ios.none": {
