diff --git a/tests/_test_base.bash b/tests/_test_base.bash
index 1581905..80d80a5 100644
--- a/tests/_test_base.bash
+++ b/tests/_test_base.bash
@@ -34,9 +34,9 @@ GPGTEST="$SECRETS_GPG_COMMAND --homedir=$TEST_GPG_HOMEDIR --no-permission-warnin
 # Personal data:
 
 # user3 created with '--quick-key-generate' and has only an email, no username.
-TEST_DEFAULT_USER="user3"
-TEST_SECOND_USER="user2"
-TEST_ATTACKER_USER="attacker1"
+TEST_DEFAULT_USER="user3@gitsecret.io"
+TEST_SECOND_USER="user2@gitsecret.io"
+TEST_ATTACKER_USER="attacker1@gitsecret.io"
 
 #TEST_DEFAULT_FILENAME="file_one"  # no spaces
 #TEST_SECOND_FILENAME="file_two"  # no spaces
@@ -48,16 +48,12 @@ TEST_THIRD_FILENAME="space file three"  # has spaces
 
 
 function test_user_password {
-  # It was set on key creation:
-  echo "${1}pass"
+  # Password for 'user3@gitsecret.io' is 'user3pass'
+  # As it was set on key creation. 
+  echo "$1" | sed -e 's/@.*/pass/' 
 }
 
 
-function test_user_email {
-  # It was set on key creation:
-  echo "${1}@gitsecret.io"
-}
-
 
 # GPG:
 
@@ -106,7 +102,7 @@ function install_fixture_full_key {
   local fp
   local fingerprint
 
-  email=$(test_user_email "$1")
+  email="$1"
 
   \cp "$FIXTURES_DIR/gpg/${1}/private.key" "$private_key"
 
@@ -127,14 +123,14 @@ function install_fixture_full_key {
 function uninstall_fixture_key {
   local email
 
-  email=$(test_user_email "$1")
+  email="$1"
   $GPGTEST --yes --delete-key "$email" > /dev/null 2>&1
 }
 
 
 function uninstall_fixture_full_key {
   local email
-  email=$(test_user_email "$1")
+  email="$1"
 
   local fingerprint="$2"
   if [[ -z "$fingerprint" ]]; then
@@ -203,7 +199,7 @@ function set_state_secret_init {
 function set_state_secret_tell {
   local email
 
-  email=$(test_user_email "$1")
+  email="$1"
   git secret tell -d "$TEST_GPG_HOMEDIR" "$email" > /dev/null 2>&1
 }
 
diff --git a/tests/fixtures/gpg/attacker1/private.key b/tests/fixtures/gpg/attacker1@gitsecret.io/private.key
similarity index 100%
rename from tests/fixtures/gpg/attacker1/private.key
rename to tests/fixtures/gpg/attacker1@gitsecret.io/private.key
diff --git a/tests/fixtures/gpg/attacker1/public.key b/tests/fixtures/gpg/attacker1@gitsecret.io/public.key
similarity index 100%
rename from tests/fixtures/gpg/attacker1/public.key
rename to tests/fixtures/gpg/attacker1@gitsecret.io/public.key
diff --git a/tests/fixtures/gpg/user1/private.key b/tests/fixtures/gpg/user1@gitsecret.io/private.key
similarity index 100%
rename from tests/fixtures/gpg/user1/private.key
rename to tests/fixtures/gpg/user1@gitsecret.io/private.key
diff --git a/tests/fixtures/gpg/user1/public.key b/tests/fixtures/gpg/user1@gitsecret.io/public.key
similarity index 100%
rename from tests/fixtures/gpg/user1/public.key
rename to tests/fixtures/gpg/user1@gitsecret.io/public.key
diff --git a/tests/fixtures/gpg/user2/private.key b/tests/fixtures/gpg/user2@gitsecret.io/private.key
similarity index 100%
rename from tests/fixtures/gpg/user2/private.key
rename to tests/fixtures/gpg/user2@gitsecret.io/private.key
diff --git a/tests/fixtures/gpg/user2/public.key b/tests/fixtures/gpg/user2@gitsecret.io/public.key
similarity index 100%
rename from tests/fixtures/gpg/user2/public.key
rename to tests/fixtures/gpg/user2@gitsecret.io/public.key
diff --git a/tests/fixtures/gpg/user3/private.key b/tests/fixtures/gpg/user3@gitsecret.io/private.key
similarity index 100%
rename from tests/fixtures/gpg/user3/private.key
rename to tests/fixtures/gpg/user3@gitsecret.io/private.key
diff --git a/tests/fixtures/gpg/user3/public.key b/tests/fixtures/gpg/user3@gitsecret.io/public.key
similarity index 100%
rename from tests/fixtures/gpg/user3/public.key
rename to tests/fixtures/gpg/user3@gitsecret.io/public.key
diff --git a/tests/test_killperson.bats b/tests/test_killperson.bats
index db523b3..65c040a 100644
--- a/tests/test_killperson.bats
+++ b/tests/test_killperson.bats
@@ -25,21 +25,25 @@ function teardown {
 }
 
 
-@test "run 'killperson' with key name" {
-  run git secret killperson "$TEST_DEFAULT_USER"
+@test "run 'killperson' with short name" {
+  local name
+  name=$(echo "$TEST_DEFAULT_USER" | sed -e 's/@.*//')
+
+  # killperson must use full email, not short name
+  run git secret killperson "$name"
+  [ "$status" -eq 1 ]
+
+  # Then whoknows will be ok because user3@gitsecret.io still knows
+  run git secret whoknows
   [ "$status" -eq 0 ]
 
   # Testing output:
   [[ "$output" == *"$TEST_DEFAULT_USER"* ]]
-
-  # Then whoknows must return an error with status code 1:
-  run git secret whoknows
-  [ "$status" -eq 1 ]
 }
 
 
 @test "run 'killperson' with email" {
-  local email=$(test_user_email "$TEST_DEFAULT_USER")
+  local email="$TEST_DEFAULT_USER"
 
   run git secret killperson "$email"
   [ "$status" -eq 0 ]
@@ -58,8 +62,8 @@ function teardown {
   install_fixture_key "$TEST_SECOND_USER"
   set_state_secret_tell "$TEST_SECOND_USER"
 
-  local default_email=$(test_user_email "$TEST_DEFAULT_USER")
-  local second_email=$(test_user_email "$TEST_SECOND_USER")
+  local default_email="$TEST_DEFAULT_USER"
+  local second_email="$TEST_SECOND_USER"
 
   run git secret killperson "$default_email" "$second_email"
   [ "$status" -eq 0 ]
diff --git a/tests/test_tell.bats b/tests/test_tell.bats
index af4e765..b9f8168 100644
--- a/tests/test_tell.bats
+++ b/tests/test_tell.bats
@@ -17,6 +17,17 @@ function teardown {
   unset_current_state
 }
 
+@test "run 'tell' on substring of emails" {
+  run git secret tell -d "$TEST_GPG_HOMEDIR" user
+  # this should give an error because there is no user named 'user', 
+  # even though there are users with the substring 'user'.
+  # See issue https://github.com/sobolevn/git-secret/issues/176 
+  [ "$status" -eq 1 ]
+
+  run git secret whoknows 
+  [ "$status" -eq 1 ]   # should error when there are no users told
+  
+}
 
 @test "fail on no users" {
   run _user_required
@@ -85,7 +96,7 @@ function teardown {
 
 
 @test "run 'tell' with '-m'" {
-  email=$(test_user_email $TEST_DEFAULT_USER)
+  local email="$TEST_DEFAULT_USER"
 
   git_set_config_email "$email"
   run git secret tell -d "$TEST_GPG_HOMEDIR" -m
diff --git a/tests/test_whoknows.bats b/tests/test_whoknows.bats
index 72747a1..0e5d500 100644
--- a/tests/test_whoknows.bats
+++ b/tests/test_whoknows.bats
@@ -61,8 +61,8 @@ function teardown {
 
 @test "run 'whoknows' without any users" {
   # Preparations, removing users:
-  local email1=$(test_user_email "$TEST_DEFAULT_USER")
-  local email2=$(test_user_email "$TEST_SECOND_USER")
+  local email1="$TEST_DEFAULT_USER"
+  local email2="$TEST_SECOND_USER"
   git secret killperson "$email1" "$email2"
 
   # Now whoknows should raise an error: there are no users.
