diff --git a/crates/loro-internal/tests/test.rs b/crates/loro-internal/tests/test.rs
index 0b8eca53..a9d3c23f 100644
--- a/crates/loro-internal/tests/test.rs
+++ b/crates/loro-internal/tests/test.rs
@@ -79,7 +79,7 @@ fn mark_with_the_same_key_value_should_be_skipped() {
 fn event_from_checkout() {
     let a = LoroDoc::new_auto_commit();
     let sub_id = a.subscribe_root(Arc::new(|event| {
-        assert!(!event.doc.from_checkout);
+        assert!(!event.event_meta.from_checkout);
     }));
     a.get_text("text").insert(0, "hello").unwrap();
     a.commit_then_renew();
@@ -90,7 +90,7 @@ fn event_from_checkout() {
     let ran = Arc::new(AtomicBool::new(false));
     let ran_cloned = ran.clone();
     a.subscribe_root(Arc::new(move |event| {
-        assert!(event.doc.from_checkout);
+        assert!(event.event_meta.from_checkout);
         ran.store(true, std::sync::atomic::Ordering::Relaxed);
     }));
     a.checkout(&version).unwrap();
@@ -101,8 +101,7 @@ fn event_from_checkout() {
 fn handler_in_event() {
     let doc = LoroDoc::new_auto_commit();
     doc.subscribe_root(Arc::new(|e| {
-        let value = e
-            .container
+        let value = e.events[0]
             .diff
             .as_list()
             .unwrap()
@@ -190,7 +189,7 @@ fn richtext_mark_event() {
     a.subscribe(
         &a.get_text("text").id(),
         Arc::new(|e| {
-            let delta = e.container.diff.as_text().unwrap();
+            let delta = e.events[0].diff.as_text().unwrap();
             assert_eq!(
                 delta.to_json_value(),
                 json!([
@@ -211,7 +210,7 @@ fn richtext_mark_event() {
     b.subscribe(
         &a.get_text("text").id(),
         Arc::new(|e| {
-            let delta = e.container.diff.as_text().unwrap();
+            let delta = e.events[0].diff.as_text().unwrap();
             assert_eq!(
                 delta.to_json_value(),
                 json!([
@@ -239,7 +238,7 @@ fn concurrent_richtext_mark_event() {
     let sub_id = a.subscribe(
         &a.get_text("text").id(),
         Arc::new(|e| {
-            let delta = e.container.diff.as_text().unwrap();
+            let delta = e.events[0].diff.as_text().unwrap();
             assert_eq!(
                 delta.to_json_value(),
                 json!([
@@ -257,7 +256,7 @@ fn concurrent_richtext_mark_event() {
     let sub_id = a.subscribe(
         &a.get_text("text").id(),
         Arc::new(|e| {
-            let delta = e.container.diff.as_text().unwrap();
+            let delta = e.events[0].diff.as_text().unwrap();
             assert_eq!(
                 delta.to_json_value(),
                 json!([
@@ -281,19 +280,21 @@ fn concurrent_richtext_mark_event() {
     a.subscribe(
         &a.get_text("text").id(),
         Arc::new(|e| {
-            let delta = e.container.diff.as_text().unwrap();
-            assert_eq!(
-                delta.to_json_value(),
-                json!([
-                    {
-                        "retain": 2,
-                    },
-                    {
-                        "insert": "A",
-                        "attributes": {"bold": true, "link": true}
-                    }
-                ])
-            )
+            for container_diff in e.events {
+                let delta = container_diff.diff.as_text().unwrap();
+                assert_eq!(
+                    delta.to_json_value(),
+                    json!([
+                        {
+                            "retain": 2,
+                        },
+                        {
+                            "insert": "A",
+                            "attributes": {"bold": true, "link": true}
+                        }
+                    ])
+                )
+            }
         }),
     );
     a.get_text("text").insert(2, "A").unwrap();
@@ -310,7 +311,7 @@ fn insert_richtext_event() {
     a.subscribe(
         &text.id(),
         Arc::new(|e| {
-            let delta = e.container.diff.as_text().unwrap();
+            let delta = e.events[0].diff.as_text().unwrap();
             assert_eq!(
                 delta.to_json_value(),
                 json!([
@@ -331,7 +332,7 @@ fn import_after_init_handlers() {
         &ContainerID::new_root("text", ContainerType::Text),
         Arc::new(|event| {
             assert!(matches!(
-                event.container.diff,
+                event.events[0].diff,
                 loro_internal::event::Diff::Text(_)
             ))
         }),
@@ -340,7 +341,7 @@ fn import_after_init_handlers() {
         &ContainerID::new_root("map", ContainerType::Map),
         Arc::new(|event| {
             assert!(matches!(
-                event.container.diff,
+                event.events[0].diff,
                 loro_internal::event::Diff::Map(_)
             ))
         }),
@@ -349,7 +350,7 @@ fn import_after_init_handlers() {
         &ContainerID::new_root("list", ContainerType::List),
         Arc::new(|event| {
             assert!(matches!(
-                event.container.diff,
+                event.events[0].diff,
                 loro_internal::event::Diff::List(_)
             ))
         }),
@@ -417,10 +418,12 @@ fn test_checkout() {
     doc_0.subscribe_root(Arc::new(move |event| {
         dbg!(&event);
         let mut root_value = root_value.lock().unwrap();
-        root_value.apply(
-            &event.container.path.iter().map(|x| x.1.clone()).collect(),
-            &[event.container.diff.clone()],
-        );
+        for container_diff in event.events {
+            root_value.apply(
+                &container_diff.path.iter().map(|x| x.1.clone()).collect(),
+                &[container_diff.diff.clone()],
+            );
+        }
     }));
 
     let map = doc_0.get_map("map");
diff --git a/crates/loro/tests/loro_rust_test.rs b/crates/loro/tests/loro_rust_test.rs
index 12824a75..5e505f8e 100644
--- a/crates/loro/tests/loro_rust_test.rs
+++ b/crates/loro/tests/loro_rust_test.rs
@@ -78,13 +78,15 @@ fn time_travel() {
     doc.subscribe(
         &text.id(),
         Arc::new(move |x| {
-            let Some(text) = x.container.diff.as_text() else {
-                return;
-            };
+            for event in x.events {
+                let Some(text) = event.diff.as_text() else {
+                    continue;
+                };
 
-            let delta: Vec<TextDelta> = text.iter().map(|x| x.into()).collect();
-            dbg!(&delta);
-            text2.apply_delta(&delta).unwrap();
+                let delta: Vec<TextDelta> = text.iter().map(|x| x.into()).collect();
+                dbg!(&delta);
+                text2.apply_delta(&delta).unwrap();
+            }
         }),
     );
 
@@ -112,13 +114,15 @@ fn travel_back_should_remove_styles() {
     doc.subscribe(
         &text.id(),
         Arc::new(move |x| {
-            let Some(text) = x.container.diff.as_text() else {
-                return;
-            };
+            for event in x.events {
+                let Some(text) = event.diff.as_text() else {
+                    continue;
+                };
 
-            let delta: Vec<TextDelta> = text.iter().map(|x| x.into()).collect();
-            // dbg!(&delta);
-            text2.apply_delta(&delta).unwrap();
+                let delta: Vec<TextDelta> = text.iter().map(|x| x.into()).collect();
+                // dbg!(&delta);
+                text2.apply_delta(&delta).unwrap();
+            }
         }),
     );
 
@@ -304,15 +308,17 @@ fn subscribe() {
     doc.subscribe(
         &text.id(),
         Arc::new(move |event: DiffEvent| {
-            assert!(event.doc.local);
-            let event = event.container.diff.as_text().unwrap();
-            let delta: Vec<_> = event.iter().cloned().collect();
-            let d = DeltaItem::Insert {
-                insert: "123".into(),
-                attributes: Default::default(),
-            };
-            assert_eq!(delta, vec![d]);
-            ran2.store(true, std::sync::atomic::Ordering::Relaxed);
+            assert!(event.event_meta.local);
+            for event in event.events {
+                let event = event.diff.as_text().unwrap();
+                let delta: Vec<_> = event.iter().cloned().collect();
+                let d = DeltaItem::Insert {
+                    insert: "123".into(),
+                    attributes: Default::default(),
+                };
+                assert_eq!(delta, vec![d]);
+                ran2.store(true, std::sync::atomic::Ordering::Relaxed);
+            }
         }),
     );
     text.insert(0, "123").unwrap();
diff --git a/loro-js/tests/event.test.ts b/loro-js/tests/event.test.ts
index e38ee67e..84fea78b 100644
--- a/loro-js/tests/event.test.ts
+++ b/loro-js/tests/event.test.ts
@@ -4,7 +4,7 @@ import {
   getType,
   ListDiff,
   Loro,
-  LoroEvent,
+  LoroEventBatch,
   LoroText,
   MapDiff,
   TextDiff,
@@ -13,7 +13,7 @@ import {
 describe("event", () => {
   it("target", async () => {
     const loro = new Loro();
-    let lastEvent: undefined | LoroEvent;
+    let lastEvent: undefined | LoroEventBatch;
     loro.subscribe((event) => {
       lastEvent = event;
     });
@@ -22,44 +22,12 @@ describe("event", () => {
     text.insert(0, "123");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.target).toEqual(id);
-  });
-
-  it("event id", async () => {
-    const loro = new Loro();
-    let lastEventId: bigint = 0n;
-    let uniqueEvent = 0;
-    loro.subscribe((event) => {
-      console.log(event.id);
-      if (event.id === lastEventId) {
-        return;
-      }
-
-      lastEventId = event.id;
-      uniqueEvent += 1;
-    });
-
-    const map = loro.getMap("map");
-    map.set("0", 100);
-    loro.commit();
-    await oneMs();
-    expect(uniqueEvent).toBe(1);
-    const bDoc = new Loro();
-    bDoc.getText("text").insert(0, "1");
-    bDoc.getText("text").insert(1, "1");
-    bDoc.commit();
-    await oneMs();
-    bDoc.getMap("map").set("0", "1");
-    bDoc.commit();
-    await oneMs();
-    loro.import(bDoc.exportFrom());
-    await oneMs();
-    expect(uniqueEvent).toBe(2);
+    expect(lastEvent?.events[0].target).toEqual(id);
   });
 
   it("path", async () => {
     const loro = new Loro();
-    let lastEvent: undefined | LoroEvent;
+    let lastEvent: undefined | LoroEventBatch;
     loro.subscribe((event) => {
       lastEvent = event;
     });
@@ -68,8 +36,7 @@ describe("event", () => {
     subMap.set("0", "1");
     loro.commit();
     await oneMs();
-
-    expect(lastEvent?.path).toStrictEqual(["map", "sub"]);
+    expect(lastEvent?.events[1].path).toStrictEqual(["map", "sub"]);
     const list = subMap.setContainer("list", "List");
     list.insert(0, "2");
     const text = list.insertContainer(1, "Text");
@@ -78,12 +45,12 @@ describe("event", () => {
     text.insert(0, "3");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.path).toStrictEqual(["map", "sub", "list", 1]);
+    expect(lastEvent?.events[0].path).toStrictEqual(["map", "sub", "list", 1]);
   });
 
   it("text diff", async () => {
     const loro = new Loro();
-    let lastEvent: undefined | LoroEvent;
+    let lastEvent: undefined | LoroEventBatch;
     loro.subscribe((event) => {
       lastEvent = event;
     });
@@ -91,14 +58,14 @@ describe("event", () => {
     text.insert(0, "3");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.diff).toStrictEqual({
+    expect(lastEvent?.events[0].diff).toStrictEqual({
       type: "text",
       diff: [{ insert: "3" }],
     } as TextDiff);
     text.insert(1, "12");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.diff).toStrictEqual({
+    expect(lastEvent?.events[0].diff).toStrictEqual({
       type: "text",
       diff: [{ retain: 1 }, { insert: "12" }],
     } as TextDiff);
@@ -106,7 +73,7 @@ describe("event", () => {
 
   it("list diff", async () => {
     const loro = new Loro();
-    let lastEvent: undefined | LoroEvent;
+    let lastEvent: undefined | LoroEventBatch;
     loro.subscribe((event) => {
       lastEvent = event;
     });
@@ -114,14 +81,14 @@ describe("event", () => {
     text.insert(0, "3");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.diff).toStrictEqual({
+    expect(lastEvent?.events[0].diff).toStrictEqual({
       type: "list",
       diff: [{ insert: ["3"] }],
     } as ListDiff);
     text.insert(1, "12");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.diff).toStrictEqual({
+    expect(lastEvent?.events[0].diff).toStrictEqual({
       type: "list",
       diff: [{ retain: 1 }, { insert: ["12"] }],
     } as ListDiff);
@@ -129,7 +96,7 @@ describe("event", () => {
 
   it("map diff", async () => {
     const loro = new Loro();
-    let lastEvent: undefined | LoroEvent;
+    let lastEvent: undefined | LoroEventBatch;
     loro.subscribe((event) => {
       lastEvent = event;
     });
@@ -138,7 +105,7 @@ describe("event", () => {
     map.set("1", "2");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.diff).toStrictEqual({
+    expect(lastEvent?.events[0].diff).toStrictEqual({
       type: "map",
       updated: {
         "0": "3",
@@ -149,7 +116,7 @@ describe("event", () => {
     map.set("1", "1");
     loro.commit();
     await oneMs();
-    expect(lastEvent?.diff).toStrictEqual({
+    expect(lastEvent?.events[0].diff).toStrictEqual({
       type: "map",
       updated: {
         "0": "0",
@@ -160,7 +127,7 @@ describe("event", () => {
 
   it("tree", async () => {
     const loro = new Loro();
-    let lastEvent: undefined | LoroEvent;
+    let lastEvent: undefined | LoroEventBatch;
     loro.subscribe((event) => {
       console.log(event);
       lastEvent = event;
@@ -170,7 +137,7 @@ describe("event", () => {
     tree.createNode();
     loro.commit();
     await oneMs();
-    expect(lastEvent?.target).toEqual(id);
+    expect(lastEvent?.events[0].target).toEqual(id);
   });
 
   describe("subscribe container events", () => {
@@ -180,12 +147,14 @@ describe("event", () => {
       let ran = 0;
       const sub = text.subscribe(loro, (event) => {
         if (!ran) {
-          expect((event.diff as any).diff).toStrictEqual([
+          expect((event.events[0].diff as any).diff).toStrictEqual([
             { insert: "123" },
           ] as Delta<string>[]);
         }
         ran += 1;
-        expect(event.target).toBe(text.id);
+        for (const containerDiff of event.events) {
+          expect(containerDiff.target).toBe(text.id);
+        }
       });
 
       text.insert(0, "123");
@@ -269,23 +238,25 @@ describe("event", () => {
       const text = loro.getText("text");
       let string = "";
       text.subscribe(loro, (event) => {
-        const diff = event.diff;
-        expect(diff.type).toBe("text");
-        if (diff.type === "text") {
-          let newString = "";
-          let pos = 0;
-          for (const delta of diff.diff) {
-            if (delta.retain != null) {
-              newString += string.slice(pos, pos + delta.retain);
-              pos += delta.retain;
-            } else if (delta.insert != null) {
-              newString += delta.insert;
-            } else {
-              pos += delta.delete;
+        for (const containerDiff of event.events) {
+          const diff = containerDiff.diff;
+          expect(diff.type).toBe("text");
+          if (diff.type === "text") {
+            let newString = "";
+            let pos = 0;
+            for (const delta of diff.diff) {
+              if (delta.retain != null) {
+                newString += string.slice(pos, pos + delta.retain);
+                pos += delta.retain;
+              } else if (delta.insert != null) {
+                newString += delta.insert;
+              } else {
+                pos += delta.delete;
+              }
             }
-          }
 
-          string = newString + string.slice(pos);
+            string = newString + string.slice(pos);
+          }
         }
       });
       text.insert(0, "你好");
@@ -317,7 +288,7 @@ describe("event", () => {
       let first = true;
       loro.subscribe((e) => {
         if (first) {
-          const diff = (e.diff as ListDiff).diff;
+          const diff = (e.events[0].diff as ListDiff).diff;
           const text = diff[0].insert![0] as LoroText;
           text.insert(0, "abc");
           first = false;
@@ -335,8 +306,8 @@ describe("event", () => {
     const list = doc.getList("list");
     let ran = false;
     doc.subscribe((event) => {
-      if (event.diff.type === "list") {
-        for (const item of event.diff.diff) {
+      if (event.events[0].diff.type === "list") {
+        for (const item of event.events[0].diff.diff) {
           const t = item.insert![0] as LoroText;
           expect(t.toString()).toBe("Hello");
           expect(item.insert?.length).toBe(2);
diff --git a/loro-js/tests/richtext.test.ts b/loro-js/tests/richtext.test.ts
index d016b521..661fa806 100644
--- a/loro-js/tests/richtext.test.ts
+++ b/loro-js/tests/richtext.test.ts
@@ -7,7 +7,7 @@ describe("richtext", () => {
     const doc = new Loro();
     doc.configTextStyle({
       bold: { expand: "after" },
-      link: { expand: "before" }
+      link: { expand: "before" },
     });
     const text = doc.getText("text");
     text.insert(0, "Hello World!");
@@ -60,9 +60,9 @@ describe("richtext", () => {
     const text = doc.getText("text");
     let called = false;
     text.subscribe(doc, (event) => {
-      if (event.diff.type == "text") {
+      if (event.events[0].diff.type == "text") {
         called = true;
-        expect(event.diff.diff).toStrictEqual([
+        expect(event.events[0].diff.diff).toStrictEqual([
           {
             insert: "Hello",
             attributes: {
@@ -113,17 +113,19 @@ describe("richtext", () => {
     doc1.configTextStyle({
       link: { expand: "none" },
       bold: { expand: "after" },
-    })
+    });
     const text1 = doc1.getText("text");
     const doc2 = new Loro();
     doc2.configTextStyle({
       link: { expand: "none" },
       bold: { expand: "after" },
-    })
+    });
     const text2 = doc2.getText("text");
     text1.subscribe(doc1, (event) => {
-      const e = event.diff as TextDiff;
-      text2.applyDelta(e.diff);
+      for (const containerDiff of event.events) {
+        const e = containerDiff.diff as TextDiff;
+        text2.applyDelta(e.diff);
+      }
     });
     text1.insert(0, "foo");
     text1.mark({ start: 0, end: 3 }, "link", true);
@@ -133,58 +135,66 @@ describe("richtext", () => {
     text1.insert(3, "baz");
     doc1.commit();
     await new Promise((r) => setTimeout(r, 1));
-    expect(text2.toDelta()).toStrictEqual([{ insert: 'foo', attributes: { link: true } }, { insert: 'baz' }]);
+    expect(text2.toDelta()).toStrictEqual([
+      { insert: "foo", attributes: { link: true } },
+      { insert: "baz" },
+    ]);
     expect(text2.toDelta()).toStrictEqual(text1.toDelta());
     text1.mark({ start: 2, end: 5 }, "bold", true);
     doc1.commit();
     await new Promise((r) => setTimeout(r, 1));
     expect(text2.toDelta()).toStrictEqual(text1.toDelta());
-  })
+  });
 
   it("custom richtext type", async () => {
     const doc = new Loro();
     doc.configTextStyle({
       myStyle: {
         expand: "none",
-      }
-    })
+      },
+    });
     const text = doc.getText("text");
     text.insert(0, "foo");
     text.mark({ start: 0, end: 3 }, "myStyle", 123);
-    expect(text.toDelta()).toStrictEqual([{ insert: 'foo', attributes: { myStyle: 123 } }]);
+    expect(text.toDelta()).toStrictEqual([
+      { insert: "foo", attributes: { myStyle: 123 } },
+    ]);
 
     expect(() => {
       text.mark({ start: 0, end: 3 }, "unknownStyle", 2);
-    }).toThrowError()
+    }).toThrowError();
 
     expect(() => {
       // default style config should be overwritten
       text.mark({ start: 0, end: 3 }, "bold", 2);
-    }).toThrowError()
-  })
+    }).toThrowError();
+  });
 
   it("allow overlapped styles", () => {
     const doc = new Loro();
     doc.configTextStyle({
-      comment: { expand: "none", }
-    })
+      comment: { expand: "none" },
+    });
     const text = doc.getText("text");
     text.insert(0, "The fox jumped.");
     text.mark({ start: 0, end: 7 }, "comment:alice", "Hi");
     text.mark({ start: 4, end: 14 }, "comment:bob", "Jump");
     expect(text.toDelta()).toStrictEqual([
       {
-        insert: "The ", attributes: { "comment:alice": "Hi" },
+        insert: "The ",
+        attributes: { "comment:alice": "Hi" },
       },
       {
-        insert: "fox", attributes: { "comment:alice": "Hi", "comment:bob": "Jump" },
+        insert: "fox",
+        attributes: { "comment:alice": "Hi", "comment:bob": "Jump" },
       },
       {
-        insert: " jumped", attributes: { "comment:bob": "Jump" },
+        insert: " jumped",
+        attributes: { "comment:bob": "Jump" },
       },
       {
         insert: ".",
-      }
-    ])
-  })
+      },
+    ]);
+  });
 });
