diff --git a/tests/capture/test_templates.py b/tests/capture/test_templates.py
index afc66e6c5..787ad6122 100644
--- a/tests/capture/test_templates.py
+++ b/tests/capture/test_templates.py
@@ -257,6 +257,8 @@ tested_modified_templates = [
     qml.MPS,
     qml.TTN,
     qml.QROM,
+    qml.PhaseAdder,
+    qml.Adder,
 ]
 
 
@@ -686,6 +688,76 @@ class TestModifiedTemplates:
         assert len(q) == 1
         qml.assert_equal(q.queue[0], qml.QROM(**kwargs))
 
+    @pytest.mark.usefixtures("new_opmath_only")
+    def test_phase_adder(self):
+        """Test the primitive bind call of PhaseAdder."""
+
+        kwargs = {
+            "k": 3,
+            "x_wires": [0, 1],
+            "mod": None,
+            "work_wire": None,
+        }
+
+        def qfunc():
+            qml.PhaseAdder(**kwargs)
+
+        # Validate inputs
+        qfunc()
+
+        # Actually test primitive bind
+        jaxpr = jax.make_jaxpr(qfunc)()
+
+        assert len(jaxpr.eqns) == 1
+
+        eqn = jaxpr.eqns[0]
+        assert eqn.primitive == qml.PhaseAdder._primitive
+        assert eqn.invars == jaxpr.jaxpr.invars
+        assert eqn.params == kwargs
+        assert len(eqn.outvars) == 1
+        assert isinstance(eqn.outvars[0], jax.core.DropVar)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts)
+
+        assert len(q) == 1
+        qml.assert_equal(q.queue[0], qml.PhaseAdder(**kwargs))
+
+    @pytest.mark.usefixtures("new_opmath_only")
+    def test_adder(self):
+        """Test the primitive bind call of Adder."""
+
+        kwargs = {
+            "k": 3,
+            "x_wires": [0, 1],
+            "mod": None,
+            "work_wires": None,
+        }
+
+        def qfunc():
+            qml.Adder(**kwargs)
+
+        # Validate inputs
+        qfunc()
+
+        # Actually test primitive bind
+        jaxpr = jax.make_jaxpr(qfunc)()
+
+        assert len(jaxpr.eqns) == 1
+
+        eqn = jaxpr.eqns[0]
+        assert eqn.primitive == qml.Adder._primitive
+        assert eqn.invars == jaxpr.jaxpr.invars
+        assert eqn.params == kwargs
+        assert len(eqn.outvars) == 1
+        assert isinstance(eqn.outvars[0], jax.core.DropVar)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts)
+
+        assert len(q) == 1
+        qml.assert_equal(q.queue[0], qml.Adder(**kwargs))
+
     @pytest.mark.parametrize(
         "template, kwargs",
         [
diff --git a/tests/templates/test_subroutines/test_adder.py b/tests/templates/test_subroutines/test_adder.py
new file mode 100644
index 000000000..0f5a0ef4f
--- /dev/null
+++ b/tests/templates/test_subroutines/test_adder.py
@@ -0,0 +1,224 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Tests for the Adder template.
+"""
+
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+def test_standard_validity_Adder():
+    """Check the operation using the assert_valid function."""
+    k = 6
+    mod = 11
+    x_wires = [0, 1, 2, 3]
+    work_wires = [4, 5]
+    op = qml.Adder(k, x_wires=x_wires, mod=mod, work_wires=work_wires)
+    qml.ops.functions.assert_valid(op)
+
+
+class TestAdder:
+    """Test the qml.Adder template."""
+
+    @pytest.mark.parametrize(
+        ("k", "x_wires", "mod", "work_wires", "x"),
+        [
+            (6, [0, 1, 2], 8, [4, 5], 1),
+            (
+                6,
+                ["a", "b", "c"],
+                8,
+                ["d", "e"],
+                2,
+            ),
+            (
+                1,
+                [0, 1, 2, 3],
+                9,
+                [4, 5],
+                3,
+            ),
+            (
+                2,
+                [0, 1, 4],
+                4,
+                [3, 2],
+                2,
+            ),
+            (
+                0,
+                [0, 1, 4],
+                4,
+                [3, 2],
+                2,
+            ),
+            (
+                1,
+                [0, 1, 4],
+                4,
+                [3, 2],
+                0,
+            ),
+            (
+                -3,
+                [0, 1, 4],
+                4,
+                [3, 2],
+                1,
+            ),
+            (
+                10,
+                [0, 1, 2, 5],
+                9,
+                [3, 4],
+                2,
+            ),
+            (
+                1,
+                [0, 1, 2],
+                7,
+                [3, 4],
+                3,
+            ),
+            (
+                6,
+                [0, 1, 2, 3],
+                None,
+                [4, 5],
+                3,
+            ),
+        ],
+    )
+    def test_operation_result(
+        self, k, x_wires, mod, work_wires, x
+    ):  # pylint: disable=too-many-arguments
+        """Test the correctness of the PhaseAdder template output."""
+        dev = qml.device("default.qubit", shots=1)
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.BasisEmbedding(x, wires=x_wires)
+            qml.Adder(k, x_wires, mod, work_wires)
+            return qml.sample(wires=x_wires)
+
+        if mod is None:
+            mod = 2 ** len(x_wires)
+
+        result = sum(bit * (2**i) for i, bit in enumerate(reversed(circuit(x))))
+        assert np.allclose(result, (x + k) % mod)
+
+    @pytest.mark.parametrize(
+        ("k", "x_wires", "mod", "work_wires", "msg_match"),
+        [
+            (
+                1,
+                [0, 1, 2],
+                9,
+                [3, 4],
+                ("Adder must have enough x_wires to represent mod."),
+            ),
+            (
+                1,
+                [0, 1, 2],
+                9,
+                None,
+                (r"If mod is not"),
+            ),
+            (
+                3,
+                [0, 1, 2, 3, 4],
+                12,
+                [4, 5],
+                "None of the wires in work_wires should be included in x_wires.",
+            ),
+        ],
+    )
+    def test_operation_and_test_wires_error(
+        self, k, x_wires, mod, work_wires, msg_match
+    ):  # pylint: disable=too-many-arguments
+        """Test that proper errors are raised"""
+
+        with pytest.raises(ValueError, match=msg_match):
+            qml.Adder(k, x_wires, mod, work_wires)
+
+    @pytest.mark.parametrize(
+        ("k", "x_wires", "mod", "work_wires", "msg_match"),
+        [
+            (
+                2.3,
+                [0, 1, 2],
+                9,
+                [3, 4],
+                ("Both k and mod must be integers"),
+            ),
+            (
+                2,
+                [0, 1, 2],
+                3.2,
+                [3, 4],
+                ("Both k and mod must be integers"),
+            ),
+        ],
+    )
+    def test_types_error(
+        self, k, x_wires, mod, work_wires, msg_match
+    ):  # pylint: disable=too-many-arguments
+        """Test errors are raised"""
+        with pytest.raises(ValueError, match=msg_match):
+            qml.Adder(k, x_wires, mod, work_wires)
+
+    def test_decomposition(self):
+        """Test that compute_decomposition and decomposition work as expected."""
+
+        k = 2
+        mod = 7
+        x_wires = [0, 1, 2]
+        work_wires = [3, 4]
+        adder_decomposition = qml.Adder(k, x_wires, mod, work_wires).compute_decomposition(
+            k, x_wires, mod, work_wires
+        )
+        op_list = []
+        op_list.append(qml.QFT(work_wires[:1] + x_wires))
+        op_list.append(qml.PhaseAdder(k, work_wires[:1] + x_wires, mod, work_wires[1:]))
+        op_list.append(qml.adjoint(qml.QFT)(work_wires[:1] + x_wires))
+
+        for op1, op2 in zip(adder_decomposition, op_list):
+            qml.assert_equal(op1, op2)
+
+    @pytest.mark.jax
+    def test_jit_compatible(self):
+        """Test that the template is compatible with the JIT compiler."""
+
+        import jax
+
+        jax.config.update("jax_enable_x64", True)
+        x = 2
+        k = 6
+        mod = 7
+        x_wires = [0, 1, 2]
+        work_wires = [3, 4]
+        dev = qml.device("default.qubit", shots=1)
+
+        @jax.jit
+        @qml.qnode(dev)
+        def circuit():
+            qml.BasisEmbedding(x, wires=x_wires)
+            qml.Adder(k, x_wires, mod, work_wires)
+            return qml.sample(wires=x_wires)
+
+        result = sum(bit * (2**i) for i, bit in enumerate(reversed(circuit())))
+        assert jax.numpy.allclose(result, (x + k) % mod)
diff --git a/tests/templates/test_subroutines/test_phase_adder.py b/tests/templates/test_subroutines/test_phase_adder.py
new file mode 100644
index 000000000..509228fc7
--- /dev/null
+++ b/tests/templates/test_subroutines/test_phase_adder.py
@@ -0,0 +1,253 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Tests for the PhaseAdder template.
+"""
+
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.templates.subroutines.phase_adder import _add_k_fourier
+
+
+def test_standard_validity_Phase_Adder():
+    """Check the operation using the assert_valid function."""
+    k = 6
+    mod = 11
+    x_wires = [0, 1, 2, 3]
+    work_wire = [4]
+    op = qml.PhaseAdder(k, x_wires=x_wires, mod=mod, work_wire=work_wire)
+    qml.ops.functions.assert_valid(op)
+
+
+def test_add_k_fourier():
+    """Test the private _add_k_fourier function."""
+
+    ops = _add_k_fourier(2, wires=range(2))
+    assert len(ops) == 2
+    assert ops[0].name == "PhaseShift"
+    assert ops[1].name == "PhaseShift"
+    assert np.isclose(ops[0].parameters[0], 2 * np.pi)
+    assert np.isclose(ops[1].parameters[0], np.pi)
+
+
+class TestPhaseAdder:
+    """Test the PhaseAdder template."""
+
+    @pytest.mark.parametrize(
+        ("k", "x_wires", "mod", "work_wire", "x"),
+        [
+            (
+                6,
+                [0, 1, 2],
+                7,
+                [4],
+                2,
+            ),
+            (
+                6,
+                ["a", "b", "c"],
+                7,
+                ["d"],
+                3,
+            ),
+            (
+                0,
+                [0, 1, 2, 3, 5],
+                9,
+                [4],
+                2,
+            ),
+            (
+                2,
+                [0, 1, 4],
+                4,
+                [3],
+                1,
+            ),
+            (
+                0,
+                [0, 1, 4],
+                4,
+                [3],
+                0,
+            ),
+            (
+                -2,
+                [0, 1, 4],
+                4,
+                [3],
+                0,
+            ),
+            (
+                10,
+                [0, 1, 2, 5],
+                9,
+                [3],
+                3,
+            ),
+            (
+                1,
+                [0, 1, 2, 4],
+                7,
+                [3],
+                3,
+            ),
+            (
+                6,
+                [0, 1, 2, 3],
+                None,
+                [4],
+                2,
+            ),
+        ],
+    )
+    def test_operation_result(
+        self, k, x_wires, mod, work_wire, x
+    ):  # pylint: disable=too-many-arguments
+        """Test the correctness of the PhaseAdder template output."""
+        dev = qml.device("default.qubit", shots=1)
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.BasisEmbedding(x, wires=x_wires)
+            qml.QFT(wires=x_wires)
+            qml.PhaseAdder(k, x_wires, mod, work_wire)
+            qml.adjoint(qml.QFT)(wires=x_wires)
+            return qml.sample(wires=x_wires)
+
+        if mod is None:
+            mod = 2 ** len(x_wires)
+
+        assert np.allclose(
+            sum(bit * (2**i) for i, bit in enumerate(reversed(circuit(x)))), (x + k) % mod
+        )
+
+    @pytest.mark.parametrize(
+        ("k", "x_wires", "mod", "work_wire", "msg_match"),
+        [
+            (
+                1,
+                [0, 1, 2],
+                9,
+                [3],
+                ("PhaseAdder must have enough x_wires to represent mod."),
+            ),
+            (
+                1,
+                [0, 1, 2],
+                9,
+                None,
+                (r"If mod is not"),
+            ),
+            (
+                3,
+                [0, 1, 2, 3, 4],
+                12,
+                [4],
+                "None of the wires in work_wire should be included in x_wires.",
+            ),
+        ],
+    )
+    def test_operation_and_wires_error(
+        self, k, x_wires, mod, work_wire, msg_match
+    ):  # pylint: disable=too-many-arguments
+        """Test errors are raised"""
+        with pytest.raises(ValueError, match=msg_match):
+            qml.PhaseAdder(k, x_wires, mod, work_wire)
+
+    @pytest.mark.parametrize(
+        ("k", "x_wires", "mod", "work_wire", "msg_match"),
+        [
+            (
+                2.3,
+                [0, 1, 2],
+                9,
+                [3],
+                ("Both k and mod must be integers"),
+            ),
+            (
+                2,
+                [0, 1, 2],
+                3.2,
+                [3],
+                ("Both k and mod must be integers"),
+            ),
+        ],
+    )
+    def test_types_error(
+        self, k, x_wires, mod, work_wire, msg_match
+    ):  # pylint: disable=too-many-arguments
+        """Test errors are raised"""
+        with pytest.raises(ValueError, match=msg_match):
+            qml.PhaseAdder(k, x_wires, mod, work_wire)
+
+    def test_decomposition(self):
+        """Test that compute_decomposition and decomposition work as expected."""
+        k = 4
+        x_wires = [1, 2, 3]
+        mod = 7
+        work_wire = [0]
+
+        phase_adder_decomposition = qml.PhaseAdder(
+            k, x_wires, mod, work_wire
+        ).compute_decomposition(k, x_wires, mod, work_wire)
+        op_list = []
+
+        if mod == 2 ** (len(x_wires)):
+            op_list.extend(_add_k_fourier(k, x_wires))
+        else:
+            aux_k = x_wires[0]
+            op_list.extend(_add_k_fourier(k, x_wires))
+            op_list.extend(qml.adjoint(_add_k_fourier)(mod, x_wires))
+            op_list.append(qml.adjoint(qml.QFT)(wires=x_wires))
+            op_list.append(qml.ctrl(qml.PauliX(work_wire), control=aux_k, control_values=1))
+            op_list.append(qml.QFT(wires=x_wires))
+            op_list.extend(qml.ctrl(op, control=work_wire) for op in _add_k_fourier(mod, x_wires))
+            op_list.extend(qml.adjoint(_add_k_fourier)(k, x_wires))
+            op_list.append(qml.adjoint(qml.QFT)(wires=x_wires))
+            op_list.append(qml.ctrl(qml.PauliX(work_wire), control=aux_k, control_values=0))
+            op_list.append(qml.QFT(wires=x_wires))
+            op_list.extend(_add_k_fourier(k, x_wires))
+
+        for op1, op2 in zip(phase_adder_decomposition, op_list):
+            qml.assert_equal(op1, op2)
+
+    @pytest.mark.jax
+    def test_jit_compatible(self):
+        """Test that the template is compatible with the JIT compiler."""
+
+        import jax
+
+        jax.config.update("jax_enable_x64", True)
+        x = 2
+        k = 6
+        mod = 7
+        x_wires = [0, 1, 2]
+        work_wire = [4]
+        dev = qml.device("default.qubit", shots=1)
+
+        @jax.jit
+        @qml.qnode(dev)
+        def circuit():
+            qml.BasisEmbedding(x, wires=x_wires)
+            qml.QFT(wires=x_wires)
+            qml.PhaseAdder(k, x_wires, mod, work_wire)
+            qml.adjoint(qml.QFT)(wires=x_wires)
+            return qml.sample(wires=x_wires)
+
+        assert jax.numpy.allclose(
+            sum(bit * (2**i) for i, bit in enumerate(reversed(circuit()))), (x + k) % mod
+        )
