diff --git a/e2etests/bats-tests.sh b/e2etests/bats-tests.sh
index 832d085..9a36628 100755
--- a/e2etests/bats-tests.sh
+++ b/e2etests/bats-tests.sh
@@ -710,6 +710,27 @@ get_value_from() {
   [[ "${count}" == "2" ]]
 }
 
+@test "readiness-port" {
+  tmp="tests/checks/readiness-port.yml"
+  cmd="${KUBE_LINTER_BIN} lint --include readiness-port --do-not-auto-add-defaults --format json ${tmp}"
+  run ${cmd}
+
+  print_info "${status}" "${output}" "${cmd}" "${tmp}"
+  [ "$status" -eq 1 ]
+
+  message1=$(get_value_from "${lines[0]}" '.Reports[0].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[0].Diagnostic.Message')
+  message2=$(get_value_from "${lines[0]}" '.Reports[1].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[1].Diagnostic.Message')
+  message3=$(get_value_from "${lines[0]}" '.Reports[2].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[2].Diagnostic.Message')
+  message4=$(get_value_from "${lines[0]}" '.Reports[3].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[3].Diagnostic.Message')
+  count=$(get_value_from "${lines[0]}" '.Reports | length')
+
+  [[ "${message1}" == "Deployment: container \"fire-deployment-name\" does not expose port http for the HTTPGet" ]]
+  [[ "${message2}" == "Deployment: container \"fire-deployment-int\" does not expose port 8080 for the HTTPGet" ]]
+  [[ "${message3}" == "Deployment: container \"fire-deployment-udp\" does not expose port udp for the TCPSocket" ]]
+  [[ "${message4}" == "Deployment: container \"fire-deployment-grpc\" does not expose port 8080 for the GRPC check" ]]
+  [[ "${count}" == "4" ]]
+}
+
 @test "required-annotation-email" {
   tmp="tests/checks/required-annotation-email.yml"
   cmd="${KUBE_LINTER_BIN} lint --include required-annotation-email --do-not-auto-add-defaults --format json ${tmp}"
@@ -812,6 +833,27 @@ get_value_from() {
   [[ "${count}" == "3" ]]
 }
 
+@test "startup-port" {
+  tmp="tests/checks/startup-port.yml"
+  cmd="${KUBE_LINTER_BIN} lint --include startup-port --do-not-auto-add-defaults --format json ${tmp}"
+  run ${cmd}
+
+  print_info "${status}" "${output}" "${cmd}" "${tmp}"
+  [ "$status" -eq 1 ]
+
+  message1=$(get_value_from "${lines[0]}" '.Reports[0].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[0].Diagnostic.Message')
+  message2=$(get_value_from "${lines[0]}" '.Reports[1].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[1].Diagnostic.Message')
+  message3=$(get_value_from "${lines[0]}" '.Reports[2].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[2].Diagnostic.Message')
+  message4=$(get_value_from "${lines[0]}" '.Reports[3].Object.K8sObject.GroupVersionKind.Kind + ": " + .Reports[3].Diagnostic.Message')
+  count=$(get_value_from "${lines[0]}" '.Reports | length')
+
+  [[ "${message1}" == "Deployment: container \"fire-deployment-name\" does not expose port http for the HTTPGet" ]]
+  [[ "${message2}" == "Deployment: container \"fire-deployment-int\" does not expose port 8080 for the HTTPGet" ]]
+  [[ "${message3}" == "Deployment: container \"fire-deployment-udp\" does not expose port udp for the TCPSocket" ]]
+  [[ "${message4}" == "Deployment: container \"fire-deployment-grpc\" does not expose port 8080 for the GRPC check" ]]
+  [[ "${count}" == "4" ]]
+}
+
 @test "unsafe-proc-mount" {
   tmp="tests/checks/unsafe-proc-mount.yml"
   cmd="${KUBE_LINTER_BIN} lint --include unsafe-proc-mount --do-not-auto-add-defaults --format json ${tmp}"
diff --git a/pkg/templates/livenessport/template_test.go b/pkg/templates/livenessport/template_test.go
index 05b4b10..942fa3f 100644
--- a/pkg/templates/livenessport/template_test.go
+++ b/pkg/templates/livenessport/template_test.go
@@ -205,6 +205,49 @@ func (s *MissingLivenessPort) TestDeploymentWith() {
 				},
 			},
 		},
+		{
+			name: "GrpcCheckWillPass",
+			container: v1.Container{
+				Name: "container",
+				Ports: []v1.ContainerPort{
+					{
+						Name:          "http",
+						ContainerPort: 8080,
+					},
+				},
+				LivenessProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						GRPC: &v1.GRPCAction{
+							Port: 8080,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "GrpcPortMissmatch",
+			container: v1.Container{
+				Name: "container",
+				Ports: []v1.ContainerPort{
+					{
+						Name:          "http",
+						ContainerPort: 8080,
+					},
+				},
+				LivenessProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						GRPC: &v1.GRPCAction{
+							Port: 9999,
+						},
+					},
+				},
+			},
+			expected: map[string][]diagnostic.Diagnostic{
+				targetName: {
+					{Message: "container \"container\" does not expose port 9999 for the GRPC check"},
+				},
+			},
+		},
 	}
 	for _, tc := range testCases {
 		s.Run(tc.name, func() {
diff --git a/pkg/templates/readinessport/template_test.go b/pkg/templates/readinessport/template_test.go
new file mode 100644
index 0000000..d3d47b1
--- /dev/null
+++ b/pkg/templates/readinessport/template_test.go
@@ -0,0 +1,106 @@
+package readinessport
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+	"golang.stackrox.io/kube-linter/pkg/diagnostic"
+	"golang.stackrox.io/kube-linter/pkg/lintcontext/mocks"
+	"golang.stackrox.io/kube-linter/pkg/templates"
+	"golang.stackrox.io/kube-linter/pkg/templates/readinessport/internal/params"
+	v1 "k8s.io/api/core/v1"
+)
+
+func TestMissingReadinessPort(t *testing.T) {
+	suite.Run(t, new(MissingReadinessPort))
+}
+
+type MissingReadinessPort struct {
+	templates.TemplateTestSuite
+
+	ctx *mocks.MockLintContext
+}
+
+func (s *MissingReadinessPort) SetupTest() {
+	s.Init(templateKey)
+	s.ctx = mocks.NewMockContext()
+}
+
+func (s *MissingReadinessPort) TestDeploymentWith() {
+	const targetName = "deployment01"
+	testCases := []struct {
+		name      string
+		container v1.Container
+		expected  map[string][]diagnostic.Diagnostic
+	}{
+		{
+			name:      "NoReadinessProbe",
+			container: v1.Container{},
+			expected:  nil,
+		},
+		{
+			name: "NoReadinessProbeExecIgnored",
+			container: v1.Container{
+				ReadinessProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						Exec: &v1.ExecAction{},
+					},
+				},
+			},
+			expected: nil,
+		},
+		{
+			name: "GrpcCheckWillPass",
+			container: v1.Container{
+				Name: "container",
+				Ports: []v1.ContainerPort{
+					{
+						Name:          "http",
+						ContainerPort: 8080,
+					},
+				},
+				ReadinessProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						GRPC: &v1.GRPCAction{
+							Port: 8080,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "GrpcPortMissmatch",
+			container: v1.Container{
+				Name: "container",
+				Ports: []v1.ContainerPort{
+					{
+						Name:          "http",
+						ContainerPort: 8080,
+					},
+				},
+				ReadinessProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						GRPC: &v1.GRPCAction{
+							Port: 9999,
+						},
+					},
+				},
+			},
+			expected: map[string][]diagnostic.Diagnostic{
+				targetName: {
+					{Message: "container \"container\" does not expose port 9999 for the GRPC check"},
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		s.Run(tc.name, func() {
+			s.ctx.AddMockDeployment(s.T(), targetName)
+			s.ctx.AddContainerToDeployment(s.T(), targetName, tc.container)
+			s.Validate(s.ctx, []templates.TestCase{{
+				Param:       params.Params{},
+				Diagnostics: tc.expected,
+			}})
+		})
+	}
+}
diff --git a/pkg/templates/startupport/template_test.go b/pkg/templates/startupport/template_test.go
new file mode 100644
index 0000000..0289e6a
--- /dev/null
+++ b/pkg/templates/startupport/template_test.go
@@ -0,0 +1,106 @@
+package startupport
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+	"golang.stackrox.io/kube-linter/pkg/diagnostic"
+	"golang.stackrox.io/kube-linter/pkg/lintcontext/mocks"
+	"golang.stackrox.io/kube-linter/pkg/templates"
+	"golang.stackrox.io/kube-linter/pkg/templates/startupport/internal/params"
+	v1 "k8s.io/api/core/v1"
+)
+
+func TestMissingStartUpPort(t *testing.T) {
+	suite.Run(t, new(MissingStartUpPort))
+}
+
+type MissingStartUpPort struct {
+	templates.TemplateTestSuite
+
+	ctx *mocks.MockLintContext
+}
+
+func (s *MissingStartUpPort) SetupTest() {
+	s.Init(templateKey)
+	s.ctx = mocks.NewMockContext()
+}
+
+func (s *MissingStartUpPort) TestDeploymentWith() {
+	const targetName = "deployment01"
+	testCases := []struct {
+		name      string
+		container v1.Container
+		expected  map[string][]diagnostic.Diagnostic
+	}{
+		{
+			name:      "NoStartUpProbe",
+			container: v1.Container{},
+			expected:  nil,
+		},
+		{
+			name: "NoStartUpProbeExecIgnored",
+			container: v1.Container{
+				StartupProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						Exec: &v1.ExecAction{},
+					},
+				},
+			},
+			expected: nil,
+		},
+		{
+			name: "GrpcCheckWillPass",
+			container: v1.Container{
+				Name: "container",
+				Ports: []v1.ContainerPort{
+					{
+						Name:          "http",
+						ContainerPort: 8080,
+					},
+				},
+				StartupProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						GRPC: &v1.GRPCAction{
+							Port: 8080,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "GrpcPortMissmatch",
+			container: v1.Container{
+				Name: "container",
+				Ports: []v1.ContainerPort{
+					{
+						Name:          "http",
+						ContainerPort: 8080,
+					},
+				},
+				StartupProbe: &v1.Probe{
+					ProbeHandler: v1.ProbeHandler{
+						GRPC: &v1.GRPCAction{
+							Port: 9999,
+						},
+					},
+				},
+			},
+			expected: map[string][]diagnostic.Diagnostic{
+				targetName: {
+					{Message: "container \"container\" does not expose port 9999 for the GRPC check"},
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		s.Run(tc.name, func() {
+			s.ctx.AddMockDeployment(s.T(), targetName)
+			s.ctx.AddContainerToDeployment(s.T(), targetName, tc.container)
+			s.Validate(s.ctx, []templates.TestCase{{
+				Param:       params.Params{},
+				Diagnostics: tc.expected,
+			}})
+		})
+	}
+}
diff --git a/tests/checks/readiness-port.yml b/tests/checks/readiness-port.yml
new file mode 100644
index 0000000..0309433
--- /dev/null
+++ b/tests/checks/readiness-port.yml
@@ -0,0 +1,53 @@
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      containers:
+        - name: dont-fire-deployment-grpc
+          ports:
+          - containerPort: 8080
+            name: http
+          readinessProbe:
+            grpc:
+              port: 8080
+        - name: dont-fire-deployment
+          ports:
+          - containerPort: 8080
+            name: http
+            protocol: TCP
+          readinessProbe:
+            httpGet:
+              path: "/"
+              port: 8080
+        - name: fire-deployment-name
+          readinessProbe:
+            httpGet:
+              path: "/"
+              port: http
+        - name: fire-deployment-int
+          ports:
+          - containerPort: 9999
+            name: http
+            protocol: TCP
+          readinessProbe:
+            httpGet:
+              path: "/"
+              port: 8080
+        - name: fire-deployment-udp
+          ports:
+          - containerPort: 9999
+            name: udp
+            protocol: UDP
+          readinessProbe:
+            tcpSocket:
+              port: udp
+        - name: fire-deployment-grpc
+          ports:
+          - containerPort: 9999
+            name: grpc
+            protocol: TCP
+          readinessProbe:
+            grpc:
+              port: 8080
+---
diff --git a/tests/checks/startup-port.yml b/tests/checks/startup-port.yml
new file mode 100644
index 0000000..ab79b5a
--- /dev/null
+++ b/tests/checks/startup-port.yml
@@ -0,0 +1,53 @@
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      containers:
+        - name: dont-fire-deployment-grpc
+          ports:
+          - containerPort: 8080
+            name: http
+          startupProbe:
+            grpc:
+              port: 8080
+        - name: dont-fire-deployment
+          ports:
+          - containerPort: 8080
+            name: http
+            protocol: TCP
+          startupProbe:
+            httpGet:
+              path: "/"
+              port: 8080
+        - name: fire-deployment-name
+          startupProbe:
+            httpGet:
+              path: "/"
+              port: http
+        - name: fire-deployment-int
+          ports:
+          - containerPort: 9999
+            name: http
+            protocol: TCP
+          startupProbe:
+            httpGet:
+              path: "/"
+              port: 8080
+        - name: fire-deployment-udp
+          ports:
+          - containerPort: 9999
+            name: udp
+            protocol: UDP
+          startupProbe:
+            tcpSocket:
+              port: udp
+        - name: fire-deployment-grpc
+          ports:
+          - containerPort: 9999
+            name: grpc
+            protocol: TCP
+          startupProbe:
+            grpc:
+              port: 8080
+---
