diff --git a/tests/src/db/mysql.rs b/tests/src/db/mysql.rs
index d35edc0..bcd825d 100644
--- a/tests/src/db/mysql.rs
+++ b/tests/src/db/mysql.rs
@@ -33,7 +33,8 @@ impl Setup for SetupMySQL {
 
     fn capability(&self) -> &Capability {
         &Capability::Sql(CapabilitySql {
-            cte_with_update: true,
+            cte_with_update: false,
+            select_for_update: true,
         })
     }
 }
diff --git a/tests/src/db/postgresql.rs b/tests/src/db/postgresql.rs
index 4e6f0a9..491a248 100644
--- a/tests/src/db/postgresql.rs
+++ b/tests/src/db/postgresql.rs
@@ -34,6 +34,7 @@ impl Setup for SetupPostgreSQL {
     fn capability(&self) -> &Capability {
         &Capability::Sql(CapabilitySql {
             cte_with_update: true,
+            select_for_update: true,
         })
     }
 }
diff --git a/tests/src/db/sqlite.rs b/tests/src/db/sqlite.rs
index 7468de1..df5340e 100644
--- a/tests/src/db/sqlite.rs
+++ b/tests/src/db/sqlite.rs
@@ -16,6 +16,7 @@ impl Setup for SetupSqlite {
     fn capability(&self) -> &Capability {
         &Capability::Sql(CapabilitySql {
             cte_with_update: false,
+            select_for_update: false,
         })
     }
 }
