diff --git a/internal/configs/version2/templates_test.go b/internal/configs/version2/templates_test.go
index 7dc0739a5..5aa546fc7 100644
--- a/internal/configs/version2/templates_test.go
+++ b/internal/configs/version2/templates_test.go
@@ -1574,10 +1574,10 @@ var (
 			},
 		},
 		Upstreams: []Upstream{},
-		AuthJwtClaimSet: []AuthJwtClaimSet{
+		AuthJWTClaimSets: []AuthJWTClaimSet{
 			{
 				Variable: "$jwt_default_webapp_group_consumer_group_type",
-				Claims:   "consumer_group type",
+				Claim:    "consumer_group type",
 			},
 		},
 		Maps: []Map{
diff --git a/internal/configs/virtualserver_test.go b/internal/configs/virtualserver_test.go
index 93af61005..45744e643 100644
--- a/internal/configs/virtualserver_test.go
+++ b/internal/configs/virtualserver_test.go
@@ -6395,6 +6395,223 @@ func TestGenerateVirtualServerConfigAPIKeyClientMaps(t *testing.T) {
 	}
 }
 
+func TestGenerateVirtualServerConfigRateLimitPolicyAuthJwt(t *testing.T) {
+	t.Parallel()
+
+	virtualServerEx := VirtualServerEx{
+		VirtualServer: &conf_v1.VirtualServer{
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name:      "cafe",
+				Namespace: "default",
+			},
+			Spec: conf_v1.VirtualServerSpec{
+				Host: "cafe.example.com",
+				Policies: []conf_v1.PolicyReference{
+					{
+						Name: "gold-rate-limit-policy",
+					},
+					{
+						Name: "silver-rate-limit-policy",
+					},
+				},
+				Upstreams: []conf_v1.Upstream{
+					{
+						Name:    "tea",
+						Service: "tea-svc",
+						Port:    80,
+					},
+					{
+						Name:    "coffee",
+						Service: "coffee-svc",
+						Port:    80,
+					},
+				},
+				Routes: []conf_v1.Route{
+					{
+						Path: "/tea",
+						Action: &conf_v1.Action{
+							Pass: "tea",
+						},
+					},
+					{
+						Path: "/coffee",
+						Action: &conf_v1.Action{
+							Pass: "coffee",
+						},
+					},
+				},
+			},
+		},
+		Policies: map[string]*conf_v1.Policy{
+			"default/gold-rate-limit-policy": {
+				Spec: conf_v1.PolicySpec{
+					RateLimit: &conf_v1.RateLimit{
+						Key:      "test",
+						ZoneSize: "10M",
+						Rate:     "10r/s",
+						Condition: &conf_v1.RateLimitCondition{
+							JWT: &conf_v1.JWTCondition{
+								Claim: "user_type.tier",
+								Match: "gold",
+							},
+						},
+					},
+				},
+			},
+			"default/silver-rate-limit-policy": {
+				Spec: conf_v1.PolicySpec{
+					RateLimit: &conf_v1.RateLimit{
+						Key:      "test",
+						ZoneSize: "20M",
+						Rate:     "20r/s",
+						Condition: &conf_v1.RateLimitCondition{
+							JWT: &conf_v1.JWTCondition{
+								Claim: "user_type.tier",
+								Match: "silver",
+							},
+						},
+					},
+				},
+			},
+		},
+		Endpoints: map[string][]string{
+			"default/tea-svc:80": {
+				"10.0.0.20:80",
+			},
+			"default/coffee-svc:80": {
+				"10.0.0.30:80",
+			},
+		},
+	}
+	expected := version2.VirtualServerConfig{
+		Maps:             nil,
+		AuthJWTClaimSets: []version2.AuthJWTClaimSet{{Variable: "$jwt_default_cafe_user_type_tier", Claim: "user_type tier"}},
+		Upstreams: []version2.Upstream{
+			{
+				UpstreamLabels: version2.UpstreamLabels{
+					Service:           "coffee-svc",
+					ResourceType:      "virtualserver",
+					ResourceName:      "cafe",
+					ResourceNamespace: "default",
+				},
+				Name: "vs_default_cafe_coffee",
+				Servers: []version2.UpstreamServer{
+					{
+						Address: "10.0.0.30:80",
+					},
+				},
+				Keepalive: 16,
+			},
+			{
+				UpstreamLabels: version2.UpstreamLabels{
+					Service:           "tea-svc",
+					ResourceType:      "virtualserver",
+					ResourceName:      "cafe",
+					ResourceNamespace: "default",
+				},
+				Name: "vs_default_cafe_tea",
+				Servers: []version2.UpstreamServer{
+					{
+						Address: "10.0.0.20:80",
+					},
+				},
+				Keepalive: 16,
+			},
+		},
+		HTTPSnippets: []string{},
+		LimitReqZones: []version2.LimitReqZone{
+			{Key: "test", ZoneName: "pol_rl_default_gold-rate-limit-policy_default_cafe", ZoneSize: "10M", Rate: "10r/s"},
+			{Key: "test", ZoneName: "pol_rl_default_silver-rate-limit-policy_default_cafe", ZoneSize: "20M", Rate: "20r/s"},
+		},
+		Server: version2.Server{
+			JWTAuthList:     nil,
+			JWTAuth:         nil,
+			JWKSAuthEnabled: false,
+			ServerName:      "cafe.example.com",
+			StatusZone:      "cafe.example.com",
+			ProxyProtocol:   true,
+			ServerTokens:    "off",
+			RealIPHeader:    "X-Real-IP",
+			SetRealIPFrom:   []string{"0.0.0.0/0"},
+			RealIPRecursive: true,
+			Snippets:        []string{"# server snippet"},
+			TLSPassthrough:  true,
+			VSNamespace:     "default",
+			VSName:          "cafe",
+			APIKeyEnabled:   false,
+			LimitReqs: []version2.LimitReq{
+				{ZoneName: "pol_rl_default_gold-rate-limit-policy_default_cafe", Burst: 0, NoDelay: false, Delay: 0},
+				{ZoneName: "pol_rl_default_silver-rate-limit-policy_default_cafe", Burst: 0, NoDelay: false, Delay: 0},
+			},
+			LimitReqOptions: version2.LimitReqOptions{
+				DryRun:     false,
+				LogLevel:   "error",
+				RejectCode: 503,
+			},
+			Locations: []version2.Location{
+				{
+					Path:                     "/tea",
+					ProxyPass:                "http://vs_default_cafe_tea",
+					ProxyNextUpstream:        "error timeout",
+					ProxyNextUpstreamTimeout: "0s",
+					ProxyNextUpstreamTries:   0,
+					HasKeepalive:             true,
+					ProxySSLName:             "tea-svc.default.svc",
+					ProxyPassRequestHeaders:  true,
+					ProxySetHeaders:          []version2.Header{{Name: "Host", Value: "$host"}},
+					ServiceName:              "tea-svc",
+				},
+				{
+					Path:                     "/coffee",
+					ProxyPass:                "http://vs_default_cafe_coffee",
+					ProxyNextUpstream:        "error timeout",
+					ProxyNextUpstreamTimeout: "0s",
+					ProxyNextUpstreamTries:   0,
+					HasKeepalive:             true,
+					ProxySSLName:             "coffee-svc.default.svc",
+					ProxyPassRequestHeaders:  true,
+					ProxySetHeaders:          []version2.Header{{Name: "Host", Value: "$host"}},
+					ServiceName:              "coffee-svc",
+				},
+			},
+		},
+	}
+
+	baseCfgParams := ConfigParams{
+		Context:         context.Background(),
+		ServerTokens:    "off",
+		Keepalive:       16,
+		ServerSnippets:  []string{"# server snippet"},
+		ProxyProtocol:   true,
+		SetRealIPFrom:   []string{"0.0.0.0/0"},
+		RealIPHeader:    "X-Real-IP",
+		RealIPRecursive: true,
+	}
+
+	vsc := newVirtualServerConfigurator(
+		&baseCfgParams,
+		false,
+		false,
+		&StaticConfigParams{TLSPassthrough: true},
+		false,
+		&fakeBV,
+	)
+
+	result, warnings := vsc.GenerateVirtualServerConfig(&virtualServerEx, nil, nil)
+
+	sort.Slice(result.Maps, func(i, j int) bool {
+		return result.Maps[i].Variable < result.Maps[j].Variable
+	})
+
+	if diff := cmp.Diff(expected, result); diff != "" {
+		t.Errorf("GenerateVirtualServerConfig() mismatch (-want +got):\n%s", diff)
+	}
+
+	if len(warnings) != 0 {
+		t.Errorf("GenerateVirtualServerConfig returned warnings: %v", vsc.warnings)
+	}
+}
+
 func TestGeneratePolicies(t *testing.T) {
 	t.Parallel()
 	ownerDetails := policyOwnerDetails{
@@ -8877,7 +9094,7 @@ func TestGeneratePoliciesFails(t *testing.T) {
 	}
 }
 
-func TestRemoveDuplicates(t *testing.T) {
+func TestRemoveDuplicateLimitReqZones(t *testing.T) {
 	t.Parallel()
 	tests := []struct {
 		rlz      []version2.LimitReqZone
@@ -8919,6 +9136,72 @@ func TestRemoveDuplicates(t *testing.T) {
 	}
 }
 
+func TestRemoveDuplicateAuthJWTClaimSets(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		ajcs     []version2.AuthJWTClaimSet
+		expected []version2.AuthJWTClaimSet
+	}{
+		{
+			ajcs: []version2.AuthJWTClaimSet{
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+			},
+			expected: []version2.AuthJWTClaimSet{
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+			},
+		},
+		{
+			ajcs: []version2.AuthJWTClaimSet{
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+			},
+			expected: []version2.AuthJWTClaimSet{
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+			},
+		},
+		{
+			ajcs: []version2.AuthJWTClaimSet{
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+				{
+					Variable: "$jwt_default_webapp_user_group_type",
+				},
+			},
+			expected: []version2.AuthJWTClaimSet{
+				{
+					Variable: "$jwt_default_webapp_consumer_group_type",
+				},
+				{
+					Variable: "$jwt_default_webapp_user_group_type",
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		result := removeDuplicateAuthJWTClaimSets(test.ajcs)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("removeDuplicateAuthJWTClaimSets() returned \n%v, but expected \n%v", result, test.expected)
+		}
+	}
+}
+
 func TestAddPoliciesCfgToLocations(t *testing.T) {
 	t.Parallel()
 	cfg := policiesCfg{
@@ -9322,6 +9605,74 @@ func TestGenerateString(t *testing.T) {
 	}
 }
 
+func TestGenerateAuthJwtClaimSetVariable(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		claim       string
+		vsNamespace string
+		vsName      string
+		expected    string
+	}{
+		{
+			claim:       "consumer_group.type",
+			vsNamespace: "default",
+			vsName:      "webapp",
+			expected:    "$jwt_default_webapp_consumer_group_type",
+		},
+		{
+			claim:       "type",
+			vsNamespace: "default",
+			vsName:      "webapp",
+			expected:    "$jwt_default_webapp_type",
+		},
+		{
+			claim:       "a.b.c",
+			vsNamespace: "default",
+			vsName:      "webapp",
+			expected:    "$jwt_default_webapp_a_b_c",
+		},
+	}
+
+	for _, test := range tests {
+		result := generateAuthJwtClaimSetVariable(test.claim, test.vsNamespace, test.vsName)
+		if result != test.expected {
+			t.Errorf("generateAuthJwtClaimSetVariable() return %v but expected %v", result, test.expected)
+		}
+	}
+}
+
+func TestGenerateAuthJwtClaimSetClaim(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		claim    string
+		expected string
+	}{
+		{
+			claim:    "consumer_group.type",
+			expected: "consumer_group type",
+		},
+		{
+			claim:    "consumer_group.type",
+			expected: "consumer_group type",
+		},
+		{
+			claim:    "type",
+			expected: "type",
+		},
+		{
+			claim:    "a.b.c",
+			expected: "a b c",
+		},
+	}
+
+	for _, test := range tests {
+		result := generateAuthJwtClaimSetClaim(test.claim)
+		if result != test.expected {
+			t.Errorf("generateAuthJwtClaimSetClaim() return %v but expected %v", result, test.expected)
+		}
+	}
+}
+
 func TestGenerateSnippets(t *testing.T) {
 	t.Parallel()
 	tests := []struct {
diff --git a/pkg/apis/configuration/validation/policy_test.go b/pkg/apis/configuration/validation/policy_test.go
index 542bfaba2..0f3840e3d 100644
--- a/pkg/apis/configuration/validation/policy_test.go
+++ b/pkg/apis/configuration/validation/policy_test.go
@@ -572,6 +572,7 @@ func TestValidateRateLimit_PassesOnValidInput(t *testing.T) {
 
 	tests := []struct {
 		rateLimit *v1.RateLimit
+		isPlus    bool
 		msg       string
 	}{
 		{
@@ -580,7 +581,8 @@ func TestValidateRateLimit_PassesOnValidInput(t *testing.T) {
 				ZoneSize: "10M",
 				Key:      "${request_uri}",
 			},
-			msg: "only required fields are set",
+			isPlus: false,
+			msg:    "only required fields are set",
 		},
 		{
 			rateLimit: &v1.RateLimit{
@@ -594,14 +596,29 @@ func TestValidateRateLimit_PassesOnValidInput(t *testing.T) {
 				LogLevel:   "info",
 				RejectCode: createPointerFromInt(505),
 			},
-			msg: "ratelimit all fields set",
+			isPlus: false,
+			msg:    "ratelimit all fields set",
+		},
+		{
+			rateLimit: &v1.RateLimit{
+				Rate:     "30r/m",
+				Key:      "${request_uri}",
+				ZoneSize: "10M",
+				Condition: &v1.RateLimitCondition{
+					JWT: &v1.JWTCondition{
+						Claim: "sub",
+						Match: "Gold",
+					},
+					Default: false,
+				},
+			},
+			isPlus: true,
+			msg:    "ratelimit JWT Condition",
 		},
 	}
 
-	isPlus := false
-
 	for _, test := range tests {
-		allErrs := validateRateLimit(test.rateLimit, field.NewPath("rateLimit"), isPlus)
+		allErrs := validateRateLimit(test.rateLimit, field.NewPath("rateLimit"), test.isPlus)
 		if len(allErrs) > 0 {
 			t.Errorf("validateRateLimit() returned errors %v for valid input for the case of %v", allErrs, test.msg)
 		}
@@ -622,56 +639,83 @@ func TestValidateRateLimit_FailsOnInvalidInput(t *testing.T) {
 	t.Parallel()
 	tests := []struct {
 		rateLimit *v1.RateLimit
+		isPlus    bool
 		msg       string
 	}{
 		{
 			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
 				r.Rate = "0r/s"
 			}),
-			msg: "invalid rateLimit rate",
+			isPlus: false,
+			msg:    "invalid rateLimit rate",
 		},
 		{
 			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
 				r.Key = "${fail}"
 			}),
-			msg: "invalid rateLimit key variable use",
+			isPlus: false,
+			msg:    "invalid rateLimit key variable use",
 		},
 		{
 			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
 				r.Delay = createPointerFromInt(0)
 			}),
-			msg: "invalid rateLimit delay",
+			isPlus: false,
+			msg:    "invalid rateLimit delay",
 		},
 		{
 			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
 				r.Burst = createPointerFromInt(0)
 			}),
-			msg: "invalid rateLimit burst",
+			isPlus: false,
+			msg:    "invalid rateLimit burst",
 		},
 		{
 			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
 				r.ZoneSize = "31k"
 			}),
-			msg: "invalid rateLimit zoneSize",
+			isPlus: false,
+			msg:    "invalid rateLimit zoneSize",
 		},
 		{
 			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
 				r.RejectCode = createPointerFromInt(600)
 			}),
-			msg: "invalid rateLimit rejectCode",
+			isPlus: false,
+			msg:    "invalid rateLimit rejectCode",
 		},
 		{
 			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
 				r.LogLevel = "invalid"
 			}),
-			msg: "invalid rateLimit logLevel",
+			isPlus: false,
+			msg:    "invalid rateLimit logLevel",
+		},
+		{
+			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
+				r.Condition = &v1.RateLimitCondition{
+					JWT: &v1.JWTCondition{
+						Claim: "sub",
+						Match: "Gold",
+					},
+				}
+			}),
+			isPlus: false,
+			msg:    "must be plus",
+		},
+		{
+			rateLimit: createInvalidRateLimit(func(r *v1.RateLimit) {
+				r.Condition = &v1.RateLimitCondition{
+					Default: false,
+				}
+			}),
+			isPlus: true,
+			msg:    "missing JWTCondition",
 		},
 	}
 
-	isPlus := false
-
 	for _, test := range tests {
-		allErrs := validateRateLimit(test.rateLimit, field.NewPath("rateLimit"), isPlus)
+		allErrs := validateRateLimit(test.rateLimit, field.NewPath("rateLimit"), test.isPlus)
 		if len(allErrs) == 0 {
 			t.Errorf("validateRateLimit() returned no errors for invalid input for the case of %v", test.msg)
 		}
