diff --git a/storage/storage_dest_test.go b/storage/storage_dest_test.go
new file mode 100644
index 00000000..93b60e69
--- /dev/null
+++ b/storage/storage_dest_test.go
@@ -0,0 +1,105 @@
+package storage
+
+import (
+	"testing"
+
+	"github.com/opencontainers/go-digest"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestLayerID(t *testing.T) {
+	blobDigest, err := digest.Parse("sha256:0000000000000000000000000000000000000000000000000000000000000000")
+	require.NoError(t, err)
+
+	for _, c := range []struct {
+		parentID        string
+		identifiedByTOC bool
+		diffID          string
+		tocDigest       string
+		expected        string
+	}{
+		{
+			parentID:        "",
+			identifiedByTOC: false,
+			diffID:          "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			tocDigest:       "",
+			expected:        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+		},
+		{
+			parentID:        "",
+			identifiedByTOC: false,
+			diffID:          "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			expected:        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			tocDigest:       "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
+		},
+		{
+			parentID:        "",
+			identifiedByTOC: true,
+			diffID:          "",
+			tocDigest:       "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
+			expected:        "07f60ddaf18a3d1fa18a71bf40f0b9889b473e26555d6fffdfbd72ba6a59469e",
+		},
+		{
+			parentID:        "",
+			identifiedByTOC: true,
+			diffID:          "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			tocDigest:       "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
+			expected:        "07f60ddaf18a3d1fa18a71bf40f0b9889b473e26555d6fffdfbd72ba6a59469e",
+		},
+		{
+			parentID:        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			identifiedByTOC: false,
+			diffID:          "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
+			tocDigest:       "",
+			expected:        "76f79efda453922cda1cecb6ec9e7cf9d86ea968c6dd199d4030dd00078a1686",
+		},
+		{
+			parentID:        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			identifiedByTOC: false,
+			diffID:          "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
+			tocDigest:       "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
+			expected:        "76f79efda453922cda1cecb6ec9e7cf9d86ea968c6dd199d4030dd00078a1686",
+		},
+		{
+			parentID:        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			identifiedByTOC: true,
+			diffID:          "",
+			tocDigest:       "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
+			expected:        "468becc3d25ee862f81fd728d229a2b2487cfc9b3e6cf3a4d0af8c3fdde0e7a9",
+		},
+		{
+			parentID:        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			identifiedByTOC: true,
+			diffID:          "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
+			tocDigest:       "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
+			expected:        "468becc3d25ee862f81fd728d229a2b2487cfc9b3e6cf3a4d0af8c3fdde0e7a9",
+		},
+	} {
+		var diffID, tocDigest digest.Digest
+		if c.diffID != "" {
+			diffID, err = digest.Parse(c.diffID)
+			require.NoError(t, err)
+		}
+		if c.tocDigest != "" {
+			tocDigest, err = digest.Parse(c.tocDigest)
+			require.NoError(t, err)
+		}
+
+		res := layerID(c.parentID, trustedLayerIdentityData{
+			layerIdentifiedByTOC: c.identifiedByTOC,
+			diffID:               diffID,
+			tocDigest:            tocDigest,
+			blobDigest:           "",
+		})
+		assert.Equal(t, c.expected, res)
+		// blobDigest does not affect the layer ID
+		res = layerID(c.parentID, trustedLayerIdentityData{
+			layerIdentifiedByTOC: c.identifiedByTOC,
+			diffID:               diffID,
+			tocDigest:            tocDigest,
+			blobDigest:           blobDigest,
+		})
+		assert.Equal(t, c.expected, res)
+	}
+}
diff --git a/storage/storage_test.go b/storage/storage_test.go
index 6f90495f..87068893 100644
--- a/storage/storage_test.go
+++ b/storage/storage_test.go
@@ -9,6 +9,7 @@ import (
 	"context"
 	"crypto/rand"
 	"crypto/sha256"
+	"encoding/json"
 	"errors"
 	"flag"
 	"fmt"
@@ -187,21 +188,25 @@ func TestParseWithGraphDriverOptions(t *testing.T) {
 	}
 }
 
-// makeLayerGoroutine writes to pwriter, and on success, updates uncompressedCount
+// makeLayerGoroutine writes to dest, and on success, updates uncompressedCount and uncompressedDigest
 // before it terminates.
-func makeLayerGoroutine(pwriter io.Writer, uncompressedCount *int64, compression archive.Compression) error {
-	var uncompressed *ioutils.WriteCounter
+func makeLayerGoroutine(dest io.Writer, uncompressedCount *int64, uncompressedDigest *digest.Digest, compression archive.Compression) (retErr error) {
 	if compression != archive.Uncompressed {
-		compressor, err := archive.CompressStream(pwriter, compression)
+		compressor, err := archive.CompressStream(dest, compression)
 		if err != nil {
 			return fmt.Errorf("compressing layer: %w", err)
 		}
-		defer compressor.Close()
-		uncompressed = ioutils.NewWriteCounter(compressor)
-	} else {
-		uncompressed = ioutils.NewWriteCounter(pwriter)
+		defer func() {
+			if err := compressor.Close(); err != nil && retErr == nil {
+				retErr = err
+			}
+		}()
+		dest = compressor
 	}
-	twriter := tar.NewWriter(uncompressed)
+
+	uncompressedCounter := ioutils.NewWriteCounter(dest)
+	uncompressedDigester := digest.Canonical.Digester()
+	twriter := tar.NewWriter(io.MultiWriter(uncompressedCounter, uncompressedDigester.Hash()))
 	// 	defer twriter.Close()
 	// should be called here to correctly terminate the archive.
 	// We do not do that, to workaround https://github.com/containers/storage/issues/1729 :
@@ -244,36 +249,40 @@ func makeLayerGoroutine(pwriter io.Writer, uncompressedCount *int64, compression
 	if err := twriter.Flush(); err != nil {
 		return fmt.Errorf("Error flushing output to tar archive: %w", err)
 	}
-	*uncompressedCount = uncompressed.Count
+	*uncompressedCount = uncompressedCounter.Count
+	*uncompressedDigest = uncompressedDigester.Digest()
 	return nil
 }
 
 type testBlob struct {
-	compressedDigest digest.Digest
-	uncompressedSize int64
-	compressedSize   int64
-	data             []byte
+	uncompressedDigest digest.Digest
+	compressedDigest   digest.Digest
+	uncompressedSize   int64
+	compressedSize     int64
+	data               []byte
 }
 
 func makeLayer(t *testing.T, compression archive.Compression) testBlob {
 	preader, pwriter := io.Pipe()
 	var uncompressedCount int64
+	var uncompressedDigest digest.Digest
 	go func() {
 		err := errors.New("Internal error: unexpected panic in makeLayer")
 		defer func() { // Note that this is not the same as {defer pipeWriter.CloseWithError(err)}; we need err to be evaluated lazily.
 			_ = pwriter.CloseWithError(err)
 		}()
-		err = makeLayerGoroutine(pwriter, &uncompressedCount, compression)
+		err = makeLayerGoroutine(pwriter, &uncompressedCount, &uncompressedDigest, compression)
 	}()
 
 	tbuffer := bytes.Buffer{}
 	_, err := io.Copy(&tbuffer, preader)
 	require.NoError(t, err)
 	return testBlob{
-		compressedDigest: digest.SHA256.FromBytes(tbuffer.Bytes()),
-		uncompressedSize: uncompressedCount,
-		compressedSize:   int64(tbuffer.Len()),
-		data:             tbuffer.Bytes(),
+		uncompressedDigest: uncompressedDigest,
+		compressedDigest:   digest.SHA256.FromBytes(tbuffer.Bytes()),
+		uncompressedSize:   uncompressedCount,
+		compressedSize:     int64(tbuffer.Len()),
+		data:               tbuffer.Bytes(),
 	}
 }
 
@@ -305,6 +314,40 @@ func ensureTestCanCreateImages(t *testing.T) {
 	}
 }
 
+// configForLayers returns a minimally-plausible config for layers
+func configForLayers(t *testing.T, layers []testBlob) testBlob {
+	rootFS := manifest.Schema2RootFS{
+		Type:    "layers",
+		DiffIDs: []digest.Digest{},
+	}
+	for _, l := range layers {
+		rootFS.DiffIDs = append(rootFS.DiffIDs, l.uncompressedDigest)
+	}
+	// Add a unique label so that different calls to configForLayers don’t try to use the same image ID.
+	randomBytes := make([]byte, digest.Canonical.Size())
+	_, err := rand.Read(randomBytes)
+	require.NoError(t, err)
+	config := manifest.Schema2Image{
+		Schema2V1Image: manifest.Schema2V1Image{
+			Config: &manifest.Schema2Config{
+				Labels: map[string]string{"unique": fmt.Sprintf("%x", randomBytes)},
+			},
+			Created: time.Now(),
+		},
+		RootFS: &rootFS,
+	}
+	configBytes, err := json.Marshal(config)
+	require.NoError(t, err)
+	configDigest := digest.Canonical.FromBytes(configBytes)
+	return testBlob{
+		uncompressedDigest: configDigest,
+		compressedDigest:   configDigest,
+		uncompressedSize:   int64(len(configBytes)),
+		compressedSize:     int64(len(configBytes)),
+		data:               configBytes,
+	}
+}
+
 func createUncommittedImageDest(t *testing.T, ref types.ImageReference, cache types.BlobInfoCache,
 	layers []testBlob, config *testBlob) (types.ImageDestination, types.UnparsedImage) {
 	dest, err := ref.NewImageDestination(context.Background(), nil)
@@ -316,21 +359,11 @@ func createUncommittedImageDest(t *testing.T, ref types.ImageReference, cache ty
 		layerDescriptors = append(layerDescriptors, desc)
 	}
 
-	var configDescriptor manifest.Schema2Descriptor
-	if config != nil {
-		configDescriptor = config.storeBlob(t, dest, cache, manifest.DockerV2Schema2ConfigMediaType, true)
-	} else {
-		// Use a random digest so that different calls to createUncommittedImageDest with config == nil don’t try to
-		// use the same image ID.
-		digestBytes := make([]byte, digest.Canonical.Size())
-		_, err := rand.Read(digestBytes)
-		require.NoError(t, err)
-		configDescriptor = manifest.Schema2Descriptor{
-			MediaType: manifest.DockerV2Schema2ConfigMediaType,
-			Size:      1,
-			Digest:    digest.NewDigestFromBytes(digest.Canonical, digestBytes),
-		}
+	if config == nil {
+		cd := configForLayers(t, layers)
+		config = &cd
 	}
+	configDescriptor := config.storeBlob(t, dest, cache, manifest.DockerV2Schema2ConfigMediaType, true)
 
 	manifest := manifest.Schema2FromComponents(configDescriptor, layerDescriptors)
 	manifestBytes, err := manifest.Serialize()
@@ -358,14 +391,6 @@ func createImage(t *testing.T, ref types.ImageReference, cache types.BlobInfoCac
 func TestWriteRead(t *testing.T) {
 	ensureTestCanCreateImages(t)
 
-	configBytes := []byte(`{"config":{"labels":{}},"created":"2006-01-02T15:04:05Z"}`)
-	config := testBlob{
-		compressedDigest: digest.SHA256.FromBytes(configBytes),
-		uncompressedSize: int64(len(configBytes)),
-		compressedSize:   int64(len(configBytes)),
-		data:             configBytes,
-	}
-
 	manifests := []string{
 		//`{
 		//    "schemaVersion": 2,
@@ -443,6 +468,7 @@ func TestWriteRead(t *testing.T) {
 		layer := makeLayer(t, compression)
 		_ = layer.storeBlob(t, dest, cache, manifest.DockerV2Schema2LayerMediaType, false)
 		t.Logf("Wrote randomly-generated layer %q (%d/%d bytes) to destination", layer.compressedDigest, layer.compressedSize, layer.uncompressedSize)
+		config := configForLayers(t, []testBlob{layer})
 		_ = config.storeBlob(t, dest, cache, manifest.DockerV2Schema2ConfigMediaType, true)
 
 		manifest := strings.ReplaceAll(manifestFmt, "%lh", layer.compressedDigest.String())
@@ -635,18 +661,13 @@ func TestSize(t *testing.T) {
 
 	layer1 := makeLayer(t, archive.Gzip)
 	layer2 := makeLayer(t, archive.Gzip)
-	configBytes := []byte(`{"config":{"labels":{}},"created":"2006-01-02T15:04:05Z"}`)
-	config := testBlob{
-		compressedDigest: digest.SHA256.FromBytes(configBytes),
-		uncompressedSize: int64(len(configBytes)),
-		compressedSize:   int64(len(configBytes)),
-		data:             configBytes,
-	}
+	layerBlobs := []testBlob{layer1, layer2}
+	config := configForLayers(t, layerBlobs)
 
 	ref, err := Transport.ParseReference("test")
 	require.NoError(t, err)
 
-	createImage(t, ref, cache, []testBlob{layer1, layer2}, &config)
+	createImage(t, ref, cache, layerBlobs, &config)
 
 	img, err := ref.NewImage(context.Background(), nil)
 	require.NoError(t, err)
@@ -673,15 +694,9 @@ func TestDuplicateBlob(t *testing.T) {
 
 	layer1 := makeLayer(t, archive.Gzip)
 	layer2 := makeLayer(t, archive.Gzip)
-	configBytes := []byte(`{"config":{"labels":{}},"created":"2006-01-02T15:04:05Z"}`)
-	config := testBlob{
-		compressedDigest: digest.SHA256.FromBytes(configBytes),
-		uncompressedSize: int64(len(configBytes)),
-		compressedSize:   int64(len(configBytes)),
-		data:             configBytes,
-	}
-
-	createImage(t, ref, cache, []testBlob{layer1, layer2, layer1, layer2}, &config)
+	layerBlobs := []testBlob{layer1, layer2, layer1, layer2}
+	config := configForLayers(t, layerBlobs)
+	createImage(t, ref, cache, layerBlobs, &config)
 
 	img, err := ref.NewImage(context.Background(), nil)
 	require.NoError(t, err)
