diff --git a/src/cmd/logic-function.test.js b/src/cmd/logic-function.test.js
index 058aedd1..9539157d 100644
--- a/src/cmd/logic-function.test.js
+++ b/src/cmd/logic-function.test.js
@@ -21,6 +21,7 @@ describe('LogicFunctionCommands', () => {
 			stderr: {
 				write: sinon.stub()
 			},
+			showBusySpinnerUntilResolved: sinon.stub().callsFake((text, promise) => promise),
 			prompt: sinon.stub(),
 			chalk: {
 				bold: sinon.stub().callsFake((str) => str),
@@ -60,7 +61,7 @@ describe('LogicFunctionCommands', () => {
 			});
 		});
 
-		it('lists logic functions in Sandbox account', async () => {
+		it('lists Logic Functions in Sandbox account', async () => {
 
 			const logicListStub = sinon.stub(LogicFunction, 'listFromCloud').resolves(logicFunc1.logic_functions);
 			await logicFunctionCommands.list({});
@@ -70,7 +71,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`- LF1 (enabled)${os.EOL}`);
 		});
 
-		it('lists logic functions in an org', async () => {
+		it('lists Logic Functions in an org', async () => {
 			const logicListStub = sinon.stub(LogicFunction, 'listFromCloud').resolves(logicFunc1.logic_functions);
 			await logicFunctionCommands.list({ org: 'particle' });
 			expect(logicListStub.calledWith({ api: logicFunctionCommands.api, org: 'particle' })).to.be.true;
@@ -79,7 +80,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`- LF1 (enabled)${os.EOL}`);
 		});
 
-		it('shows help if no logic functions are found', async () => {
+		it('shows help if no Logic Functions are found', async () => {
 			const logicListStub = sinon.stub(LogicFunction, 'listFromCloud').resolves([]);
 			await logicFunctionCommands.list({});
 			expect(logicListStub.calledWith({ api: logicFunctionCommands.api, org: undefined })).to.be.true;
@@ -87,7 +88,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`No Logic Functions deployed in your Sandbox.${os.EOL}`);
 		});
 
-		it('shows help if no logic functions are found', async () => {
+		it('shows help if no Logic Functions are found', async () => {
 			const logicListStub = sinon.stub(LogicFunction, 'listFromCloud').resolves([]);
 			await logicFunctionCommands.list({ api: logicFunctionCommands.api, org: 'particle' });
 			expect(logicListStub.calledWith({ api: logicFunctionCommands.api, org: 'particle' })).to.be.true;
@@ -112,7 +113,7 @@ describe('LogicFunctionCommands', () => {
 			};
 		});
 
-		it('gets a logic function with an specific name from Sandbox account', async () => {
+		it('gets a Logic Function with an specific name from Sandbox account', async () => {
 			const logicGetStub = sinon.stub(LogicFunction, 'getByIdOrName').resolves(lf);
 			lf.saveToDisk = sinon.stub().resolves(true);
 			sinon.stub(LogicFunction, 'listFromCloud').resolves(logicFunc1.logic_functions);
@@ -124,7 +125,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(` - ${lf.files.configuration.name}${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(` - ${lf.files.sourceCode.name}${os.EOL}`);
 		});
-		it('gets a logic function with an specific id from Sandbox account', async () => {
+		it('gets a Logic Function with an specific id from Sandbox account', async () => {
 			const logicGetStub = sinon.stub(LogicFunction, 'getByIdOrName').resolves(lf);
 			lf.saveToDisk = sinon.stub().resolves(true);
 			sinon.stub(LogicFunction, 'listFromCloud').resolves(logicFunc1.logic_functions);
@@ -137,7 +138,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(` - ${lf.files.sourceCode.name}${os.EOL}`);
 
 		});
-		it('shows error if logic function is not found', async () => {
+		it('shows error if Logic Function is not found', async () => {
 			const logicGetStub = sinon.stub(LogicFunction, 'getByIdOrName').rejects(new Error('Logic function not found'));
 			lf.saveToDisk = sinon.stub().resolves(true);
 			sinon.stub(LogicFunction, 'listFromCloud').resolves(logicFunc1.logic_functions);
@@ -152,7 +153,7 @@ describe('LogicFunctionCommands', () => {
 			expect(error.message).to.equal('Logic function not found');
 		});
 
-		it('gets a logic function with an specific name from an org', async () => {
+		it('gets a Logic Function with an specific name from an org', async () => {
 			const logicGetStub = sinon.stub(LogicFunction, 'getByIdOrName').resolves(lf);
 			lf.saveToDisk = sinon.stub().resolves(true);
 			sinon.stub(LogicFunction, 'listFromCloud').resolves(logicFunc1.logic_functions);
@@ -176,7 +177,7 @@ describe('LogicFunctionCommands', () => {
 			sinon.restore();
 		});
 
-		it('creates a logic function locally for Sandbox account', async () => {
+		it('creates a Logic Function locally for Sandbox account', async () => {
 			logicFunctionCommands.ui.prompt = sinon.stub();
 			logicFunctionCommands.ui.prompt.onCall(0).resolves({ name: 'logic func 1' });
 			logicFunctionCommands.ui.prompt.onCall(1).resolves({ description: 'Logic Function 1' });
@@ -204,7 +205,7 @@ describe('LogicFunctionCommands', () => {
 	});
 
 	describe('execute', () => {
-		it('executes a logic function with user provided data', async () => {
+		it('executes a Logic Function with user provided data', async () => {
 			const logicFunction = new LogicFunction({
 				name: 'LF1',
 				description: 'Logic Function 1',
@@ -224,12 +225,12 @@ describe('LogicFunctionCommands', () => {
 			expect(logicStub.calledOnce).to.be.true;
 			expect(executeStub.calledOnce).to.be.true;
 			expect(logicFunctionCommands.ui.prompt.callCount).to.equal(0);
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF1 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF1 for your Sandbox...', logicFunction.execute());
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Execution Status: Success${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Logs from Execution:${os.EOL}`);
 		});
 
-		it('throws an error if there is no logic function in the directory', async () => {
+		it('throws an error if there is no Logic Function in the directory', async () => {
 			const logicStub = sinon.stub(LogicFunction, 'listFromDisk').resolves({
 				malformedLogicFunctions: [],
 				logicFunctions: []
@@ -245,7 +246,7 @@ describe('LogicFunctionCommands', () => {
 			expect(error.message).to.equal('No Logic Functions found');
 		});
 
-		it('executes a logic function with user provided data from file', async () => {
+		it('executes a Logic Function with user provided data from file', async () => {
 			const filepath = PATH_TMP_DIR;
 			const logicFunction = new LogicFunction({
 				name: 'LF1',
@@ -265,7 +266,7 @@ describe('LogicFunctionCommands', () => {
 			// called with the file path
 			expect(logicStub).to.have.been.calledWith({ filepath, api: logicFunctionCommands.api, org: logicFunctionCommands.org });
 		});
-		it('executes a logic function with user provided data and shows error', async () => {
+		it('executes a Logic Function with user provided data and shows error', async () => {
 			const logicFunction = new LogicFunction({
 				name: 'LF1',
 				description: 'Logic Function 1',
@@ -284,11 +285,11 @@ describe('LogicFunctionCommands', () => {
 			await logicFunctionCommands.execute({ data: '{"eventData": "someData"}' , params: {} });
 
 			expect(logicStub.calledOnce).to.be.true;
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF1 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF1 for your Sandbox...', logicFunction.execute());
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Execution Status: Exception${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Error during Execution:${os.EOL}`);
 		});
-		it('prompts if found multiple logic functions', async () => {
+		it('prompts if found multiple Logic Functions', async () => {
 			const logicFunction1 = new LogicFunction({
 				name: 'LF1',
 				description: 'Logic Function 1',
@@ -318,7 +319,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicStub.calledOnce).to.be.true;
 			expect(logicFunctionCommands.ui.prompt.callCount).to.equal(1);
 		});
-		it('executes a logic function with provided logic function name', async () => {
+		it('executes a Logic function with provided Logic Function name', async () => {
 			const lf1 = new LogicFunction({
 				name: 'LF1',
 				description: 'Logic Function 1',
@@ -346,11 +347,11 @@ describe('LogicFunctionCommands', () => {
 			expect(logicStub.calledOnce).to.be.true;
 			expect(executeStub.calledOnce).to.be.true;
 			expect(logicFunctionCommands.ui.prompt.callCount).to.equal(0);
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF1 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF1 for your Sandbox...', lf1.execute());
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Execution Status: Success${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Logs from Execution:${os.EOL}`);
 		});
-		it('executes a logic function with provided logic function id', async () => {
+		it('executes a Logic Function with provided Logic Function id', async () => {
 			const lf1 = new LogicFunction({
 				name: 'LF1',
 				description: 'Logic Function 1',
@@ -378,7 +379,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicStub.calledOnce).to.be.true;
 			expect(executeStub.calledOnce).to.be.true;
 			expect(logicFunctionCommands.ui.prompt.callCount).to.equal(0);
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF2 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF2 for your Sandbox...', lf2.execute());
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Execution Status: Success${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Logs from Execution:${os.EOL}`);
 		});
@@ -404,7 +405,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicStub.calledOnce).to.be.true;
 			expect(executeStub.calledOnce).to.be.true;
 			expect(logicFunctionCommands.ui.prompt.callCount).to.equal(0);
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF1 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF1 for your Sandbox...', lf1.execute());
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Execution Status: Success${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Logs from Execution:${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`The following Logic Functions are not valid:${os.EOL}`);
@@ -451,7 +452,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicStub.calledOnce).to.be.true;
 			expect(executeStub.calledOnce).to.be.true;
 			expect(logicFunctionCommands.ui.prompt.callCount).to.equal(0);
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF1 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF1 for your Sandbox...', lf1.execute());
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Execution Status: Success${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Logs from Execution:${os.EOL}`);
 			expect(logicFunctionCommands.ui.stdout.write).to.not.have.been.calledWith(`The following Logic Functions are not valid:${os.EOL}`);
@@ -524,7 +525,7 @@ describe('LogicFunctionCommands', () => {
 	});
 
 	describe('_selectLogicFunctionName', () => {
-		it('selects logic function from a list', async () => {
+		it('selects Logic Function from a list', async () => {
 			const logicFunctions = ['logicFunc1', 'logicFunc2'];
 			const selectedLF = 'logicFunc2';
 			const promptStub = sinon.stub(logicFunctionCommands, '_prompt');
@@ -536,9 +537,9 @@ describe('LogicFunctionCommands', () => {
 			sinon.assert.calledOnceWithExactly(promptStub, {
 				type: 'list',
 				name: 'logic_function',
-				message: 'Which logic function would you like to download?',
+				message: 'Which Logic Function would you like to download?',
 				choices: logicFunctions,
-				nonInteractiveError: 'Provide name for the logic function',
+				nonInteractiveError: 'Provide name for the Logic Function',
 			});
 		});
 
@@ -553,7 +554,7 @@ describe('LogicFunctionCommands', () => {
 			}
 
 			expect(error).to.be.an.instanceOf(Error);
-			expect(error.message).to.eql('No logic functions found');
+			expect(error.message).to.eql('No Logic Functions found');
 		});
 	});
 
@@ -636,7 +637,7 @@ describe('LogicFunctionCommands', () => {
 			fs.rmSync('lf1', { recursive: true, force: true });
 		});
 
-		it('deploys a new logic function', async() => {
+		it('deploys a new Logic Function', async() => {
 			const logicFunction = new LogicFunction({
 				name: 'LF1',
 				description: 'Logic Function 1',
@@ -663,10 +664,10 @@ describe('LogicFunctionCommands', () => {
 			expect(executeStub).to.have.been.calledOnce;
 			expect(logicStub).to.have.been.calledOnce;
 			expect(saveStub).to.have.been.calledOnce;
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF1 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF1 for your Sandbox...', logicFunction.execute());
 		});
 
-		it('deploys an existent logic function', async() => {
+		it('deploys an existent Logic Function', async() => {
 			const logicFunction = new LogicFunction({
 				name: 'LF1',
 				description: 'Logic Function 1',
@@ -693,7 +694,7 @@ describe('LogicFunctionCommands', () => {
 			expect(executeStub).to.have.been.calledOnce;
 			expect(logicStub).to.have.been.calledOnce;
 			expect(saveStub).to.have.been.calledOnce;
-			expect(logicFunctionCommands.ui.stdout.write).calledWith(`Executing Logic Function LF1 for your Sandbox...${os.EOL}`);
+			expect(logicFunctionCommands.ui.showBusySpinnerUntilResolved).calledWith('Executing Logic Function LF1 for your Sandbox...', logicFunction.execute());
 			expect(logicFunctionCommands._prompt).calledWith(sinon.match(expectedPromptCall));
 			expect(logicFunction.id).to.equal('0021e8f4-64ee-416d-83f3-898aa909fb1b');
 		});
@@ -750,7 +751,7 @@ describe('LogicFunctionCommands', () => {
 			sinon.restore();
 		});
 
-		it('enable a logic function with name', async() => {
+		it('enable a Logic Function with name', async() => {
 			const updateStub = sinon.stub(logicFunction, 'updateToCloud').resolves(undefined);
 			const listLocalStub = sinon.stub(LogicFunction, 'listFromDisk').resolves({
 				malformedLogicFunctions: [],
@@ -764,7 +765,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicFunction.enabled).to.be.true;
 		});
 
-		it('enable a logic function with id', async() => {
+		it('enable a Logic Function with id', async() => {
 			const updateStub = sinon.stub(logicFunction, 'updateToCloud').resolves(undefined);
 			const listLocalStub = sinon.stub(LogicFunction, 'listFromDisk').resolves({
 				malformedLogicFunctions: [],
@@ -778,7 +779,7 @@ describe('LogicFunctionCommands', () => {
 			expect(logicFunction.enabled).to.be.true;
 		});
 
-		it('fails to enable a logic function', async() => {
+		it('fails to enable a Logic Function', async() => {
 			let error;
 			const updateStub = sinon.stub(logicFunction, 'updateToCloud').rejects(new Error('Error enabling Logic Function LF1'));
 			const listLocalStub = sinon.stub(LogicFunction, 'listFromDisk').resolves({
@@ -797,7 +798,7 @@ describe('LogicFunctionCommands', () => {
 			expect(error.message).to.contain('Error enabling Logic Function LF1');
 		});
 
-		it('updates the local logic function if it exists in the path', async() => {
+		it('updates the local Logic Function if it exists in the path', async() => {
 			const updateStub = sinon.stub(logicFunction, 'updateToCloud').resolves(undefined);
 			const listLocalStub = sinon.stub(LogicFunction, 'listFromDisk').resolves({
 				malformedLogicFunctions: [],
diff --git a/src/lib/logic-function.test.js b/src/lib/logic-function.test.js
index 6b06319a..90136b7c 100644
--- a/src/lib/logic-function.test.js
+++ b/src/lib/logic-function.test.js
@@ -25,25 +25,25 @@ describe('LogicFunction', () => {
 		return logicFunction;
 	}
 	describe('list', () => {
-		it('returns an empty array if there are no logic functions', async () => {
+		it('returns an empty array if there are no Logic Functions', async () => {
 			nock('https://api.particle.io/v1/', )
 				.intercept('/logic/functions', 'GET')
 				.reply(200, { logic_functions : [] } );
 			const logicFunctions = await LogicFunction.listFromCloud();
 			expect(logicFunctions).to.have.lengthOf(0);
 		});
-		it('returns a list of logic functions', async () => {
+		it('returns a list of Logic Functions', async () => {
 			nock('https://api.particle.io/v1/', )
 				.intercept('/logic/functions', 'GET')
 				.reply(200, { logic_functions : [{ name: 'my-logic-function' }] } );
 
 			const logicFunctions = await LogicFunction.listFromCloud();
 			expect(logicFunctions).to.have.lengthOf(1);
-			// check if the logic function is an instance of LogicFunction
+			// check if the Logic Function is an instance of LogicFunction
 			expect(logicFunctions[0]).to.be.an.instanceof(LogicFunction);
 		});
 
-		it('returns a list of logic functions from an specific org', async () => {
+		it('returns a list of Logic Functions from an specific org', async () => {
 			nock('https://api.particle.io/v1/orgs/', )
 				.intercept('/my-org/logic/functions', 'GET')
 				.reply(200, { logic_functions : [{ name: 'my-logic-function' }] } );
@@ -60,20 +60,20 @@ describe('LogicFunction', () => {
 			try {
 				await LogicFunction.listFromCloud({ org: 'my-org' });
 			} catch (error) {
-				expect(error.message).to.equal('Error listing logic functions: Internal Server Error');
+				expect(error.message).to.equal('Error listing Logic Functions: Internal Server Error');
 			}
 		});
 	});
 	describe('getByIdOrName', () => {
-		it ('throws an error if the logic function is not found', async () => {
+		it ('throws an error if the Logic Function is not found', async () => {
 			try {
 				const logicFunctions = [];
 				await LogicFunction.getByIdOrName({ org: 'my-org', name: 'my-logic-function', list: logicFunctions });
 			} catch (error) {
-				expect(error.message).to.equal('Logic function not found');
+				expect(error.message).to.equal('Logic Function not found');
 			}
 		});
-		it('returns a logic function by name', async () => {
+		it('returns a Logic Function by name', async () => {
 			const logicFunctions = [new LogicFunction(logicFunc1.logic_functions[0])];
 			logicFunctions[0].files.sourceCode.content = logicFunc1.logic_functions[0].source.code;
 			const logicFunction = await LogicFunction.getByIdOrName({ name: 'LF1', list: logicFunctions });
@@ -103,7 +103,7 @@ describe('LogicFunction', () => {
 		});
 	});
 	describe('saveToDisk', () => {
-		it('saves a logic function to disk', async () => {
+		it('saves a Logic Function to disk', async () => {
 			const logicFunction = new LogicFunction(logicFunc1.logic_functions[0]);
 			logicFunction.path = path.join(PATH_TMP_DIR, 'logic-functions', 'lf1');
 			await logicFunction.saveToDisk();
@@ -128,10 +128,10 @@ describe('LogicFunction', () => {
 			sinon.restore();
 		});
 
-		it('initializes a logic function from a template', async () => {
+		it('initializes a Logic Function from a template', async () => {
 			const config = {
 				logic_function: {
-					'name': 'my logic function',
+					'name': 'my Logic Function',
 					'description': 'my test description',
 					'enabled': true,
 					'source': {
@@ -141,7 +141,7 @@ describe('LogicFunction', () => {
 				}
 			};
 			const loadStub = sinon.stub(templateProcessor, 'loadTemplateFiles').resolves([
-				{ fileName: path.join('my_path', 'lf1.js'), content: 'logic function content' },
+				{ fileName: path.join('my_path', 'lf1.js'), content: 'Logic Function content' },
 				{ fileName: path.join('my_path', 'lf1.logic.json'), content: JSON.stringify(config) }
 			]);
 			const logicFunction = new LogicFunction({
@@ -187,58 +187,58 @@ describe('LogicFunction', () => {
 			sinon.restore();
 			fs.emptyDirSync(PATH_TMP_DIR);
 		});
-		it('returns a list of logic functions', async () => {
+		it('returns a list of Logic Functions', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions') });
 			expect(logicFunctions).to.have.lengthOf(1);
 			expect(logicFunctions[0]).to.be.an.instanceof(LogicFunction);
 		});
-		it('returns a list of logic functions with org', async () => {
+		it('returns a list of Logic Functions with org', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions'), org: 'my-org' });
 			expect(logicFunctions).to.have.lengthOf(1);
 			expect(logicFunctions[0]).to.be.an.instanceof(LogicFunction);
 			expect(logicFunctions[0]).to.have.property('org', 'my-org');
 		});
-		it('returns a list of more than one logic functions', async () => {
+		it('returns a list of more than one Logic Functions', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			await createLogicFunction({ name: 'lf2', description: 'Logic Function 2 on SandBox' });
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions') });
 			expect(logicFunctions).to.have.lengthOf(2);
 			expect(logicFunctions[0]).to.be.an.instanceof(LogicFunction);
 		});
-		it('filter out files that are not logic functions', async () => {
+		it('filter out files that are not Logic Functions', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			await fs.writeFile(path.join(PATH_TMP_DIR, 'logic_functions', 'lf2.json'), '{}');
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions') });
 			expect(logicFunctions).to.have.lengthOf(1);
 			expect(logicFunctions[0]).to.be.an.instanceof(LogicFunction);
 		});
-		it('filter out no well formed logic functions', async () => {
+		it('filter out no well formed Logic Functions', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			await fs.writeFile(path.join(PATH_TMP_DIR, 'logic_functions', 'lf2.logic.json'), '{ "name": "lf2", "description": "Logic Function 2 on SandBox" }');
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions') });
 			expect(logicFunctions).to.have.lengthOf(1);
 			expect(logicFunctions[0]).to.be.an.instanceof(LogicFunction);
 		});
-		it('returns a list of one element if the path is a file and it is a logic function', async () => {
+		it('returns a list of one element if the path is a file and it is a Logic Function', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions', 'lf1.js') });
 			expect(logicFunctions).to.have.lengthOf(1);
 			expect(logicFunctions[0]).to.be.an.instanceof(LogicFunction);
 		});
-		it ('returns an empty list if the path is a file and it is not a logic function', async () => {
+		it ('returns an empty list if the path is a file and it is not a Logic Function', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			await fs.writeFile(path.join(PATH_TMP_DIR, 'logic_functions', 'lf2.js'), '{ "name": "lf2", "description": "Logic Function 2 on SandBox" }');
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions', 'lf2.js') });
 			expect(logicFunctions).to.have.lengthOf(0);
 		});
-		it('returns an empty list if there are no logic functions', async () => {
+		it('returns an empty list if there are no Logic Functions', async () => {
 			const { logicFunctions } = await LogicFunction.listFromDisk({ filepath: PATH_TMP_DIR });
 			expect(logicFunctions).to.have.lengthOf(0);
 		});
 
-		it('returns a list of logic functions and a list of malformed logic functions', async () => {
+		it('returns a list of Logic Functions and a list of malformed Logic Functions', async () => {
 			await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			await fs.writeFile(path.join(PATH_TMP_DIR, 'logic_functions', 'lf2.logic.json'), '{ "name": "lf2", "description": "Logic Function 2 on SandBox" }');
 			const { logicFunctions, malformedLogicFunctions } = await LogicFunction.listFromDisk({ filepath: path.join(PATH_TMP_DIR, 'logic_functions') });
@@ -261,7 +261,7 @@ describe('LogicFunction', () => {
 			sinon.restore();
 			fs.emptyDirSync(PATH_TMP_DIR);
 		});
-		it('executes a logic function', async () => {
+		it('executes a Logic Function', async () => {
 			nock('https://api.particle.io/v1/', )
 				.intercept('/logic/execute', 'POST')
 				.reply(200, { result: { status: 'Success', logs: ['abc1'] } });
@@ -315,7 +315,7 @@ describe('LogicFunction', () => {
 				await lf1.execute(trigger);
 				expect.fail('Should have thrown an error');
 			} catch (error) {
-				expect(error.message).to.equal('Error executing logic function: Internal Server Error');
+				expect(error.message).to.equal('Error executing Logic Function: Internal Server Error');
 			}
 		});
 	});
@@ -325,19 +325,19 @@ describe('LogicFunction', () => {
 			sinon.restore();
 			fs.emptyDirSync(PATH_TMP_DIR);
 		});
-		it('deploys a new logic function', async () => {
+		it('deploys a new Logic Function', async () => {
 			nock('https://api.particle.io/v1/', )
 				.intercept('/logic/functions', 'POST')
-				.reply(200, { id: '1234', version: 1 } );
+				.reply(200, { logic_function: { id: '1234', version: 1 } });
 			const lf1 = await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			await lf1.deploy();
 			expect(lf1).to.have.property('id', '1234');
 			expect(lf1).to.have.property('version', 1);
 		});
-		it ('deploys existent logic function', async () => {
+		it ('deploys existent Logic Function', async () => {
 			nock('https://api.particle.io/v1/', )
 				.intercept('/logic/functions/1234', 'PUT')
-				.reply(200, { id: '1234', version: 1 } );
+				.reply(200, { logic_function: { id: '1234', version: 1 } } );
 			const lf1 = await createLogicFunction({ name: 'lf1', description: 'Logic Function 1 on SandBox' });
 			lf1.id = '1234';
 			await lf1.deploy();
@@ -353,7 +353,7 @@ describe('LogicFunction', () => {
 				await lf1.deploy();
 				expect.fail('Should have thrown an error');
 			} catch (error) {
-				expect(error.message).to.equal('Error deploying logic function: Internal Server Error');
+				expect(error.message).to.equal('Error deploying Logic Function: Internal Server Error');
 			}
 		});
 		it('propagates errors when updating', async () => {
@@ -366,7 +366,7 @@ describe('LogicFunction', () => {
 				await lf1.deploy();
 				expect.fail('Should have thrown an error');
 			} catch (error) {
-				expect(error.message).to.equal('Error deploying logic function: Internal Server Error');
+				expect(error.message).to.equal('Error deploying Logic Function: Internal Server Error');
 			}
 		});
 	});
diff --git a/test/e2e/help.e2e.js b/test/e2e/help.e2e.js
index bc9af93b..6cfff7cb 100644
--- a/test/e2e/help.e2e.js
+++ b/test/e2e/help.e2e.js
@@ -23,7 +23,7 @@ describe('Help & Unknown Command / Argument Handling', () => {
 		'  keys            Manage your device\'s key pair and server public key',
 		'  library         Manage firmware libraries',
 		'  list            Display a list of your devices, as well as their variables and functions',
-		'  logic-function  Create, execute, and deploy logic functions',
+		'  logic-function  Create, execute, and deploy Logic Functions',
 		'  login           Login to the cloud and store an access token locally',
 		'  logout          Log out of your session and clear your saved access token',
 		'  monitor         Connect and display messages from a device',
diff --git a/test/e2e/logic-function.e2e.js b/test/e2e/logic-function.e2e.js
index a9d45681..8b74595a 100644
--- a/test/e2e/logic-function.e2e.js
+++ b/test/e2e/logic-function.e2e.js
@@ -8,19 +8,19 @@ describe('Logic Function Commands', () => {
 	let id;
 
 	const help = [
-		'Create, execute, and deploy logic functions',
+		'Create, execute, and deploy Logic Functions',
 		'Usage: particle logic-function <command>',
 		'Help:  particle help logic-function <command>',
 		'',
 		'Commands:',
-		'  list     Lists the deployed logic functions',
-		'  get      Downloads the logic function',
-		'  create   Creates a logic function',
-		'  execute  Executes a logic function with user provided data',
-		'  deploy   Deploys a logic function to the cloud',
-		'  disable  Disables a logic function in the cloud',
-		'  enable   Enables a logic function in the cloud',
-		'  delete   Deletes a logic function from the cloud',
+		'  list     Lists the deployed Logic Functions',
+		'  get      Downloads the Logic Function',
+		'  create   Creates a Logic Function',
+		'  execute  Executes a Logic Function with user provided data',
+		'  deploy   Deploys a Logic Function to the cloud',
+		'  disable  Disables a Logic Function in the cloud',
+		'  enable   Enables a Logic Function in the cloud',
+		'  delete   Deletes a Logic Function from the cloud',
 		'  logs     Shows logs from a Logic Function',
 		'',
 		'Global Options:',
@@ -48,13 +48,12 @@ describe('Logic Function Commands', () => {
 		'',
 		'Guidelines for creating your Logic Function can be found here https://docs.particle.io/getting-started/cloud/logic/',
 		'Once you have written your Logic Function, run',
-		'- \'particle logic-function execute\' to run your Function',
-		'- \'particle logic-function deploy\' to deploy your new changes',
+		'- particle logic-function execute to run your Function',
+		'- particle logic-function deploy to deploy your new changes',
 		''
 	];
 
 	const executeOutput = [
-		'Executing Logic Function newlf for cyberdyne-systems...',
 		'',
 		'Execution Status: Success',
 		'Logs from Execution:',
@@ -63,7 +62,6 @@ describe('Logic Function Commands', () => {
 	];
 
 	const deployOutput = [
-		'Executing Logic Function newlf for cyberdyne-systems...',
 		'',
 		'Execution Status: Success',
 		'Logs from Execution:',
@@ -74,7 +72,7 @@ describe('Logic Function Commands', () => {
 		// 'Deploying Logic Function lf3 to for cyberdyne-systems...',
 		// 'Success! Logic Function name deployed with ID: bbd75c65-0db2-44bd-9d35-8ce6db9885e3',
 		'',
-		'Visit \'console.particle.io\' to view results from your device(s)!',
+		'Visit console.particle.io to view results from your device(s)!',
 	];
 
 	const getOutput = [
@@ -123,7 +121,7 @@ describe('Logic Function Commands', () => {
 	});
 
 	it('Lists Logic Functions', async () => {
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'list', '--org', 'cyberdyne-systems']);
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'list', '--quiet', '--org', 'cyberdyne-systems']);
 		// FIXME: This would pass even if listOutput was empty
 		expect(stdout.split('\n')).to.include.members(listOutput);
 		expect(stderr).to.equal('');
@@ -133,7 +131,7 @@ describe('Logic Function Commands', () => {
 	it('Creates a blank Logic Function locally', async () => {
 		await fs.remove(path.join(PATH_TMP_DIR, 'newlf'));
 
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'create', '--name', 'newLF', '--org', 'cyberdyne-systems', '--force'], { cwd: PATH_TMP_DIR });
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'create', '--quiet', '--name', 'newLF', '--org', 'cyberdyne-systems', '--force'], { cwd: PATH_TMP_DIR });
 
 		expect(stdout.split('\n')).to.include.members(createOutput);
 		expect(stderr).to.equal('');
@@ -144,7 +142,7 @@ describe('Logic Function Commands', () => {
 		await fs.copy(path.join(PATH_FIXTURES_LOGIC_FUNCTIONS, 'lf3_proj', 'config.json'), path.join(PATH_TMP_DIR, 'newlf', 'newlf.logic.json'));
 		await fs.copy(path.join(PATH_FIXTURES_LOGIC_FUNCTIONS, 'lf3_proj', 'code.js'), path.join(PATH_TMP_DIR, 'newlf', 'newlf.js'));
 
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'execute', '--org', 'cyberdyne-systems', '--data', '1234'], { cwd: path.join(PATH_TMP_DIR, 'newlf') });
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'execute', '--quiet', '--org', 'cyberdyne-systems', '--data', '1234'], { cwd: path.join(PATH_TMP_DIR, 'newlf') });
 
 		expect(stdout.split('\n')).to.include.members(executeOutput);
 		expect(stderr).to.equal('');
@@ -155,7 +153,7 @@ describe('Logic Function Commands', () => {
 		await fs.copy(path.join(PATH_FIXTURES_LOGIC_FUNCTIONS, 'lf3_proj', 'config.json'), path.join(PATH_TMP_DIR, 'newlf', 'newlf.logic.json'));
 		await fs.copy(path.join(PATH_FIXTURES_LOGIC_FUNCTIONS, 'lf3_proj', 'code.js'), path.join(PATH_TMP_DIR, 'newlf', 'newlf.js'));
 
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'deploy', '--org', 'cyberdyne-systems', '--data', '1234', '--force'], { cwd: path.join(PATH_TMP_DIR, 'newlf') });
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'deploy', '--quiet', '--org', 'cyberdyne-systems', '--data', '1234', '--force'], { cwd: path.join(PATH_TMP_DIR, 'newlf') });
 
 		stdout.split('\n').forEach((line) => {
 			if (line.includes('Deploying Logic Function')) {
@@ -174,10 +172,10 @@ describe('Logic Function Commands', () => {
 			await fs.copy(path.join(PATH_FIXTURES_LOGIC_FUNCTIONS, 'lf3_proj', 'config.json'), path.join(PATH_TMP_DIR, 'newlf', 'newlf.logic.json'));
 			await fs.copy(path.join(PATH_FIXTURES_LOGIC_FUNCTIONS, 'lf3_proj', 'code.js'), path.join(PATH_TMP_DIR, 'newlf', 'newlf.js'));
 		}
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'deploy', '--org', 'cyberdyne-systems', '--data', '1234', '--force'], { cwd: path.join(PATH_TMP_DIR, 'newlf') });
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'deploy', '--quiet', '--org', 'cyberdyne-systems', '--data', '1234', '--force'], { cwd: path.join(PATH_TMP_DIR, 'newlf') });
 
 		stdout.split('\n').forEach((line) => {
-			if (line.includes('Deploying Logic Function')) {
+			if (line.includes('deployed to')) {
 				id = line.substring(line.indexOf('(') + 1, line.indexOf(')'));
 			}
 		});
@@ -189,14 +187,14 @@ describe('Logic Function Commands', () => {
 	});
 
 	it('Disables a Logic Function', async () => {
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'disable', '--org', 'cyberdyne-systems', '--id', id]);
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'disable', '--quiet', '--org', 'cyberdyne-systems', '--id', id]);
 		expect(stdout).to.equal(`Logic Function newlf (${id}) is now disabled.`);
 		expect(stderr).to.equal('');
 		expect(exitCode).to.equal(0);
 	});
 
 	it('Enables a Logic Function', async () => {
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'enable', '--org', 'cyberdyne-systems', '--id', id]);
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'enable', '--quiet', '--org', 'cyberdyne-systems', '--id', id]);
 
 		expect(stdout).to.equal(`Logic Function newlf (${id}) is now enabled.`);
 		expect(stderr).to.equal('');
@@ -206,7 +204,7 @@ describe('Logic Function Commands', () => {
 	it('Downloads a Logic Function', async () => {
 		await fs.remove(path.join(PATH_TMP_DIR, 'newlf'));
 
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'get', '--id', id, '--org', 'cyberdyne-systems'], { cwd: PATH_TMP_DIR });
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'get', '--quiet', '--id', id, '--org', 'cyberdyne-systems'], { cwd: PATH_TMP_DIR });
 
 		expect(stdout.split('\n')).to.include.members(getOutput);
 		expect(stderr).to.equal('');
@@ -217,7 +215,7 @@ describe('Logic Function Commands', () => {
 	});
 
 	it('Deletes a Logic Function', async () => {
-		const { stdout, stderr, exitCode } = await cli.run(['lf', 'delete', '--org', 'cyberdyne-systems', '--id', id, '--force']);
+		const { stdout, stderr, exitCode } = await cli.run(['lf', 'delete', '--quiet', '--org', 'cyberdyne-systems', '--id', id, '--force']);
 
 		expect(stdout).to.contain('has been successfully deleted.');
 		expect(stderr).to.equal('');
