diff --git a/cdc/owner/ddl_manager_test.go b/cdc/owner/ddl_manager_test.go
index 3255608be..520b6e0f0 100644
--- a/cdc/owner/ddl_manager_test.go
+++ b/cdc/owner/ddl_manager_test.go
@@ -44,6 +44,7 @@ func createDDLManagerForTest(t *testing.T) *ddlManager {
 		startTs,
 		checkpointTs,
 		ddlSink,
+		f,
 		ddlPuller,
 		schema,
 		redo.NewDisabledDDLManager(),
diff --git a/cdc/puller/ddl_puller_test.go b/cdc/puller/ddl_puller_test.go
index 75c0c4e53..aa51bf1ef 100644
--- a/cdc/puller/ddl_puller_test.go
+++ b/cdc/puller/ddl_puller_test.go
@@ -440,7 +440,7 @@ func TestHandleJob(t *testing.T) {
 		job = helper.DDL2Job("alter table test1.t1 add column c1 int")
 		skip, err = ddlJobPullerImpl.handleJob(job)
 		require.NoError(t, err)
-		require.True(t, skip)
+		require.False(t, skip)
 
 		job = helper.DDL2Job("create table test1.testStartTs(id int)")
 		skip, err = ddlJobPullerImpl.handleJob(job)
@@ -451,7 +451,7 @@ func TestHandleJob(t *testing.T) {
 		job.StartTS = 1
 		skip, err = ddlJobPullerImpl.handleJob(job)
 		require.NoError(t, err)
-		require.True(t, skip)
+		require.False(t, skip)
 
 		job = helper.DDL2Job("create table test1.t2(id int)")
 		skip, err = ddlJobPullerImpl.handleJob(job)
diff --git a/pkg/filter/filter_test.go b/pkg/filter/filter_test.go
index 608aadb8b..c4227bbce 100644
--- a/pkg/filter/filter_test.go
+++ b/pkg/filter/filter_test.go
@@ -138,8 +138,9 @@ func TestShouldIgnoreDMLEvent(t *testing.T) {
 	}
 }
 
-func TestShouldDiscardDDL(t *testing.T) {
+func TestShouldIgnoreDDL(t *testing.T) {
 	t.Parallel()
+
 	testCases := []struct {
 		cases []struct {
 			startTs uint64
@@ -152,137 +153,199 @@ func TestShouldDiscardDDL(t *testing.T) {
 		rules        []string
 		ignoredTs    []uint64
 		eventFilters []*config.EventFilterRule
-	}{{
-		// Ignore by table name cases.
-		cases: []struct {
-			startTs uint64
-			schema  string
-			table   string
-			query   string
-			ddlType timodel.ActionType
-			ignore  bool
-		}{
-			{1, "sns", "", "create database test", timodel.ActionCreateSchema, false},
-			{1, "sns", "", "drop database test", timodel.ActionDropSchema, false},
-			{
-				1, "sns", "", "ALTER DATABASE dbname CHARACTER SET utf8 COLLATE utf8_general_ci",
-				timodel.ActionModifySchemaCharsetAndCollate, false,
+	}{
+		{ // cases ignore by startTs
+			cases: []struct {
+				startTs uint64
+				schema  string
+				table   string
+				query   string
+				ddlType timodel.ActionType
+				ignore  bool
+			}{
+				{1, "ts", "", "create database test", timodel.ActionCreateSchema, true},
+				{2, "ts", "student", "drop database test2", timodel.ActionDropSchema, true},
+				{
+					3, "ts", "teacher", "ALTER DATABASE dbname CHARACTER SET utf8 COLLATE utf8_general_ci",
+					timodel.ActionModifySchemaCharsetAndCollate, true,
+				},
+				{4, "ts", "man", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
+				{5, "ts", "fruit", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
+				{6, "ts", "insect", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
 			},
-			{1, "ecom", "", "create database test", timodel.ActionCreateSchema, false},
-			{1, "ecom", "aa", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
-			{1, "ecom", "", "create database test", timodel.ActionCreateSchema, false},
-			{1, "test", "", "create database test", timodel.ActionCreateSchema, true},
+			rules:     []string{"*.*"},
+			ignoredTs: []uint64{1, 2, 3},
 		},
-		rules: []string{"sns.*", "ecom.*", "!sns.log", "!ecom.test"},
-		// Ignore by schema name cases.
-	}, {
-		cases: []struct {
-			startTs uint64
-			schema  string
-			table   string
-			query   string
-			ddlType timodel.ActionType
-			ignore  bool
-		}{
-			{1, "schema", "C1", "create database test", timodel.ActionCreateSchema, false},
-			{1, "test", "", "drop database test1", timodel.ActionDropSchema, true},
-			{
-				1, "dbname", "", "ALTER DATABASE dbname CHARACTER SET utf8 COLLATE utf8_general_ci",
-				timodel.ActionModifySchemaCharsetAndCollate, true,
+		{ // cases ignore by ddl type.
+			cases: []struct {
+				startTs uint64
+				schema  string
+				table   string
+				query   string
+				ddlType timodel.ActionType
+				ignore  bool
+			}{
+				{1, "event", "", "drop table t1", timodel.ActionDropTable, true},
+				{1, "event", "January", "drop index i on t1", timodel.ActionDropIndex, true},
+				{1, "event", "February", "drop index x2 on t2", timodel.ActionDropIndex, true},
+				{1, "event", "March", "create table t2(age int)", timodel.ActionCreateTable, false},
+				{1, "event", "April", "create table t2(age int)", timodel.ActionCreateTable, false},
+				{1, "event", "May", "create table t2(age int)", timodel.ActionCreateTable, false},
 			},
-			{1, "test", "aa", "create table test.t1(a int primary key)", timodel.ActionCreateTable, true},
-			{1, "schema", "C1", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
-			{1, "schema", "", "create table test.t1(a int primary key)", timodel.ActionCreateTable, true},
-		},
-		rules: []string{"schema.C1"},
-	}, { // cases ignore by startTs
-		cases: []struct {
-			startTs uint64
-			schema  string
-			table   string
-			query   string
-			ddlType timodel.ActionType
-			ignore  bool
-		}{
-			{1, "ts", "", "create database test", timodel.ActionCreateSchema, true},
-			{2, "ts", "student", "drop database test2", timodel.ActionDropSchema, true},
-			{
-				3, "ts", "teacher", "ALTER DATABASE dbname CHARACTER SET utf8 COLLATE utf8_general_ci",
-				timodel.ActionModifySchemaCharsetAndCollate, true,
+			rules: []string{"*.*"},
+			eventFilters: []*config.EventFilterRule{
+				{
+					Matcher: []string{"event.*"},
+					IgnoreEvent: []bf.EventType{
+						bf.AlterTable, bf.DropTable,
+					},
+				},
 			},
-			{4, "ts", "man", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
-			{5, "ts", "fruit", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
-			{6, "ts", "insect", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
+			ignoredTs: []uint64{},
 		},
-		rules:     []string{"*.*"},
-		ignoredTs: []uint64{1, 2, 3},
-	}, { // cases ignore by ddl type.
-		cases: []struct {
-			startTs uint64
-			schema  string
-			table   string
-			query   string
-			ddlType timodel.ActionType
-			ignore  bool
-		}{
-			{1, "event", "", "drop table t1", timodel.ActionDropTable, true},
-			{1, "event", "January", "drop index i on t1", timodel.ActionDropIndex, true},
-			{1, "event", "February", "drop index x2 on t2", timodel.ActionDropIndex, true},
-			{1, "event", "March", "create table t2(age int)", timodel.ActionCreateTable, false},
-			{1, "event", "April", "create table t2(age int)", timodel.ActionCreateTable, false},
-			{1, "event", "May", "create table t2(age int)", timodel.ActionCreateTable, false},
-		},
-		rules: []string{"*.*"},
-		eventFilters: []*config.EventFilterRule{
-			{
-				Matcher: []string{"event.*"},
-				IgnoreEvent: []bf.EventType{
-					bf.AlterTable, bf.DropTable,
+		{ // cases ignore by ddl query
+			cases: []struct {
+				startTs uint64
+				schema  string
+				table   string
+				query   string
+				ddlType timodel.ActionType
+				ignore  bool
+			}{
+				{1, "sql_pattern", "t1", "CREATE DATABASE sql_pattern", timodel.ActionCreateSchema, false},
+				{1, "sql_pattern", "t1", "DROP DATABASE sql_pattern", timodel.ActionDropSchema, true},
+				{
+					1, "sql_pattern", "t1",
+					"ALTER DATABASE `test_db` CHARACTER SET 'utf8' COLLATE 'utf8_general_ci'",
+					timodel.ActionModifySchemaCharsetAndCollate,
+					false,
 				},
+				{1, "sql_pattern", "t1", "CREATE TABLE t1(id int primary key)", timodel.ActionCreateTable, false},
+				{1, "sql_pattern", "t1", "DROP TABLE t1", timodel.ActionDropTable, true},
+				{1, "sql_pattern", "t1", "CREATE VIEW test.v AS SELECT * FROM t", timodel.ActionCreateView, true},
 			},
+			rules: []string{"*.*"},
+			eventFilters: []*config.EventFilterRule{
+				{
+					Matcher:   []string{"sql_pattern.*"},
+					IgnoreSQL: []string{"^DROP TABLE", "^CREATE VIEW", "^DROP DATABASE"},
+				},
+			},
+			ignoredTs: []uint64{},
 		},
-	}, { // cases ignore by ddl query
-		cases: []struct {
-			startTs uint64
+	}
+
+	for _, ftc := range testCases {
+		filter, err := NewFilter(&config.ReplicaConfig{
+			Filter: &config.FilterConfig{
+				Rules:            ftc.rules,
+				EventFilters:     ftc.eventFilters,
+				IgnoreTxnStartTs: ftc.ignoredTs,
+			},
+		}, "")
+		require.Nil(t, err)
+		for _, tc := range ftc.cases {
+			ddl := &model.DDLEvent{
+				StartTs: tc.startTs,
+				TableInfo: &model.TableInfo{
+					TableName: model.TableName{
+						Schema: tc.schema,
+						Table:  tc.table,
+					},
+				},
+				Query: tc.query,
+				Type:  tc.ddlType,
+			}
+			ignore, err := filter.ShouldIgnoreDDLEvent(ddl)
+			require.NoError(t, err)
+			require.Equal(t, tc.ignore, ignore, "%#v", tc)
+		}
+	}
+}
+
+func TestShouldDiscardDDL(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		cases []struct {
 			schema  string
 			table   string
 			query   string
 			ddlType timodel.ActionType
 			ignore  bool
-		}{
-			{1, "sql_pattern", "t1", "CREATE DATABASE sql_pattern", timodel.ActionCreateSchema, false},
-			{1, "sql_pattern", "t1", "DROP DATABASE sql_pattern", timodel.ActionDropSchema, true},
-			{
-				1, "sql_pattern", "t1",
-				"ALTER DATABASE `test_db` CHARACTER SET 'utf8' COLLATE 'utf8_general_ci'",
-				timodel.ActionModifySchemaCharsetAndCollate,
-				false,
+		}
+		rules        []string
+		eventFilters []*config.EventFilterRule
+	}{
+		{
+			// Discard by not allowed DDL type cases.
+			cases: []struct {
+				schema  string
+				table   string
+				query   string
+				ddlType timodel.ActionType
+				ignore  bool
+			}{
+				{"sns", "", "create database test", timodel.ActionCreateSchema, false},
+				{"sns", "", "drop database test", timodel.ActionDropSchema, false},
+				{"test", "", "create database test", timodel.ActionCreateSequence, true},
 			},
-			{1, "sql_pattern", "t1", "CREATE TABLE t1(id int primary key)", timodel.ActionCreateTable, false},
-			{1, "sql_pattern", "t1", "DROP TABLE t1", timodel.ActionDropTable, true},
-			{1, "sql_pattern", "t1", "CREATE VIEW test.v AS SELECT * FROM t", timodel.ActionCreateView, true},
+			rules: []string{"*.*"},
 		},
-		rules: []string{"*.*"},
-		eventFilters: []*config.EventFilterRule{
-			{
-				Matcher:   []string{"sql_pattern.*"},
-				IgnoreSQL: []string{"^DROP TABLE", "^CREATE VIEW", "^DROP DATABASE"},
+		{
+			// Discard by table name cases.
+			cases: []struct {
+				schema  string
+				table   string
+				query   string
+				ddlType timodel.ActionType
+				ignore  bool
+			}{
+				{"sns", "", "create database test", timodel.ActionCreateSchema, false},
+				{"sns", "", "drop database test", timodel.ActionDropSchema, false},
+				{
+					"sns", "", "ALTER DATABASE dbname CHARACTER SET utf8 COLLATE utf8_general_ci",
+					timodel.ActionModifySchemaCharsetAndCollate, false,
+				},
+				{"ecom", "", "create database test", timodel.ActionCreateSchema, false},
+				{"ecom", "aa", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
+				{"ecom", "", "create database test", timodel.ActionCreateSchema, false},
+				{"test", "", "create database test", timodel.ActionCreateSchema, true},
 			},
+			rules: []string{"sns.*", "ecom.*", "!sns.log", "!ecom.test"},
 		},
-	}}
+		{
+			// Discard by schema name cases.
+			cases: []struct {
+				schema  string
+				table   string
+				query   string
+				ddlType timodel.ActionType
+				ignore  bool
+			}{
+				{"schema", "C1", "create database test", timodel.ActionCreateSchema, false},
+				{"test", "", "drop database test1", timodel.ActionDropSchema, true},
+				{
+					"dbname", "", "ALTER DATABASE dbname CHARACTER SET utf8 COLLATE utf8_general_ci",
+					timodel.ActionModifySchemaCharsetAndCollate, true,
+				},
+				{"test", "aa", "create table test.t1(a int primary key)", timodel.ActionCreateTable, true},
+				{"schema", "C1", "create table test.t1(a int primary key)", timodel.ActionCreateTable, false},
+				{"schema", "", "create table test.t1(a int primary key)", timodel.ActionCreateTable, true},
+			},
+			rules: []string{"schema.C1"},
+		},
+	}
 
 	for _, ftc := range testCases {
 		filter, err := NewFilter(&config.ReplicaConfig{
 			Filter: &config.FilterConfig{
-				Rules:            ftc.rules,
-				IgnoreTxnStartTs: ftc.ignoredTs,
-				EventFilters:     ftc.eventFilters,
+				Rules:        ftc.rules,
+				EventFilters: ftc.eventFilters,
 			},
 		}, "")
 		require.Nil(t, err)
 		for _, tc := range ftc.cases {
-			ignore, err := filter.ShouldDiscardDDL(tc.startTs, tc.ddlType, tc.schema, tc.table, tc.query)
-			require.Nil(t, err, "%#v", tc)
+			ignore := filter.ShouldDiscardDDL(tc.ddlType, tc.schema, tc.table)
 			require.Equal(t, tc.ignore, ignore, "%#v", tc)
 		}
 	}
diff --git a/pkg/filter/sql_event_filter_test.go b/pkg/filter/sql_event_filter_test.go
index 685d9767b..a564e29d7 100644
--- a/pkg/filter/sql_event_filter_test.go
+++ b/pkg/filter/sql_event_filter_test.go
@@ -183,8 +183,7 @@ func TestShouldSkipDDL(t *testing.T) {
 				},
 				Query: c.query,
 			}
-			skip, err := f.shouldSkipDDL(ddl.Type,
-				ddl.TableInfo.TableName.Schema, ddl.TableInfo.TableName.Table, ddl.Query)
+			skip, err := f.shouldSkipDDL(ddl)
 			require.NoError(t, err)
 			require.Equal(t, c.skip, skip, "case: %+v", c)
 		}
diff --git a/tests/integration_tests/event_filter/conf/cf.toml b/tests/integration_tests/event_filter/conf/cf.toml
index 2ea6083c5..d3a36188b 100644
--- a/tests/integration_tests/event_filter/conf/cf.toml
+++ b/tests/integration_tests/event_filter/conf/cf.toml
@@ -6,3 +6,10 @@ matcher = ["event_filter.t1"]
 ignore-event = ["drop table", "delete"]
 ignore-insert-value-expr = "id = 2 or city = 'tokyo'"
 
+[[filter.event-filters]]
+matcher = ["event_filter.t_truncate"]
+ignore-event = ["truncate table"]
+
+[[filter.event-filters]]
+matcher = ["event_filter.t_alter"]
+ignore-event = ["alter table"]
diff --git a/tests/integration_tests/event_filter/conf/diff_config.toml b/tests/integration_tests/event_filter/conf/diff_config.toml
index 75593ed89..fff11e2ee 100644
--- a/tests/integration_tests/event_filter/conf/diff_config.toml
+++ b/tests/integration_tests/event_filter/conf/diff_config.toml
@@ -13,7 +13,7 @@ source-instances = ["tidb0"]
 
 target-instance = "mysql1"
 
-target-check-tables = ["event_filter.t2"]
+target-check-tables = ["event_filter.t_normal", "event_filter.t_truncate", "event_filter.t_alter"]
 
 [data-sources]
 [data-sources.tidb0]
diff --git a/tests/integration_tests/event_filter/data/test.sql b/tests/integration_tests/event_filter/data/test.sql
index a3ff2e479..6af0efed2 100644
--- a/tests/integration_tests/event_filter/data/test.sql
+++ b/tests/integration_tests/event_filter/data/test.sql
@@ -38,8 +38,8 @@ WHERE id = 4;
 /* ignore by event type*/
 DROP TABLE t1;
 
-/* all event of t2 will be replicated to downstream */
-CREATE TABLE t2 (
+/* all event of t_normal will be replicated to downstream */
+CREATE TABLE t_normal (
                     id INT,
                     name varchar(128),
                     country char(32),
@@ -48,16 +48,45 @@ CREATE TABLE t2 (
                     gender char(32),
                     PRIMARY KEY (id)
 );
-INSERT INTO t2
+INSERT INTO t_normal
 VALUES (1, 'guagua', "china", "chengdu", 1, "female");
 
-INSERT INTO t2
+INSERT INTO t_normal
 VALUES (2, 'huahua', "china", "chengdu", 2, "female");
 
-INSERT INTO t2
+INSERT INTO t_normal
 VALUES (3, 'xigua', "japan", "tokyo", 2, "male");
 
-INSERT INTO t2
+INSERT INTO t_normal
 VALUES (4, 'yuko', "japan", "nagoya", 33, "female");
 
-create table finish_mark(id int primary key);
+CREATE TABLE t_truncate (
+                    id INT,
+                    name varchar(128),
+                    PRIMARY KEY (id)
+);
+CREATE TABLE t_alter
+(
+    id          INT AUTO_INCREMENT,
+    t_boolean   BOOLEAN,
+    t_bigint    DECIMAL(38, 19),
+    t_double    DOUBLE,
+    t_decimal   DECIMAL(38, 19),
+    t_bit       BIT(64),
+    t_date      DATE,
+    t_datetime  DATETIME,
+    t_timestamp TIMESTAMP NULL,
+    t_time      TIME,
+    t_year      YEAR,
+    t_char      CHAR,
+    t_varchar   VARCHAR(10),
+    t_blob      BLOB,
+    t_text      TEXT,
+    t_enum      ENUM ('enum1', 'enum2', 'enum3'),
+    t_set       SET ('a', 'b', 'c'),
+    t_json      JSON,
+    PRIMARY KEY (id)
+) ENGINE = InnoDB
+  DEFAULT CHARSET = utf8
+  COLLATE = utf8_bin;
+  
\ No newline at end of file
diff --git a/tests/integration_tests/event_filter/data/test_alter.sql b/tests/integration_tests/event_filter/data/test_alter.sql
new file mode 100644
index 000000000..0e6fbabd2
--- /dev/null
+++ b/tests/integration_tests/event_filter/data/test_alter.sql
@@ -0,0 +1,49 @@
+USE `event_filter`;
+
+/* test  */
+ALTER TABLE t_alter MODIFY t_bigint BIGINT;
+
+-- make sure `nullable` can be handled by the mounter and mq encoding protocol
+INSERT INTO t_alter() VALUES ();
+
+INSERT INTO t_alter( t_boolean, t_bigint, t_double, t_decimal, t_bit
+                           , t_date, t_datetime, t_timestamp, t_time, t_year
+                           , t_char, t_varchar, t_blob, t_text, t_enum
+                           , t_set, t_json)
+VALUES ( true, 9223372036854775807, 123.123, 123456789012.123456789012, b'1000001'
+       , '1000-01-01', '9999-12-31 23:59:59', '19731230153000', '23:59:59', 1970
+       , '测', '测试', 'blob', '测试text', 'enum2'
+       , 'a,b', NULL);
+
+INSERT INTO t_alter( t_boolean, t_bigint, t_double, t_decimal, t_bit
+                           , t_date, t_datetime, t_timestamp, t_time, t_year
+                           , t_char, t_varchar, t_blob, t_text, t_enum
+                           , t_set, t_json)
+VALUES ( false, 666, 123.777, 123456789012.123456789012, b'1000001'
+       , '1000-01-01', '9999-12-31 23:59:59', '19731230153000', '23:59:59', 1970
+       , '测', '测试', 'blob', '测试text11', 'enum3'
+       , 'a,b', NULL);
+
+UPDATE t_alter
+SET t_bigint = 555
+WHERE id = 1;
+
+INSERT INTO t_alter( t_boolean, t_bigint, t_double, t_decimal, t_bit
+                           , t_date, t_datetime, t_timestamp, t_time, t_year
+                           , t_char, t_varchar, t_blob, t_text, t_enum
+                           , t_set, t_json)
+VALUES ( true, 9223372036875807, 153.123, 123456669012.123456789012, b'1010001'
+       , '2000-01-01', '9999-12-31 23:59:59', '19731230153000', '23:59:59', 1970
+       , '测', '测试', 'blob', '测试text', 'enum1'
+       , 'a,b', '{
+    "key1": "value1",
+    "key2": "value2"
+  }');
+
+UPDATE t_alter
+SET t_bigint = 888,
+    t_json   = '{
+      "key0": "value0",
+      "key2": "value2"
+    }'
+WHERE id = 2;
diff --git a/tests/integration_tests/event_filter/data/test_truncate.sql b/tests/integration_tests/event_filter/data/test_truncate.sql
new file mode 100644
index 000000000..170d94c9c
--- /dev/null
+++ b/tests/integration_tests/event_filter/data/test_truncate.sql
@@ -0,0 +1,15 @@
+USE `event_filter`;
+
+/* test new physical table is replicated */
+TRUNCATE TABLE t_truncate;
+INSERT INTO t_truncate
+VALUES (1, 'guagua');
+
+INSERT INTO t_truncate
+VALUES (2, 'huahua');
+
+INSERT INTO t_truncate
+VALUES (3, 'xigua');
+
+INSERT INTO t_truncate
+VALUES (4, 'yuko');
diff --git a/tests/integration_tests/event_filter/run.sh b/tests/integration_tests/event_filter/run.sh
index c25f06a23..6e9385052 100644
--- a/tests/integration_tests/event_filter/run.sh
+++ b/tests/integration_tests/event_filter/run.sh
@@ -34,8 +34,9 @@ function run() {
 	# make suer table t1 is deleted in upstream and exists in downstream
 	check_table_not_exists "event_filter.t1" ${UP_TIDB_HOST} ${UP_TIDB_PORT}
 	check_table_exists "event_filter.t1" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
-	check_table_exists "event_filter.t2" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
-	check_table_exists "event_filter.finish_mark" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
+	check_table_exists "event_filter.t_normal" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
+	check_table_exists "event_filter.t_truncate" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
+	check_table_exists "event_filter.t_alter" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
 
 	# check those rows that are not filtered are synced to downstream
 	run_sql "select count(1) from event_filter.t1;" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
@@ -49,7 +50,14 @@ function run() {
 	run_sql "select count(5) from event_filter.t1 where id=4;" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
 	check_contains "count(5): 1"
 
-	# check table t2 is replicated
+	run_sql "TRUNCATE TABLE event_filter.t_truncate;" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
+	run_sql_file $CUR/data/test_truncate.sql ${UP_TIDB_HOST} ${UP_TIDB_PORT}
+	run_sql "ALTER TABLE event_filter.t_alter MODIFY t_bigint BIGINT;" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
+	run_sql_file $CUR/data/test_alter.sql ${UP_TIDB_HOST} ${UP_TIDB_PORT}
+	run_sql "create table event_filter.finish_mark(id int primary key);" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
+	check_table_exists "event_filter.finish_mark" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
+
+	# check table t_normal is replicated
 	check_sync_diff $WORK_DIR $CUR/conf/diff_config.toml
 
 	cleanup_process $CDC_BINARY
