diff --git a/app/pages/project/classify.spec.js b/app/pages/project/classify.spec.js
new file mode 100644
index 00000000..1c7d82e3
--- /dev/null
+++ b/app/pages/project/classify.spec.js
@@ -0,0 +1,25 @@
+import React from 'react';
+import assert from 'assert';
+import { shallow } from 'enzyme';
+
+import { FakeApiClient } from '../../../test/fake-api-client';
+import ProjectClassifyPage from './classify';
+import FinishedBanner from './finished-banner';
+
+describe('ProjectClassifyPage', function () {
+  let wrapper;
+  let project = {};
+  let apiUser = {id: 123};
+
+  before(function () {
+    wrapper = shallow(<ProjectClassifyPage project={project} />);
+  });
+
+  it('renders when not logged in', function () {
+    assert.equal(wrapper.find('.classify-page').length, 1);
+  });
+
+  it('never tells you when the project is finished', () => {
+    assert.equal(wrapper.find(FinishedBanner).length, 0);
+  });
+});
