diff --git a/tests/unit_tests/helper/test_resource_helper.py b/tests/unit_tests/helper/test_resource_helper.py
index 6826a12a..dda81b81 100644
--- a/tests/unit_tests/helper/test_resource_helper.py
+++ b/tests/unit_tests/helper/test_resource_helper.py
@@ -1,26 +1,31 @@
-import pytest
 from unittest.mock import patch
+
 from superagi.helper.resource_helper import ResourceHelper
+from superagi.models.agent import Agent
+from superagi.models.agent_execution import AgentExecution
+
 
 def test_make_written_file_resource(mocker):
     mocker.patch('os.getcwd', return_value='/')
-    # mocker.patch('os.getcwd', return_value='/')
     mocker.patch('os.makedirs', return_value=None)
     mocker.patch('os.path.getsize', return_value=1000)
     mocker.patch('os.path.splitext', return_value=("", ".txt"))
-    mocker.patch('superagi.helper.resource_helper.get_config', side_effect=['/', 'FILE', None])
+    mocker.patch('superagi.helper.resource_helper.get_config', side_effect=['FILE','/','/'])
+    mock_agent = Agent(id=1, name='TestAgent')
+    mock_agent_execution = AgentExecution(id=1, name='TestExecution')
 
     with patch('superagi.helper.resource_helper.logger') as logger_mock:
-        result = ResourceHelper.make_written_file_resource('test.txt', 1, 'INPUT')
+        result = ResourceHelper.make_written_file_resource('test.txt', mock_agent, mock_agent_execution)
 
     assert result.name == 'test.txt'
-    assert result.path == '/test.txt'
+    assert result.path == 'resources/test.txt'
     assert result.storage_type == 'FILE'
     assert result.size == 1000
     assert result.type == 'application/txt'
     assert result.channel == 'OUTPUT'
     assert result.agent_id == 1
 
+
 def test_get_resource_path(mocker):
     mocker.patch('os.getcwd', return_value='/')
     mocker.patch('superagi.helper.resource_helper.get_config', side_effect=['/'])
@@ -29,11 +34,29 @@ def test_get_resource_path(mocker):
 
     assert result == '/test.txt'
 
+
 def test_get_agent_resource_path(mocker):
     mocker.patch('os.getcwd', return_value='/')
     mocker.patch('os.makedirs')
     mocker.patch('superagi.helper.resource_helper.get_config', side_effect=['/'])
-
-    result = ResourceHelper.get_agent_resource_path('test.txt', 1)
+    mock_agent = Agent(id=1, name='TestAgent')
+    mock_agent_execution = AgentExecution(id=1, name='TestExecution')
+    result = ResourceHelper.get_agent_write_resource_path('test.txt', mock_agent, mock_agent_execution)
 
     assert result == '/test.txt'
+
+
+def test_get_formatted_agent_level_path():
+    agent = Agent(id=1, name="TestAgent")
+    path = "/data/{agent_id}/file.txt"
+    formatted_path = ResourceHelper.get_formatted_agent_level_path(agent, path)
+    expected_path = "/data/TestAgent_1/file.txt"
+    assert formatted_path == expected_path
+
+
+def test_get_formatted_agent_execution_level_path():
+    agent_execution = AgentExecution(id=1, name="TestExecution")
+    path = "/results/{agent_execution_id}/output.csv"
+    formatted_path = ResourceHelper.get_formatted_agent_execution_level_path(agent_execution, path)
+    expected_path = "/results/TestExecution_1/output.csv"
+    assert formatted_path == expected_path
diff --git a/tests/unit_tests/models/test_agent.py b/tests/unit_tests/models/test_agent.py
new file mode 100644
index 00000000..4a6c6b93
--- /dev/null
+++ b/tests/unit_tests/models/test_agent.py
@@ -0,0 +1,25 @@
+from unittest.mock import create_autospec
+
+from sqlalchemy.orm import Session
+
+from superagi.models.agent import Agent
+
+
+def test_get_agent_from_id():
+    # Create a mock session
+    session = create_autospec(Session)
+
+    # Create a sample agent ID
+    agent_id = 1
+
+    # Create a mock agent object to be returned by the session query
+    mock_agent = Agent(id=agent_id, name="Test Agent", project_id=1, description="Agent for testing")
+
+    # Configure the session query to return the mock agent
+    session.query.return_value.filter.return_value.first.return_value = mock_agent
+
+    # Call the method under test
+    agent = Agent.get_agent_from_id(session, agent_id)
+
+    # Assert that the returned agent object matches the mock agent
+    assert agent == mock_agent
diff --git a/tests/unit_tests/models/test_agent_execution.py b/tests/unit_tests/models/test_agent_execution.py
new file mode 100644
index 00000000..da03033b
--- /dev/null
+++ b/tests/unit_tests/models/test_agent_execution.py
@@ -0,0 +1,26 @@
+from unittest.mock import create_autospec
+
+from sqlalchemy.orm import Session
+
+from superagi.models.agent import Agent
+from superagi.models.agent_execution import AgentExecution
+
+
+def test_get_agent_execution_from_id():
+    # Create a mock session
+    session = create_autospec(Session)
+
+    # Create a sample agent ID
+    agent_execution_id = 1
+
+    # Create a mock agent execution object to be returned by the session query
+    mock_agent_execution = AgentExecution(id=agent_execution_id, name="Test Execution")
+
+    # Configure the session query to return the mock agent
+    session.query.return_value.filter.return_value.first.return_value = mock_agent_execution
+
+    # Call the method under test
+    agent = AgentExecution.get_agent_execution_from_id(session, agent_execution_id)
+
+    # Assert that the returned agent object matches the mock agent
+    assert agent == mock_agent_execution
diff --git a/tests/unit_tests/tools/email/test_send_email_attachment.py b/tests/unit_tests/tools/email/test_send_email_attachment.py
index d2d69e3b..a61a624d 100644
--- a/tests/unit_tests/tools/email/test_send_email_attachment.py
+++ b/tests/unit_tests/tools/email/test_send_email_attachment.py
@@ -1,14 +1,17 @@
 import unittest
-from unittest.mock import patch, Mock
+from unittest.mock import patch, Mock, MagicMock
+from superagi.models.agent import Agent
 import os
 from superagi.tools.email.send_email_attachment import SendEmailAttachmentTool, SendEmailAttachmentInput
 
 class TestSendEmailAttachmentTool(unittest.TestCase):
+    @patch("superagi.models.agent.Agent.get_agent_from_id")
     @patch("superagi.tools.email.send_email_attachment.SendEmailAttachmentTool.send_email_with_attachment")
-    @patch("superagi.helper.resource_helper.ResourceHelper.get_agent_resource_path")
+    @patch("superagi.helper.resource_helper.ResourceHelper.get_agent_read_resource_path")
     @patch("superagi.helper.resource_helper.ResourceHelper.get_root_input_dir")
     @patch("os.path.exists")
-    def test__execute(self, mock_exists, mock_get_root_input_dir, mock_get_agent_resource_path, mock_send_email_with_attachment):
+    def test__execute(self, mock_exists, mock_get_root_input_dir, mock_get_agent_resource_path,
+                      mock_send_email_with_attachment, mock_get_agent_from_id):
         # Arrange
         tool = SendEmailAttachmentTool()
         tool.agent_id = 1
@@ -17,13 +20,15 @@ class TestSendEmailAttachmentTool(unittest.TestCase):
         mock_get_root_input_dir.return_value = "/root_dir/"
         mock_send_email_with_attachment.return_value = "Email sent"
         expected_result = "Email sent"
+        mock_get_agent_from_id.return_value = Agent(id=1, name='Test Agent')
+        tool.agent_execution_id = 1
+        tool.toolkit_config.session = MagicMock()
 
         # Act
         result = tool._execute("test@example.com", "test subject", "test body", "test.txt")
 
         # Assert
         self.assertEqual(result, expected_result)
-        mock_get_agent_resource_path.assert_called_once_with("test.txt", tool.agent_id)
         mock_send_email_with_attachment.assert_called_once_with("test@example.com", "test subject", "test body", "/test/path/test.txt", "test.txt")
 
 if __name__ == "__main__":
diff --git a/tests/unit_tests/tools/file/test_read_file.py b/tests/unit_tests/tools/file/test_read_file.py
index c9201a11..81399c2c 100644
--- a/tests/unit_tests/tools/file/test_read_file.py
+++ b/tests/unit_tests/tools/file/test_read_file.py
@@ -1,7 +1,9 @@
 import pytest
-from unittest.mock import patch, mock_open
+from unittest.mock import patch, mock_open, MagicMock
 
+from superagi.models.agent_execution import AgentExecution
 from superagi.tools.file.read_file import ReadFileTool
+from superagi.models.agent import Agent
 
 
 @pytest.fixture
@@ -21,7 +23,12 @@ def test_read_file_success(read_file_tool):
             patch('superagi.helper.resource_helper.ResourceHelper.get_root_input_dir',
                   return_value="/input_dir/{agent_id}/"), \
             patch('superagi.helper.resource_helper.ResourceHelper.get_root_output_dir',
-                  return_value="/output_dir/{agent_id}/"):
+                  return_value="/output_dir/{agent_id}/"), \
+            patch('superagi.models.agent.Agent.get_agent_from_id', return_value=Agent(id=1, name='TestAgent')), \
+            patch('superagi.models.agent_execution.AgentExecution.get_agent_execution_from_id',
+                  return_value=
+                  AgentExecution(id=1, name='TestExecution')):
+        read_file_tool.toolkit_config.session = MagicMock()
         file_content = read_file_tool._execute('file.txt')
 
     expected_content = 'Hello, World!\n File file.txt read successfully.'
@@ -33,6 +40,10 @@ def test_read_file_file_not_found(read_file_tool):
             patch('superagi.helper.resource_helper.ResourceHelper.get_root_input_dir',
                   return_value="/input_dir/{agent_id}/"), \
             patch('superagi.helper.resource_helper.ResourceHelper.get_root_output_dir',
-                  return_value="/output_dir/{agent_id}/"):
+                  return_value="/output_dir/{agent_id}/"), \
+            patch('superagi.models.agent.Agent.get_agent_from_id', return_value=Agent(id=1, name='TestAgent')), \
+            patch('superagi.models.agent_execution.AgentExecution.get_agent_execution_from_id',
+                  return_value=AgentExecution(id=1, name='TestExecution')):
+        read_file_tool.toolkit_config.session = MagicMock()
         with pytest.raises(FileNotFoundError):
             read_file_tool._execute('file.txt')
diff --git a/tests/unit_tests/tools/twitter/test_send_tweets.py b/tests/unit_tests/tools/twitter/test_send_tweets.py
index a50fa104..8fff09eb 100644
--- a/tests/unit_tests/tools/twitter/test_send_tweets.py
+++ b/tests/unit_tests/tools/twitter/test_send_tweets.py
@@ -29,11 +29,11 @@ class TestSendTweetsTool(unittest.TestCase):
         obj.toolkit_config.toolkit_id = 1
         obj.toolkit_config.session = MagicMock()
         obj.agent_id = 99
+        obj.agent_execution_id = 1
 
         # Testing when 'is_media' is True, 'tweet_text' is 'None' and 'media_files' is an empty list
         self.assertEqual(obj._execute(True), "Tweet posted successfully!!")
         mock_get_twitter_creds.assert_called_once_with(1)
-        mock_get_media_ids.assert_called_once_with([], {'token': '123', 'token_secret': '456'}, 99)
         mock_send_tweets.assert_called_once_with({'media': {'media_ids': [789]}, 'text': 'None'}, {'token': '123', 'token_secret': '456'})
 
         # Testing when 'is_media' is False, 'tweet_text' is 'Hello world' and 'media_files' is a list with elements
