diff --git a/tests/e2e_test.rs b/tests/e2e_test.rs
index d9f006cf3..92bb42731 100644
--- a/tests/e2e_test.rs
+++ b/tests/e2e_test.rs
@@ -28,7 +28,9 @@ use once_cell::sync::Lazy;
 /// use the cached queries that rely on the corelib's code, which vastly reduces the tests runtime.
 static SHARED_DB_WITH_GAS: Lazy<Mutex<RootDatabase>> = Lazy::new(|| {
     let mut db = RootDatabase::builder().detect_corelib().build().unwrap();
-    db.set_optimization_config(Arc::new(OptimizationConfig::default()));
+    db.set_optimization_config(Arc::new(
+        OptimizationConfig::default().with_minimal_movable_functions(),
+    ));
     Mutex::new(db)
 });
 static SHARED_DB_NO_GAS: Lazy<Mutex<RootDatabase>> = Lazy::new(|| {
@@ -36,7 +38,9 @@ static SHARED_DB_NO_GAS: Lazy<Mutex<RootDatabase>> = Lazy::new(|| {
     let add_withdraw_gas_flag_id = FlagId::new(db.upcast(), "add_withdraw_gas");
     db.set_flag(add_withdraw_gas_flag_id, Some(Arc::new(Flag::AddWithdrawGas(false))));
 
-    db.set_optimization_config(Arc::new(OptimizationConfig::default()));
+    db.set_optimization_config(Arc::new(
+        OptimizationConfig::default().with_minimal_movable_functions(),
+    ));
     Mutex::new(db)
 });
