diff --git a/internal/service/securitylake/aws_log_source_test.go b/internal/service/securitylake/aws_log_source_test.go
index 0824f095c40..779598094c0 100644
--- a/internal/service/securitylake/aws_log_source_test.go
+++ b/internal/service/securitylake/aws_log_source_test.go
@@ -5,7 +5,6 @@ package securitylake_test
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"testing"
 
@@ -15,7 +14,6 @@ import (
 	"github.com/hashicorp/terraform-plugin-testing/terraform"
 	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
 	"github.com/hashicorp/terraform-provider-aws/internal/conns"
-	"github.com/hashicorp/terraform-provider-aws/internal/create"
 	tfsecuritylake "github.com/hashicorp/terraform-provider-aws/internal/service/securitylake"
 	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
 	"github.com/hashicorp/terraform-provider-aws/names"
@@ -23,12 +21,7 @@ import (
 
 func testAccAWSLogSource_basic(t *testing.T) {
 	ctx := acctest.Context(t)
-	if testing.Short() {
-		t.Skip("skipping long-running test in short mode")
-	}
-
 	resourceName := "aws_securitylake_aws_log_source.test"
-	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	var logSource types.AwsLogSourceConfiguration
 
 	resource.Test(t, resource.TestCase{
@@ -41,7 +34,7 @@ func testAccAWSLogSource_basic(t *testing.T) {
 		CheckDestroy:             testAccCheckAWSLogSourceDestroy(ctx),
 		Steps: []resource.TestStep{
 			{
-				Config: testAccLogSourceConfig_basic(rName),
+				Config: testAccAWSLogSourceConfig_basic(),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckAWSLogSourceExists(ctx, resourceName, &logSource),
 					resource.TestCheckResourceAttr(resourceName, "source.#", "1"),
@@ -53,10 +46,9 @@ func testAccAWSLogSource_basic(t *testing.T) {
 				),
 			},
 			{
-				ResourceName:            resourceName,
-				ImportState:             true,
-				ImportStateVerify:       true,
-				ImportStateVerifyIgnore: []string{""},
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
 			},
 		},
 	})
@@ -64,10 +56,6 @@ func testAccAWSLogSource_basic(t *testing.T) {
 
 func testAccAWSLogSource_multiRegion(t *testing.T) {
 	ctx := acctest.Context(t)
-	if testing.Short() {
-		t.Skip("skipping long-running test in short mode")
-	}
-
 	resourceName := "aws_securitylake_aws_log_source.test"
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	var logSource types.AwsLogSourceConfiguration
@@ -82,7 +70,7 @@ func testAccAWSLogSource_multiRegion(t *testing.T) {
 		CheckDestroy:             testAccCheckAWSLogSourceDestroy(ctx),
 		Steps: []resource.TestStep{
 			{
-				Config: testAccLogSourceConfig_multiRegion(rName),
+				Config: testAccAWSLogSourceConfig_multiRegion(rName),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckAWSLogSourceExists(ctx, resourceName, &logSource),
 					resource.TestCheckResourceAttr(resourceName, "source.#", "1"),
@@ -95,10 +83,9 @@ func testAccAWSLogSource_multiRegion(t *testing.T) {
 				),
 			},
 			{
-				ResourceName:            resourceName,
-				ImportState:             true,
-				ImportStateVerify:       true,
-				ImportStateVerifyIgnore: []string{""},
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
 			},
 		},
 	})
@@ -106,9 +93,7 @@ func testAccAWSLogSource_multiRegion(t *testing.T) {
 
 func testAccAWSLogSource_disappears(t *testing.T) {
 	ctx := acctest.Context(t)
-
 	resourceName := "aws_securitylake_aws_log_source.test"
-	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	var logSource types.AwsLogSourceConfiguration
 
 	resource.Test(t, resource.TestCase{
@@ -122,7 +107,7 @@ func testAccAWSLogSource_disappears(t *testing.T) {
 		CheckDestroy:             testAccCheckAWSLogSourceDestroy(ctx),
 		Steps: []resource.TestStep{
 			{
-				Config: testAccLogSourceConfig_basic(rName),
+				Config: testAccAWSLogSourceConfig_basic(),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAWSLogSourceExists(ctx, resourceName, &logSource),
 					acctest.CheckFrameworkResourceDisappears(ctx, acctest.Provider, tfsecuritylake.ResourceAWSLogSource, resourceName),
@@ -152,66 +137,63 @@ func testAccCheckAWSLogSourceDestroy(ctx context.Context) resource.TestCheckFunc
 				return err
 			}
 
-			return create.Error(names.SecurityLake, create.ErrActionCheckingDestroyed, tfsecuritylake.ResNameAWSLogSource, rs.Primary.ID, errors.New("not destroyed"))
+			return fmt.Errorf("Security Lake AWS Log Source %s still exists", rs.Primary.ID)
 		}
 
 		return nil
 	}
 }
 
-func testAccCheckAWSLogSourceExists(ctx context.Context, name string, logSource *types.AwsLogSourceConfiguration) resource.TestCheckFunc {
+func testAccCheckAWSLogSourceExists(ctx context.Context, n string, v *types.AwsLogSourceConfiguration) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
-		rs, ok := s.RootModule().Resources[name]
+		rs, ok := s.RootModule().Resources[n]
 		if !ok {
-			return create.Error(names.SecurityLake, create.ErrActionCheckingExistence, tfsecuritylake.ResNameAWSLogSource, name, errors.New("not found"))
-		}
-
-		if rs.Primary.ID == "" {
-			return create.Error(names.SecurityLake, create.ErrActionCheckingExistence, tfsecuritylake.ResNameAWSLogSource, name, errors.New("not set"))
+			return fmt.Errorf("Not found: %s", n)
 		}
 
 		conn := acctest.Provider.Meta().(*conns.AWSClient).SecurityLakeClient(ctx)
 
-		resp, err := tfsecuritylake.FindAWSLogSourceBySourceName(ctx, conn, types.AwsLogSourceName(rs.Primary.ID))
+		output, err := tfsecuritylake.FindAWSLogSourceBySourceName(ctx, conn, types.AwsLogSourceName(rs.Primary.ID))
+
 		if err != nil {
-			return create.Error(names.SecurityLake, create.ErrActionCheckingExistence, tfsecuritylake.ResNameAWSLogSource, rs.Primary.ID, err)
+			return err
 		}
 
-		*logSource = *resp
+		*v = *output
 
 		return nil
 	}
 }
 
-func testAccLogSourceConfig_basic(rName string) string {
-	return acctest.ConfigCompose(testAccDataLakeConfig_basic(rName), fmt.Sprintf(`
+func testAccAWSLogSourceConfig_basic() string {
+	return acctest.ConfigCompose(testAccDataLakeConfig_basic(), fmt.Sprintf(`
 data "aws_caller_identity" "test" {}
 
 resource "aws_securitylake_aws_log_source" "test" {
   source {
     accounts       = [data.aws_caller_identity.test.account_id]
-    regions        = [%[2]q]
+    regions        = [%[1]q]
     source_name    = "ROUTE53"
     source_version = "1.0"
   }
   depends_on = [aws_securitylake_data_lake.test]
 }
-`, rName, acctest.Region()))
+`, acctest.Region()))
 }
 
-func testAccLogSourceConfig_multiRegion(rName string) string {
+func testAccAWSLogSourceConfig_multiRegion(rName string) string {
 	return acctest.ConfigCompose(testAccDataLakeConfig_replication(rName), fmt.Sprintf(`
 data "aws_caller_identity" "test" {}
 
 resource "aws_securitylake_aws_log_source" "test" {
   source {
     accounts       = [data.aws_caller_identity.test.account_id]
-    regions        = [%[2]q,%[3]q]
+    regions        = [%[1]q, %[2]q]
     source_name    = "ROUTE53"
     source_version = "1.0"
   }
 
   depends_on = [aws_securitylake_data_lake.test, aws_securitylake_data_lake.region_2]
 }
-`, rName, acctest.Region(), acctest.AlternateRegion()))
+`, acctest.Region(), acctest.AlternateRegion()))
 }
diff --git a/internal/service/securitylake/custom_log_source_test.go b/internal/service/securitylake/custom_log_source_test.go
new file mode 100644
index 00000000000..f6d300559ee
--- /dev/null
+++ b/internal/service/securitylake/custom_log_source_test.go
@@ -0,0 +1,199 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package securitylake_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/service/securitylake/types"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	tfsecuritylake "github.com/hashicorp/terraform-provider-aws/internal/service/securitylake"
+	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+	"github.com/hashicorp/terraform-provider-aws/names"
+)
+
+func testAccCustomLogSource_basic(t *testing.T) {
+	ctx := acctest.Context(t)
+	resourceName := "aws_securitylake_custom_log_source.test"
+	var customLogSource types.CustomLogSourceResource
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.SecurityLake)
+		},
+		ErrorCheck:               acctest.ErrorCheck(t, names.SecurityLakeEndpointID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckCustomLogSourceDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCustomLogSourceConfig_basic(),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckCustomLogSourceExists(ctx, resourceName, &customLogSource),
+					resource.TestCheckResourceAttr(resourceName, "configuration.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "configuration.0.crawler_configuration.#", "1"),
+					resource.TestCheckResourceAttrPair(resourceName, "configuration.0.crawler_configuration.0.role_arn", "aws_iam_role.test", "arn"),
+					resource.TestCheckResourceAttr(resourceName, "configuration.0.provider_identity.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "configuration.0.provider_identity.0.external_id", "windows-sysmon-test"),
+					resource.TestCheckResourceAttr(resourceName, "event_classes.#", "1"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "event_classes.*", "FILE_ACTIVITY"),
+					resource.TestCheckResourceAttr(resourceName, "source_name", "windows-sysmon"),
+					resource.TestCheckResourceAttr(resourceName, "source_version", "1.0"),
+				),
+			},
+			{
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"configuration", "event_classes"},
+			},
+		},
+	})
+}
+
+func testAccCustomLogSource_disappears(t *testing.T) {
+	ctx := acctest.Context(t)
+	resourceName := "aws_securitylake_custom_log_source.test"
+	var customLogSource types.CustomLogSourceResource
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.SecurityLake)
+			acctest.PreCheckOrganizationsAccount(ctx, t)
+		},
+		ErrorCheck:               acctest.ErrorCheck(t, names.SecurityLakeEndpointID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckCustomLogSourceDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCustomLogSourceConfig_basic(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCustomLogSourceExists(ctx, resourceName, &customLogSource),
+					acctest.CheckFrameworkResourceDisappears(ctx, acctest.Provider, tfsecuritylake.ResourceCustomLogSource, resourceName),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
+func testAccCheckCustomLogSourceDestroy(ctx context.Context) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := acctest.Provider.Meta().(*conns.AWSClient).SecurityLakeClient(ctx)
+
+		for _, rs := range s.RootModule().Resources {
+			if rs.Type != "aws_securitylake_custom_log_source" {
+				continue
+			}
+
+			_, err := tfsecuritylake.FindCustomLogSourceBySourceName(ctx, conn, rs.Primary.ID)
+
+			if tfresource.NotFound(err) {
+				continue
+			}
+
+			if err != nil {
+				return err
+			}
+
+			return fmt.Errorf("Security Lake Custom Log Source %s still exists", rs.Primary.ID)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckCustomLogSourceExists(ctx context.Context, n string, v *types.CustomLogSourceResource) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		conn := acctest.Provider.Meta().(*conns.AWSClient).SecurityLakeClient(ctx)
+
+		output, err := tfsecuritylake.FindCustomLogSourceBySourceName(ctx, conn, rs.Primary.ID)
+
+		if err != nil {
+			return err
+		}
+
+		*v = *output
+
+		return nil
+	}
+}
+
+func testAccCustomLogSourceConfig_basic() string {
+	return acctest.ConfigCompose(testAccDataLakeConfig_basic(), `
+resource "aws_iam_role" "test" {
+  name = "AmazonSecurityLakeCustomDataGlueCrawler-windows-sysmon"
+  path = "/service-role/"
+
+  assume_role_policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [{
+    "Action": "sts:AssumeRole",
+    "Principal": {
+      "Service": "glue.amazonaws.com"
+    },
+    "Effect": "Allow"
+  }]
+}
+POLICY
+}
+
+resource "aws_iam_role_policy" "test" {
+  name = "AmazonSecurityLakeCustomDataGlueCrawler-windows-sysmon"
+  role = aws_iam_role.test.name
+
+  policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [{
+    "Effect": "Allow",
+    "Action": [
+      "s3:GetObject",
+      "s3:PutObject"
+    ],
+    "Resource": "*"
+  }]
+}
+POLICY
+
+  depends_on = [aws_securitylake_data_lake.test]
+}
+
+resource "aws_iam_role_policy_attachment" "test" {
+  policy_arn = "arn:${data.aws_partition.current.partition}:iam::aws:policy/service-role/AWSGlueServiceRole"
+  role       = aws_iam_role.test.name
+}
+
+resource "aws_securitylake_custom_log_source" "test" {
+  source_name    = "windows-sysmon"
+  source_version = "1.0"
+  event_classes  = ["FILE_ACTIVITY"]
+
+  configuration {
+    crawler_configuration {
+      role_arn = aws_iam_role.test.arn
+    }
+
+    provider_identity {
+      external_id = "windows-sysmon-test"
+      principal   = data.aws_caller_identity.current.account_id
+    }
+  }
+
+  depends_on = [aws_securitylake_data_lake.test, aws_iam_role.test]
+}
+`)
+}
diff --git a/internal/service/securitylake/data_lake_test.go b/internal/service/securitylake/data_lake_test.go
index 93aa04ea80c..9e674be2526 100644
--- a/internal/service/securitylake/data_lake_test.go
+++ b/internal/service/securitylake/data_lake_test.go
@@ -5,7 +5,6 @@ package securitylake_test
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"testing"
 
@@ -15,7 +14,6 @@ import (
 	"github.com/hashicorp/terraform-plugin-testing/terraform"
 	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
 	"github.com/hashicorp/terraform-provider-aws/internal/conns"
-	"github.com/hashicorp/terraform-provider-aws/internal/create"
 	tfsecuritylake "github.com/hashicorp/terraform-provider-aws/internal/service/securitylake"
 	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
 	"github.com/hashicorp/terraform-provider-aws/names"
@@ -23,12 +21,7 @@ import (
 
 func testAccDataLake_basic(t *testing.T) {
 	ctx := acctest.Context(t)
-	if testing.Short() {
-		t.Skip("skipping long-running test in short mode")
-	}
-
 	var datalake types.DataLakeResource
-	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_securitylake_data_lake.test"
 
 	resource.Test(t, resource.TestCase{
@@ -42,7 +35,7 @@ func testAccDataLake_basic(t *testing.T) {
 		CheckDestroy:             testAccCheckDataLakeDestroy(ctx),
 		Steps: []resource.TestStep{
 			{
-				Config: testAccDataLakeConfig_basic(rName),
+				Config: testAccDataLakeConfig_basic(),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckDataLakeExists(ctx, resourceName, &datalake),
 					resource.TestCheckResourceAttr(resourceName, "configuration.#", "1"),
@@ -69,7 +62,6 @@ func testAccDataLake_basic(t *testing.T) {
 func testAccDataLake_disappears(t *testing.T) {
 	ctx := acctest.Context(t)
 	var datalake types.DataLakeResource
-	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_securitylake_data_lake.test"
 
 	resource.Test(t, resource.TestCase{
@@ -83,7 +75,7 @@ func testAccDataLake_disappears(t *testing.T) {
 		CheckDestroy:             testAccCheckDataLakeDestroy(ctx),
 		Steps: []resource.TestStep{
 			{
-				Config: testAccDataLakeConfig_basic(rName),
+				Config: testAccDataLakeConfig_basic(),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckDataLakeExists(ctx, resourceName, &datalake),
 					acctest.CheckFrameworkResourceDisappears(ctx, acctest.Provider, tfsecuritylake.ResourceDataLake, resourceName),
@@ -96,12 +88,7 @@ func testAccDataLake_disappears(t *testing.T) {
 
 func testAccDataLake_tags(t *testing.T) {
 	ctx := acctest.Context(t)
-	if testing.Short() {
-		t.Skip("skipping long-running test in short mode")
-	}
-
 	var datalake types.DataLakeResource
-	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_securitylake_data_lake.test"
 
 	resource.Test(t, resource.TestCase{
@@ -115,7 +102,7 @@ func testAccDataLake_tags(t *testing.T) {
 		CheckDestroy:             testAccCheckDataLakeDestroy(ctx),
 		Steps: []resource.TestStep{
 			{
-				Config: testAccDataLakeConfig_tags1(rName, "key1", "value1"),
+				Config: testAccDataLakeConfig_tags1("key1", "value1"),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckDataLakeExists(ctx, resourceName, &datalake),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
@@ -129,7 +116,7 @@ func testAccDataLake_tags(t *testing.T) {
 				ImportStateVerifyIgnore: []string{"meta_store_manager_role_arn"},
 			},
 			{
-				Config: testAccDataLakeConfig_tags2(rName, "key1", "value1updated", "key2", "value2"),
+				Config: testAccDataLakeConfig_tags2("key1", "value1updated", "key2", "value2"),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckDataLakeExists(ctx, resourceName, &datalake),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "2"),
@@ -138,7 +125,7 @@ func testAccDataLake_tags(t *testing.T) {
 				),
 			},
 			{
-				Config: testAccDataLakeConfig_tags1(rName, "key2", "value2"),
+				Config: testAccDataLakeConfig_tags1("key2", "value2"),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckDataLakeExists(ctx, resourceName, &datalake),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
@@ -151,10 +138,6 @@ func testAccDataLake_tags(t *testing.T) {
 
 func testAccDataLake_lifeCycle(t *testing.T) {
 	ctx := acctest.Context(t)
-	if testing.Short() {
-		t.Skip("skipping long-running test in short mode")
-	}
-
 	var datalake types.DataLakeResource
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_securitylake_data_lake.test"
@@ -199,10 +182,6 @@ func testAccDataLake_lifeCycle(t *testing.T) {
 
 func testAccDataLake_lifeCycleUpdate(t *testing.T) {
 	ctx := acctest.Context(t)
-	if testing.Short() {
-		t.Skip("skipping long-running test in short mode")
-	}
-
 	var datalake types.DataLakeResource
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_securitylake_data_lake.test"
@@ -269,10 +248,6 @@ func testAccDataLake_lifeCycleUpdate(t *testing.T) {
 
 func testAccDataLake_replication(t *testing.T) {
 	ctx := acctest.Context(t)
-	if testing.Short() {
-		t.Skip("skipping long-running test in short mode")
-	}
-
 	var datalake types.DataLakeResource
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	resourceName := "aws_securitylake_data_lake.region_2"
@@ -337,38 +312,35 @@ func testAccCheckDataLakeDestroy(ctx context.Context) resource.TestCheckFunc {
 				return err
 			}
 
-			return create.Error(names.SecurityLake, create.ErrActionCheckingDestroyed, tfsecuritylake.ResNameDataLake, rs.Primary.ID, errors.New("not destroyed"))
+			return fmt.Errorf("Security Lake Data Lake %s still exists", rs.Primary.ID)
 		}
 
 		return nil
 	}
 }
 
-func testAccCheckDataLakeExists(ctx context.Context, name string, datalake *types.DataLakeResource) resource.TestCheckFunc {
+func testAccCheckDataLakeExists(ctx context.Context, n string, v *types.DataLakeResource) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
-		rs, ok := s.RootModule().Resources[name]
+		rs, ok := s.RootModule().Resources[n]
 		if !ok {
-			return create.Error(names.SecurityLake, create.ErrActionCheckingExistence, tfsecuritylake.ResNameDataLake, name, errors.New("not found"))
-		}
-
-		if rs.Primary.ID == "" {
-			return create.Error(names.SecurityLake, create.ErrActionCheckingExistence, tfsecuritylake.ResNameDataLake, name, errors.New("not set"))
+			return fmt.Errorf("Not found: %s", n)
 		}
 
 		conn := acctest.Provider.Meta().(*conns.AWSClient).SecurityLakeClient(ctx)
-		resp, err := tfsecuritylake.FindDataLakeByARN(ctx, conn, rs.Primary.ID)
+
+		output, err := tfsecuritylake.FindDataLakeByARN(ctx, conn, rs.Primary.ID)
+
 		if err != nil {
-			return create.Error(names.SecurityLake, create.ErrActionCheckingExistence, tfsecuritylake.ResNameDataLake, rs.Primary.ID, err)
+			return err
 		}
 
-		*datalake = *resp
+		*v = *output
 
 		return nil
 	}
 }
 
-func testAccDataLakeConfigConfig_base(rName string) string {
-	return `
+const testAccDataLakeConfigConfig_base = `
 data "aws_caller_identity" "current" {}
 data "aws_partition" "current" {}
 
@@ -538,61 +510,60 @@ resource "aws_kms_key" "test" {
 POLICY
 }
 `
-}
 
-func testAccDataLakeConfig_basic(rName string) string {
-	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base(rName), fmt.Sprintf(`
+func testAccDataLakeConfig_basic() string {
+	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base, fmt.Sprintf(`
 resource "aws_securitylake_data_lake" "test" {
   meta_store_manager_role_arn = aws_iam_role.meta_store_manager.arn
 
   configuration {
-    region = %[2]q
+    region = %[1]q
   }
 
   depends_on = [aws_iam_role.meta_store_manager]
 }
-`, rName, acctest.Region()))
+`, acctest.Region()))
 }
 
-func testAccDataLakeConfig_tags1(rName, tag1Key, tag1Value string) string {
-	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base(rName), fmt.Sprintf(`
+func testAccDataLakeConfig_tags1(tag1Key, tag1Value string) string {
+	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base, fmt.Sprintf(`
 resource "aws_securitylake_data_lake" "test" {
   meta_store_manager_role_arn = aws_iam_role.meta_store_manager.arn
 
   configuration {
-    region = %[4]q
+    region = %[3]q
   }
 
   tags = {
-    %[2]q = %[3]q
+    %[1]q = %[2]q
   }
 
   depends_on = [aws_iam_role.meta_store_manager]
 }
-`, rName, tag1Key, tag1Value, acctest.Region()))
+`, tag1Key, tag1Value, acctest.Region()))
 }
 
-func testAccDataLakeConfig_tags2(rName, tag1Key, tag1Value, tag2Key, tag2Value string) string {
-	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base(rName), fmt.Sprintf(`
+func testAccDataLakeConfig_tags2(tag1Key, tag1Value, tag2Key, tag2Value string) string {
+	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base, fmt.Sprintf(`
 resource "aws_securitylake_data_lake" "test" {
   meta_store_manager_role_arn = aws_iam_role.meta_store_manager.arn
 
   configuration {
-    region = %[6]q
+    region = %[5]q
   }
 
   tags = {
-    %[2]q = %[3]q
-    %[4]q = %[5]q
+    %[1]q = %[2]q
+    %[3]q = %[4]q
   }
 
   depends_on = [aws_iam_role.meta_store_manager]
 }
-`, rName, tag1Key, tag1Value, tag2Key, tag2Value, acctest.Region()))
+`, tag1Key, tag1Value, tag2Key, tag2Value, acctest.Region()))
 }
 
 func testAccDataLakeConfig_lifeCycle(rName string) string {
-	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base(rName), fmt.Sprintf(`
+	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base, fmt.Sprintf(`
 resource "aws_securitylake_data_lake" "test" {
   meta_store_manager_role_arn = aws_iam_role.meta_store_manager.arn
 
@@ -628,7 +599,7 @@ resource "aws_securitylake_data_lake" "test" {
 }
 
 func testAccDataLakeConfig_lifeCycleUpdate(rName string) string {
-	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base(rName), fmt.Sprintf(`
+	return acctest.ConfigCompose(testAccDataLakeConfigConfig_base, fmt.Sprintf(`
 resource "aws_securitylake_data_lake" "test" {
   meta_store_manager_role_arn = aws_iam_role.meta_store_manager.arn
 
@@ -660,7 +631,7 @@ resource "aws_securitylake_data_lake" "test" {
 }
 
 func testAccDataLakeConfig_replication(rName string) string {
-	return acctest.ConfigCompose(testAccDataLakeConfig_basic(rName), fmt.Sprintf(`
+	return acctest.ConfigCompose(testAccDataLakeConfig_basic(), fmt.Sprintf(`
 resource "aws_securitylake_data_lake" "region_2" {
   meta_store_manager_role_arn = aws_iam_role.meta_store_manager.arn
 
diff --git a/internal/service/securitylake/exports_test.go b/internal/service/securitylake/exports_test.go
index a49ab1e406f..e48e68ea517 100644
--- a/internal/service/securitylake/exports_test.go
+++ b/internal/service/securitylake/exports_test.go
@@ -5,9 +5,11 @@ package securitylake
 
 // Exports for use in tests only.
 var (
-	ResourceAWSLogSource = newAWSLogSourceResource
-	ResourceDataLake     = newDataLakeResource
+	ResourceAWSLogSource    = newAWSLogSourceResource
+	ResourceCustomLogSource = newCustomLogSourceResource
+	ResourceDataLake        = newDataLakeResource
 
-	FindAWSLogSourceBySourceName = findAWSLogSourceBySourceName
-	FindDataLakeByARN            = findDataLakeByARN
+	FindAWSLogSourceBySourceName    = findAWSLogSourceBySourceName
+	FindCustomLogSourceBySourceName = findCustomLogSourceBySourceName
+	FindDataLakeByARN               = findDataLakeByARN
 )
diff --git a/internal/service/securitylake/securitylake_test.go b/internal/service/securitylake/securitylake_test.go
index 7718df725fe..cef7ca99d14 100644
--- a/internal/service/securitylake/securitylake_test.go
+++ b/internal/service/securitylake/securitylake_test.go
@@ -18,6 +18,10 @@ func TestAccSecurityLake_serial(t *testing.T) {
 			"disappears":  testAccAWSLogSource_disappears,
 			"multiRegion": testAccAWSLogSource_multiRegion,
 		},
+		"CustomLogSource": {
+			"basic":      testAccCustomLogSource_basic,
+			"disappears": testAccCustomLogSource_disappears,
+		},
 		"DataLake": {
 			"basic":           testAccDataLake_basic,
 			"disappears":      testAccDataLake_disappears,
