diff --git a/cli/tests/extension/mod.rs b/cli/tests/extension/mod.rs
index d269b1721..1094f09fe 100644
--- a/cli/tests/extension/mod.rs
+++ b/cli/tests/extension/mod.rs
@@ -37,12 +37,12 @@ fn init_resolver() {
     codegen-units = 1
 
     [dependencies]
-    grafbase-sdk = "0.5.2"
+    grafbase-sdk = "0.5.4"
 
     [dev-dependencies]
     indoc = "2"
     insta = { version = "1.42.1", features = ["json"] }
-    grafbase-sdk = { version = "0.5.2", features = ["test-utils"] }
+    grafbase-sdk = { version = "0.5.4", features = ["test-utils"] }
     tokio = { version = "1", features = ["rt-multi-thread", "macros", "test-util"] }
     serde_json = "1"
     "#);
@@ -281,12 +281,12 @@ fn init_auth() {
     codegen-units = 1
 
     [dependencies]
-    grafbase-sdk = "0.5.2"
+    grafbase-sdk = "0.5.4"
 
     [dev-dependencies]
     indoc = "2"
     insta = { version = "1.42.1", features = ["json"] }
-    grafbase-sdk = { version = "0.5.2", features = ["test-utils"] }
+    grafbase-sdk = { version = "0.5.4", features = ["test-utils"] }
     tokio = { version = "1", features = ["rt-multi-thread", "macros", "test-util"] }
     serde_json = "1"
     "#);
diff --git a/crates/grafbase-sdk/src/test/config.rs b/crates/grafbase-sdk/src/test/config.rs
index d71a54781..30580e771 100644
--- a/crates/grafbase-sdk/src/test/config.rs
+++ b/crates/grafbase-sdk/src/test/config.rs
@@ -19,6 +19,8 @@ pub enum LogLevel {
     Warn,
     /// Show only error messages.
     Error,
+    /// Show all output from engine, debug upwards.
+    EngineDebug,
 }
 
 impl AsRef<str> for LogLevel {
@@ -29,6 +31,7 @@ impl AsRef<str> for LogLevel {
             LogLevel::Info => "info",
             LogLevel::Warn => "warn",
             LogLevel::Error => "error",
+            LogLevel::EngineDebug => "engine=debug",
         }
     }
 }
diff --git a/crates/grafbase-sdk/src/test/runner.rs b/crates/grafbase-sdk/src/test/runner.rs
index a4762921c..5b7690fef 100644
--- a/crates/grafbase-sdk/src/test/runner.rs
+++ b/crates/grafbase-sdk/src/test/runner.rs
@@ -142,7 +142,7 @@ impl TestRunner {
 
         let mut expr = duct::cmd(&self.config.gateway_path, args);
 
-        if !self.config.enable_stderr {
+        if !dbg!(self.config.enable_stderr) {
             expr = expr.stderr_null();
         }
 
diff --git a/extensions/nats/tests/integration_tests.rs b/extensions/nats/tests/integration_tests.rs
index bdace603f..67534384a 100644
--- a/extensions/nats/tests/integration_tests.rs
+++ b/extensions/nats/tests/integration_tests.rs
@@ -33,19 +33,10 @@ fn subgraph() -> ExtensionOnlySubgraph {
               "@natsSubscription",
               "@natsRequest",
               "@natsKeyValue",
-              "NatsPublishResult",
-              "NatsStreamDeliverPolicy"
+              "NatsStreamDeliverPolicy",
             ]
           )
 
-        input RequestReplyInput {{
-          message: String!
-        }}
-
-        type RequestReplyResult {{
-          message: String!
-        }}
-
         type Query {{
           hello: String!
 
@@ -75,23 +66,23 @@ fn subgraph() -> ExtensionOnlySubgraph {
         }}
 
         type Mutation {{
-          publishUserEvent(id: Int!, input: UserEventInput!): NatsPublishResult! @natsPublish(
+          publishUserEvent(id: Int!, input: UserEventInput!): Boolean! @natsPublish(
             subject: "publish.user.{{{{args.id}}}}.events"
           )
 
-          kvPutUser(id: Int!, input: UserEventInput!): NatsKeyValueResult! @natsKeyValue(
+          kvPutUser(id: Int!, input: UserEventInput!): String! @natsKeyValue(
             bucket: "putUsers"
             key: "user.{{{{ args.id }}}}"
             action: PUT
           )
 
-          kvCreateUser(id: Int!, input: UserEventInput!): NatsKeyValueResult! @natsKeyValue(
+          kvCreateUser(id: Int!, input: UserEventInput!): String! @natsKeyValue(
             bucket: "createUsers"
             key: "user.{{{{ args.id }}}}"
             action: CREATE
           )
 
-          kvDeleteUser(id: Int!): NatsPublishResult! @natsKeyValue(
+          kvDeleteUser(id: Int!): Boolean! @natsKeyValue(
             bucket: "deleteUsers"
             key: "user.{{{{ args.id }}}}"
             action: DELETE
@@ -127,12 +118,12 @@ fn subgraph() -> ExtensionOnlySubgraph {
           )
         }}
 
-        type NatsPublishResult {{
-          success: Boolean!
+        input RequestReplyInput {{
+          message: String!
         }}
 
-        type NatsKeyValueResult {{
-          sequence: Int!
+        type RequestReplyResult {{
+          message: String!
         }}
 
         input UserEventInput {{
@@ -304,9 +295,7 @@ async fn test_publish() {
 
     let query = indoc! {r#"
         mutation {
-          publishUserEvent(id: 1, input: { email: "alice@example.com", name: "Alice" }) {
-            success
-          }
+          publishUserEvent(id: 1, input: { email: "alice@example.com", name: "Alice" })
         }
     "#};
 
@@ -314,9 +303,7 @@ async fn test_publish() {
     insta::assert_json_snapshot!(result, @r#"
     {
       "data": {
-        "publishUserEvent": {
-          "success": true
-        }
+        "publishUserEvent": true
       }
     }
     "#);
@@ -690,9 +677,7 @@ async fn kv_put() {
 
     let query = indoc! {r#"
         mutation {
-          kvPutUser(id: 1,input: { email: "user1@example.com", name: "User One" }) {
-            sequence
-          }
+          kvPutUser(id: 1,input: { email: "user1@example.com", name: "User One" })
         }
     "#};
 
@@ -700,9 +685,7 @@ async fn kv_put() {
     insta::assert_json_snapshot!(result, @r#"
     {
       "data": {
-        "kvPutUser": {
-          "sequence": 1
-        }
+        "kvPutUser": "1"
       }
     }
     "#);
@@ -744,9 +727,7 @@ async fn kv_create() {
 
     let query = indoc! {r#"
         mutation {
-          kvCreateUser(id: 1,input: { email: "user1@example.com", name: "User One" }) {
-            sequence
-          }
+          kvCreateUser(id: 1,input: { email: "user1@example.com", name: "User One" })
         }
     "#};
 
@@ -754,9 +735,7 @@ async fn kv_create() {
     insta::assert_json_snapshot!(result, @r#"
     {
       "data": {
-        "kvCreateUser": {
-          "sequence": 1
-        }
+        "kvCreateUser": "1"
       }
     }
     "#);
@@ -808,9 +787,7 @@ async fn kv_delete() {
 
     let query = indoc! {r#"
         mutation {
-          kvDeleteUser(id: 1) {
-            success
-          }
+          kvDeleteUser(id: 1)
         }
     "#};
 
@@ -818,9 +795,7 @@ async fn kv_delete() {
     insta::assert_json_snapshot!(result, @r#"
     {
       "data": {
-        "kvDeleteUser": {
-          "success": true
-        }
+        "kvDeleteUser": true
       }
     }
     "#);
diff --git a/extensions/rest/tests/integration_tests.rs b/extensions/rest/tests/integration_tests.rs
index d869fa8ae..9c23d9573 100644
--- a/extensions/rest/tests/integration_tests.rs
+++ b/extensions/rest/tests/integration_tests.rs
@@ -2,8 +2,8 @@ use grafbase_sdk::test::{DynamicSchema, ExtensionOnlySubgraph, TestConfig, TestR
 use indoc::{formatdoc, indoc};
 use serde_json::json;
 use wiremock::{
-    matchers::{body_json, header, method, path},
     Mock, MockServer, ResponseTemplate,
+    matchers::{body_json, header, method, path},
 };
 
 const CLI_PATH: &str = "../../target/debug/grafbase";
@@ -85,7 +85,7 @@ fn subgraph(rest_endpoint: &str) -> ExtensionOnlySubgraph {
         type User {{
           id: ID!
           name: String!
-          age: Int!
+          age: Int
         }}
 
         input UserInput {{
@@ -221,6 +221,92 @@ async fn get_one() {
     "#);
 }
 
+#[tokio::test]
+async fn get_one_missing() {
+    let response_body = json!(null);
+
+    let template = ResponseTemplate::new(200).set_body_json(response_body);
+    let mock_server = mock_server("/users/1", template).await;
+    let subgraph = subgraph(&mock_server.uri());
+
+    let config = TestConfig::builder()
+        .with_cli(CLI_PATH)
+        .with_gateway(GATEWAY_PATH)
+        .with_subgraph(subgraph)
+        .enable_networking()
+        .build("")
+        .unwrap();
+
+    let runner = TestRunner::new(config).await.unwrap();
+
+    let query = indoc! {r#"
+        query {
+          user(id: 1) {
+            id
+            name
+            age
+          }
+        }
+    "#};
+
+    let result: serde_json::Value = runner.graphql_query(query).send().await.unwrap();
+
+    insta::assert_json_snapshot!(result, @r#"
+    {
+      "data": {
+        "user": null
+      }
+    }
+    "#);
+}
+
+#[tokio::test]
+async fn get_one_nested_null() {
+    let response_body = json!({
+        "id": "1",
+        "name": "John Doe",
+        "age": null,
+    });
+
+    let template = ResponseTemplate::new(200).set_body_json(response_body);
+    let mock_server = mock_server("/users/1", template).await;
+    let subgraph = subgraph(&mock_server.uri());
+
+    let config = TestConfig::builder()
+        .with_cli(CLI_PATH)
+        .with_gateway(GATEWAY_PATH)
+        .with_subgraph(subgraph)
+        .enable_networking()
+        .build("")
+        .unwrap();
+
+    let runner = TestRunner::new(config).await.unwrap();
+
+    let query = indoc! {r#"
+        query {
+          user(id: 1) {
+            id
+            name
+            age
+          }
+        }
+    "#};
+
+    let result: serde_json::Value = runner.graphql_query(query).send().await.unwrap();
+
+    insta::assert_json_snapshot!(result, @r#"
+    {
+      "data": {
+        "user": {
+          "id": "1",
+          "name": "John Doe",
+          "age": null
+        }
+      }
+    }
+    "#);
+}
+
 #[tokio::test]
 async fn get_some_fields() {
     let response_body = json!([
